<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>ANNOgesic - Transcriptome annotation pipeline &mdash; ANNOgesic 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ANNOgesic 1.0.0 documentation" href="#" />
    <link rel="next" title="ANNOgesic - Transcriptome annotation pipeline" href="" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="#" title="ANNOgesic - Transcriptome annotation pipeline"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">ANNOgesic 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="annogesic-transcriptome-annotation-pipeline">
<h1>ANNOgesic - Transcriptome annotation pipeline<a class="headerlink" href="#annogesic-transcriptome-annotation-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="table-of-content">
<h2>Table of content<a class="headerlink" href="#table-of-content" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="">ANNOgesic - Transcriptome annotation pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="prerequired.html">Pre-required tools or databases</a></li>
<li class="toctree-l1"><a class="reference internal" href="docker.html">Dockfiles</a></li>
<li class="toctree-l1"><a class="reference internal" href="subcommands.html">ANNOgesic&#8217;s subcommands</a></li>
<li class="toctree-l1"><a class="reference internal" href="test_case.html">Performing an test cast</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
</ul>
</div>
</div>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>ANNOgesic is a bacterial transcriptome annotation pipeline based on RNA-Seq.
ANNOgesic covers different aspects of the genome annotation. In order to get the
best results, ANNOgesic also can optimize the parameters of some tools.
ANNOgesic can automatically generate high-quality annotation information for
query strains. Moreover, it is modular and its subcommands can be separately
used. Currently, the pipeline already can detect or integrate a) fasta, CDS,
tRNA, rRNA and genes of query genome, b) transcription starting sites
(TSSs), c) rho-independent terminators, d) transcript assembly,
e) untranslated region (UTRs), f) sRNA, g) promoters, h) processing sites,
i) circular RNAs, j) protein-protein interaction networks, k) potential
sRNA target, l) single-nucleotide polymorphism (SNP), m) operons,
n) GO terms, o) subcellular localization, p) riboswitch, q) potential sORF.</p>
<div class="highlight-python"><div class="highlight"><pre>usage: ANNOgesic.py [-h] [--version]
                 {create,get_input_files,get_target_fasta,annotation_transfer,expression_analysis,tsspredator,optimize_tsspredator,color_png,terminator,transcript_assembly,utr,srna,sorf,promoter,operon,circrna,go_term,srna_target,snp,ppi_network,subcellular_localization,riboswitch,screenshot}
                 ...

positional arguments:
{create,get_input_files,get_target_fasta,annotation_transfer,expression_analysis,tsspredator,optimize_tsspredator,color_png,terminator,transcript_assembly,utr,srna,sorf,promoter,operon,circrna,go_term,srna_target,snp,ppi_network,subcellular_localization,riboswitch,screenshot}
                     commands
 create              Create a project
 get_input_files     Get required files (i.e. annotation files, fasta
                     files)
 get_target_fasta    Get target fasta.
 annotation_transfer
                     Run RATT to transfer the annotation files from
                     reference to target.
 expression_analysis
                     Run gene expression analysis to compare which CDS is
                     expressing in which libraries
 tsspredator         Run TSSpredator to predict TSSs or processing sites.
 optimize_tsspredator
                     Optimize TSSpredator based on (partial)manual detect
                     one.
 color_png           Generating color screenshots of TSS or processing
                     site. It only works after running batch script.
 terminator          Detect Terminators.
 transcript_assembly
                     Run Transcript for doing transcriptome assembly.
 utr                 Run UTR detection to detect 5&#39;UTR and 3&#39;UTR.
 srna                Run sRNA detection to detect sRNA candidates.
 sorf                Run sORF detection to detect sORF candidates which has
                     expression.
 promoter            Run MEME to dicover promoter.
 operon              Detect operon and combine features together.
 circrna             Detect circular RNA.
 go_term             Extract and find Go terms.
 srna_target         sRNA target prediction.
 snp                 Detection of SNP of transcripts.
 ppi_network         Generate protein-protein interaction with literature
                     supported.
 subcellular_localization
                     prediction of subcellular localization of genomic CDS.
 riboswitch          prediction of riboswitch.
 screenshot          Generate screenshot for selected feature.

optional arguments:
-h, --help            show this help message and exit
--version, -v         show version
</pre></div>
</div>
</div>
<div class="section" id="download">
<h2>Download<a class="headerlink" href="#download" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre>git clone git@github.com:Sung-Huan/ANNOgesic.git
</pre></div>
</div>
<p>or</p>
<div class="highlight-python"><div class="highlight"><pre>pip3 install annogesic
</pre></div>
</div>
</div>
<div class="section" id="source-code">
<h2>Source code<a class="headerlink" href="#source-code" title="Permalink to this headline">¶</a></h2>
<p>The source code of ANNOgesic can be found at <a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master">https://github.com/Sung-Huan/ANNOgesic/tree/master</a>.</p>
</div>
<div class="section" id="cite">
<h2>Cite<a class="headerlink" href="#cite" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="contact">
<h2>Contact<a class="headerlink" href="#contact" title="Permalink to this headline">¶</a></h2>
<p>For question and requests feel free to contact <a class="reference external" href="mailto:sung-huan&#46;yu&#37;&#52;&#48;uni-wuerzburg&#46;de">Sung-Huan Yu</a></p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">ANNOgesic - Transcriptome annotation pipeline</a><ul>
<li><a class="reference internal" href="#table-of-content">Table of content</a></li>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#download">Download</a></li>
<li><a class="reference internal" href="#source-code">Source code</a></li>
<li><a class="reference internal" href="#cite">Cite</a></li>
<li><a class="reference internal" href="#contact">Contact</a></li>
</ul>
</li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href=""
                        title="next chapter">ANNOgesic - Transcriptome annotation pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, Sung-Huan Yu.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.4</a>
      
      |
      <a href="_sources/index.txt"
          rel="nofollow">Page source</a></li>
    </div>

    

    
  </body>
</html>