<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Required tools or databases &mdash; ANNOgesic 0.5.7 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.5.7',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ANNOgesic 0.5.7 documentation" href="index.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="ANNOgesic - Transcriptome annotation pipeline" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="installation.html" title="Installation"
             accesskey="N">next</a></li>
        <li class="right" >
          <a href="index.html" title="ANNOgesic - Transcriptome annotation pipeline"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.5.7 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo_annogesic.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Required tools or databases</a><ul>
<li><a class="reference internal" href="#tools">Tools</a></li>
<li><a class="reference internal" href="#databases">Databases</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">ANNOgesic - Transcriptome annotation pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="installation.html"
                        title="next chapter">Installation</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="required-tools-or-databases">
<span id="id1"></span><h1>Required tools or databases<a class="headerlink" href="#required-tools-or-databases" title="Permalink to this headline">¶</a></h1>
<p>Please also check our <a class="reference external" href="https://hub.docker.com/r/silasysh/annogesic/">Docker image</a>.
Via Docker image, all the required tools are installed.</p>
<div class="section" id="tools">
<h2>Tools<a class="headerlink" href="#tools" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://www.python.org/">Python</a> : version higher or equal to 3.4</p>
<p><a class="reference external" href="http://biopython.org/wiki/Main_Page">BioPython</a>: version higher or equal to 1.65</p>
<p><a class="reference external" href="http://www.bioperl.org/wiki/Main_Page">BioPerl</a>:  version higher or equal to 1.6.1</p>
<p><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">Blast+</a> : version higher or equal to 2.2.28+ (for sRNA_detection)</p>
<p><a class="reference external" href="http://www.imagemagick.org/script/index.php">ImageMagick</a> : version higher or equal to 6.9.0-0 (for generate screenshot)</p>
<p><a class="reference external" href="http://infernal.janelia.org/">Infernal</a> : version higher or equal o 1.1.1 (for riboswitch and RNA thermometer detection)</p>
<p><a class="reference external" href="http://matplotlib.org/">Matplotlib</a> : version higher or equal to 1.5.0 (for generate all statistics figure)</p>
<p><a class="reference external" href="https://networkx.github.io/">Networkx</a> : version higher or equal to 1.10 (for generate PPI figure)</p>
<p><a class="reference external" href="http://meme-suite.org/tools/meme">MEME</a> : version higher or equal to 4.11.1 (for promoter detection)</p>
<p><a class="reference external" href="http://meme-suite.org/tools/glam2">GLAM2</a> : version higher or equal to 4.11.1 (for promoter detection)</p>
<p><a class="reference external" href="http://www.mpich.org/">MPICH</a> : version higher or equal to 3.2 (for parallel version of promoter detection)</p>
<p><a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">RATT</a> : version higher or equal to 1.64 (for annotation transfer)
Please be attation, before you start to run RATT (annotation transfer), run <code class="docutils literal"><span class="pre">source</span> <span class="pre">$PAGIT_HOME/sourceme.pagit</span></code> first. it will
modify the path for execute RATT. If you run ANNOgesic through Docker container, you can skip this step.</p>
<p><a class="reference external" href="http://www.psort.org/psortb/">Psortb</a> : version higher or equal to 3.0 (for subcellular localization detection)</p>
<p><a class="reference external" href="https://github.com/samtools">Samtools</a> : version higher or equal to 1.3.1 (using htslib 1.3.1) (for SNP calling, CircRNA_detection)</p>
<p><a class="reference external" href="https://github.com/samtools">Bcftools</a> : version higher or equal to 1.3.1 (using htslib 1.3.1) (for SNP calling)</p>
<p><a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> : version higher or equal to 0.1.9 (for CircRNA_detection)
When you install Segemehl, please type &#8216;make all&#8217; instead of &#8216;make&#8217; after running configure. Otherwise, the testrealign.x won&#8217;t appear.</p>
<p><a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a> : version higher or equal to 2.09 (for rho-independent terminator prediction)</p>
<p><a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a> : version higher or equal to 1.06 (for TSS and processing site prediction, TSS optimization)</p>
<p><a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a> : version higher or equal to 2.2.5 (for rho-independent terminator prediction, sRNA detection, sRNA_target detection)</p>
<p><a class="reference external" href="http://pages.cs.wisc.edu/~ghost/doc/AFPL/6.50/Ps2pdf.htm">Ps2pdf14</a> (for sRNA detection)</p>
<p><a class="reference external" href="https://www.broadinstitute.org/software/igv/home">IGV</a> (for screenshot and color_png)</p>
<p><a class="reference external" href="http://www.room220.com/crt/">CRT</a>: version higher or equal to 1.2 (for CRISPR detection)</p>
</div>
<div class="section" id="databases">
<h2>Databases<a class="headerlink" href="#databases" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a> (for sRNA detection)</p>
<p><a class="reference external" href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/">nr database</a> (for sRNA detection)</p>
<p><a class="reference external" href="http://rfam.xfam.org/">Rfam</a> (for riboswitch and RNA thermometer detection)</p>
<p><a class="reference external" href="http://www.uniprot.org/downloads">idmapping_selected.tab from Uniprot</a> (for Go term detection)</p>
<p><a class="reference external" href="http://geneontology.org/page/go-slim-and-subset-guide">goslim.obo</a> (for Go term detection)</p>
<p><a class="reference external" href="http://geneontology.org/page/download-ontology">go.obo</a> (for Go term detection)</p>
<p><a class="reference external" href="http://string-db.org/cgi/download.pl">species.v${VERSION}.txt from STRING</a> (for protein-protein
interaction prediction; replace ${VERSION} by the version you would
like to use)</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="installation.html" title="Installation"
             >next</a></li>
        <li class="right" >
          <a href="index.html" title="ANNOgesic - Transcriptome annotation pipeline"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.5.7 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Sung-Huan Yu.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.3.
    </div>
  </body>
</html>