<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>ANNOgesic’s subcommands &mdash; ANNOgesic 0.1.3 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ANNOgesic 0.1.3 documentation" href="index.html" />
    <link rel="next" title="Performing a test cast" href="test_case.html" />
    <link rel="prev" title="Dockfiles" href="docker.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="test_case.html" title="Performing a test cast"
             accesskey="N">next</a></li>
        <li class="right" >
          <a href="docker.html" title="Dockfiles"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.1.3 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo_annogesic.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">ANNOgesic&#8217;s subcommands</a><ul>
<li><a class="reference internal" href="#the-format-of-filename">The format of filename</a></li>
<li><a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic">The input format of libraries for running ANNOgesic</a></li>
<li><a class="reference internal" href="#the-format-of-srna-database">The format of sRNA database</a></li>
<li><a class="reference internal" href="#definition-of-reference-strain-and-target-strain">Definition of reference strain and target strain</a></li>
<li><a class="reference internal" href="#riboswitch-and-rna-thermometer-dataset-of-rfam">Riboswitch and RNA thermometer dataset of Rfam</a></li>
<li><a class="reference internal" href="#create">create</a></li>
<li><a class="reference internal" href="#get-input-files">get_input_files</a></li>
<li><a class="reference internal" href="#get-target-fasta">get_target_fasta</a></li>
<li><a class="reference internal" href="#annotation-transfer">annotation_transfer</a></li>
<li><a class="reference internal" href="#snp">snp</a></li>
<li><a class="reference internal" href="#tsspredator-tss-and-processing-site-prediction">tsspredator(TSS and processing site prediction)</a></li>
<li><a class="reference internal" href="#transcript-assembly">transcript_assembly</a></li>
<li><a class="reference internal" href="#terminator">terminator</a></li>
<li><a class="reference internal" href="#utr">utr</a></li>
<li><a class="reference internal" href="#srna">srna</a></li>
<li><a class="reference internal" href="#sorf">sorf</a></li>
<li><a class="reference internal" href="#promoter">promoter</a></li>
<li><a class="reference internal" href="#operon">operon</a></li>
<li><a class="reference internal" href="#circrna">circrna</a></li>
<li><a class="reference internal" href="#go-term">go_term</a></li>
<li><a class="reference internal" href="#srna-target">srna_target</a></li>
<li><a class="reference internal" href="#ppi-network">ppi_network</a></li>
<li><a class="reference internal" href="#subcellular-localization">subcellular_localization</a></li>
<li><a class="reference internal" href="#riboswitch-thermometer">riboswitch_thermometer</a></li>
<li><a class="reference internal" href="#crispr">crispr</a></li>
<li><a class="reference internal" href="#optimize-tsspredator">optimize_tsspredator</a></li>
<li><a class="reference internal" href="#screenshot">screenshot</a></li>
<li><a class="reference internal" href="#color-png">color_png</a></li>
<li><a class="reference internal" href="#merge-features">merge_features</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="docker.html"
                        title="previous chapter">Dockfiles</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="test_case.html"
                        title="next chapter">Performing a test cast</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="annogesic-s-subcommands">
<h1>ANNOgesic&#8217;s subcommands<a class="headerlink" href="#annogesic-s-subcommands" title="Permalink to this headline">¶</a></h1>
<p>In general, the subcommands need at least one argument - the analysis
folder. If this is not given, ANNOgesic assumes that the current
folder is the analysis folder.</p>
<div class="section" id="the-format-of-filename">
<h2>The format of filename<a class="headerlink" href="#the-format-of-filename" title="Permalink to this headline">¶</a></h2>
<p>In order to recognize the file types and the relationship between different files,
please follow the principle of the format of filename.</p>
<p>Keep the filename of genome fasta file the same as annotation file. For example,
<code class="docutils literal"><span class="pre">NC_007795.fa,</span> <span class="pre">NC_007795.gff,</span> <span class="pre">NC_007795.ptt,</span> <span class="pre">NC_007795.rnt,</span> <span class="pre">NC_007705.gbk</span></code>.</p>
<p>All the <code class="docutils literal"><span class="pre">.gff</span></code> files which are input of <code class="docutils literal"><span class="pre">ANNOgesic</span></code>, please follow the format.
<code class="docutils literal"><span class="pre">$STRAINNAME_$FEATURE.gff</span></code>. For example, <code class="docutils literal"><span class="pre">NC_007795_TSS.gff,</span> <span class="pre">NC_007795_transcript.gff</span></code>.</p>
<p>The names of features are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="36%" />
<col width="64%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Feature name</td>
<td>meaning</td>
</tr>
<tr class="row-even"><td>TSS</td>
<td>transcription starting site</td>
</tr>
<tr class="row-odd"><td>processing</td>
<td>processing site</td>
</tr>
<tr class="row-even"><td>transcript</td>
<td>transcript assembly</td>
</tr>
<tr class="row-odd"><td>sRNA</td>
<td>small RNA</td>
</tr>
<tr class="row-even"><td>sORF</td>
<td>small open reading frame</td>
</tr>
<tr class="row-odd"><td>term</td>
<td>terminator</td>
</tr>
<tr class="row-even"><td>5UTR</td>
<td>5&#8217;UTR</td>
</tr>
<tr class="row-odd"><td>3UTR</td>
<td>3&#8217;UTR</td>
</tr>
<tr class="row-even"><td>circRNA</td>
<td>circular RNA</td>
</tr>
<tr class="row-odd"><td>riboswitch</td>
<td>riboswitch</td>
</tr>
<tr class="row-even"><td>RNA_thermometer</td>
<td>RNA_thermometer</td>
</tr>
<tr class="row-odd"><td>CRISPR</td>
<td>CRISPR</td>
</tr>
</tbody>
</table>
<p>Please avoid <code class="docutils literal"><span class="pre">|</span></code> in the filename, strain name of Gff3 files or fasta file.</p>
</div>
<div class="section" id="the-input-format-of-libraries-for-running-annogesic">
<h2>The input format of libraries for running ANNOgesic<a class="headerlink" href="#the-input-format-of-libraries-for-running-annogesic" title="Permalink to this headline">¶</a></h2>
<p>Some modules of <code class="docutils literal"><span class="pre">ANNOgesic</span></code> need the information of libraries.
please follow the following format:</p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_FILENAME:$LIBRARY_TYPE:$CONDITION:$REPLICATE:$STRAND</span></code></p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_FILENAME</span></code> means the <code class="docutils literal"><span class="pre">.wig</span></code> file.</p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_TYPE</span></code> can be <code class="docutils literal"><span class="pre">tex</span></code> (TEX+) or <code class="docutils literal"><span class="pre">notex</span></code> (TEX-) or <code class="docutils literal"><span class="pre">frag</span></code> (fragmented).</p>
<p><code class="docutils literal"><span class="pre">$CONDITION</span></code> is the index of conditions. Please use 1, 2, 3, ... to represent different conditions.</p>
<p><code class="docutils literal"><span class="pre">$REPLICATE</span></code> is the index of replicates. Please use a, b, c, ... to represent different replicates.</p>
<p><code class="docutils literal"><span class="pre">$STRAND</span></code> is the strand of wiggle file. Please use + or -.</p>
<p>All the libraries should be separated by comma like the above.
The libraries should be listed in one line.</p>
<p>For example,</p>
<p>for TEX +/- treated libraries:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_TEX_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span><span class="p">,</span>
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_TEX_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span><span class="p">,</span>
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_TEX_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span><span class="p">,</span>
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_TEX_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span><span class="p">,</span>
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span><span class="p">,</span>
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span><span class="p">,</span>
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span><span class="p">,</span>
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span><span class="p">,</span>
</pre></div>
</div>
<p>And for fragmented libraries:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">fragmented_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">frag</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span><span class="p">,</span><span class="n">fragmented_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">frag</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span>
</pre></div>
</div>
<p>If only conventional RNA-seq data without fragementation or TEX treated can be provided,
it can still be assigned to fragmented libraries and run ANNOgesic.
However, it may influnece on the results.</p>
</div>
<div class="section" id="the-format-of-srna-database">
<h2>The format of sRNA database<a class="headerlink" href="#the-format-of-srna-database" title="Permalink to this headline">¶</a></h2>
<p>If searching sRNA database is assigned to sRNA detection, please follow the format. Or the output will become
nonsense. The format is</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ID|$STRAIN|$SRNANAME</dd>
</dl>
<p>for example</p>
<dl class="docutils">
<dt>::</dt>
<dd>srn_4840|S._aureus_NCTC8325|RsaOV</dd>
</dl>
<p>You can also download sRNA database <a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a> from our
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/database">Github</a> easily.</p>
</div>
<div class="section" id="definition-of-reference-strain-and-target-strain">
<h2>Definition of reference strain and target strain<a class="headerlink" href="#definition-of-reference-strain-and-target-strain" title="Permalink to this headline">¶</a></h2>
<p>&#8220;target strain&#8221; represents the strain which user want to annotate.
&#8220;reference strain&#8221; represents the strain which is close to the &#8220;target strain&#8221;.
If user have no fasta file or genome annotation files of &#8220;target strain&#8221;,
ANNOgesic can generate them by applying the mutations between &#8220;reference strain&#8221; and &#8220;target strain&#8221;.</p>
</div>
<div class="section" id="riboswitch-and-rna-thermometer-dataset-of-rfam">
<h2>Riboswitch and RNA thermometer dataset of Rfam<a class="headerlink" href="#riboswitch-and-rna-thermometer-dataset-of-rfam" title="Permalink to this headline">¶</a></h2>
<p>For riboswitch and RNA thermometer detection, it needs the information of riboswitch and RNA thermometer in Rfam.
The input format is the following.</p>
<p>All columns are splited by <code class="docutils literal"><span class="pre">tab</span></code>. You can also download the data from our
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/database">Github</a>.</p>
</div>
<div class="section" id="create">
<h2>create<a class="headerlink" href="#create" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">create</span></code> generates the folders for analysis. Once these folders are created,
please put the required files into the proper folders.</p>
<p>BAMs: For <code class="docutils literal"><span class="pre">.bam</span></code> files. <code class="docutils literal"><span class="pre">BAMs_map_reference</span></code>
is for the <code class="docutils literal"><span class="pre">.bam</span></code> files which mapped on &#8220;reference strain&#8221;.
<code class="docutils literal"><span class="pre">BAMs_map_target</span></code> is for the <code class="docutils literal"><span class="pre">.bam</span></code> files which mapped on &#8220;target strain&#8221;.</p>
<p>database: For all databases</p>
<p>manual_TSS: If the manual detected transcription starting sites(TSSs) is provided,
it can be stored here for running <code class="docutils literal"><span class="pre">TSS_optimization</span></code> or merging
the automatic predicted ones and manual detected ones. Please use gff3 format.</p>
<p>manual_processing_site: It is similar with <code class="docutils literal"><span class="pre">manual_TSS</span></code>, it is for
processing sites.</p>
<p>mutation_table: If the mutations table between &#8220;reference strain&#8221; and
&#8220;target strain&#8221; is provided, please put the file here. Please refer
to the section of <code class="docutils literal"><span class="pre">get_target_fasta</span></code> for the format of
mutation table.</p>
<p>reads: For running <code class="docutils literal"><span class="pre">circrna</span></code> with mapping reads by ANNOgesic,
please put the reads here. It can also deal with <code class="docutils literal"><span class="pre">.bzip2</span></code> and <code class="docutils literal"><span class="pre">.gzip</span></code>.</p>
<p>reference: For annotation files and fasta files of &#8220;reference strain&#8221;.
If they can be downloaded from NCBI, the files can also be gain via running <code class="docutils literal"><span class="pre">get_input_files</span></code>.</p>
<p>riboswitch_ID: For storing the file which contains all the Rfam ID of riboswitch.
For the details of format, please refer to the section of
<code class="docutils literal"><span class="pre">Riboswitch</span> <span class="pre">and</span> <span class="pre">RNA</span> <span class="pre">thermometer</span> <span class="pre">dataset</span> <span class="pre">of</span> <span class="pre">Rfam</span></code>.</p>
<p>RNA_thermometer_ID: For storing the file which contains all the Rfam ID of RNA thermometer.
For the details of format, please refer to the section of
<code class="docutils literal"><span class="pre">Riboswitch</span> <span class="pre">and</span> <span class="pre">RNA</span> <span class="pre">thermometer</span> <span class="pre">dataset</span> <span class="pre">of</span> <span class="pre">Rfam</span></code>.</p>
<p>wigs: For wiggle files. Based on the methods of RNA-Seq, wiggle files can be stored in
<code class="docutils literal"><span class="pre">fragment</span></code> (fragmented libraries) or <code class="docutils literal"><span class="pre">tex_notex</span></code> (TEX +/- treated libraries).</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">ANNOgesic</span><span class="o">.</span><span class="n">py</span> <span class="n">create</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="n">project_path</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>  <span class="n">Name</span><span class="o">/</span><span class="n">path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>    <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
</pre></div>
</div>
</div>
<div class="section" id="get-input-files">
<h2>get_input_files<a class="headerlink" href="#get-input-files" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">get_input_files</span></code> is the subcommand for downloading required files (fasta, annotation files) from NCBI.
Therefore, user needs to assign the IP of the reference genome in NCBI. For example,
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000013425.1_ASM1342v1">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000013425.1_ASM1342v1</a>
Then, user can assign the file type for download.</p>
<ul class="simple">
<li>Pre-reqired information</li>
</ul>
<p><code class="docutils literal"><span class="pre">FTP</span> <span class="pre">source</span></code>: The IP of NCBI.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">get_input_files</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">FTP_path</span> <span class="n">FTP_PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_fasta</span><span class="p">]</span>
                                 <span class="p">[</span><span class="o">--</span><span class="n">ref_gff</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_ptt</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_rnt</span><span class="p">]</span>
                                 <span class="p">[</span><span class="o">--</span><span class="n">ref_gbk</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">convert_embl</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">for_target</span><span class="p">]</span>
                                 <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">FTP_path</span> <span class="n">FTP_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">F</span> <span class="n">FTP_PATH</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">NCBI</span> <span class="n">FTP</span> <span class="n">which</span> <span class="n">can</span> <span class="n">download</span> <span class="n">the</span> <span class="n">required</span>
                        <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_fasta</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span>       <span class="n">Download</span> <span class="n">fasta</span> <span class="n">files</span> <span class="n">of</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_gff</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span>         <span class="n">Download</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">of</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_ptt</span><span class="p">,</span> <span class="o">-</span><span class="n">p</span>         <span class="n">Download</span> <span class="n">ptt</span> <span class="n">files</span> <span class="n">of</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_rnt</span><span class="p">,</span> <span class="o">-</span><span class="n">r</span>         <span class="n">Download</span> <span class="n">rnt</span> <span class="n">files</span> <span class="n">of</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_gbk</span><span class="p">,</span> <span class="o">-</span><span class="n">k</span>         <span class="n">Download</span> <span class="n">genbank</span> <span class="n">files</span> <span class="n">of</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">convert_embl</span><span class="p">,</span> <span class="o">-</span><span class="n">e</span>    <span class="n">Convert</span> <span class="n">gbk</span> <span class="n">to</span> <span class="n">embl</span> <span class="n">files</span> <span class="n">of</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">for_target</span><span class="p">,</span> <span class="o">-</span><span class="n">t</span>      <span class="n">If</span> <span class="n">the</span> <span class="n">required</span> <span class="n">files</span> <span class="n">of</span> <span class="n">query</span> <span class="n">strain</span> <span class="n">can</span> <span class="n">be</span>
                        <span class="n">downloaded</span> <span class="kn">from</span> <span class="nn">NCBI</span> <span class="p">(</span><span class="n">you</span> <span class="n">won</span><span class="s1">&#39;t modify the genome),</span>
                        <span class="n">The</span> <span class="n">files</span> <span class="n">can</span> <span class="n">store</span> <span class="ow">in</span> <span class="n">target</span> <span class="n">folder</span> <span class="ow">in</span> <span class="n">stead</span> <span class="n">of</span>
                        <span class="n">reference</span> <span class="n">folder</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/input/reference</span></code> if <code class="docutils literal"><span class="pre">--for_target</span></code> is False.
The output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/target</span></code> if <code class="docutils literal"><span class="pre">--for_target</span></code> is True.</p>
<p><code class="docutils literal"><span class="pre">fasta</span></code>: Fasta files.</p>
<p><code class="docutils literal"><span class="pre">annotation</span></code>: Annotation files.</p>
</div>
<div class="section" id="get-target-fasta">
<h2>get_target_fasta<a class="headerlink" href="#get-target-fasta" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">get_target_fasta</span></code> is the subcommand for generating fasta files of &#8220;target strain&#8221; from
&#8220;reference strain&#8221;. The format of mutation table is following:</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="9%" />
<col width="12%" />
<col width="8%" />
<col width="9%" />
<col width="20%" />
<col width="13%" />
<col width="4%" />
<col width="12%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>#reference_id</td>
<td>target_id</td>
<td>reference_nt</td>
<td>position</td>
<td>target_nt</td>
<td>impact_of_correction</td>
<td>locus_tag</td>
<td>gene</td>
<td>Description</td>
</tr>
<tr class="row-even"><td>NC_007795.1</td>
<td>HG003</td>
<td>a</td>
<td>333</td>
<td>c</td>
<td>&nbsp;</td>
<td>SAOUHSC_00002</td>
<td>dnaA</td>
<td>XXXXXX</td>
</tr>
<tr class="row-odd"><td>NC_007795.1</td>
<td>HG003</td>
<td>t</td>
<td>543</td>
<td>-</td>
<td>deletion</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>YYYYYY</td>
</tr>
<tr class="row-even"><td>NC_007795.1</td>
<td>HG003</td>
<td>-</td>
<td>600</td>
<td>g</td>
<td>insertion</td>
<td>SAOUHSC_00132</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
<p>If the titles of columns is presented on the top, they need to start from <code class="docutils literal"><span class="pre">#</span></code>.
Each column is separated by <code class="docutils literal"><span class="pre">tab</span></code>. If the mutation type is deletion or insertion,
user can put <code class="docutils literal"><span class="pre">-</span></code> to represent them. The information of <code class="docutils literal"><span class="pre">target_id</span></code>, <code class="docutils literal"><span class="pre">reference_id</span></code>,
<code class="docutils literal"><span class="pre">reference_nt</span></code>, <code class="docutils literal"><span class="pre">position</span></code>, <code class="docutils literal"><span class="pre">target_nt</span></code> is required. The others can be blank.
However, please still use <code class="docutils literal"><span class="pre">tab</span></code> to separate all blank columns.</p>
<p>If no mutation information is provided, <code class="docutils literal"><span class="pre">SNP_calling</span></code> can be used for detecting mutations.
(one module of <code class="docutils literal"><span class="pre">ANNOgesic</span></code>). Please refer to the section of <code class="docutils literal"><span class="pre">SNP_calling</span></code>.</p>
<ul class="simple">
<li>Pre-required files</li>
</ul>
<p>Fasta files of reference genome.</p>
<p>Mutation table: it indicates the information of mutations between reference and target strain.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">get_target_fasta</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_fasta_folder</span> <span class="n">REF_FASTA_FOLDER</span><span class="p">]</span>
                                  <span class="p">[</span><span class="o">--</span><span class="n">mutation_table</span> <span class="n">MUTATION_TABLE</span><span class="p">]</span>
                                  <span class="p">[</span><span class="o">--</span><span class="n">output_format</span> <span class="n">OUTPUT_FORMAT</span><span class="p">]</span>
                                  <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">ref_fasta_folder</span> <span class="n">REF_FASTA_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">r</span> <span class="n">REF_FASTA_FOLDER</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">mutation_table</span> <span class="n">MUTATION_TABLE</span><span class="p">,</span> <span class="o">-</span><span class="n">m</span> <span class="n">MUTATION_TABLE</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">mutation</span> <span class="n">table</span><span class="o">.</span>
  <span class="o">--</span><span class="n">output_format</span> <span class="n">OUTPUT_FORMAT</span><span class="p">,</span> <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FORMAT</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">output</span> <span class="n">filename</span> <span class="ow">and</span> <span class="n">the</span> <span class="n">strain</span> <span class="n">which</span>
                        <span class="n">should</span> <span class="n">be</span> <span class="n">included</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">file</span><span class="o">.</span> <span class="n">For</span> <span class="n">example</span><span class="p">:</span>
                        <span class="n">FILE1</span><span class="p">:</span><span class="n">strain1_and_strain2</span><span class="p">,</span><span class="n">FILE2</span><span class="p">:</span><span class="n">strain3</span><span class="o">.</span> <span class="n">FILE1</span> <span class="ow">is</span> <span class="n">a</span>
                        <span class="n">output</span> <span class="n">fasta</span> <span class="n">file</span> <span class="n">which</span> <span class="n">include</span> <span class="n">the</span> <span class="n">information</span> <span class="n">of</span>
                        <span class="n">strain1</span> <span class="ow">and</span> <span class="n">strain2</span> <span class="p">(</span><span class="kn">import</span> <span class="nn">multi</span><span class="o">-</span><span class="n">strains</span> <span class="n">to</span> <span class="n">one</span> <span class="n">file</span>
                        <span class="n">should</span> <span class="n">be</span> <span class="n">separated</span> <span class="n">by</span> <span class="s2">&quot;_and_&quot;</span><span class="o">.</span><span class="p">)</span> <span class="n">And</span> <span class="n">FILE2</span> <span class="ow">is</span> <span class="k">for</span>
                        <span class="n">strain3</span><span class="o">.</span> <span class="n">The</span> <span class="n">files</span> <span class="n">are</span> <span class="n">splitted</span> <span class="n">by</span> <span class="n">comma</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>Fasta files of target genome will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/target/fasta</span></code>.</p>
</div>
<div class="section" id="annotation-transfer">
<h2>annotation_transfer<a class="headerlink" href="#annotation-transfer" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">annotation</span> <span class="pre">transfer</span></code> is the subcommand for transfering the annotation from &#8220;reference strain&#8221;
to target &#8220;strain&#8221;. In this subcommand, <a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">RATT</a>
is integrated to achieve it. The similarity of &#8220;reference strain&#8221; and &#8220;target strain&#8221; should be closed enough or
it will influence the final results.
Be attation, before running RATT (annotation transfer),
please run <code class="docutils literal"><span class="pre">source</span> <span class="pre">$PAGIT_HOME/sourceme.pagit</span></code> first. it will modify the path for execute RATT.
If you use Dockerfile to execute ANNOgesic, the path modification can be skipped.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">RATT</a>.</p>
<p>The genbank files of reference genome.</p>
<p>The fasta files of reference genome.</p>
<p>The fasta files of target genome.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">annotation_transfer</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">RATT_path</span> <span class="n">RATT_PATH</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">compare_pair</span> <span class="n">COMPARE_PAIR</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">element</span> <span class="n">ELEMENT</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">transfer_type</span> <span class="n">TRANSFER_TYPE</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">ref_embl</span> <span class="n">REF_EMBL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_gbk</span> <span class="n">REF_GBK</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">ref_fasta</span> <span class="n">REF_FASTA</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">target_fasta</span> <span class="n">TARGET_FASTA</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">convert_to_gff_rnt_ptt</span><span class="p">]</span>
                                     <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">RATT_path</span> <span class="n">RATT_PATH</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">start</span><span class="o">.</span><span class="n">ratt</span><span class="o">.</span><span class="n">sh</span> <span class="n">file</span> <span class="n">of</span> <span class="n">RATT</span> <span class="n">folder</span><span class="o">.</span> <span class="n">Default</span>
                        <span class="ow">is</span> <span class="n">start</span><span class="o">.</span><span class="n">ratt</span><span class="o">.</span><span class="n">sh</span><span class="o">.</span>
  <span class="o">--</span><span class="n">compare_pair</span> <span class="n">COMPARE_PAIR</span><span class="p">,</span> <span class="o">-</span><span class="n">p</span> <span class="n">COMPARE_PAIR</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">name</span> <span class="n">of</span> <span class="n">strain</span> <span class="n">pairs</span><span class="o">.</span> <span class="n">ex</span><span class="o">.</span>
                        <span class="n">NC_007795</span><span class="p">:</span><span class="n">NEW_NC_007795</span><span class="o">.</span> <span class="n">The</span> <span class="n">reference</span> <span class="n">strain</span> <span class="ow">is</span>
                        <span class="n">NC_007795</span> <span class="ow">and</span> <span class="n">the</span> <span class="n">target</span> <span class="n">strain</span> <span class="ow">is</span> <span class="n">NEW_NC_007795</span><span class="o">.</span> <span class="n">the</span>
                        <span class="n">assigned</span> <span class="n">names</span> <span class="ow">is</span> <span class="n">the</span> <span class="n">strain</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">fasta</span> <span class="n">file</span><span class="p">,</span>
                        <span class="ow">not</span> <span class="n">the</span> <span class="n">filenames</span> <span class="n">of</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span> <span class="n">For</span> <span class="n">multiple</span>
                        <span class="n">strains</span><span class="p">,</span> <span class="n">please</span> <span class="n">use</span> <span class="n">comma</span> <span class="n">to</span> <span class="n">separate</span> <span class="n">the</span> <span class="n">strains</span><span class="o">.</span>
  <span class="o">--</span><span class="n">element</span> <span class="n">ELEMENT</span><span class="p">,</span> <span class="o">-</span><span class="n">e</span> <span class="n">ELEMENT</span>
                        <span class="n">It</span> <span class="n">will</span> <span class="n">be</span> <span class="n">assigned</span> <span class="n">to</span> <span class="n">the</span> <span class="n">prefix</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">output</span> <span class="n">file</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transfer_type</span> <span class="n">TRANSFER_TYPE</span><span class="p">,</span> <span class="o">-</span><span class="n">t</span> <span class="n">TRANSFER_TYPE</span>
                        <span class="n">The</span> <span class="n">transfer</span> <span class="nb">type</span> <span class="k">for</span> <span class="n">running</span> <span class="n">RATT</span><span class="o">.</span> <span class="p">(</span><span class="n">For</span> <span class="n">the</span> <span class="n">details</span><span class="p">,</span>
                        <span class="n">please</span> <span class="n">refer</span> <span class="n">to</span> <span class="n">the</span> <span class="n">manual</span> <span class="n">of</span> <span class="n">RATT</span><span class="o">.</span><span class="p">)</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="n">Strain</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_embl</span> <span class="n">REF_EMBL</span><span class="p">,</span> <span class="o">-</span><span class="n">re</span> <span class="n">REF_EMBL</span>
                        <span class="n">The</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">embl</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_gbk</span> <span class="n">REF_GBK</span><span class="p">,</span> <span class="o">-</span><span class="n">rg</span> <span class="n">REF_GBK</span>
                        <span class="n">If</span> <span class="n">you</span> <span class="n">have</span> <span class="n">no</span> <span class="n">embl</span> <span class="n">file</span><span class="p">,</span> <span class="n">you</span> <span class="n">can</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">folder</span> <span class="n">of</span>
                        <span class="n">genbank</span> <span class="n">files</span><span class="o">.</span> <span class="n">The</span> <span class="n">genbank</span> <span class="n">can</span> <span class="n">be</span> <span class="n">ended</span> <span class="n">by</span> <span class="o">.</span><span class="n">gbk</span><span class="p">,</span> <span class="o">.</span><span class="n">gbff</span>
                        <span class="ow">or</span> <span class="o">.</span><span class="n">gb</span>
  <span class="o">--</span><span class="n">ref_fasta</span> <span class="n">REF_FASTA</span><span class="p">,</span> <span class="o">-</span><span class="n">rf</span> <span class="n">REF_FASTA</span>
                        <span class="n">The</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">reference</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">target_fasta</span> <span class="n">TARGET_FASTA</span><span class="p">,</span> <span class="o">-</span><span class="n">tf</span> <span class="n">TARGET_FASTA</span>
                        <span class="n">The</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">target</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">convert_to_gff_rnt_ptt</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span>
                        <span class="n">Convert</span> <span class="n">the</span> <span class="n">annotation</span> <span class="n">to</span> <span class="n">gff</span><span class="p">,</span> <span class="n">rnt</span> <span class="ow">and</span> <span class="n">ptt</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All the output files from <a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">RATT</a>
will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/annotation_transfer</span></code>.</p>
<p>All annotation files (<code class="docutils literal"><span class="pre">.gff</span></code>, <code class="docutils literal"><span class="pre">.ptt</span></code>, <code class="docutils literal"><span class="pre">.rnt</span></code>) will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/target/annotation</span></code>.</p>
</div>
<div class="section" id="snp">
<h2>snp<a class="headerlink" href="#snp" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">snp</span></code> can compare the alignment files and fasta files to detect the mutations by
<a class="reference external" href="https://github.com/samtools">Samtools</a>, <a class="reference external" href="https://github.com/samtools">Bcftools</a>.
There are multiple programs which can be applied (with BAQ, without BAQ and extend BAQ) and set the filters
(QUAL, DP, DP4, etc.) to run <code class="docutils literal"><span class="pre">snp</span></code>. Moreover,
It can also be used for generating the fasta file of &#8220;target strain&#8221;.</p>
<ul class="simple">
<li>Pre-required files and tools:</li>
</ul>
<p><a class="reference external" href="https://github.com/samtools">Samtools</a>.</p>
<p><a class="reference external" href="https://github.com/samtools">Bcftools</a>.</p>
<p>BAM files for fragmented libraries or TEX +/- treated libraries.</p>
<p>Reference or target genome fasta files.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic snp [-h] [--samtools_path SAMTOOLS_PATH]
                     [--bcftools_path BCFTOOLS_PATH] [--bam_type BAM_TYPE]
                     [--program PROGRAM] [--fasta_path FASTA_PATH]
                     [--tex_bam_path TEX_BAM_PATH]
                     [--frag_bam_path FRAG_BAM_PATH] [--quality QUALITY]
                     [--read_depth_range READ_DEPTH_RANGE] [--ploidy PLOIDY]
                     [--RG_tag] [--min_sample_number MIN_SAMPLE_NUMBER]
                     [--caller CALLER] [--DP4_cutoff DP4_CUTOFF]
                     [--indel_fraction INDEL_FRACTION]
                     [--filter_tag_info FILTER_TAG_INFO]
                     [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --samtools_path SAMTOOLS_PATH
                        If you want to assign the path of samtools, please
                        assign here.
  --bcftools_path BCFTOOLS_PATH
                        If you want to assign the path of bcftools, please
                        assign here.
  --bam_type BAM_TYPE, -t BAM_TYPE
                        Please assign the type of BAM. If the BAM files are
                        produced by mapping to reference strain, please keyin
                        &quot;reference&quot;. It is for detecting the mutations between
                        refenece strain and target strain, and generate
                        potential target fasta file. If the BAM files are
                        produced by mapping to target strain, please keyin
                        &#39;target&#39;. It is for detecting the mutations of target
                        genome sequence.
  --program PROGRAM, -p PROGRAM
                        Please assign the program for detecting SNP of
                        transcript: 1: calculate with BAQ, 2: calculate
                        without BAQ, 3: calculate with extend BAQ. Multi-
                        programs can be executed at the same time. For
                        example: 1,2,3. Default is 1,2,3.
  --fasta_path FASTA_PATH, -f FASTA_PATH
                        The path of genome fasta folder.
  --tex_bam_path TEX_BAM_PATH, -tw TEX_BAM_PATH
                        The path of tex+/- wig folder.
  --frag_bam_path FRAG_BAM_PATH, -fw FRAG_BAM_PATH
                        The path of fragmented wig folder.
  --quality QUALITY, -q QUALITY
                        The minimum quality which considers a real snp.
                        Default is 40.
  --read_depth_range READ_DEPTH_RANGE, -d READ_DEPTH_RANGE
                        The range of read depth. If the read depth is higher
                        or lower than this range, it will be excluded. The
                        format is $MIN,$MAX. --read_depth_range can be
                        assigned by different types: 1. real number (r), 2.
                        the read depth based on the number of samples (n) or
                        3. times of average read depth (a). For example,
                        n_10,a_2 means the range of read depth should be
                        higher than average 10 reads of --min_sample_number
                        (if --min_sample_number is 2, DP value in output will
                        be higher than 20) and lower than 2 times of average
                        read depth.If r_10,a_2 is assigned, it means that the
                        minimum read depth becomes 10 without considering the
                        number of samples. Default is n_10,a_2.
  --ploidy PLOIDY, -pl PLOIDY
                        haploid or diploid. Default is haploid.
  --RG_tag, -R          It is opposite of --ignore-RG in samtools. (one BAM
                        file includes multi samples). Default is False.
  --min_sample_number MIN_SAMPLE_NUMBER, -ms MIN_SAMPLE_NUMBER
                        The minimum numbers of samples will effect
                        --read_depth_range, --DP4_cutoff and --indel_fraction.
  --caller CALLER, -c CALLER
                        The types of caller - consensus-caller or
                        multiallelic-caller. For details, please check
                        bcftools. &quot;c&quot; represents consensus-caller. &quot;m&quot;
                        represents multiallelic-caller. Default is m.
  --DP4_cutoff DP4_CUTOFF, -D DP4_CUTOFF
                        The cutoff of DP4. DP4 is compose of four numbers:
                        high-quality reference forward bases (number 1),
                        reference reverse bases (number 2), alternate forward
                        bases (number 3) and alternative reverse bases (number
                        4). Two cutoff values can be assigned, ex: n_10,0.8.
                        First cutoff is for (number 3 + number 4). It can be
                        assigned based on 1. real number (r), 2. the read
                        depth based on the number of samples (n) or 3. the
                        times of average read depth (a). The second cutoff is
                        for (number 3 + number 4) / (number 1 + number 2 +
                        number 3 + number 4). These two cutoff is splited by
                        comma. For example, n_10,0.8 means the sum of read
                        depth of number 3 and number 4 should be higher than
                        average 10 reads of --min_sample_number (if
                        --min_sample_number is 2, DP value in output will be
                        higher than 20). And the fraction should be higher
                        than 0.8. If r_10,0.8 is assigned, it means that the
                        sum of read depth of number 3 and number 4 become 10
                        without considering the number of samples. Default is
                        n_10,0.8.
  --indel_fraction INDEL_FRACTION, -if INDEL_FRACTION
                        The fraction of maximum read depth (IMF) and read
                        number of each sample (IDV), which supports insertion
                        of deletion. There are different types can be
                        assigned: 1. real number (r), 2. the read depth based
                        on the number of samples (n) or 3. the times of
                        average read depth (a) for IDV. For example, n_10,0.8
                        means the IDV should be higher than average 10 reads
                        of --min_sample_number (if --min_sample_number, DP
                        value in output will be higher than 20). And IMF
                        should be higher than 0.8. If r_10,0.8 is assigned, it
                        means that IDV become 10 without considering the
                        number of samples. Default is n_10,0.8.
  --filter_tag_info FILTER_TAG_INFO, -ft FILTER_TAG_INFO
                        Please assign 1. the tag, 2. bigger or samller and 3.
                        value for filters. For example, &quot;RPB_b0.1,MQ0F_s0&quot;
                        means RPB should bigger than 0.1 and MQ0F should
                        smaller than 0. Default is
                        RPB_b0.1,MQSB_b0.1,MQB_b0.1,BQB_b0.1.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>If <code class="docutils literal"><span class="pre">bam_type</span></code> is <code class="docutils literal"><span class="pre">reference</span></code>,
the results will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/compare_reference</span></code>.
If it is <code class="docutils literal"><span class="pre">target</span></code>, the results will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/validate_target</span></code>.</p>
<p>The raw data from <a class="reference external" href="https://github.com/samtools">Samtools</a> and <a class="reference external" href="https://github.com/samtools">Bcftools</a>
will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/SNP_raw_outputs</span></code>.</p>
<p>The results will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/SNP_table</span></code>.</p>
<p>The meaning of filenames are:</p>
<p><code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_best.vcf</span></code> which is in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/SNP_table/$STRAIN</span></code>.
It means the results after filtering by cutoff.</p>
<p><code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM.vcf</span></code> which is in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/SNP_raw_output/$STRAIN</span></code>.
It means the results match the condition of read depth and quality.</p>
<p><code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_seq_reference.csv</span></code> is the index of fasta files which generated by <code class="docutils literal"><span class="pre">snp</span></code>.</p>
<p>For example,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Staphylococcus_aureus_HG003</span>     <span class="mi">1632629</span> <span class="o">.</span>       <span class="n">AaA</span>     <span class="n">AA</span>      <span class="mi">57</span>      <span class="o">.</span>
<span class="n">Staphylococcus_aureus_HG003</span>     <span class="mi">1632630</span> <span class="o">.</span>       <span class="n">aA</span>      <span class="n">a</span>       <span class="mi">57</span>      <span class="o">.</span>
<span class="n">Staphylococcus_aureus_HG003</span>     <span class="mi">1499572</span> <span class="o">.</span>       <span class="n">T</span>       <span class="n">TT</span><span class="p">,</span><span class="n">TTTTT</span>        <span class="mf">43.8525</span> <span class="o">.</span>
</pre></div>
</div>
<p>These mutations will cause conflict. Then, the conflict will effect the positions of other mutations.
Therefore, it will generate four different fasta files. The fisrt two lines are &#8220;position conflict&#8221;, and
the last line is &#8220;mutation conflict&#8221;.
<code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_seq_reference.csv</span></code> is the index for these four fasta files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="mi">1</span>       <span class="mi">1632629</span> <span class="mi">1</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TT</span>      <span class="n">Staphylococcus_aureus_HG003</span>
<span class="mi">1</span>       <span class="mi">1632629</span> <span class="mi">2</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TTTTT</span>   <span class="n">Staphylococcus_aureus_HG003</span>
<span class="mi">2</span>       <span class="mi">1632630</span> <span class="mi">1</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TT</span>      <span class="n">Staphylococcus_aureus_HG003</span>
<span class="mi">2</span>       <span class="mi">1632630</span> <span class="mi">2</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TTTTT</span>   <span class="n">Staphylococcus_aureus_HG003</span>
</pre></div>
</div>
<p>The first column is the index of position conflict. The second column is the position which be selected.
The third one is the index of mutations conflict. The fourth one is
the position and nucleotides of selected mutation. The last column is the name of strain.
If you refer to <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/seqs</span></code>, the filename of fasta is like
<code class="docutils literal"><span class="pre">$FILENAME_$STRIANNAME_$INDEXofPOSITIONCONNFLICT_$INDEXofMUTATIONCONFLICT.fa</span></code>. Therefore, the first line of
<code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_seq_reference.csv</span></code> will generate
<code class="docutils literal"><span class="pre">Staphylococcus_aureus_HG003_Staphylococcus_aureus_HG003_1_1.fa</span></code>
(if the file name of genome is Staphylococcus_aureus_HG003). The second line will generate
<code class="docutils literal"><span class="pre">Staphylococcus_aureus_HG003_Staphylococcus_aureus_HG003_1_2.fa</span></code> and so forth.</p>
<p>The statistics files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/statistics</span></code>.</p>
</div>
<div class="section" id="tsspredator-tss-and-processing-site-prediction">
<h2>tsspredator(TSS and processing site prediction)<a class="headerlink" href="#tsspredator-tss-and-processing-site-prediction" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">tsspredator</span></code> can generate the candidates of TSSs and processing sites via
<a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>. The parameters
of <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a> can be assigned. For
optimization of parameters of <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>,
please refer to the section of <code class="docutils literal"><span class="pre">optimize_tsspredator</span></code>.</p>
<p>For the information of libraries, please refer to the section
<code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p>The libraries and wiggle files of TEX +/-. Please refer to <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>Fasta file of genome sequence.</p>
<p>Gff file of genome annotation.</p>
<p>If the gff file of manual detected TSSs is provided, <code class="docutils literal"><span class="pre">tsspredator</span></code> can merge the manual one
and predicted one.</p>
<p>If comparing TSSs with transcripts is needed, the gff files of transcripts were required.
For the transcripts, please refer to the section of <code class="docutils literal"><span class="pre">transcript_assembly</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic tsspredator [-h] [--TSSpredator_path TSSPREDATOR_PATH]
                             [--fasta_folder FASTA_FOLDER]
                             [--annotation_folder ANNOTATION_FOLDER]
                             [--wig_folder WIG_FOLDER] [--height HEIGHT]
                             [--height_reduction HEIGHT_REDUCTION]
                             [--factor FACTOR]
                             [--factor_reduction FACTOR_REDUCTION]
                             [--enrichment_factor ENRICHMENT_FACTOR]
                             [--processing_factor PROCESSING_FACTOR]
                             [--base_height BASE_HEIGHT]
                             [--replicate_match REPLICATE_MATCH]
                             [--utr_length UTR_LENGTH] [--lib LIB]
                             [--output_prefix OUTPUT_PREFIX]
                             [--merge_manual MERGE_MANUAL] [--statistics]
                             [--validate_gene]
                             [--compute_program COMPUTE_PROGRAM]
                             [--compare_transcript_assembly COMPARE_TRANSCRIPT_ASSEMBLY]
                             [--fuzzy FUZZY] [--cluster CLUSTER]
                             [--length LENGTH] [--re_check_orphan]
                             [--overlap_feature OVERLAP_FEATURE]
                             [--reference_gff_folder REFERENCE_GFF_FOLDER]
                             [--remove_low_expression REMOVE_LOW_EXPRESSION]
                             [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TSSpredator_path TSSPREDATOR_PATH
                        If you want to assign the path of TSSpredator, please
                        assign here. Default is /usr/local/bin/TSSpredator.jar
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        Path of the target genome fasta folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        Path of the target genome gff folder.
  --wig_folder WIG_FOLDER, -w WIG_FOLDER
                        The folder of the TEX+/- wig folder.
  --height HEIGHT, -he HEIGHT
                        This value relates to the minimal number of read
                        starts at a certain genomic position to be considered
                        as a TSS candidate. Default is 0.3.
  --height_reduction HEIGHT_REDUCTION, -rh HEIGHT_REDUCTION
                        When comparing different strains/conditions and the
                        step height threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step height threshold.
                        Default is 0.2.
  --factor FACTOR, -fa FACTOR
                        This is the minimal factor by which the TSS height has
                        to exceed the local expression background. Default is
                        2.0.
  --factor_reduction FACTOR_REDUCTION, -rf FACTOR_REDUCTION
                        When comparing different strains/conditions and the
                        step factor threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step factor threshold.
                        Default is 0.5.
  --enrichment_factor ENRICHMENT_FACTOR, -ef ENRICHMENT_FACTOR
                        This is the minimal enrichment factor. Default is 2.0.
  --processing_factor PROCESSING_FACTOR, -pf PROCESSING_FACTOR
                        This is the minimal processing factor. If untreated
                        library is higher than the treated library and above
                        which the TSS candidate is considered as a processing
                        site and not annotated as detected. Default is 1.5.
  --base_height BASE_HEIGHT, -bh BASE_HEIGHT
                        This is the minimal number of reads should be mapped
                        on TSS. Default is 0.0.
  --replicate_match REPLICATE_MATCH, -rm REPLICATE_MATCH
                        The TSS candidates should be detected at least in the
                        number of the replicates. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATED. For assigning
                        different --replicate_match to different conditions,
                        please use comma to separate them. For example,
                        1_2,2_2,3_3 means number 1 and 2 condition assign 2 to
                        --replicate_match, and number 3 condition assign 3 to
                        --replcate_match. For assigning the same
                        --replicate_match to all conditions, just assign like
                        all_1 (all condition use 1 --replicate_match). Default
                        is all_1.
  --utr_length UTR_LENGTH, -u UTR_LENGTH
                        The length of UTR. It is for Primary and Secondary
                        definition. Default is 300.
  --lib LIB, -l LIB     The libraries of TEX+/- wig files for TSSpredator. The
                        format is: wig_file_name:tex_treat_or_not(tex or notex
                        ):condition_id(integer):replicate_id(alphabet):strand(
                        + or -). For multiple wig files, please use comma to
                        separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --output_prefix OUTPUT_PREFIX, -p OUTPUT_PREFIX
                        The output prefix of all conditions. For multiple
                        conditions, please use comma to separate them. For
                        example, prefix_condition1,prefix_condition2.
  --merge_manual MERGE_MANUAL, -m MERGE_MANUAL
                        If the gff file of manual checked TSS is provided, it
                        will merge manual checked ones and predicted ones.
                        please assign the path of gff file of manual checked
                        TSS.
  --statistics, -s      Doing statistics for TSS candidates. it will store in
                        statistics folder. Default is False.
  --validate_gene, -v   Using TSS candidates to validate genes in annotation
                        file. it will store in statistics folder. Default is
                        False.
  --compute_program COMPUTE_PROGRAM, -t COMPUTE_PROGRAM
                        The program for prediction (TSS or processing_site).
                        Default is TSS.
  --compare_transcript_assembly COMPARE_TRANSCRIPT_ASSEMBLY, -ta COMPARE_TRANSCRIPT_ASSEMBLY
                        For comparing with transcriptome assembly, please
                        assign the folder of gff file of transcript assembly.
                        Default is False.
  --fuzzy FUZZY, -fu FUZZY
                        The fuzzy for comparing TSS and transcript assembly.
                        Default is 5.
  --cluster CLUSTER, -c CLUSTER
                        This number is for comparing manual detected TSS and
                        prediced one. If the position between manual checked
                        one and predicted one is smaller or equal than this
                        value, It will only print one of them. Default is 2.
  --length LENGTH, -le LENGTH
                        The length of genome for comparing between predicted
                        one and manual checked one for statistics. For
                        comparing whole genome, please don&#39;t use this function
                        (Default). The default is comparing whole genome.
  --re_check_orphan, -ro
                        If the annotation file lacks information of gene or
                        locus_tag, all TSSs will be assigned to orphan TSSs.
                        The function can compare with CDS to classify the TSS.
                        Default is False.
  --overlap_feature OVERLAP_FEATURE, -of OVERLAP_FEATURE
                        If processing site and TSS are overlap, you can keep
                        &quot;TSS&quot; or &quot;processing_site&quot; or &quot;both&quot;. Default is both.
  --reference_gff_folder REFERENCE_GFF_FOLDER, -rg REFERENCE_GFF_FOLDER
                        If --overlap_feature is &quot;TSS&quot; or &quot;processing_site&quot;,
                        --reference_gff_folder need to be assigned. For
                        running TSS, please assign the folder of processing
                        site. For running processing_site, please assign the
                        folder of TSS. If --overlap_feature is &quot;both&quot;, please
                        don&#39;t use this function (Default). Default is None
                        (for keep both).
  --remove_low_expression REMOVE_LOW_EXPRESSION, -rl REMOVE_LOW_EXPRESSION
                        If removing low expressed TSS/processing site is
                        needed, please assign the file of manual checked gff
                        file here. It will remove the low expressed ones based
                        on comparison of manual checked ones. Please Be
                        ATTENTION: this parameter may remove some True
                        positive, too. So, please make sure you want to do it.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/TSS</span></code>.</p>
<p><code class="docutils literal"><span class="pre">MasterTables</span></code>: The MasterTable from <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p>The output files of processing sites are similar. Just replace <code class="docutils literal"><span class="pre">TSS</span></code> to <code class="docutils literal"><span class="pre">processing_site</span></code>
like <code class="docutils literal"><span class="pre">$ANNOgesic/output/processing_site</span></code>.</p>
<p><code class="docutils literal"><span class="pre">configs</span></code>: The configuration files for running TSSpredator.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files of TSSs.</p>
<p>There are some useful tags in the attributes of gff files:</p>
<p><code class="docutils literal"><span class="pre">method</span></code>: The TSSs are from manual detection or <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p><code class="docutils literal"><span class="pre">type</span></code>: The type of TSSs. It could be Primary, Secondary, Internal, Antisense or Orphan.</p>
<p><code class="docutils literal"><span class="pre">utr_length</span></code>: The length of UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_gene</span></code>: Which genes are associated with this TSS.</p>
<p><code class="docutils literal"><span class="pre">Parent</span></code>: Which transcript are associated with this TSS, if user has compared with transcript.</p>
<p>If the comparison between TSSs and genome annotation files is done, the tag - <code class="docutils literal"><span class="pre">start_TSS</span></code> will appear in the gff files
of genome annotation. It represents the TSSs which associates with the CDS/tRNA/rRNA.</p>
<p>If the comparison between TSSs and transcripts is done, the tag - <code class="docutils literal"><span class="pre">associated_tss</span></code> will appear in the gff files
of transcript. It will show the associated TSSs which is in the transcript.</p>
</div>
<div class="section" id="transcript-assembly">
<h2>transcript_assembly<a class="headerlink" href="#transcript-assembly" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">transcript_assembly</span></code> will detect transcripts based on the coverage.</p>
<p>For importing the information of libraries, please refer to the section of
<code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p>Wiggle files of fragmented libraries or TEX+/- treated libraries.</p>
<p>If user wants to compare transcripts with TSSs, it requires <code class="docutils literal"><span class="pre">.gff</span></code> files of TSSs.
If user wants to compare transcripts with genome anntation, it requires <code class="docutils literal"><span class="pre">.gff</span></code> files of genomes.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic transcript_assembly [-h]
                                     [--annotation_folder ANNOTATION_FOLDER]
                                     [--length LENGTH]
                                     [--tex_wig_path TEX_WIG_PATH]
                                     [--frag_wig_path FRAG_WIG_PATH]
                                     [--height HEIGHT] [--width WIDTH]
                                     [--tolerance TOLERANCE]
                                     [--tolerance_coverage TOLERANCE_COVERAGE]
                                     [-rt REPLICATES_TEX]
                                     [--replicates_frag REPLICATES_FRAG]
                                     [--tex_notex TEX_NOTEX]
                                     [--compare_TSS COMPARE_TSS]
                                     [--compare_genome_annotation COMPARE_GENOME_ANNOTATION]
                                     [--compare_feature_genome COMPARE_FEATURE_GENOME]
                                     [--TSS_fuzzy TSS_FUZZY]
                                     [--Tex_treated_libs TEX_TREATED_LIBS]
                                     [--fragmented_libs FRAGMENTED_LIBS]
                                     [--table_best]
                                     [--terminator_folder TERMINATOR_FOLDER]
                                     [--fuzzy_term FUZZY_TERM]
                                     [--max_length_distribution MAX_LENGTH_DISTRIBUTION]
                                     [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        It is for comparing transcript assembly and genome
                        annotation gff file. This function can use annotation
                        gff file as reference and modify transcript assembly
                        file, If the genome annotation gff folder is provided.
  --length LENGTH, -l LENGTH
                        The minimum width of transcript. It is for comparing
                        to annotation file (--annotation_folder). If
                        --annotation_folder is assigned, it will be the final
                        output. Otherwise, --width would be minimum length for
                        the final output. The default is 20.
  --tex_wig_path TEX_WIG_PATH, -tw TEX_WIG_PATH
                        The path of TEX+/- wig folder.
  --frag_wig_path FRAG_WIG_PATH, -fw FRAG_WIG_PATH
                        The path of fragment wig folder.
  --height HEIGHT, -he HEIGHT
                        The minimum height of coverage to be a transcript. The
                        default is 10.
  --width WIDTH, -w WIDTH
                        The minimum width of transcript. It is for not
                        comparing to annotation file (--annotation_folder). It
                        will be the final output if --annotation_folder is not
                        provided. Otherwise, --length would be the minimum
                        length of transcript for the final output. The default
                        is 20.
  --tolerance TOLERANCE, -t TOLERANCE
                        It indicates the number of nucleotides which coverages
                        drop below --height can be ignore. The default is 5.
  --tolerance_coverage TOLERANCE_COVERAGE, -tc TOLERANCE_COVERAGE
                        If the coverage is lower than tolerance_coverage, even
                        the length is within --tolerance, it will still
                        terminate the current transcript. The default is 0.
  -rt REPLICATES_TEX, --replicates_tex REPLICATES_TEX
                        The transcript should be detected at least in the
                        number of the replicates. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATED. For assigning
                        different --replicates_tex to different conditions,
                        please use comma to separate it. For example,
                        1_2,2_2,3_3 means number 1 and 2 condition assign 2 to
                        --replicate_tex, and number 3 condition assign 3 to
                        --replcate_tex. For assigning the same
                        --replicates_tex to all conditions, just assign like
                        all_1 (all condition use 1 --replicate_tex).
  --replicates_frag REPLICATES_FRAG, -rf REPLICATES_FRAG
                        The meaning and input type is the same as
                        --replicates_tex.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        For TEX+/- library, transcript should be detected in
                        both (TEX+ and TEX-) or can be detected in only one
                        library (TEX+ or TEX-). Please assign 1 or 2. Default
                        is 1.
  --compare_TSS COMPARE_TSS, -ct COMPARE_TSS
                        If comparing with TSS is needed, please assign TSS
                        folder.
  --compare_genome_annotation COMPARE_GENOME_ANNOTATION, -cg COMPARE_GENOME_ANNOTATION
                        If comparing with genome annotation file and searching
                        the parent transcript of gene is needed, please assign
                        annotation folder.
  --compare_feature_genome COMPARE_FEATURE_GENOME, -cf COMPARE_FEATURE_GENOME
                        If --compare_genome_annotation is provided, please
                        assign the feature which you want to compare. Default
                        is gene,CDS. For multiple features, just insert comma
                        between each feature, such as gene,CDS.
  --TSS_fuzzy TSS_FUZZY, -fu TSS_FUZZY
                        The fuzzy for comparing TSS and transcript assembly.
                        Default is 5.
  --Tex_treated_libs TEX_TREATED_LIBS, -tl TEX_TREATED_LIBS
                        Tex+/- library. The format is:
                        wig_file_name:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). For multiple
                        wig files, please use comma to separate the wig files.
                        For example, wig1:tex:1:a:+,wig2:tex:1:a:-.
  --fragmented_libs FRAGMENTED_LIBS, -fl FRAGMENTED_LIBS
                        Fragmented library. The format is: wig_file_name:fragm
                        ented(frag):condition_id(integer):replicate_id(alphabe
                        t):strand(+ or -). For multiple wig files, please use
                        comma to separate the wig files. For example,
                        wig1:frag:1:a:+,wig2:frag:1:a:-.
  --table_best, -tb     The output table only includes the best library.
                        Default is False.
  --terminator_folder TERMINATOR_FOLDER, -tr TERMINATOR_FOLDER
                        If comparing between transcripts and terminators is
                        needed, please assign the folder of gff files of
                        terminator here. Default is None.
  --fuzzy_term FUZZY_TERM, -fz FUZZY_TERM
                        If --terminator_folder is provided, please assign the
                        fuzzy here. Default is 30.
  --max_length_distribution MAX_LENGTH_DISTRIBUTION, -mb MAX_LENGTH_DISTRIBUTION
                        For generating the figure of distribution of
                        transcript length, please assign the maximum length
                        that you want to include. Default is 2000.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/transcriptome_assembly</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files of transcript.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The table of transcript with more details.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p>There are some useful tags in gff files.</p>
<p><code class="docutils literal"><span class="pre">compare_FEATURE</span></code>: The situation of overlap between transcripts and features (&#8211;compare_feature_genome)
(If &#8211;compare_genome_annotation is assigned.)</p>
<p><code class="docutils literal"><span class="pre">associated_tss</span></code>: Which TSSs are located in this transcripts.
(If &#8211;compare_TSS is assigned.)</p>
<p><code class="docutils literal"><span class="pre">associated_$FEATURE</span></code>: It shows the feature (&#8211;compare_feature_genome) which are located in this transcripts.
(If &#8211;compare_genome_annotation is assigned.)</p>
<p><code class="docutils literal"><span class="pre">detect_lib</span></code>: The transcript is detected by tex-treated libraries or fragmented libraries.</p>
<p><code class="docutils literal"><span class="pre">best_avg_coverage</span></code>: The average coverage of highest expressed library.</p>
<p>If &#8211;compare_genome_annotation is assigned, the associated transcript will be assigned in <code class="docutils literal"><span class="pre">Parent</span></code>
of genome annotations.</p>
<p>If &#8211;compare_TSS is assigned, the tag - <code class="docutils literal"><span class="pre">Parent</span></code> will appear
in the gff files of TSSs. It will show which transcripts that TSSs are located.</p>
</div>
<div class="section" id="terminator">
<h2>terminator<a class="headerlink" href="#terminator" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">terminator</span></code> will predict the rho-independent terminators. <code class="docutils literal"><span class="pre">ANNOgesic</span></code> combine the results of
two methods in order to get more reliable candidates. First one is using <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a>.
The other one is detect the specific secondary structure between converging pairs
of transcripts and CDSs. <code class="docutils literal"><span class="pre">ANNOgesic</span></code> can also compare with coverages in order to generate the terminators
which has coverage significant decrease.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a></p>
<p>RNAfold of <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a>.</p>
<p>Gff files target genome annotation.</p>
<p>Fasta files of target genome sequence.</p>
<p>Wiggle files of TEX +/- treated libraries or fragmented libraries.</p>
<p>Gff files of transcript.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic terminator [-h] [--TransTermHP_path TRANSTERMHP_PATH]
                            [--expterm_path EXPTERM_PATH]
                            [--RNAfold_path RNAFOLD_PATH]
                            [--fasta_folder FASTA_FOLDER]
                            [--annotation_folder ANNOTATION_FOLDER]
                            [--transcript_folder TRANSCRIPT_FOLDER]
                            [--sRNA SRNA] [--statistics]
                            [--tex_wig_folder TEX_WIG_FOLDER]
                            [--frag_wig_folder FRAG_WIG_FOLDER]
                            [--decrease DECREASE]
                            [--fuzzy_detect_coverage FUZZY_DETECT_COVERAGE]
                            [--fuzzy_within_transcript FUZZY_WITHIN_TRANSCRIPT]
                            [--fuzzy_downstream_transcript FUZZY_DOWNSTREAM_TRANSCRIPT]
                            [--fuzzy_within_gene FUZZY_WITHIN_GENE]
                            [--fuzzy_downstream_gene FUZZY_DOWNSTREAM_GENE]
                            [--highest_coverage HIGHEST_COVERAGE]
                            [-tl TEX_NOTEX_LIBS] [-fl FRAG_LIBS]
                            [-te TEX_NOTEX] [-rt REPLICATES_TEX]
                            [-rf REPLICATES_FRAG] [-tb] [-ml MIN_LOOP_LENGTH]
                            [-Ml MAX_LOOP_LENGTH] [-ms MIN_STEM_LENGTH]
                            [-Ms MAX_STEM_LENGTH] [-mr MISS_RATE]
                            [-mu MIN_U_TAIL_LENGTH] [-ru RANGE_U_TAIL] [-kp]
                            [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TransTermHP_path TRANSTERMHP_PATH
                        Please assign the path of &quot;transterm&quot; in TransTermHP.
  --expterm_path EXPTERM_PATH
                        Please assign the path of expterm.dat for TransTermHP.
                        Default is /usr/local/bin/expterm.dat
  --RNAfold_path RNAFOLD_PATH
                        If you want to assign the path of &quot;RNAfold&quot; of Vienna
                        package, please assign here.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        The path of genome fasta folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --transcript_folder TRANSCRIPT_FOLDER, -a TRANSCRIPT_FOLDER
                        The path of transcript assembly gff folder.
  --sRNA SRNA, -sr SRNA
                        For including sRNA information, please assign the
                        folder of sRNA gff files.
  --statistics, -s      Doing statistics for terminator. The name of
                        statistics file is - stat_terminator_$STRAIN_NAME.csv.
                        Default is False.
  --tex_wig_folder TEX_WIG_FOLDER, -tw TEX_WIG_FOLDER
                        If TEX+/- libraries are provided, please assign TEX+/-
                        wig folder.
  --frag_wig_folder FRAG_WIG_FOLDER, -fw FRAG_WIG_FOLDER
                        If fragmented libraries are provided, please assign
                        fragmented wig folder.
  --decrease DECREASE, -d DECREASE
                        If the (lowest coverage / highest coverage) in the
                        terminator is smaller than this number, it will
                        consider this terminator have coverage dramatic
                        decreasing in it. Default is 0.5.
  --fuzzy_detect_coverage FUZZY_DETECT_COVERAGE, -fc FUZZY_DETECT_COVERAGE
                        For extending the region of detection of coverage
                        significant decreasing. Ex: the location of terminator
                        is 300-400, and --fuzzy_detect_coverage is 30. If the
                        coverage decrease is detected within 270-430, it will
                        still consider the terminator have coverage dramatic
                        decrease. Default is 30.
  --fuzzy_within_transcript FUZZY_WITHIN_TRANSCRIPT, -fut FUZZY_WITHIN_TRANSCRIPT
                        If the candidates are within transcript and the
                        distance between the end of gene/transcript and
                        terminator candidate is within this number, it will be
                        consider as terminator. Default is 30.
  --fuzzy_downstream_transcript FUZZY_DOWNSTREAM_TRANSCRIPT, -fdt FUZZY_DOWNSTREAM_TRANSCRIPT
                        The meaning is similar with --fuzzy_within_transcript.
                        It is for the candidates which are downstream of
                        transcript. Default is 30.
  --fuzzy_within_gene FUZZY_WITHIN_GENE, -fuc FUZZY_WITHIN_GENE
                        The meaning is similar with --fuzzy_within_transcript.
                        It is for gene in stead of transcript. Default is 10.
  --fuzzy_downstream_gene FUZZY_DOWNSTREAM_GENE, -fdg FUZZY_DOWNSTREAM_GENE
                        The meaning is similar with
                        --fuzzy_downstream_transcript. It is for gene in stead
                        of transcript. Default is 310.
  --highest_coverage HIGHEST_COVERAGE, -hc HIGHEST_COVERAGE
                        If the highest coverage of terminator is below to this
                        number, the terminator will be classify to non-detect
                        and not included in &quot;best&quot; results, but still included
                        in &quot;all_candidates&quot;. Default is 10.
  -tl TEX_NOTEX_LIBS, --tex_notex_libs TEX_NOTEX_LIBS
                        Library name of TEX+/- library. The format is:
                        wig_file_name:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). For multiple
                        wig files, please use comma to separate the wig files.
                        For example, wig1:tex:1:a:+,wig2:tex:1:a:-.
  -fl FRAG_LIBS, --frag_libs FRAG_LIBS
                        Library name of fragmented library. The format is: wig
                        _file_name:fragmented(frag):condition_id(integer):repl
                        icate_id(alphabet):strand(+ or -). For multiple wig
                        files, please use comma to separate the wig files. For
                        example, wig1:frag:1:a:+,wig2:frag:1:a:-.
  -te TEX_NOTEX, --tex_notex TEX_NOTEX
                        For TEX+/- library, terminators should be detected in
                        both (TEX+ and TEX-) or can be detected in only one
                        library (TEX+ or TEX-). Please assign 1 or 2. Default
                        is 1.
  -rt REPLICATES_TEX, --replicates_tex REPLICATES_TEX
                        The Terminator candidates should be detected at least
                        in the number of the replicates. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATED. For assigning
                        different --replicate_tex to different conditions,
                        please use comma to separate it. For example,
                        1_2,2_2,3_3 means number 1 and 2 condition assign 2 to
                        --replicate_tex, and number 3 condition assign 3 to
                        --replcate_tex. For assigning the same
                        --replicates_tex to all conditions, just assign like
                        all_1 (all condition use 1 --replicate_tex).
  -rf REPLICATES_FRAG, --replicates_frag REPLICATES_FRAG
                        The meaning and input type is the same as
                        --replicates_tex.
  -tb, --table_best     Output table only contains the library which has
                        coverage most significant decreasing. Default is
                        False.
  -ml MIN_LOOP_LENGTH, --min_loop_length MIN_LOOP_LENGTH
                        The minimum length of loop for terminator. Default is
                        3 nts.
  -Ml MAX_LOOP_LENGTH, --max_loop_length MAX_LOOP_LENGTH
                        The maximum length of loop for terminator. Default is
                        10 nts.
  -ms MIN_STEM_LENGTH, --min_stem_length MIN_STEM_LENGTH
                        The minimum length of stem for terminator. Default is
                        4 nts.
  -Ms MAX_STEM_LENGTH, --max_stem_length MAX_STEM_LENGTH
                        The maximum length of stem for terminator. Default is
                        20 nts.
  -mr MISS_RATE, --miss_rate MISS_RATE
                        The percentage of nucleotides which can be no base
                        pair in the stem. Default is 0.25.
  -mu MIN_U_TAIL_LENGTH, --min_U_tail_length MIN_U_TAIL_LENGTH
                        The minimum length of U tail for terminator. Default
                        is 3 nts.
  -ru RANGE_U_TAIL, --range_U_tail RANGE_U_TAIL
                        The range of nucleotides for detection of U tail. For
                        example, if --range_U_tail is 6 and
                        --min_U_tail_length is 3, and there are 3 U within 6
                        nts, it will be assigned to detecting U tail
                        successfully. Default is 6.
  -kp, --keep_multi_term
                        Sometimes, one gene is associated with more terminator
                        candidates. In default, it will only keep the high
                        confident one. The function can keep all terminators
                        which associated with the same gene. Default is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/terminator</span></code>.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p><code class="docutils literal"><span class="pre">transtermhp</span></code>: All output of <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of terminator.
There are four different sub-folders to store terminators.</p>
<p><code class="docutils literal"><span class="pre">all_candidates</span></code> will store all terminators which <code class="docutils literal"><span class="pre">ANNOgesic</span></code> can detect.</p>
<p><code class="docutils literal"><span class="pre">express</span></code> will store the terminators which has gene expression.</p>
<p><code class="docutils literal"><span class="pre">best</span></code> will store the terminators which not only has gene expression but also
has coverage dramatic decrease.</p>
<p><code class="docutils literal"><span class="pre">non_express</span></code> will store the terminators which has no gene expression.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of terminators with more details.</p>
<p>The tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">method</span></code>: The method that this terminator be detected.</p>
<p><code class="docutils literal"><span class="pre">coverage_decrease</span></code>: The coverage of the terminator has dramatic decreasing or not.</p>
<p><code class="docutils literal"><span class="pre">express</span></code>: The terminator has gene expression or not.</p>
<p><code class="docutils literal"><span class="pre">diff_coverage</span></code>: The highest coverage and lowest coverage of the library which expresses highest.
The numbers in parens are highest coverage and lowest coverage.</p>
<p><code class="docutils literal"><span class="pre">Parent</span></code>: This tag presents the parent transcript of terminator.</p>
</div>
<div class="section" id="utr">
<h2>utr<a class="headerlink" href="#utr" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">utr</span></code> can compare with TSSs, CDSs/tRNAs/sRNAs, transcripts and terminators
to generate proper UTRs. 5&#8217;UTRs are based on detecting the regions between TSSs and CDSs/tRNAs/sRNAs.
3&#8217;UTRs are based on detecting the
regions between the end of transcripts and CDSs/tRNAs/sRNAs. If the gff files of TSSs are not computed by
ANNOgesic, please use &#8211;TSS_source. <code class="docutils literal"><span class="pre">utr</span></code> would classify TSSs for the analysis.</p>
<ul class="simple">
<li>Pre-required files</li>
</ul>
<p>Gff files of genome annotations, TSSs and transcripts.</p>
<p>If the information of terminators is needed, the gff files of terminators are required.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">utr</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">annotation_folder</span> <span class="n">ANNOTATION_FOLDER</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">TSS_folder</span> <span class="n">TSS_FOLDER</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">transcript_assembly_folder</span> <span class="n">TRANSCRIPT_ASSEMBLY_FOLDER</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">terminator_folder</span> <span class="n">TERMINATOR_FOLDER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">TSS_source</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">base_5UTR</span> <span class="n">BASE_5UTR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">UTR_length</span> <span class="n">UTR_LENGTH</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">base_3UTR</span> <span class="n">BASE_3UTR</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">terminator_fuzzy</span> <span class="n">TERMINATOR_FUZZY</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_3utr</span> <span class="n">FUZZY_3UTR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_5utr</span> <span class="n">FUZZY_5UTR</span><span class="p">]</span>
                     <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">annotation_folder</span> <span class="n">ANNOTATION_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_FOLDER</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">TSS_folder</span> <span class="n">TSS_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">t</span> <span class="n">TSS_FOLDER</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">TSS</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_assembly_folder</span> <span class="n">TRANSCRIPT_ASSEMBLY_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_ASSEMBLY_FOLDER</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">transcriptome</span> <span class="n">assembly</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">terminator_folder</span> <span class="n">TERMINATOR_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">e</span> <span class="n">TERMINATOR_FOLDER</span>
                        <span class="n">If</span> <span class="n">including</span> <span class="n">the</span> <span class="n">information</span> <span class="n">of</span> <span class="n">terminator</span> <span class="ow">is</span>
                        <span class="n">required</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">terminator</span> <span class="n">folder</span>
                        <span class="n">here</span><span class="o">.</span>
  <span class="o">--</span><span class="n">TSS_source</span><span class="p">,</span> <span class="o">-</span><span class="n">s</span>      <span class="n">If</span> <span class="n">TSS</span> <span class="n">file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">generated</span> <span class="kn">from</span> <span class="nn">ANNOgesic</span><span class="p">,</span> <span class="n">please</span>
                        <span class="n">turn</span> <span class="n">it</span> <span class="n">on</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">(</span><span class="n">ANNOgesic</span><span class="p">)</span><span class="o">.</span>
  <span class="o">--</span><span class="n">base_5UTR</span> <span class="n">BASE_5UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">b5</span> <span class="n">BASE_5UTR</span>
                        <span class="n">The</span> <span class="n">information</span> <span class="k">for</span> <span class="n">detection</span> <span class="n">of</span> <span class="mi">5</span><span class="s1">&#39;UTR. It can be</span>
                        <span class="s2">&quot;TSS&quot;</span> <span class="ow">or</span> <span class="s2">&quot;transcript&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="n">both</span><span class="o">.</span>
  <span class="o">--</span><span class="n">UTR_length</span> <span class="n">UTR_LENGTH</span><span class="p">,</span> <span class="o">-</span><span class="n">l</span> <span class="n">UTR_LENGTH</span>
                        <span class="n">The</span> <span class="n">maximum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">UTR</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">300.</span>
  <span class="o">--</span><span class="n">base_3UTR</span> <span class="n">BASE_3UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">b3</span> <span class="n">BASE_3UTR</span>
                        <span class="n">the</span> <span class="n">information</span> <span class="k">for</span> <span class="n">detection</span> <span class="n">of</span> <span class="mi">3</span><span class="s1">&#39;UTR. It can be</span>
                        <span class="s2">&quot;transcript&quot;</span> <span class="ow">or</span> <span class="s2">&quot;terminator&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="n">transcript</span><span class="o">.</span>
  <span class="o">--</span><span class="n">terminator_fuzzy</span> <span class="n">TERMINATOR_FUZZY</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">TERMINATOR_FUZZY</span>
                        <span class="n">This</span> <span class="ow">is</span> <span class="n">only</span> <span class="k">for</span> <span class="o">--</span><span class="n">base_3UTR</span> <span class="n">which</span> <span class="n">assigned</span> <span class="n">by</span>
                        <span class="s2">&quot;transcript&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="ow">and</span> <span class="n">terminator</span> <span class="n">file</span> <span class="n">are</span>
                        <span class="n">provided</span><span class="o">.</span> <span class="n">If</span> <span class="n">the</span> <span class="n">distance</span> <span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span> <span class="n">between</span>
                        <span class="n">terminator</span> <span class="ow">and</span> <span class="n">the</span> <span class="n">end</span> <span class="n">of</span> <span class="n">transcript</span> <span class="n">lower</span> <span class="n">than</span> <span class="n">this</span>
                        <span class="n">value</span><span class="p">,</span> <span class="n">it</span> <span class="n">will</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">terminator</span> <span class="n">associated</span> <span class="k">with</span>
                        <span class="n">the</span> <span class="mi">3</span><span class="s1">&#39;UTR. Therefore, the relationship between 3&#39;</span><span class="n">UTRs</span>
                        <span class="ow">and</span> <span class="n">terminator</span> <span class="n">can</span> <span class="n">be</span> <span class="n">gain</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">30.</span>
  <span class="o">--</span><span class="n">fuzzy_3utr</span> <span class="n">FUZZY_3UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">f3</span> <span class="n">FUZZY_3UTR</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">base_3UTR</span> <span class="n">includes</span> <span class="n">transcript</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span>
                        <span class="n">fuzzy</span> <span class="n">of</span> <span class="mi">3</span><span class="s1">&#39;UTR. Default is 10 nucleotides.</span>
  <span class="o">--</span><span class="n">fuzzy_5utr</span> <span class="n">FUZZY_5UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">f5</span> <span class="n">FUZZY_5UTR</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">base_5UTR</span> <span class="n">includes</span> <span class="n">transcript</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span>
                        <span class="n">fuzzy</span> <span class="n">of</span> <span class="mi">5</span><span class="s1">&#39;UTR. Default is 5 nucleotides.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output of 5&#8217;UTRs will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/UTR/5UTR</span></code>.</p>
<p>All output of 3&#8217;UTRs will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/UTR/3UTR</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of 5&#8217;UTR/3&#8217;UTR</p>
<p>The tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">length</span></code>: UTR length.</p>
<p><code class="docutils literal"><span class="pre">associated_cds</span></code>: Which CDSs/rRNAs/tRNAs are associated with this UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_gene</span></code>: Which genes are associated with this UTR.</p>
<p><code class="docutils literal"><span class="pre">Parent</span></code>: Which transcript is associated with this UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_tss</span></code>: Which TSSs are associated with this 5&#8217;UTR.</p>
<p><code class="docutils literal"><span class="pre">tss_type</span></code>: What types of TSSs are associated with this 5&#8217;UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_term</span></code>: Which terminators are associated with this 3&#8217;UTR.</p>
</div>
<div class="section" id="srna">
<h2>srna<a class="headerlink" href="#srna" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">srna</span></code> can predict different types of sRNAs. For intergenic and antisense sRNA, it
is detected via comparison of the transcripts and annotation profile.
For UTR-derived sRNA, the detection is based on the TSSs and processing sites,
transcript and genome annotation.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p>Gff files of genome annotation and Transcript data.</p>
<p>wiggle files: The libraries and wiggle files, Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>If you want to detect the UTR-derived sRNAs, it is necessary to input
TSS information. It is for the detection of 5&#8217;UTR-derived sRNA and interCDS-derived sRNA.
If you don&#8217;t want to detect UTR-derived sRNAs,
TSS information still can be provided as a filter.</p>
<p>Optional input file:</p>
<p>processing site: It is for checking the sRNAs which end with processing sites. Moreover,
Some 3&#8217;UTR-derived and interCDS-derived sRNA candidates start
from processing sites not TSSs. If you don&#8217;t want to detect UTR-derived sRNAs,
This information still can be provided to increase the accuracy, especially for some
long non-coding RNAs.</p>
<p>There are some filters can improve the prediction, and they need some input information.</p>
<p>Secondary structure: <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a>,
<a class="reference external" href="http://pages.cs.wisc.edu/~ghost/doc/AFPL/6.50/Ps2pdf.htm">Ps2pdf14</a> and
Fasta files of genome sequence is required.</p>
<p>TSS: gff file of TSS is necessary. Based on different types of TSS, the cutoff of coverage can be
assigned separately.</p>
<p>Searching sRNA database: <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">Blast+</a> and
sRNA database, such as <a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a> are necessary.
The format of header should be <code class="docutils literal"><span class="pre">$ID|$STRAIN|$SRNANAME</span></code>.
For example, <code class="docutils literal"><span class="pre">&gt;saci2813.1|Acinetobacter</span> <span class="pre">sp.</span> <span class="pre">ADP1|Aar</span></code>.
The ID is saci403.1; the strain of this sRNA is Acinetobacter sp. ADP1 and the name of sRNA is Aar.
If the database does not follow the format, it will occur error when the user runs with <code class="docutils literal"><span class="pre">--sRNA_blast_stat,</span> <span class="pre">-sb</span></code>.
Or the results will be meaningless. User can also assign <code class="docutils literal"><span class="pre">--best_with_all_sRNAhit,</span> <span class="pre">-ba</span></code> for including
all candidates which can find hits in sRNA database (even without matching other filters).</p>
<p>Searching nr database: <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">Blast+</a> and
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/">nr database</a> are required.</p>
<p>Terminator: gff file of terminator is needed. If sRNA must be associated with terminator,
please add <code class="docutils literal"><span class="pre">-bt</span></code>. Without <code class="docutils literal"><span class="pre">-bt</span></code>, it will only provide to the relationship between terminator and sRNA
but does not remove the the candidates which not associated with terminator.</p>
<p>sORF: gff file of sORF needs to be assigned. If sRNA must be not overlap with sORF,
please add <code class="docutils literal"><span class="pre">-bs</span></code>. Without <code class="docutils literal"><span class="pre">-bs</span></code>, it will only provide to the relationship between sORF and sRNA
but does not remove the the candidates which are overlap with sORF.</p>
<p>Promoter: Table of promoters have to be input. It can compare between promoters and sRNA for ranking
the sRNA candidate as well. Moreover, If sRNA must be associated with promoter,
please add <code class="docutils literal"><span class="pre">-bp</span></code>. Without <code class="docutils literal"><span class="pre">-bp</span></code>, it will only provide to the relationship between promoter and sRNA
but does not remove the the candidates which not associated with promoter.</p>
<p>The format should be</p>
<table border="1" class="docutils">
<colgroup>
<col width="28%" />
<col width="30%" />
<col width="25%" />
<col width="18%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>strain</td>
<td>TSS_position</td>
<td>TSS_strand</td>
<td>Motif</td>
</tr>
<tr class="row-even"><td>NC_000915.1</td>
<td>237118</td>
<td>-</td>
<td>MOTIF_1</td>
</tr>
<tr class="row-odd"><td>NC_000915.1</td>
<td>729009</td>
<td>-</td>
<td>MOTIF_1</td>
</tr>
</tbody>
</table>
<p>First row is the header of table, the last column is the name of motif/promoter.
If subcommand <code class="docutils literal"><span class="pre">promoter</span></code> was used for detecting promoter, the table will be generated automatically.
Please refer to the section of <code class="docutils literal"><span class="pre">promoter</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic srna [-h] [--Vienna_folder VIENNA_FOLDER]
                      [--Vienna_utils VIENNA_UTILS]
                      [--blast_plus_folder BLAST_PLUS_FOLDER]
                      [--ps2pdf14_path PS2PDF14_PATH] [--UTR_derived_sRNA]
                      [--import_info IMPORT_INFO]
                      [--transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER]
                      [--annotation_folder ANNOTATION_FOLDER]
                      [--TSS_folder TSS_FOLDER]
                      [--processing_site_folder PROCESSING_SITE_FOLDER]
                      [--promoter_table PROMOTER_TABLE]
                      [--promoter_name PROMOTER_NAME] [--TSS_source]
                      [--TSS_intergenic_fuzzy TSS_INTERGENIC_FUZZY]
                      [--TSS_5UTR_fuzzy TSS_5UTR_FUZZY]
                      [--TSS_3UTR_fuzzy TSS_3UTR_FUZZY]
                      [--TSS_interCDS_fuzzy TSS_INTERCDS_FUZZY]
                      [--terminator_folder TERMINATOR_FOLDER]
                      [--terminator_fuzzy_in_CDS TERMINATOR_FUZZY_IN_CDS]
                      [--terminator_fuzzy_out_CDS TERMINATOR_FUZZY_OUT_CDS]
                      [--min_length MIN_LENGTH] [--max_length MAX_LENGTH]
                      [--tex_wig_folder TEX_WIG_FOLDER]
                      [--frag_wig_folder FRAG_WIG_FOLDER]
                      [--run_intergenic_TEX_coverage RUN_INTERGENIC_TEX_COVERAGE]
                      [--run_intergenic_noTEX_coverage RUN_INTERGENIC_NOTEX_COVERAGE]
                      [--run_intergenic_fragmented_coverage RUN_INTERGENIC_FRAGMENTED_COVERAGE]
                      [--run_antisense_TEX_coverage RUN_ANTISENSE_TEX_COVERAGE]
                      [--run_antisense_noTEX_coverage RUN_ANTISENSE_NOTEX_COVERAGE]
                      [--run_antisense_fragmented_coverage RUN_ANTISENSE_FRAGMENTED_COVERAGE]
                      [--intergenic_tolerance INTERGENIC_TOLERANCE]
                      [--run_utr_TEX_coverage RUN_UTR_TEX_COVERAGE]
                      [--run_utr_noTEX_coverage RUN_UTR_NOTEX_COVERAGE]
                      [--run_utr_fragmented_coverage RUN_UTR_FRAGMENTED_COVERAGE]
                      [--min_utr_coverage MIN_UTR_COVERAGE]
                      [--fasta_folder FASTA_FOLDER]
                      [--cutoff_energy CUTOFF_ENERGY] [--mountain_plot]
                      [--nr_format] [--srna_format]
                      [--sRNA_database_path SRNA_DATABASE_PATH]
                      [--nr_database_path NR_DATABASE_PATH]
                      [--tex_notex_libs TEX_NOTEX_LIBS]
                      [--frag_libs FRAG_LIBS] [--tex_notex TEX_NOTEX]
                      [-rt REPLICATES_TEX] [--replicates_frag REPLICATES_FRAG]
                      [--table_best]
                      [--decrease_intergenic DECREASE_INTERGENIC]
                      [--decrease_utr DECREASE_UTR]
                      [--fuzzy_intergenic FUZZY_INTERGENIC]
                      [--fuzzy_utr FUZZY_UTR] [--cutoff_nr_hit CUTOFF_NR_HIT]
                      [--blast_e_nr BLAST_E_NR] [--blast_e_srna BLAST_E_SRNA]
                      [--sORF SORF] [--best_with_all_sRNAhit]
                      [--best_without_sORF_candidate] [--best_with_terminator]
                      [--best_with_promoter] [--detect_sRNA_in_CDS]
                      [--overlap_percent_CDS OVERLAP_PERCENT_CDS]
                      [--ignore_hypothetical_protein]
                      [--ranking_time_promoter RANKING_TIME_PROMOTER]
                      [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --Vienna_folder VIENNA_FOLDER
                        Please assign the folder of Vienna package. It should
                        include RNAfold.
  --Vienna_utils VIENNA_UTILS
                        Please assign the folder of Utils of Vienna package.
                        It should include relplot.pl and mountain.pl.
  --blast_plus_folder BLAST_PLUS_FOLDER
                        Please assign the folder of blast+ which include
                        blastn, blastx, makeblastdb.
  --ps2pdf14_path PS2PDF14_PATH
                        Please assign the path of ps2pdf14.
  --UTR_derived_sRNA, -u
                        The function is for detecting UTR-derived sRNA.
                        Default is False.
  --import_info IMPORT_INFO, -d IMPORT_INFO
                        There are several types of information that you can
                        import to detect and filter sRNAs: tss (the sRNA
                        should start from a TSS), sec_str (free energy change
                        of secondary structure (normalized by length)),
                        blast_nr (blast to non-redundant database), blast_srna
                        (blast to sRNA), sorf (compare with sORF), term
                        (compare with terminator), promoter (compare with
                        promoter motif). ATTENTION: without filters, the
                        results may include many false positives. Please
                        assign the information that you want to import (comma
                        for separate the filters), ex: tss,sec_str,blast_nr -
                        means it used 1. TSS, 2. free energy change of
                        secondary structure and 3. blast to nr database to
                        detect sRNA. Besides these, it will also consider the
                        sequence length of sRNA. ATTENTION: if you want to
                        import sRNA database, please follow the format:
                        $ID|$STRAIN|$SRNANAME. Default is
                        tss,sec_str,blast_nr,blast_srna.
  --transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER, -a TRANSCRIPT_ASSEMBLY_FOLDER
                        The path of transcriptome assembly folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        If TSS information is needed, please assign the path
                        of gff folder of TSS. For detection of UTR-derived
                        sRNA, TSS information MUST be provided.
  --processing_site_folder PROCESSING_SITE_FOLDER, -p PROCESSING_SITE_FOLDER
                        If processing site information is needed, please
                        assign the path of gff folder of processing site.For
                        detection of UTR-derived sRNA, processing site
                        information can improve the results.
  --promoter_table PROMOTER_TABLE, -pt PROMOTER_TABLE
                        If promoter information is needed, please assign the
                        path of promoter table. The format of table is $STRAIN
                        $TSS_POSITION $TSS_STRAND $PROMOTER_NAME. TSS
                        information is also required.
  --promoter_name PROMOTER_NAME, -pn PROMOTER_NAME
                        If --promoter_table is provided, please assign the
                        promoter name (the last column of promoter table)
                        which you want to compare. For multiple promoters,
                        please put comma between the promoters. Default is
                        None.
  --TSS_source, -ts     If the gff file of TSS is not generated by ANNOgesic,
                        please use this function to classify TSSs and generate
                        the proper format for sRNA prediction. Default is
                        True.
  --TSS_intergenic_fuzzy TSS_INTERGENIC_FUZZY, -ft TSS_INTERGENIC_FUZZY
                        If --TSS_folder is provided, please assign the fuzzy
                        for comparing TSS and transcript. It is for intergenic
                        sRNA. Default is 3.
  --TSS_5UTR_fuzzy TSS_5UTR_FUZZY, -f5 TSS_5UTR_FUZZY
                        If --TSS_folder is provided, please assign the fuzzy
                        for comparing TSS and transcript. It is for 5&#39;UTR of
                        UTR derived sRNA.The type of input can be percentage
                        or the real amount of reads. Ex: p_0.05 means the
                        fuzzy is 5 percent of the length of 5&#39;UTR. n_10 means
                        the fuzzy is 10 base pair. Default is n_3.
  --TSS_3UTR_fuzzy TSS_3UTR_FUZZY, -f3 TSS_3UTR_FUZZY
                        The meaning is similar with --TSS_5UTR_fuzzy. It is
                        for 3&#39;UTR instead of 5&#39;UTR. Default is p_0.04.
  --TSS_interCDS_fuzzy TSS_INTERCDS_FUZZY, -fc TSS_INTERCDS_FUZZY
                        The meaning is similar with --TSS_5UTR_fuzzy. It is
                        for interCDS instead of 5&#39;UTR. Default is p_0.04.
  --terminator_folder TERMINATOR_FOLDER, -tf TERMINATOR_FOLDER
                        If terminator information is needed, please assign the
                        path of gff folder of terminator.
  --terminator_fuzzy_in_CDS TERMINATOR_FUZZY_IN_CDS, -tfi TERMINATOR_FUZZY_IN_CDS
                        If --terminator_folder is provided, please assign the
                        fuzzy for comparing terminator and transcript. It is
                        the fuzzy for the terminator which is within CDS.
                        Default is 30.
  --terminator_fuzzy_out_CDS TERMINATOR_FUZZY_OUT_CDS, -tfo TERMINATOR_FUZZY_OUT_CDS
                        If --terminator_folder is provided, please assign the
                        fuzzy for comparing terminator and transcript. It is
                        the fuzzy for the terminator which is outside of CDS.
                        Default is 30.
  --min_length MIN_LENGTH, -lm MIN_LENGTH
                        Please assign the minimum length of sRNA. Default is
                        30.
  --max_length MAX_LENGTH, -lM MAX_LENGTH
                        Please assign the maximum length of sRNA. Default is
                        500.
  --tex_wig_folder TEX_WIG_FOLDER, -tw TEX_WIG_FOLDER
                        The path of TEX+/- wig folder.
  --frag_wig_folder FRAG_WIG_FOLDER, -fw FRAG_WIG_FOLDER
                        The path of fragment wig folder.
  --run_intergenic_TEX_coverage RUN_INTERGENIC_TEX_COVERAGE, -it RUN_INTERGENIC_TEX_COVERAGE
                        The minimum average coverage of intergenic sRNA
                        candidates for TEX+. The cutoff of coverage for sRNA
                        prediction is based on different types of TSSs. The
                        order of numbers is &quot;Primary Secondary Internal
                        Antisense Orphan&quot; (separated by comma). Ex: The cutoff
                        is 0,0,0,50,10, it means that antisense (cutoff
                        coverage is 50) TSS and orphan (cutoff coverage is 10)
                        TSS are used for sRNA prediction. 0 means that not use
                        it for prediction. If TSS information is not provided,
                        it will choose the lowest one as a general cutoff for
                        prediction. Ex: if the cutoff is 0,0,0,50,10 and
                        --TSS_folder is not provided, 10 will be the general
                        cutoff for prediction. Default is 0,0,0,40,20.
  --run_intergenic_noTEX_coverage RUN_INTERGENIC_NOTEX_COVERAGE, -in RUN_INTERGENIC_NOTEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_TEX_coverage. This is for TEX-
                        library. Default is 0,0,0,30,10.
  --run_intergenic_fragmented_coverage RUN_INTERGENIC_FRAGMENTED_COVERAGE, -if RUN_INTERGENIC_FRAGMENTED_COVERAGE
                        The meaning is the same as
                        --run_intergenic_TEX_coverage. This is for fragmented
                        library. Default is 400,200,0,50,20.
  --run_antisense_TEX_coverage RUN_ANTISENSE_TEX_COVERAGE, -at RUN_ANTISENSE_TEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_TEX_coverage. Just apply to
                        antisense. Default is 0,0,0,40,20.
  --run_antisense_noTEX_coverage RUN_ANTISENSE_NOTEX_COVERAGE, -an RUN_ANTISENSE_NOTEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_noTEX_coverage. Just apply to
                        antisense. Default is 0,0,0,30,10.
  --run_antisense_fragmented_coverage RUN_ANTISENSE_FRAGMENTED_COVERAGE, -af RUN_ANTISENSE_FRAGMENTED_COVERAGE
                        The meaning is the same as
                        --run_intergenic_fragmented_coverage. Just apply to
                        antisense. Default is 400,200,0,50,20.
  --intergenic_tolerance INTERGENIC_TOLERANCE, -ti INTERGENIC_TOLERANCE
                        This number indicates the tolerance of temporary drop
                        below cutoff of coverage. Default is 5.
  --run_utr_TEX_coverage RUN_UTR_TEX_COVERAGE, -ut RUN_UTR_TEX_COVERAGE
                        The minimum average coverage of UTR-derived sRNA
                        candidates for TEX+. The cutoff can be assigned by the
                        percentile or real number of coverage. The order of
                        numbers are &quot;5&#39;UTR, 3&#39;UTR and interCDS&quot; (separated by
                        comma). Ex: if the cutoff is &quot;p_0.7,p_0.5,p_0.5&quot;, it
                        will use 70 percentile of coverage as cutoff for
                        5&#39;UTR, median of coverage as cutoff for 3&#39;UTR and
                        interCDS. Ex: if the cutoff is &quot;n_30,n_10,n_20 &quot; it
                        will use 30 as cutoff for 5&#39;UTR and 10 as cutoff for
                        3&#39;UTR and 20 for interCDS. Default is
                        p_0.8,p_0.6,p_0.7.
  --run_utr_noTEX_coverage RUN_UTR_NOTEX_COVERAGE, -un RUN_UTR_NOTEX_COVERAGE
                        The meaning is the same as --run_utr_TEX_coverage.
                        This is for TEX- library. Default is
                        p_0.7,p_0.5,p_0.6.
  --run_utr_fragmented_coverage RUN_UTR_FRAGMENTED_COVERAGE, -uf RUN_UTR_FRAGMENTED_COVERAGE
                        The meaning is the same as --run_utr_TEX_coverage.
                        This is for fragmented library. Default is
                        p_0.7,p_0.5,p_0.6.
  --min_utr_coverage MIN_UTR_COVERAGE, -mu MIN_UTR_COVERAGE
                        The minimum coverage of UTR-derived sRNA. The coverage
                        should not only fit the --run_utr_TEX_coverage,
                        --run_utr_noTEX_coverage and
                        --run_utr_fragmented_coverage, but also this value.
                        Defaul is 50.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        If &quot;sec_str&quot; or &quot;blast_nr&quot; or &quot;blast_srna&quot; is assigned
                        to --import_info, please assign the path of genome
                        fasta folder.
  --cutoff_energy CUTOFF_ENERGY, -e CUTOFF_ENERGY
                        If secondary structure information is needed, please
                        assign the cutoff of folding energy change (normalized
                        by length of gene). Default is -0.05.
  --mountain_plot, -m   It is for generating mountain plot of sRNA candidate.
                        Default is False.
  --nr_format, -nf      If nr database is not formatted, it is for formating
                        nr database. Default is False.
  --srna_format, -sf    If sRNA databse is not formatted, it is for formating
                        sRNA database. Default is False.
  --sRNA_database_path SRNA_DATABASE_PATH, -sd SRNA_DATABASE_PATH
                        If blast results of sRNA is needed, please assign the
                        path of sRNA database.
  --nr_database_path NR_DATABASE_PATH, -nd NR_DATABASE_PATH
                        If blast results of nr is needed, please assign the
                        path of nr database.
  --tex_notex_libs TEX_NOTEX_LIBS, -tl TEX_NOTEX_LIBS
                        library name of TEX+/- libraries. The format is:
                        wig_file_name:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). For multiple
                        wig files, please use comma to separate the wig files.
                        For example, wig1:tex:1:a:+,wig2:tex:1:a:-.
  --frag_libs FRAG_LIBS, -fl FRAG_LIBS
                        library name of fragmented libraries. The format is: w
                        ig_file_name:fragmented(frag):condition_id(integer):re
                        plicate_id(alphabet):strand(+ or -). For multiple wig
                        files, please use comma to separate the wig files. For
                        example, wig1:frag:1:a:+,wig2:frag:1:a:-.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        For TEX+/- library, sRNA should be detected in both
                        (TEX+ and TEX-) or can be detected in only one library
                        (TEX+ or TEX-). Please assign 1 or 2. Default is 2.
  -rt REPLICATES_TEX, --replicates_tex REPLICATES_TEX
                        The sRNA should be detected at least in the number of
                        the replicates. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATED. For assigning
                        different --replicates_tex to different conditions,
                        please use comma to separate it. For example,
                        1_2,2_2,3_3 means number 1 and 2 condition assign 2 to
                        --replicates_tex, and number 3 condition assign 3 to
                        --replcates_tex. For assigning the same
                        --replicates_tex to all conditions, just assign like
                        all_1 (all condition use 1 --replicates_tex).
  --replicates_frag REPLICATES_FRAG, -rf REPLICATES_FRAG
                        The meaning and input type is the same as
                        --replicates_tex.
  --table_best, -tb     The output table of sRNA candidates only prints the
                        best library. Default is False.
  --decrease_intergenic DECREASE_INTERGENIC, -di DECREASE_INTERGENIC
                        If the length of intergenic potential sRNA is longer
                        than the max_length, it will check the sRNA candidates
                        based on coverage. If (the lowest coverage / the
                        highest coverage) is smaller than this number, it will
                        consider that the spot of lowest coverage as end
                        point. If new length is suitable for a sRNA candidate,
                        this candiate is included as a sRNA. Default is 0.1.
  --decrease_utr DECREASE_UTR, -du DECREASE_UTR
                        It is similar with --decrease_intergenic. This is for
                        UTR-derived sRNAs. Default is 0.05.
  --fuzzy_intergenic FUZZY_INTERGENIC, -fi FUZZY_INTERGENIC
                        If the situation is like --decrease_intergenic
                        mentioned, This is a fuzzy value between the end of
                        sRNA. Default is 10.
  --fuzzy_utr FUZZY_UTR, -fu FUZZY_UTR
                        It is simliar with --fuzzy_intergenic. This is for
                        UTR-derived sRNAs. Default is 10.
  --cutoff_nr_hit CUTOFF_NR_HIT, -cn CUTOFF_NR_HIT
                        The cutoff of hits number in nr database. If the
                        number of nr hits more than this cutoff, it will be
                        excluded. Default is 0.
  --blast_e_nr BLAST_E_NR, -en BLAST_E_NR
                        The cutoff of blast e value for nr alignment. Default
                        is 0.0001.
  --blast_e_srna BLAST_E_SRNA, -es BLAST_E_SRNA
                        The cutoff of blast e value for sRNA alignment.
                        Default is 0.0001.
  --sORF SORF, -O SORF  If comparing sORF and sRNA is needed, please assign
                        the path of sORF gff folder.
  --best_with_all_sRNAhit, -ba
                        The sRNA candidates which have the homology in sRNA
                        database can be included in best results without
                        fitting other information (ex. TSS, blast in nr...) if
                        this parameter is True. Or it will just select the
                        best candidates based on all filter conditions.
                        Default is False.
  --best_without_sORF_candidate, -bs
                        It is for generating the best sRNA candidates without
                        including the sRNA candidates which are overlap with
                        sORFs.Default is False.
  --best_with_terminator, -bt
                        It is for generating the best sRNA candidates which
                        must be associated with terminator. It also includes
                        the sRNA which ends with processing site. Default is
                        False.
  --best_with_promoter, -bp
                        It is for generating the best sRNA candidates which is
                        must be associated with promoter.Default is False.
  --detect_sRNA_in_CDS, -ds
                        It is for searching sRNA in CDS (ex: the genome
                        annotation is not correct). It may find more sRNA
                        candidates which overlap with CDS. Default is False.
  --overlap_percent_CDS OVERLAP_PERCENT_CDS, -oc OVERLAP_PERCENT_CDS
                        If --detect_sRNA_in_CDS is True, please assign the
                        cutoff of the ratio of overlap between CDS and sRNA
                        candidates. Default is 0.5
  --ignore_hypothetical_protein, -ih
                        For ignoring hypothetical protein in genome annotation
                        file. Default is False.
  --ranking_time_promoter RANKING_TIME_PROMOTER, -rp RANKING_TIME_PROMOTER
                        If --promoter_table is provided, it will also use for
                        ranking sRNA candidates. The ranking will base on
                        --ranking_time_promoter * average coverage. For
                        example, one candidates which average coverage is 10,
                        associated with promoter and --ranking_time_promoter
                        is 2, the score for ranking will be 20 (2*10). The
                        candidates which are not associated with promoters,
                        the --ranking_time_promoter is 1. Default is 2. This
                        number can not be smaller than 1.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sRNA</span></code>.</p>
<p><code class="docutils literal"><span class="pre">sRNA_2d_$STRAIN_NAME</span></code>: The secondary structure of all sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">sRNA_seq_$STRAIN_NAME</span></code>: The sequence of all sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">blast_result_and_misc</span></code>: The results of blast.</p>
<p><code class="docutils literal"><span class="pre">mountain_plot</span></code>: The mountain plots of sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">sec_structure</span></code>: The dot plots and secondary structure plots of sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files. <code class="docutils literal"><span class="pre">stat_$STRAIN_NAME_sRNA_blast.csv</span></code> is the results of analysis of blast sRNA databases.
<code class="docutils literal"><span class="pre">stat_sRNA_class_Staphylococcus_aureus_HG003.csv</span></code> is the results of classification of sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: sRNA tables with more details. It also includes the ranking of sRNA candidates.
<code class="docutils literal"><span class="pre">for</span> <span class="pre">class</span></code> is for different classes of sRNAs.
<code class="docutils literal"><span class="pre">best</span></code> is the best results of sRNAs after filtering. <code class="docutils literal"><span class="pre">all_candidates</span></code> is for all candidates without filtering.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of sRNAs. The meanings of <code class="docutils literal"><span class="pre">for</span> <span class="pre">class</span></code>, <code class="docutils literal"><span class="pre">best</span></code>, <code class="docutils literal"><span class="pre">all_candidates</span></code> are the same as <code class="docutils literal"><span class="pre">tables</span></code>.</p>
<p>The useful tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">sRNA_type</span></code>: The sRNA is from 5&#8217;UTR, 3&#8217;UTR, interCDS, intergenic, antisense or within CDS.</p>
<p><code class="docutils literal"><span class="pre">with_TSS</span></code>: Which TSSs are related to this sRNA. &#8220;NA&#8221; means the sRNA is not related to any TSSs.</p>
<p><code class="docutils literal"><span class="pre">sORF</span></code>: Which sORFs are overlap with this sRNA.</p>
<p><code class="docutils literal"><span class="pre">sRNA_hit</span></code>: The blast hit of sRNA database.</p>
<p><code class="docutils literal"><span class="pre">nr_hit</span></code>: The blast hit of nr database.</p>
<p><code class="docutils literal"><span class="pre">2d_energy</span></code>: The normalized (by the length of sRNA) free energy change of secondary structure of sRNA candidate.</p>
<p><code class="docutils literal"><span class="pre">with_term</span></code>: The terminators which are associated with the sRNA candidate.</p>
<p>If you assigned <code class="docutils literal"><span class="pre">--TSS_source</span></code> for sRNA prediction, <code class="docutils literal"><span class="pre">TSS_class</span></code> will be generated and store the gff files of TSSs.</p>
<p><code class="docutils literal"><span class="pre">promoter</span></code>: The promoters which are associated with this sRNA candidate.</p>
<p><code class="docutils literal"><span class="pre">overlap_cds</span></code>: The CDSs which are overlap with sRNA.</p>
<p><code class="docutils literal"><span class="pre">overlap_percent</span></code>: If there are CDSs overlap with sRNA, it will shows the percentage of overlap.</p>
</div>
<div class="section" id="sorf">
<h2>sorf<a class="headerlink" href="#sorf" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">sorf</span></code> can detect sORF based on searching ribosome binding sites, start codons and stop codons within the non-annotated transcripts.
Since non-annotated region may be sRNAs or sORFs, it also provides the function to compare sORFs and sRNAs.
If there are some sORFs overlaped, it will merge them together. Therefore, one region may contain more than one sORF.
BE CAREFUL, The position of start codon is assigned to the first nucleotide. The position of stop codon is the last nucleotide.
<code class="docutils literal"><span class="pre">sorf</span></code> provides the region which covers all possible sORFs. Thus, the region may contain different frame shift.
Ex: (200, 202, 203) are the positions of three start codons and (241, 243) are two stop codons in
a small transcript. Therefore, there are three possible ORFs(200-241, 203-241 and 202-243).
Please be aware this point for using the results.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p>The gff files of genome annotation and transcripts.</p>
<p>The libraries and wiggle files, Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>The fasta files of genome sequence for detection of ribosome binding sites, start codons and stop codons.</p>
<p>Some useful information can be used to improve the prediction:</p>
<p>gff files of TSSs for checking the sORFs start from TSS or not.</p>
<p>gff files of sRNAs for checking the overlap of sRNAs and sORFs.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic sorf [-h] [--UTR_derived_sORF]
                      [--transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER]
                      [--annotation_folder ANNOTATION_FOLDER]
                      [--TSS_folder TSS_FOLDER] [--utr_length UTR_LENGTH]
                      [--min_length MIN_LENGTH] [--max_length MAX_LENGTH]
                      [--tex_wig_folder TEX_WIG_FOLDER]
                      [--frag_wig_folder FRAG_WIG_FOLDER]
                      [--cutoff_intergenic_coverage CUTOFF_INTERGENIC_COVERAGE]
                      [--cutoff_antisense_coverage CUTOFF_ANTISENSE_COVERAGE]
                      [--cutoff_5utr_coverage CUTOFF_5UTR_COVERAGE]
                      [--cutoff_3utr_coverage CUTOFF_3UTR_COVERAGE]
                      [--cutoff_interCDS_coverage CUTOFF_INTERCDS_COVERAGE]
                      [--cutoff_background CUTOFF_BACKGROUND]
                      [--fasta_folder FASTA_FOLDER]
                      [--tex_notex_libs TEX_NOTEX_LIBS]
                      [--frag_libs FRAG_LIBS] [--tex_notex TEX_NOTEX]
                      [-rt REPLICATES_TEX] [--replicates_frag REPLICATES_FRAG]
                      [--table_best] [--sRNA_folder SRNA_FOLDER]
                      [--start_codon START_CODON] [--stop_codon STOP_CODON]
                      [--min_rbs_distance MIN_RBS_DISTANCE]
                      [--max_rbs_distance MAX_RBS_DISTANCE]
                      [--rbs_not_after_TSS] [--fuzzy_rbs FUZZY_RBS]
                      [--print_all_combination] [--best_no_sRNA]
                      [--best_no_TSS]
                      [--ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN]
                      [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --UTR_derived_sORF, -u
                        It is for detecting UTR-derived sORF. Default is
                        False.
  --transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER, -a TRANSCRIPT_ASSEMBLY_FOLDER
                        The path of transcriptome assembly folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        If TSS information is needed, please assign the path
                        of gff folder of TSS.
  --utr_length UTR_LENGTH, -ul UTR_LENGTH
                        If TSS information is needed, please assign the utr
                        length for comparing TSS and sORF. The default number
                        is 300.
  --min_length MIN_LENGTH, -lm MIN_LENGTH
                        Please assign the minimum residue length of sORF.
                        Default is 30.
  --max_length MAX_LENGTH, -lM MAX_LENGTH
                        Please assign the maximum residue length of sORF.
                        Default is 150.
  --tex_wig_folder TEX_WIG_FOLDER, -tw TEX_WIG_FOLDER
                        The path of TEX+/- wig folder.
  --frag_wig_folder FRAG_WIG_FOLDER, -fw FRAG_WIG_FOLDER
                        The path of fragment wig folder.
  --cutoff_intergenic_coverage CUTOFF_INTERGENIC_COVERAGE, -ci CUTOFF_INTERGENIC_COVERAGE
                        The cutoff of minimum coverage of intergenic sORF
                        candidates.
  --cutoff_antisense_coverage CUTOFF_ANTISENSE_COVERAGE, -ai CUTOFF_ANTISENSE_COVERAGE
                        The cutoff of minimum coverage of antisense sORF
                        candidates.
  --cutoff_5utr_coverage CUTOFF_5UTR_COVERAGE, -cu5 CUTOFF_5UTR_COVERAGE
                        The cutoff of minimum coverage of 5&#39;UTR derived sORF
                        candidates. It can be assigned by percentage or the
                        amount of reads. p_0.05 means the coverage of sORF
                        candidates should higher than 5 percentile of all
                        5&#39;UTR transcripts. n_10 means the coverage of sORF
                        candidates should be 10. Default is p_0.5.
  --cutoff_3utr_coverage CUTOFF_3UTR_COVERAGE, -cu3 CUTOFF_3UTR_COVERAGE
                        The meaning is the same as --cutoff_5utr_coverage.
                        This is for 3&#39;UTR. Default is p_0.5.
  --cutoff_interCDS_coverage CUTOFF_INTERCDS_COVERAGE, -cuf CUTOFF_INTERCDS_COVERAGE
                        The meaning is the same as --cutoff_5utr_coverage.
                        This is for interCDS. Default is p_0.5.
  --cutoff_background CUTOFF_BACKGROUND, -cub CUTOFF_BACKGROUND
                        The cutoff of minimum coverage of all sORF candidates.
                        Default is 10.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        The folder of genome fasta file.
  --tex_notex_libs TEX_NOTEX_LIBS, -tl TEX_NOTEX_LIBS
                        Library name of TEX+/- library. The format is:
                        wig_file_name:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). For multiple
                        wig files, please use comma to separate the wig files.
                        For example, wig1:tex:1:a:+,wig2:tex:1:a:-.
  --frag_libs FRAG_LIBS, -fl FRAG_LIBS
                        Library name of fragmented library The format is: wig_
                        file_name:fragmented(frag):condition_id(integer):repli
                        cate_id(alphabet):strand(+ or -). For multiple wig
                        files, please use comma to separate the wig files. For
                        example, wig1:frag:1:a:+,wig2:frag:1:a:-.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        For TEX+/- library, sORF should be detected in both
                        (TEX+ and TEX-) or can be detected in only one library
                        (TEX+ or TEX-). Please assign 1 or 2. Default is 2.
  -rt REPLICATES_TEX, --replicates_tex REPLICATES_TEX
                        The sORF should be detected at least in the number of
                        the replicates. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATED. For assigning
                        different --replicates_tex to different conditions,
                        please use comma to separate it. For example,
                        1_2,2_2,3_3 means number 1 and 2 condition assign 2 to
                        --replicate_tex, and number 3 condition assign 3 to
                        --replcate_tex. For assigning the same
                        --replicates_tex to all conditions, just assign like
                        all_1 (all condition use 1 --replicate_tex).
  --replicates_frag REPLICATES_FRAG, -rf REPLICATES_FRAG
                        The meaning and input type is the same as
                        --replicates_tex.
  --table_best, -tb     The output table of sORF candidates only includes the
                        best library. Default is False.
  --sRNA_folder SRNA_FOLDER, -s SRNA_FOLDER
                        If comparing sORF and sRNA is needed, please assign
                        the path of sORF gff folder.
  --start_codon START_CODON, -ac START_CODON
                        The types of start coden. For assigning multiple types
                        of start codon, please use comma to separate them.
                        Default is ATG.
  --stop_codon STOP_CODON, -oc STOP_CODON
                        The types of stop coden. For assigning multiple types
                        of stop codon, please use comma to separate them.
                        Default is TTA,TAG,TGA.
  --min_rbs_distance MIN_RBS_DISTANCE, -mr MIN_RBS_DISTANCE
                        The minimum distance between the ribosome binding site
                        and start codon. Default is 3.
  --max_rbs_distance MAX_RBS_DISTANCE, -Mr MAX_RBS_DISTANCE
                        The maximum distance between the ribosome binding site
                        and start codon. Default is 15.
  --rbs_not_after_TSS, -at
                        This function can generate best results which also
                        include ribosome binding site not after TSS, Default
                        is False.
  --fuzzy_rbs FUZZY_RBS, -zr FUZZY_RBS
                        The number of nucleotides of ribosome binding site can
                        be different with AGGAGG? Default is 2.
  --print_all_combination, -pa
                        Non-annotated transcript may has many start codons and
                        stop codons. This function can print all combinations
                        of start codons and stop codons. Default is False.
  --best_no_sRNA, -bs   This function can generate best results which excluded
                        the candidate overlap with sRNA, please turn it on.
                        Default is False.
  --best_no_TSS, -bt    This function can generate best results without
                        referring to TSS. Default is False.
  --ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN, -ih IGNORE_HYPOTHETICAL_PROTEIN
                        For ignoring hypothetical protein in genome annotation
                        file. Default is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sORF</span></code>.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of sORFs with more details. <code class="docutils literal"><span class="pre">all_candidates</span></code> is for all sORF candidates without filtering.
<code class="docutils literal"><span class="pre">best</span></code> is for the best sORF candidates with filtering.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of sORFs. The meanings of <code class="docutils literal"><span class="pre">all_candidates</span></code> and <code class="docutils literal"><span class="pre">best</span></code> are the same as <code class="docutils literal"><span class="pre">tables</span></code>.</p>
<p>The tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">start_TSS</span></code>: The starting TSS of this sORF.</p>
<p><code class="docutils literal"><span class="pre">with_TSS</span></code>: Which TSSs are associated with this sORFs.</p>
<p><code class="docutils literal"><span class="pre">sORF_type</span></code>: The type of the sORF (5&#8217;UTR, 3&#8217;UTR, interCDS, intergenic, antisense or within CDS).</p>
<p><code class="docutils literal"><span class="pre">sRNA</span></code>: Which sRNAs are overlap with this sORFs.</p>
<p><code class="docutils literal"><span class="pre">rbs</span></code>: The ribosome binding sites of this sORFs.</p>
<p><code class="docutils literal"><span class="pre">frame_shift</span></code>: The number of frame shifts in the regions.</p>
</div>
<div class="section" id="promoter">
<h2>promoter<a class="headerlink" href="#promoter" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">promoter</span></code> can scan the upstream of TSSs to discover the promoter motifs.
the regions of upstream TSSs for extraction can be assigned. We integrate
<a class="reference external" href="http://meme-suite.org/tools/meme">MEME</a> to compute the promoters.
Visulization HTML files will be generated. If the gff files of TSSs is not computed by
ANNOgesic, please use &#8211;TSS_source. <code class="docutils literal"><span class="pre">promoter</span></code> will classify the TSSs for computing
promoter motifs.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://meme-suite.org/tools/meme">MEME</a>.</p>
<p><a class="reference external" href="https://http://www.mpich.org/">MPICH</a> (if parallel runs are required)</p>
<p>Fasta files of genome sequence.</p>
<p>Gff files of genome annotation.</p>
<p>Gff files of TSSs.</p>
<p>If the gff files of TSSs is not computed by ANNOgesic, the libraries and wiggle files are necessary.
Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code> in order to assign the correct format.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic promoter [-h] [--MEME_path MEME_PATH]
                          [--fasta_folder FASTA_FOLDER]
                          [--TSS_folder TSS_FOLDER] [--num_motif NUM_MOTIF]
                          [--nt_before_TSS NT_BEFORE_TSS] [--e_value E_VALUE]
                          [--motif_width MOTIF_WIDTH] [--parallel PARALLEL]
                          [--TSS_source] [--tex_libs TEX_LIBS]
                          [--tex_wig_path TEX_WIG_PATH]
                          [--annotation_folder ANNOTATION_FOLDER]
                          [--combine_all]
                          [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --MEME_path MEME_PATH
                        path of MEME.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        Please assign the folder of genome fasta file.
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        The folder of TSS gff file.
  --num_motif NUM_MOTIF, -n NUM_MOTIF
                        The number of motifs. Default is 10.
  --nt_before_TSS NT_BEFORE_TSS, -b NT_BEFORE_TSS
                        The number of upstream nucleotides of TSS for promoter
                        prediction? Default is 50.
  --e_value E_VALUE, -e E_VALUE
                        The cutoff of e value. Default is 0.05.
  --motif_width MOTIF_WIDTH, -w MOTIF_WIDTH
                        MEME will try to search motifs based on this length.
                        For detecting a range of width, please insert &quot;-&quot;
                        between two values. Moreover, for computing more than
                        one motif length, please use comma to separate them.
                        for example, 50,2-10. It means the width for detection
                        is 50 and within 2 to 10. The number should be less or
                        equal than --nt_before_TSS. Default is 50.
  --parallel PARALLEL, -pl PARALLEL
                        It is for the parallel running promoter. Please input
                        the number of parallel runs.
  --TSS_source, -s      It is for the TSS gff file which is generated from
                        other metho. It can classify TSS and generate the
                        proper format for promoter detection. Default is True
                        (from ANNOgesic)
  --tex_libs TEX_LIBS, -tl TEX_LIBS
                        Library name of TEX+/- library. If --TSS_source is
                        False, please must assign it.The format is:
                        wig_file_name:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). For multiple
                        wig files, please use comma to separate the wig files.
                        For example, wig1:tex:1:a:+,wig2:tex:1:a:-.
  --tex_wig_path TEX_WIG_PATH, -tw TEX_WIG_PATH
                        The path of TEX+/- wig folder. If --TSS_source is
                        False, please must assign it.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder. If
                        --TSS_source is False, please must assign it..
  --combine_all, -c     It will combine all TSS files in &quot;TSS_folder&quot; to
                        generate global promoter motifs. Default is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/promoter_analysis</span></code>.</p>
<p><code class="docutils literal"><span class="pre">allfasta</span></code>: If <code class="docutils literal"><span class="pre">--combine_all</span></code> is True, it will combine all TSS files in <code class="docutils literal"><span class="pre">--TSS_folder</span></code>
to generate promoter motifs. The results will be stored in this folder.</p>
<p><code class="docutils literal"><span class="pre">fasta_class</span></code>: The fasta files of different types of TSSs.</p>
<p>Each strain will generate one folder for storing the output of promoter motifs.
The format of sub-folder is <code class="docutils literal"><span class="pre">promoter_motifs_$FILENAME_$STRAINNAME_$TSSTYPE_$PROMOTERLEGNTH</span></code>.
Each sub-folder stores PNG files and HTML files of promoters. It also contains a table (<code class="docutils literal"><span class="pre">meme.csv</span></code>) of
prometer which can be used to sRNA detection (please refer to <code class="docutils literal"><span class="pre">srna</span></code>).</p>
<p>If the TSSs are not computed by ANNOgesic, <code class="docutils literal"><span class="pre">TSS_class</span></code> will be generated. It will classify the
TSSs and store as gff files.</p>
</div>
<div class="section" id="operon">
<h2>operon<a class="headerlink" href="#operon" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">operon</span></code> will group TSSs, genes/CDSs/tRNAs/rRNAs, transcripts, terminators and UTRs to operons and
sub-operons.</p>
<ul class="simple">
<li>Pre-required tools or files</li>
</ul>
<p>Gff files of TSSs, annotations, transcripts, 5&#8217;UTRs, and 3&#8217;UTRs.</p>
<p><code class="docutils literal"><span class="pre">operon</span></code> can integrate terminators as well, but it is not necessary.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic operon [-h] [--TSS_folder TSS_FOLDER]
                        [--annotation_folder ANNOTATION_FOLDER]
                        [--transcript_folder TRANSCRIPT_FOLDER]
                        [--UTR5_folder UTR5_FOLDER]
                        [--UTR3_folder UTR3_FOLDER]
                        [--term_folder TERM_FOLDER] [--TSS_fuzzy TSS_FUZZY]
                        [--term_fuzzy TERM_FUZZY] [--min_length MIN_LENGTH]
                        [--statistics] [--combine_gff]
                        [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        The path of TSS gff folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --transcript_folder TRANSCRIPT_FOLDER, -a TRANSCRIPT_FOLDER
                        The path of transcript gff folder.
  --UTR5_folder UTR5_FOLDER, -u5 UTR5_FOLDER
                        The path of 5&#39;UTR gff folder.
  --UTR3_folder UTR3_FOLDER, -u3 UTR3_FOLDER
                        The path of 3&#39;UTR gff folder.
  --term_folder TERM_FOLDER, -e TERM_FOLDER
                        If the information of terminator is needed, please
                        assign the path of terminator gff folder.
  --TSS_fuzzy TSS_FUZZY, -tf TSS_FUZZY
                        The fuzzy for comparing between TSS and transcript
                        assembly. Default is 5.
  --term_fuzzy TERM_FUZZY, -ef TERM_FUZZY
                        The fuzzy for comparing bewteen terminator and
                        transcript assembly. Default is 30.
  --min_length MIN_LENGTH, -l MIN_LENGTH
                        The minimum length of operon. Default is 20.
  --statistics, -s      Doing statistics for operon analysis. Default is
                        False. The name of statistics file is -
                        stat_operon_$STRAIN_NAME.csv.
  --combine_gff, -c     Convert all assigned features to one gff file. Default
                        is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/operon</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files which integrate the information of TSSs, annotations,
transcripts, 5&#8217;UTRs, and 3&#8217;UTRs and assign parent transcript to all features (presented by
<code class="docutils literal"><span class="pre">Parent</span></code>).</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of operons which store all information of operons and sub-operons.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
</div>
<div class="section" id="circrna">
<h2>circrna<a class="headerlink" href="#circrna" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">circrna</span></code> can detect the potential circular RNAs via <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>.
Moreover, it remove the false positive by checking genome annotation files and quality of splicing site detection.
User can assign reads for mapping and detecting circular RNAs or assign alignment files to skip mapping.
But BE CAREFUL, the alignment files must be mapped by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>
with <code class="docutils literal"><span class="pre">-S</span></code> or <code class="docutils literal"><span class="pre">circrna</span></code> can&#8217;t find the proper candidates.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>.</p>
<p>Fasta files of reads or alignment files (BAM or SAM file). If you want to use alignment files directly, they should be
mapped by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> with <code class="docutils literal"><span class="pre">-S</span></code>.</p>
<p>Fasta files and gff files of genome.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">circrna</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">segemehl_folder</span> <span class="n">SEGEMEHL_FOLDER</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">samtools_path</span> <span class="n">SAMTOOLS_PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">align</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">tex_bam_path</span> <span class="n">TEX_BAM_PATH</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">fragmented_bam_path</span> <span class="n">FRAGMENTED_BAM_PATH</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">read_path</span> <span class="n">READ_PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">process</span> <span class="n">PROCESS</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">annotation_path</span> <span class="n">ANNOTATION_PATH</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">support_reads</span> <span class="n">SUPPORT_READS</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">start_ratio</span> <span class="n">START_RATIO</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">end_ratio</span> <span class="n">END_RATIO</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">ignore_hypothetical_protein</span> <span class="n">IGNORE_HYPOTHETICAL_PROTEIN</span><span class="p">]</span>
                         <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">segemehl_folder</span> <span class="n">SEGEMEHL_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">sg</span> <span class="n">SEGEMEHL_FOLDER</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">segemehl</span><span class="o">.</span>
  <span class="o">--</span><span class="n">samtools_path</span> <span class="n">SAMTOOLS_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">st</span> <span class="n">SAMTOOLS_PATH</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">samtools</span><span class="o">.</span>
  <span class="o">--</span><span class="n">align</span><span class="p">,</span> <span class="o">-</span><span class="n">a</span>           <span class="n">Using</span> <span class="n">segemehl</span> <span class="n">to</span> <span class="nb">map</span> <span class="n">reads</span> <span class="p">(</span><span class="n">included</span> <span class="n">splice</span>
                        <span class="n">detection</span><span class="p">)</span><span class="o">.</span> <span class="n">If</span> <span class="n">you</span> <span class="n">already</span> <span class="n">usd</span> <span class="n">segemehl</span> <span class="k">with</span> <span class="o">-</span><span class="n">S</span> <span class="n">to</span> <span class="nb">map</span>
                        <span class="n">your</span> <span class="n">reads</span><span class="p">,</span> <span class="n">you</span> <span class="n">can</span> <span class="n">skip</span> <span class="n">this</span> <span class="n">step</span><span class="o">.</span> <span class="n">Please</span> <span class="n">be</span>
                        <span class="n">attention</span><span class="p">,</span> <span class="n">it</span> <span class="n">only</span> <span class="n">use</span> <span class="n">default</span> <span class="n">parameters</span> <span class="n">of</span> <span class="n">segemehl</span>
                        <span class="n">to</span> <span class="nb">map</span> <span class="n">the</span> <span class="n">reads</span><span class="o">.</span> <span class="n">Moreover</span><span class="p">,</span> <span class="n">it</span> <span class="n">will</span> <span class="nb">map</span> <span class="nb">all</span> <span class="n">read</span> <span class="n">files</span>
                        <span class="ow">in</span> <span class="n">ANNOgesic</span><span class="o">/</span><span class="nb">input</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span> <span class="n">For</span> <span class="n">running</span> <span class="n">some</span> <span class="n">specific</span>
                        <span class="n">functions</span> <span class="n">of</span> <span class="n">segemehl</span><span class="p">,</span> <span class="n">please</span> <span class="n">directly</span> <span class="n">run</span> <span class="n">segemehl</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">tex_bam_path</span> <span class="n">TEX_BAM_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">tb</span> <span class="n">TEX_BAM_PATH</span>
                        <span class="n">If</span> <span class="n">Bam</span> <span class="n">files</span> <span class="n">can</span> <span class="n">be</span> <span class="n">provided</span><span class="p">,</span> <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">TEX</span><span class="o">+/-</span>
                        <span class="n">Bam</span> <span class="n">path</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fragmented_bam_path</span> <span class="n">FRAGMENTED_BAM_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">fb</span> <span class="n">FRAGMENTED_BAM_PATH</span>
                        <span class="n">If</span> <span class="n">Bam</span> <span class="n">files</span> <span class="n">can</span> <span class="n">be</span> <span class="n">provided</span><span class="p">,</span> <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span>
                        <span class="n">fragmented</span> <span class="n">Bam</span> <span class="n">path</span><span class="o">.</span>
  <span class="o">--</span><span class="n">read_path</span> <span class="n">READ_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">rp</span> <span class="n">READ_PATH</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">align</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">reads</span><span class="o">.</span>
  <span class="o">--</span><span class="n">process</span> <span class="n">PROCESS</span><span class="p">,</span> <span class="o">-</span><span class="n">p</span> <span class="n">PROCESS</span>
                        <span class="n">The</span> <span class="n">number</span> <span class="n">of</span> <span class="n">parallels</span> <span class="n">processes</span> <span class="k">for</span> <span class="o">--</span><span class="n">align</span><span class="o">.</span> <span class="n">Default</span>
                        <span class="ow">is</span> <span class="mf">10.</span>
  <span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA_PATH</span>
                        <span class="n">The</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">fasta</span><span class="o">.</span>
  <span class="o">--</span><span class="n">annotation_path</span> <span class="n">ANNOTATION_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_PATH</span>
                        <span class="n">The</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">support_reads</span> <span class="n">SUPPORT_READS</span><span class="p">,</span> <span class="o">-</span><span class="n">s</span> <span class="n">SUPPORT_READS</span>
                        <span class="n">The</span> <span class="n">cutoff</span> <span class="n">of</span> <span class="n">supported</span> <span class="n">reads</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">10.</span>
  <span class="o">--</span><span class="n">start_ratio</span> <span class="n">START_RATIO</span><span class="p">,</span> <span class="o">-</span><span class="n">sr</span> <span class="n">START_RATIO</span>
                        <span class="n">The</span> <span class="n">ratio</span> <span class="n">of</span> <span class="p">(</span><span class="n">read</span> <span class="n">support</span> <span class="n">circ</span> <span class="o">/</span> <span class="nb">all</span> <span class="n">read</span><span class="p">)</span> <span class="n">at</span>
                        <span class="n">starting</span> <span class="n">point</span><span class="o">.</span> <span class="n">The</span> <span class="n">ratio</span> <span class="n">of</span> <span class="n">candidates</span> <span class="n">should</span> <span class="n">higher</span>
                        <span class="n">than</span> <span class="n">this</span> <span class="n">cutoff</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">0.5</span><span class="o">.</span>
  <span class="o">--</span><span class="n">end_ratio</span> <span class="n">END_RATIO</span><span class="p">,</span> <span class="o">-</span><span class="n">er</span> <span class="n">END_RATIO</span>
                        <span class="n">The</span> <span class="n">ratio</span> <span class="n">of</span> <span class="p">(</span><span class="n">read</span> <span class="n">support</span> <span class="n">circ</span> <span class="o">/</span> <span class="nb">all</span> <span class="n">read</span><span class="p">)</span> <span class="n">at</span> <span class="n">end</span>
                        <span class="n">point</span><span class="o">.</span> <span class="n">The</span> <span class="n">ratio</span> <span class="n">of</span> <span class="n">candidates</span> <span class="n">should</span> <span class="n">higher</span> <span class="n">than</span> <span class="n">this</span>
                        <span class="n">cutoff</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">0.5</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ignore_hypothetical_protein</span> <span class="n">IGNORE_HYPOTHETICAL_PROTEIN</span><span class="p">,</span> <span class="o">-</span><span class="n">ih</span> <span class="n">IGNORE_HYPOTHETICAL_PROTEIN</span>
                        <span class="n">For</span> <span class="n">ignoring</span> <span class="n">hypothetical</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">genome</span> <span class="n">annotation</span>
                        <span class="n">file</span><span class="p">,</span> <span class="n">please</span> <span class="n">turn</span> <span class="n">it</span> <span class="n">on</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All the output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/circRNA</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of circular RNAs. <code class="docutils literal"><span class="pre">$STRAINNAME_best.gff</span></code> is the gff files for best result after comparing
with genome annotation and quality of splicing. <code class="docutils literal"><span class="pre">$STRAINNAME_all.gff</span></code> is for all candidates without filering.</p>
<p><code class="docutils literal"><span class="pre">circRNA_tables</span></code>: The tables for circular RNAs with more details.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p><code class="docutils literal"><span class="pre">segemehl_align</span></code>: If <code class="docutils literal"><span class="pre">circrna</span></code> starts from read mapping, the folder is for results of mapping.</p>
<p><code class="docutils literal"><span class="pre">segemehl_splice</span></code>: The results of splicing detection. The information of the splicing tables, please
refer to <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>.</p>
</div>
<div class="section" id="go-term">
<h2>go_term<a class="headerlink" href="#go-term" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">go_term</span></code> can retreive the information of Gene Ontology from Uniprot.
It also provides some analyses of Go terms.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.uniprot.org/downloads">idmapping_selected.tab from Uniprot</a>.</p>
<p><a class="reference external" href="http://geneontology.org/page/go-slim-and-subset-guide">goslim.obo</a>.</p>
<p><a class="reference external" href="http://geneontology.org/page/download-ontology">go.obo</a>.</p>
<p>Gff files of genome annotation.</p>
<p>For detecting based on the expressed CDSs, the transcript gff file is required.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">go_term</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">annotation_path</span> <span class="n">ANNOTATION_PATH</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">UniProt_id</span> <span class="n">UNIPROT_ID</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">go_obo</span> <span class="n">GO_OBO</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">goslim_obo</span> <span class="n">GOSLIM_OBO</span><span class="p">]</span>
                         <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">annotation_path</span> <span class="n">ANNOTATION_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_PATH</span>
                        <span class="n">If</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">folder</span> <span class="ow">is</span> <span class="n">provided</span><span class="p">,</span> <span class="n">it</span>
                        <span class="n">can</span> <span class="n">retrieve</span> <span class="n">GO</span> <span class="n">term</span> <span class="n">based</span> <span class="n">on</span> <span class="n">expressed</span> <span class="n">CDS</span> <span class="ow">and</span> <span class="nb">all</span>
                        <span class="n">CDS</span><span class="o">.</span>
  <span class="o">--</span><span class="n">UniProt_id</span> <span class="n">UNIPROT_ID</span><span class="p">,</span> <span class="o">-</span><span class="n">u</span> <span class="n">UNIPROT_ID</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">UniProt</span> <span class="n">ID</span> <span class="n">mapping</span> <span class="n">database</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="n">ANNOgesic</span><span class="o">/</span><span class="nb">input</span><span class="o">/</span><span class="n">database</span><span class="o">/</span><span class="n">idmapping_selected</span><span class="o">.</span><span class="n">tab</span><span class="o">.</span>
  <span class="o">--</span><span class="n">go_obo</span> <span class="n">GO_OBO</span><span class="p">,</span> <span class="o">-</span><span class="n">go</span> <span class="n">GO_OBO</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">go</span><span class="o">.</span><span class="n">obo</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="n">ANNOgesic</span><span class="o">/</span><span class="nb">input</span><span class="o">/</span><span class="n">database</span><span class="o">/</span><span class="n">go</span><span class="o">.</span><span class="n">obo</span><span class="o">.</span>
  <span class="o">--</span><span class="n">goslim_obo</span> <span class="n">GOSLIM_OBO</span><span class="p">,</span> <span class="o">-</span><span class="n">gs</span> <span class="n">GOSLIM_OBO</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">goslim</span><span class="o">.</span><span class="n">obo</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="n">ANNOgesic</span><span class="o">/</span><span class="nb">input</span><span class="o">/</span><span class="n">database</span><span class="o">/</span><span class="n">goslim_generic</span><span class="o">.</span><span class="n">obo</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">ANNOgesic/output/Go_term</span></code>.</p>
<p><code class="docutils literal"><span class="pre">Go_term_results</span></code>: The tables of Go terms information.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files and figures.</p>
</div>
<div class="section" id="srna-target">
<h2>srna_target<a class="headerlink" href="#srna-target" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">srna_target</span></code> will search the potential targets of sRNA. Different programs can
be assigned for running (RNAup or RNAplex or both). We recommand running with both
programs. <code class="docutils literal"><span class="pre">srna_target</span></code> can compare the both results and provide the best ones.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a> .</p>
<p>Gff files of genome annotation.</p>
<p>Gff files of sRNAs.</p>
<p>Fasta files of genome.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic srna_target [-h] [--Vienna_folder VIENNA_FOLDER]
                             [--annotation_path ANNOTATION_PATH]
                             [--fasta_path FASTA_PATH] [--sRNA_path SRNA_PATH]
                             [--query_sRNA QUERY_SRNA] [--program PROGRAM]
                             [--interaction_length INTERACTION_LENGTH]
                             [--window_size_target WINDOW_SIZE_TARGET]
                             [--span_target SPAN_TARGET]
                             [--window_size_srna WINDOW_SIZE_SRNA]
                             [--span_srna SPAN_SRNA]
                             [--unstructured_region_RNAplex_target UNSTRUCTURED_REGION_RNAPLEX_TARGET]
                             [--unstructured_region_RNAplex_srna UNSTRUCTURED_REGION_RNAPLEX_SRNA]
                             [--unstructured_region_RNAup UNSTRUCTURED_REGION_RNAUP]
                             [--energy_threshold ENERGY_THRESHOLD]
                             [--duplex_distance DUPLEX_DISTANCE] [--top TOP]
                             [--process_rnaplex PROCESS_RNAPLEX]
                             [--process_rnaup PROCESS_RNAUP]
                             [--continue_rnaup]
                             [--potential_target_start POTENTIAL_TARGET_START]
                             [--potential_target_end POTENTIAL_TARGET_END]
                             [--target_feature TARGET_FEATURE]
                             [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --Vienna_folder VIENNA_FOLDER
                        Please assign the folder of Vienna package. It should
                        include RNAplfold, RNAup and RNAplex.
  --annotation_path ANNOTATION_PATH, -g ANNOTATION_PATH
                        The path of genome annotation gff folder.
  --fasta_path FASTA_PATH, -f FASTA_PATH
                        The path of genome fasta folder.
  --sRNA_path SRNA_PATH, -r SRNA_PATH
                        The path of sRNA gff folder.
  --query_sRNA QUERY_SRNA, -q QUERY_SRNA
                        Please assign the query sRNA. For computing all sRNA
                        in gff file, please keyin &#39;all&#39;.the input format
                        should be like, $STRAIN:$STRAND:$START:$END. For
                        assigning more than one sRNA, please use comma to
                        separate them.For example,
                        NC_007795.1:+:200:534,NC_007795.1:-:6767:6900. Default
                        is all.
  --program PROGRAM, -p PROGRAM
                        Using RNAplex, RNAup or both. Default is both.
  --interaction_length INTERACTION_LENGTH, -i INTERACTION_LENGTH
                        Maximum length of an interaction. Default is 30.
  --window_size_target WINDOW_SIZE_TARGET, -wt WINDOW_SIZE_TARGET
                        Only works when --program is RNAplex or both. Average
                        the pair probabilities over windows of given size for
                        RNAplex. Default is 240.
  --span_target SPAN_TARGET, -st SPAN_TARGET
                        Only works when --program is RNAplex or both. Set the
                        maximum allowed separation of a base pair to span for
                        RNAplex. Default is 160.
  --window_size_srna WINDOW_SIZE_SRNA, -ws WINDOW_SIZE_SRNA
                        Only works when --program is RNAplex or both. Average
                        the pair probabilities over windows of given size for
                        RNAplex. Default is 30.
  --span_srna SPAN_SRNA, -ss SPAN_SRNA
                        Only works when --program is RNAplex or both. Set the
                        maximum allowed separation of a base pair to span for
                        RNAplex. Default is 30.
  --unstructured_region_RNAplex_target UNSTRUCTURED_REGION_RNAPLEX_TARGET, -ut UNSTRUCTURED_REGION_RNAPLEX_TARGET
                        Only works when --program is RNAplex or both. Compute
                        the mean probability that regions of length 1 to a
                        given length are unpaired for RNAplex. Default is 30.
  --unstructured_region_RNAplex_srna UNSTRUCTURED_REGION_RNAPLEX_SRNA, -us UNSTRUCTURED_REGION_RNAPLEX_SRNA
                        Only works when --program is RNAplex or both. Compute
                        the mean probability that regions of length 1 to a
                        given length are unpaired for RNAplex. Default is 30.
  --unstructured_region_RNAup UNSTRUCTURED_REGION_RNAUP, -uu UNSTRUCTURED_REGION_RNAUP
                        Only works when --program is RNAup or both. Compute
                        the mean probability that regions of length 1 to a
                        given length are unpaired for RNAplex. Default is 30.
  --energy_threshold ENERGY_THRESHOLD, -e ENERGY_THRESHOLD
                        Only works when --program is RNAplex or both. Minimum
                        energy for a duplex to be returned for RNAplex.
                        Default is -8.
  --duplex_distance DUPLEX_DISTANCE, -d DUPLEX_DISTANCE
                        Only works when --program is RNAplex or both. Distance
                        between target 3&#39; ends of two consecutive duplexes for
                        RNAplex. Default is 20.
  --top TOP, -t TOP     The output file only includes the candidates which
                        ranking number is smaller or equal than --top. Default
                        is 20.
  --process_rnaplex PROCESS_RNAPLEX, -pp PROCESS_RNAPLEX
                        The amount of parallel processes for running RNAplex
                        prediction. Default is 5.
  --process_rnaup PROCESS_RNAUP, -pu PROCESS_RNAUP
                        The amount of parallel processes for running RNAup
                        prediction. Default is 20.
  --continue_rnaup, -cr
                        The running time of RNAup is long, when numerous sRNAs
                        are assigned. It can run RNAup based on the previous
                        run if the process was crushed. Default is False.
  --potential_target_start POTENTIAL_TARGET_START, -ps POTENTIAL_TARGET_START
                        The number of upstream nucleotides of the start point
                        of --target_feature for extracting as potential
                        target. Default is 200.
  --potential_target_end POTENTIAL_TARGET_END, -pe POTENTIAL_TARGET_END
                        The number of downstream nucleotides of the start
                        point of --target_feature for extracting as potential
                        target. Default is 150.
  --target_feature TARGET_FEATURE, -tf TARGET_FEATURE
                        The features which are applied for extracting as
                        potential targets. If multi-features need to be
                        assigned, just use comma to separate them. Ex:
                        CDS,gene. Default is CDS.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sRNA_targets</span></code>.</p>
<p><code class="docutils literal"><span class="pre">RNAplex</span></code>: All results of RNAplex. <code class="docutils literal"><span class="pre">$STRAIN_RNAplex.txt</span></code> is raw results of RNAplex.
It includes the information of binding situations. <code class="docutils literal"><span class="pre">$STRAIN_RNAplex_rank.csv</span></code> is the results
that sorted by binding energy.</p>
<p><code class="docutils literal"><span class="pre">RNAup</span></code>: All results of RNAup. <code class="docutils literal"><span class="pre">$STRAIN_RNAup.txt</span></code> is raw results of RNAup.
It includes the information of binding situations. <code class="docutils literal"><span class="pre">$STRAIN_RNAup_rank.csv</span></code> is the results
that sorted by binding energy.</p>
<p><code class="docutils literal"><span class="pre">merge</span></code>: The results which merge <code class="docutils literal"><span class="pre">RNAplex</span></code> and <code class="docutils literal"><span class="pre">RNAup</span></code>. <code class="docutils literal"><span class="pre">$STRAIN_merge.csv</span></code> merges all candidates of both programs.
<code class="docutils literal"><span class="pre">$STRAIN_overlap.csv</span></code> lists the results which be top 20 (default) in both methods.</p>
<p><code class="docutils literal"><span class="pre">sRNA_seqs</span></code>: The fasta sequences of sRNAs.</p>
<p><code class="docutils literal"><span class="pre">target_seqs</span></code>: The fasta sequences of potential targets.</p>
</div>
<div class="section" id="ppi-network">
<h2>ppi_network<a class="headerlink" href="#ppi-network" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">ppi_network</span></code> will retrieve the data from <a class="reference external" href="http://string-db.org/">STRING</a>.
Then using <a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a> to search
the literatures to support the protein-protein interaction networks. Therefore,
<code class="docutils literal"><span class="pre">ppi_network</span></code> can generate the protein-protein interaction networks with supported literatures.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://string-db.org/cgi/download.pl">species.vXXXX.txt from STRING</a>.</p>
<p>Ptt files of genome annotation.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic ppi_network [-h] [--gff_path GFF_PATH]
                             [--proteinID_strains PROTEINID_STRAINS]
                             [--without_strain_pubmed]
                             [--species_STRING SPECIES_STRING] [--score SCORE]
                             [--node_size NODE_SIZE] [--query QUERY]
                             [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --gff_path GFF_PATH, -g GFF_PATH
                        The path of genome annotation gff folder. Please
                        confirm the gff files which have proper locus_tag item
                        in attributes. The locus_tag items can be real locus
                        tag like locus_tag=SAOUHSC_00003, or they can be
                        assigned by gene name, such as, locus_tag=dnaA.
  --proteinID_strains PROTEINID_STRAINS, -s PROTEINID_STRAINS
                        This is for query strain and filename. In order to
                        retrieve the data from STRING and Pubmed, The close
                        reference must be assigned. For example, the query
                        strain is Staphylococcus aureus HG003, but there is no
                        Staphylococcus aureus HG003 in STRING database.
                        However, there is Staphylococcus aureus 8325 which is
                        close to Staphylococcus aureus HG003. Therefore, it
                        can be used as reference. The format is like Staphyloc
                        occus_aureus_HG003.gff:Staphylococcus_aureus_HG003:&quot;St
                        aphylococcus aureus 8325&quot;:&quot;Staphylococcus aureus&quot;. or
                        Staphylococcus_aureus_HG003.gff:Staphylococcus_aureus_
                        HG003:&quot;93061&quot;:&quot;Staphylococcus aureus&quot;. or Staphylococc
                        us_aureus_HG003.gff:Staphylococcus_aureus_HG003:&quot;Staph
                        ylococcus aureus NCTC 8325&quot;:&quot;Staphylococcus aureus&quot;. (
                        gff_filename:gff_strain_name:STRING_name:Pubmed_name).
                        First one is the gff file name. Second one is the
                        strain name of gff files. Third one is for STRING
                        database, and the fourth one is for Pubmed. Running
                        multiple strains at the same time is acceptable, just
                        put comma between these strains. Before running it,
                        please check the species table which should be located
                        in ANNOgesic/input/database .If the file was not
                        downloaded before, please download it. taxon_id,
                        STRING_name_compact or official_name_NCBI can be
                        assigned to STRING_name.BE CAREFUL, if the assigned
                        name with spaces, please put &quot;&quot; at two ends. The name
                        of Pubmed can be not so specific. If a specific name
                        is assigned, it may not be able to find the related
                        literatures.
  --without_strain_pubmed, -n
                        Retrieving pubmed without assigning strains, Default
                        is False.
  --species_STRING SPECIES_STRING, -d SPECIES_STRING
                        Please assign the path of species table of STRING.
  --score SCORE, -ps SCORE
                        Please assign the cutoff of text-mining score. The
                        value is from -1 to 1. Default is 0.
  --node_size NODE_SIZE, -ns NODE_SIZE
                        Please assign the size of nodes in figure, default is
                        4000.
  --query QUERY, -q QUERY
                        Please assign the query protein here. The format is
                        $STRAINOFGFF:$START_POINT:$END_POINT:$STRAND.For
                        multiple strains, please use comma to separate them.
                        For example, Staphylococcus_aureus_HG003:345:456:+,Sta
                        phylococcus_aureus_HG003:2000:3211:-. For computing
                        all protein, just type all. Default is all.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All the output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/PPI</span></code>.</p>
<p><code class="docutils literal"><span class="pre">best_results</span></code>: The results which the score of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a>
is high enough.
<code class="docutils literal"><span class="pre">$STRAIN_without_strain.csv</span></code> is the results of searching literatures without specific strain.
<code class="docutils literal"><span class="pre">$STRAIN_with_strain.csv</span></code> is the results of searching literatures with specific strain.
For example, Staphylococcus_aureus_8325_without_strain.csv is search with Staphylococcus aureus;
Staphylococcus_aureus_8325_without_strain.csv is search without Staphylococcus aureus.
<code class="docutils literal"><span class="pre">without_strain</span></code> stores all interaction information which search without specific strain.
<code class="docutils literal"><span class="pre">with_strain</span></code> stores all interaction information which search with specific strain.</p>
<p><code class="docutils literal"><span class="pre">all_results</span></code>: The results of all protein-protein interactions.
(Even the text-mining(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a>) score is too low)</p>
<p><code class="docutils literal"><span class="pre">figures</span></code>: The thickness represents how many literatures can be found for the interactions.
The solid line means there is some literatures which strongly support the interactions. The dash-dot line
means the supported literatures are very weak. The dot line means there is no literatures which can support the
interactions. The color is the best score of the literatures of the interactions.</p>
</div>
<div class="section" id="subcellular-localization">
<h2>subcellular_localization<a class="headerlink" href="#subcellular-localization" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">subcellular</span> <span class="pre">localization</span></code> can predict the subcellular localization of CDSs. It also provides some
statistics and visualization files.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.psort.org/psortb/">Psortb</a>.</p>
<p>Gff files of genome annotation.</p>
<p>Fasta files of genome sequence.</p>
<p>For computing based on the expressed CDSs, the transcript gff file is required.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">subcellular_localization</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">Psortb_path</span> <span class="n">PSORTB_PATH</span><span class="p">]</span>
                                          <span class="p">[</span><span class="o">--</span><span class="n">gff_path</span> <span class="n">GFF_PATH</span><span class="p">]</span>
                                          <span class="p">[</span><span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">]</span>
                                          <span class="p">[</span><span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">]</span>
                                          <span class="p">[</span><span class="o">--</span><span class="n">bacteria_type</span> <span class="n">BACTERIA_TYPE</span><span class="p">]</span>
                                          <span class="p">[</span><span class="o">--</span><span class="n">difference_multi</span> <span class="n">DIFFERENCE_MULTI</span><span class="p">]</span>
                                          <span class="p">[</span><span class="o">--</span><span class="n">merge_to_gff</span><span class="p">]</span>
                                          <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">Psortb_path</span> <span class="n">PSORTB_PATH</span>
                        <span class="n">If</span> <span class="n">you</span> <span class="n">want</span> <span class="n">to</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">Psortb</span><span class="p">,</span> <span class="n">please</span>
                        <span class="n">assign</span> <span class="n">here</span><span class="o">.</span>
  <span class="o">--</span><span class="n">gff_path</span> <span class="n">GFF_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span> <span class="n">GFF_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">fasta</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_PATH</span>
                        <span class="n">If</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">folder</span> <span class="ow">is</span> <span class="n">provided</span><span class="p">,</span> <span class="n">it</span>
                        <span class="n">can</span> <span class="n">get</span> <span class="n">the</span> <span class="n">results</span> <span class="n">based</span> <span class="n">on</span> <span class="n">expressed</span> <span class="n">CDS</span> <span class="ow">and</span> <span class="nb">all</span>
                        <span class="n">CDS</span><span class="o">.</span>
  <span class="o">--</span><span class="n">bacteria_type</span> <span class="n">BACTERIA_TYPE</span><span class="p">,</span> <span class="o">-</span><span class="n">b</span> <span class="n">BACTERIA_TYPE</span>
                        <span class="n">The</span> <span class="nb">type</span> <span class="n">of</span> <span class="n">bacteria</span> <span class="p">(</span><span class="n">Gram</span><span class="o">-</span><span class="n">positive</span> <span class="ow">or</span> <span class="n">Gram</span><span class="o">-</span><span class="n">negative</span><span class="p">)</span><span class="o">.</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="s1">&#39;positive&#39;</span> <span class="ow">or</span> <span class="s1">&#39;negative&#39;</span><span class="o">.</span>
  <span class="o">--</span><span class="n">difference_multi</span> <span class="n">DIFFERENCE_MULTI</span><span class="p">,</span> <span class="o">-</span><span class="n">d</span> <span class="n">DIFFERENCE_MULTI</span>
                        <span class="n">For</span> <span class="n">the</span> <span class="n">protein</span> <span class="n">which</span> <span class="n">have</span> <span class="n">multiple</span> <span class="n">locations</span><span class="p">,</span> <span class="k">if</span> <span class="n">the</span>
                        <span class="n">difference</span> <span class="n">of</span> <span class="n">psorb</span> <span class="n">scores</span> <span class="ow">is</span> <span class="n">smaller</span> <span class="n">than</span>
                        <span class="o">--</span><span class="n">difference_multi</span><span class="p">,</span> <span class="n">it</span> <span class="n">will</span> <span class="n">be</span> <span class="n">printed</span> <span class="n">out</span> <span class="k">as</span> <span class="n">well</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="mf">0.5</span><span class="o">.</span> <span class="n">The</span> <span class="n">maximum</span> <span class="n">value</span> <span class="ow">is</span> <span class="mf">10.</span>
  <span class="o">--</span><span class="n">merge_to_gff</span><span class="p">,</span> <span class="o">-</span><span class="n">m</span>    <span class="n">Merging</span> <span class="n">the</span> <span class="n">information</span> <span class="n">to</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">file</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/subcellular_localization</span></code>.</p>
<p><code class="docutils literal"><span class="pre">psortb_results</span></code>: The results of Psortb.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files and figures.</p>
</div>
<div class="section" id="riboswitch-thermometer">
<h2>riboswitch_thermometer<a class="headerlink" href="#riboswitch-thermometer" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">riboswitch_thermometer</span></code> will search ribosome binding sites in the region between
a TSS（the starting point of transcript was assigned if no TSS was detected)
and a downstream CDSs. Then using <a class="reference external" href="http://infernal.janelia.org/">Infernal</a> to scan
riboswitch or RNA thermometer in <a class="reference external" href="http://rfam.xfam.org/">Rfam</a>.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://infernal.janelia.org/">Infernal</a>.</p>
<p><a class="reference external" href="http://rfam.xfam.org/">Rfam</a>.</p>
<p>Gff files of genome annotation.</p>
<p>Fasta files of genome sequence.</p>
<p>File of <code class="docutils literal"><span class="pre">riboswitch_ID</span></code> or <code class="docutils literal"><span class="pre">RNA_thermometer_ID</span></code>. The file should contain Rfam ID,
Name and Description of riboswitchs or RNA thermometer (refer to
<code class="docutils literal"><span class="pre">Riboswitch</span> <span class="pre">and</span> <span class="pre">RNA</span> <span class="pre">thermometer</span> <span class="pre">dataset</span> <span class="pre">of</span> <span class="pre">Rfam</span></code>). You can download the file from our
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic">Github</a> (Rfam_riboswitch_ID.csv or Rfam_RNA_thermometer_ID.csv).</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">riboswitch_thermometer</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">program</span> <span class="n">PROGRAM</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">infernal_path</span> <span class="n">INFERNAL_PATH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">riboswitch_ID</span> <span class="n">RIBOSWITCH_ID</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">RNA_thermometer_ID</span> <span class="n">RNA_THERMOMETER_ID</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">gff_path</span> <span class="n">GFF_PATH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">tss_path</span> <span class="n">TSS_PATH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">UTR_length</span> <span class="n">UTR_LENGTH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">Rfam</span> <span class="n">RFAM</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">e_value</span> <span class="n">E_VALUE</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">output_all</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">fuzzy</span> <span class="n">FUZZY</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_rbs</span> <span class="n">FUZZY_RBS</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">start_codon</span> <span class="n">START_CODON</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">max_dist_rbs</span> <span class="n">MAX_DIST_RBS</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">min_dist_rbs</span> <span class="n">MIN_DIST_RBS</span><span class="p">]</span>
                                        <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">program</span> <span class="n">PROGRAM</span><span class="p">,</span> <span class="o">-</span><span class="n">p</span> <span class="n">PROGRAM</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">query</span> <span class="n">feature</span><span class="o">.</span> <span class="n">The</span> <span class="n">options</span> <span class="n">are</span>
                        <span class="s2">&quot;riboswitch&quot;</span><span class="p">,</span> <span class="s2">&quot;thermometer&quot;</span><span class="p">,</span> <span class="s2">&quot;both&quot;</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="n">both</span><span class="o">.</span>
  <span class="o">--</span><span class="n">infernal_path</span> <span class="n">INFERNAL_PATH</span><span class="p">,</span> <span class="o">-</span><span class="k">if</span> <span class="n">INFERNAL_PATH</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">folder</span> <span class="n">of</span> <span class="n">Infernal</span> <span class="p">(</span><span class="n">where</span> <span class="ow">is</span> <span class="n">cmscan</span>
                        <span class="ow">and</span> <span class="n">cmsearch</span> <span class="n">located</span><span class="p">)</span><span class="o">.</span>
  <span class="o">--</span><span class="n">riboswitch_ID</span> <span class="n">RIBOSWITCH_ID</span><span class="p">,</span> <span class="o">-</span><span class="n">ri</span> <span class="n">RIBOSWITCH_ID</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">program</span> <span class="ow">is</span> <span class="s2">&quot;riboswitch&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="n">please</span> <span class="n">assigh</span>
                        <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">riboswitch</span> <span class="n">ID</span> <span class="n">of</span> <span class="n">Rfam</span><span class="o">.</span> <span class="n">The</span> <span class="n">file</span> <span class="n">should</span>
                        <span class="n">include</span> <span class="n">the</span> <span class="n">Accession</span><span class="p">,</span> <span class="n">ID</span> <span class="ow">and</span> <span class="n">Description</span> <span class="n">of</span>
                        <span class="n">riboswitch</span> <span class="ow">in</span> <span class="n">Rfam</span><span class="o">.</span>
  <span class="o">--</span><span class="n">RNA_thermometer_ID</span> <span class="n">RNA_THERMOMETER_ID</span><span class="p">,</span> <span class="o">-</span><span class="n">ti</span> <span class="n">RNA_THERMOMETER_ID</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">program</span> <span class="ow">is</span> <span class="s2">&quot;thermometer&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="n">please</span> <span class="n">assigh</span>
                        <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">RNA</span> <span class="n">thermometer</span> <span class="n">ID</span> <span class="n">of</span> <span class="n">Rfam</span><span class="o">.</span> <span class="n">The</span> <span class="n">file</span>
                        <span class="n">should</span> <span class="n">include</span> <span class="n">the</span> <span class="n">Accession</span><span class="p">,</span> <span class="n">ID</span> <span class="ow">and</span> <span class="n">Description</span> <span class="n">of</span>
                        <span class="n">RNA</span> <span class="n">thermometer</span> <span class="ow">in</span> <span class="n">Rfam</span><span class="o">.</span>
  <span class="o">--</span><span class="n">gff_path</span> <span class="n">GFF_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span> <span class="n">GFF_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">tss_path</span> <span class="n">TSS_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">t</span> <span class="n">TSS_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">TSS</span> <span class="n">gff</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">UTR_length</span> <span class="n">UTR_LENGTH</span><span class="p">,</span> <span class="o">-</span><span class="n">u</span> <span class="n">UTR_LENGTH</span>
                        <span class="n">The</span> <span class="n">length</span> <span class="n">of</span> <span class="n">UTR</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">300.</span>
  <span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">fasta</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">Rfam</span> <span class="n">RFAM</span><span class="p">,</span> <span class="o">-</span><span class="n">R</span> <span class="n">RFAM</span>  <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">Rfam</span> <span class="n">CM</span> <span class="n">database</span><span class="o">.</span>
  <span class="o">--</span><span class="n">e_value</span> <span class="n">E_VALUE</span><span class="p">,</span> <span class="o">-</span><span class="n">e</span> <span class="n">E_VALUE</span>
                        <span class="n">The</span> <span class="n">cutoff</span> <span class="n">of</span> <span class="n">e</span> <span class="n">value</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">0.001</span><span class="o">.</span>
  <span class="o">--</span><span class="n">output_all</span><span class="p">,</span> <span class="o">-</span><span class="n">o</span>      <span class="n">One</span> <span class="n">query</span> <span class="n">sequence</span> <span class="n">may</span> <span class="n">fit</span> <span class="n">multiple</span> <span class="n">riboswitches</span> <span class="ow">or</span>
                        <span class="n">RNA</span> <span class="n">thermometers</span><span class="o">.</span> <span class="n">If</span> <span class="o">--</span><span class="n">output_all</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">,</span> <span class="nb">all</span> <span class="n">results</span>
                        <span class="nb">all</span> <span class="n">be</span> <span class="n">printed</span> <span class="n">out</span><span class="o">.</span> <span class="n">Or</span> <span class="n">it</span> <span class="n">will</span> <span class="n">only</span> <span class="nb">print</span> <span class="n">the</span> <span class="n">best</span>
                        <span class="n">one</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fuzzy</span> <span class="n">FUZZY</span><span class="p">,</span> <span class="o">-</span><span class="n">z</span> <span class="n">FUZZY</span>
                        <span class="n">The</span> <span class="n">fuzzy</span> <span class="n">of</span> <span class="n">nucleotides</span> <span class="k">for</span> <span class="mi">3</span><span class="s1">&#39; or 5&#39;</span> <span class="n">end</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="mf">10.</span>
  <span class="o">--</span><span class="n">fuzzy_rbs</span> <span class="n">FUZZY_RBS</span><span class="p">,</span> <span class="o">-</span><span class="n">zr</span> <span class="n">FUZZY_RBS</span>
                        <span class="n">The</span> <span class="n">number</span> <span class="n">of</span> <span class="n">nucleotides</span> <span class="n">of</span> <span class="n">ribosome</span> <span class="n">binding</span> <span class="n">site</span> <span class="n">can</span>
                        <span class="n">be</span> <span class="n">different</span> <span class="k">with</span> <span class="n">AGGAGG</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">2.</span>
  <span class="o">--</span><span class="n">start_codon</span> <span class="n">START_CODON</span><span class="p">,</span> <span class="o">-</span><span class="n">ac</span> <span class="n">START_CODON</span>
                        <span class="n">The</span> <span class="n">types</span> <span class="n">of</span> <span class="n">start</span> <span class="n">coden</span><span class="o">.</span> <span class="n">If</span> <span class="n">more</span> <span class="n">than</span> <span class="n">one</span> <span class="nb">type</span> <span class="n">are</span>
                        <span class="n">assigned</span><span class="p">,</span> <span class="n">please</span> <span class="n">use</span> <span class="n">comma</span> <span class="n">to</span> <span class="n">separate</span> <span class="n">them</span><span class="o">.</span> <span class="n">Default</span>
                        <span class="ow">is</span> <span class="n">ATG</span><span class="o">.</span>
  <span class="o">--</span><span class="n">max_dist_rbs</span> <span class="n">MAX_DIST_RBS</span><span class="p">,</span> <span class="o">-</span><span class="n">Mr</span> <span class="n">MAX_DIST_RBS</span>
                        <span class="n">The</span> <span class="n">maximum</span> <span class="n">distance</span> <span class="n">between</span> <span class="n">ribosome</span> <span class="n">binding</span> <span class="n">site</span> <span class="ow">and</span>
                        <span class="n">start</span> <span class="n">codon</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">14.</span>
  <span class="o">--</span><span class="n">min_dist_rbs</span> <span class="n">MIN_DIST_RBS</span><span class="p">,</span> <span class="o">-</span><span class="n">mr</span> <span class="n">MIN_DIST_RBS</span>
                        <span class="n">The</span> <span class="n">minmum</span> <span class="n">distance</span> <span class="n">between</span> <span class="n">ribosome</span> <span class="n">binding</span> <span class="n">site</span> <span class="ow">and</span>
                        <span class="n">start</span> <span class="n">codon</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">5.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files of riboswitch will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/riboswitch</span></code> and
all output files of RNA thermometer will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/RNA_thermometer</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files of riboswitchs / RNA_thermometer.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of riboswichs / RNA_thermometer with more details.</p>
<p><code class="docutils literal"><span class="pre">scan_Rfam</span></code>: The results of searching to Rfam with <code class="docutils literal"><span class="pre">cmscan</span></code> (<a class="reference external" href="http://infernal.janelia.org/">Infernal</a>).
User can check the results of blast.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: The statistics files and figures.</p>
</div>
<div class="section" id="crispr">
<h2>crispr<a class="headerlink" href="#crispr" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">crispr</span></code> integrates CRISPR Recognition Tool (<a class="reference external" href="http://www.room220.com/crt/">CRT</a>) which can detect the repeat
units and spacer of CRISPR. It can also compare to genome annotation to exclude some false positives.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.room220.com/crt/">CRT</a>.</p>
<p>Fasta file of genome sequence.</p>
<p>Some useful information can be assigned to improve the prediction:</p>
<p>Gff files of genome annotation.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">crispr</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">CRT_path</span> <span class="n">CRT_PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gff_path</span> <span class="n">GFF_PATH</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">window_size</span> <span class="n">WINDOW_SIZE</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">min_number_repeat</span> <span class="n">MIN_NUMBER_REPEAT</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">min_length_repeat</span> <span class="n">MIN_LENGTH_REPEAT</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">Max_length_repeat</span> <span class="n">MAX_LENGTH_REPEAT</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">min_length_spacer</span> <span class="n">MIN_LENGTH_SPACER</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">Max_length_spacer</span> <span class="n">MAX_LENGTH_SPACER</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">ignore_hypothetical_protein</span><span class="p">]</span>
                        <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">CRT_path</span> <span class="n">CRT_PATH</span>   <span class="n">If</span> <span class="n">you</span> <span class="n">want</span> <span class="n">to</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">CRT</span><span class="o">.</span><span class="n">jar</span> <span class="p">(</span><span class="n">the</span> <span class="n">execute</span>
                        <span class="n">file</span><span class="p">),</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">here</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">CRT</span><span class="o">.</span><span class="n">jar</span>
  <span class="o">--</span><span class="n">gff_path</span> <span class="n">GFF_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span> <span class="n">GFF_PATH</span>
                        <span class="n">If</span> <span class="n">genome</span> <span class="n">gff</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">provided</span><span class="p">,</span> <span class="n">it</span> <span class="n">will</span> <span class="n">compare</span> <span class="k">with</span>
                        <span class="n">CRISPR</span> <span class="k">for</span> <span class="n">removing</span> <span class="n">the</span> <span class="n">false</span> <span class="n">positives</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">None</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fasta_path</span> <span class="n">FASTA_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA_PATH</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">fasta</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">window_size</span> <span class="n">WINDOW_SIZE</span><span class="p">,</span> <span class="o">-</span><span class="n">w</span> <span class="n">WINDOW_SIZE</span>
                        <span class="n">Length</span> <span class="n">of</span> <span class="n">window</span> <span class="n">size</span> <span class="k">for</span> <span class="n">searching</span> <span class="p">(</span><span class="nb">range</span><span class="p">:</span> <span class="mi">6</span><span class="o">-</span><span class="mi">9</span><span class="p">)</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="mf">8.</span>
  <span class="o">--</span><span class="n">min_number_repeat</span> <span class="n">MIN_NUMBER_REPEAT</span><span class="p">,</span> <span class="o">-</span><span class="n">mn</span> <span class="n">MIN_NUMBER_REPEAT</span>
                        <span class="n">Minimum</span> <span class="n">number</span> <span class="n">of</span> <span class="n">repeats</span> <span class="n">that</span> <span class="n">a</span> <span class="n">CRISPR</span> <span class="n">must</span> <span class="n">contain</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="mf">3.</span>
  <span class="o">--</span><span class="n">min_length_repeat</span> <span class="n">MIN_LENGTH_REPEAT</span><span class="p">,</span> <span class="o">-</span><span class="n">ml</span> <span class="n">MIN_LENGTH_REPEAT</span>
                        <span class="n">Minimum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">repeats</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">23.</span>
  <span class="o">--</span><span class="n">Max_length_repeat</span> <span class="n">MAX_LENGTH_REPEAT</span><span class="p">,</span> <span class="o">-</span><span class="n">Ml</span> <span class="n">MAX_LENGTH_REPEAT</span>
                        <span class="n">Maximum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">repeats</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">47.</span>
  <span class="o">--</span><span class="n">min_length_spacer</span> <span class="n">MIN_LENGTH_SPACER</span><span class="p">,</span> <span class="o">-</span><span class="n">ms</span> <span class="n">MIN_LENGTH_SPACER</span>
                        <span class="n">Minimum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">spacers</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">26.</span>
  <span class="o">--</span><span class="n">Max_length_spacer</span> <span class="n">MAX_LENGTH_SPACER</span><span class="p">,</span> <span class="o">-</span><span class="n">Ms</span> <span class="n">MAX_LENGTH_SPACER</span>
                        <span class="n">Maximum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">spacers</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">50.</span>
  <span class="o">--</span><span class="n">ignore_hypothetical_protein</span><span class="p">,</span> <span class="o">-</span><span class="ow">in</span>
                        <span class="n">For</span> <span class="n">ignoring</span> <span class="n">the</span> <span class="n">hypothetical</span> <span class="n">proteins</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>It will generate three folders - <code class="docutils literal"><span class="pre">CRT_output</span></code>, <code class="docutils literal"><span class="pre">gffs</span></code> and <code class="docutils literal"><span class="pre">statistics</span></code>.</p>
<p><code class="docutils literal"><span class="pre">CRT_output</span></code> stores the output of <a class="reference external" href="http://www.room220.com/crt/">CRT</a>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code> stores the gff files of CRISPR. <code class="docutils literal"><span class="pre">all_candidates</span></code> is for the all CRISPRs which ANNOgesic can detect.
<code class="docutils literal"><span class="pre">best</span></code> is for the CRISPRs which are not overlap with genome annotations.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code> is for the statistics files.</p>
</div>
<div class="section" id="optimize-tsspredator">
<h2>optimize_tsspredator<a class="headerlink" href="#optimize-tsspredator" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">optimize_tsspredator</span></code> can adapt the parameter set of <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.
For running it, only manual checking around 200kb is required (using gff format).
We suggest the manual detected TSSs should more than 50. If there are less than 50 TSSs within 200kb,
please continue checking until 50 TSSs detected.
Then <code class="docutils literal"><span class="pre">optimize_tsspredator</span></code> can scan whole genome based on the principle of manual detection to get the optimized parameters.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p>The libraries and wiggle files of TEX +/-.
Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>Fasta file of genome sequence.</p>
<p>Gff file of genome annotation.</p>
<p>Gff file of manual detection.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic optimize_tsspredator [-h]
                                      [--TSSpredator_path TSSPREDATOR_PATH]
                                      [--fasta_file FASTA_FILE]
                                      [--annotation_file ANNOTATION_FILE]
                                      [--wig_folder WIG_FOLDER]
                                      [--manual MANUAL]
                                      [--strain_name STRAIN_NAME]
                                      [--max_height MAX_HEIGHT]
                                      [--max_height_reduction MAX_HEIGHT_REDUCTION]
                                      [--max_factor MAX_FACTOR]
                                      [--max_factor_reduction MAX_FACTOR_REDUCTION]
                                      [--max_base_height MAX_BASE_HEIGHT]
                                      [--max_enrichment_factor MAX_ENRICHMENT_FACTOR]
                                      [--max_processing_factor MAX_PROCESSING_FACTOR]
                                      [--utr_length UTR_LENGTH] [--lib LIB]
                                      [--output_prefix OUTPUT_PREFIX]
                                      [--cluster CLUSTER] [--length LENGTH]
                                      [--core CORE] [--program PROGRAM]
                                      [--replicate_match REPLICATE_MATCH]
                                      [--steps STEPS]
                                      [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TSSpredator_path TSSPREDATOR_PATH
                        If you want to assign the path of TSSpredator, please
                        assign here. Default is /usr/local/bin/TSSpredator.jar
  --fasta_file FASTA_FILE, -fs FASTA_FILE
                        Path of one target genome fasta file.
  --annotation_file ANNOTATION_FILE, -g ANNOTATION_FILE
                        Path of one target genome annotation gff file.
  --wig_folder WIG_FOLDER, -w WIG_FOLDER
                        The folder of the TEX+/- wig folder.
  --manual MANUAL, -m MANUAL
                        The file of manual checked gff file.
  --strain_name STRAIN_NAME, -n STRAIN_NAME
                        The name of the strain for optimization.
  --max_height MAX_HEIGHT, -he MAX_HEIGHT
                        This value relates to the minimum number of read
                        starts at a certain genomic position to be considered
                        as a TSS candidate. During optimization --max_height
                        will be never larger than this value. Default is 2.5.
  --max_height_reduction MAX_HEIGHT_REDUCTION, -rh MAX_HEIGHT_REDUCTION
                        When comparing different strains/conditions and the
                        step height threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step height threshold.
                        During optimization, --max_height_reduction will be
                        never larger than this value. Default is 2.4.
  --max_factor MAX_FACTOR, -fa MAX_FACTOR
                        This is the minimum factor by which the TSS height has
                        to exceed the local expression background. During
                        optimization, --max_factor will be never larger than
                        this value. Default is 10.
  --max_factor_reduction MAX_FACTOR_REDUCTION, -rf MAX_FACTOR_REDUCTION
                        When comparing different strains/conditions and the
                        step factor threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step factor threshold.
                        During optimization, --max_factor_reduction will be
                        never larger than this value. Default is 9.9.
  --max_base_height MAX_BASE_HEIGHT, -bh MAX_BASE_HEIGHT
                        This is the minimum number of reads should be mapped
                        on TSS. During optimization, --max_base_height will be
                        never larger than this value. Default is 0.06.
  --max_enrichment_factor MAX_ENRICHMENT_FACTOR, -ef MAX_ENRICHMENT_FACTOR
                        This is the minimum enrichment factor. During
                        optimization, --max_enrichment_factor will be never
                        larger than this value. Default is 6.0.
  --max_processing_factor MAX_PROCESSING_FACTOR, -pf MAX_PROCESSING_FACTOR
                        This is the minimum processing factor. If untreated
                        library is higher than the treated library and above
                        which the TSS candidate is considered as a processing
                        site and not annotated as detected. During
                        optimization, --max_processing_factor will be never
                        larger than this value. Default is 6.0
  --utr_length UTR_LENGTH, -u UTR_LENGTH
                        The length of UTR. It is for Primary and Secondary
                        TSSs. Default is 300.
  --lib LIB, -l LIB     The libraries of TEX+/- wig files for TSSpredator. The
                        format is: wig_file_name:TEX+/-(tex or notex):conditio
                        n_id(integer):replicate_id(alphabet):strand(+ or -).
                        For multiple wig files, please use comma to separate
                        the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --output_prefix OUTPUT_PREFIX, -p OUTPUT_PREFIX
                        The output prefix of all conditions. For multiple
                        conditions, please use comma to separate them. For
                        example, prefix_condition1,prefix_condition2.
  --cluster CLUSTER, -cu CLUSTER
                        If the position between manual one and predicted one
                        is smaller or equal than this value, it will only
                        print one of them. Default is 2.
  --length LENGTH, -le LENGTH
                        The length of genome for running optimization. If
                        running for whole genome is not the case, please
                        assign the length of partial genome. Default is
                        comparing whole genome.
  --core CORE, -c CORE  Paralle runs for optimization. Default is 4.
  --program PROGRAM, -t PROGRAM
                        The feature for optimization. Please assign &quot;TSS&quot; or
                        &quot;Processint_site&quot;. Default is TSS.
  --replicate_match REPLICATE_MATCH, -rm REPLICATE_MATCH
                        The TSS candidates should be detected at lease in the
                        number of replicates. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATED. For assigning
                        different --replicate_match to different conditions,
                        please use comma to separate it. For example,
                        1_2,2_2,3_3 means number 1 and 2 condition assign 2 to
                        --replicate_match, and number 3 condition assign 3 to
                        --replcate_match. For assigning the same
                        --replicate_match to all conditions, just assign like
                        all_1 (all condition use 1 --replicate_match). Default
                        is all_1.
  --steps STEPS, -s STEPS
                        The total runs for optimization. Default is 4000 runs.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>Based on the programs (TSS/processing site), the output files will be stored in
<code class="docutils literal"><span class="pre">$ANNOgesic/output/TSS/optimized_TSSpredator</span></code> or <code class="docutils literal"><span class="pre">$ANNOgesic/output/processing_site/optimized_TSSpredator</span></code>.</p>
<p><code class="docutils literal"><span class="pre">stat.csv</span></code> stores the information of every step. The first column is the number of run.
The second column is the parameter set. <code class="docutils literal"><span class="pre">he</span></code> represents height; <code class="docutils literal"><span class="pre">rh</span></code> represents
height reduction; <code class="docutils literal"><span class="pre">fa</span></code> means factor; <code class="docutils literal"><span class="pre">rf</span></code> means factor reduction; <code class="docutils literal"><span class="pre">bh</span></code> is
base height; <code class="docutils literal"><span class="pre">ef</span></code> is enrichment factor; <code class="docutils literal"><span class="pre">pf</span></code> is processing factor. About the details
of parameters, please refer to <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.
For example, <code class="docutils literal"><span class="pre">he_2.0_rh_1.8_fa_4.4_rf_2.8_bh_0.08_ef_3.0_pf_2.6</span></code> means height is 2.0,
height reduction is 1.8, factor is 4.4, factor reduction is 2.8, base height is 0.08,
enrichment factor is 3.0 and processing factor is 2.6. The third and fourth columns are
the number of true positive. The fifth and sixth columns are true positive rate. The seventh
and eighth columns are the number of false positive. The ninth and tenth false positive rate.
The eleventh and twelfth columns are the number of false negative. The thirteenth and fourteenth
columns are missing rate.</p>
<p><code class="docutils literal"><span class="pre">best.csv</span></code> stores the best parameter set. The meanings of all columns are the same as <code class="docutils literal"><span class="pre">stat.csv</span></code>.</p>
</div>
<div class="section" id="screenshot">
<h2>screenshot<a class="headerlink" href="#screenshot" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">screenshot</span></code> will generate batch files for producing screenshot of <a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>.
It can reduce the time for checking the results in genome browser.
When the batch files is produced, user just needs to open <a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>, then presses <code class="docutils literal"><span class="pre">tools</span></code>
on the top tags and choose <code class="docutils literal"><span class="pre">run</span> <span class="pre">batch</span> <span class="pre">script</span></code>. The program will automatically produce screenshots.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>.</p>
<p>Gff files that user want to produce screenshots. All screenshots will be produced based on the positions of <code class="docutils literal"><span class="pre">main_gff</span></code>.
<code class="docutils literal"><span class="pre">side_gffs</span></code> are the gff files that user want to compare with <code class="docutils literal"><span class="pre">main_gff</span></code>. They will also appear in the screenshots.</p>
<p>Fasta files of genome.</p>
<p>The libraries and wiggle files. Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">screenshot</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">main_gff</span> <span class="n">MAIN_GFF</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">side_gffs</span> <span class="n">SIDE_GFFS</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">fasta</span> <span class="n">FASTA</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">frag_wig_folder</span> <span class="n">FRAG_WIG_FOLDER</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">tex_wig_folder</span> <span class="n">TEX_WIG_FOLDER</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">height</span> <span class="n">HEIGHT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tex_libs</span> <span class="n">TEX_LIBS</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">frag_libs</span> <span class="n">FRAG_LIBS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">present</span> <span class="n">PRESENT</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">output_folder</span> <span class="n">OUTPUT_FOLDER</span><span class="p">]</span>
                            <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">main_gff</span> <span class="n">MAIN_GFF</span><span class="p">,</span> <span class="o">-</span><span class="n">mg</span> <span class="n">MAIN_GFF</span>
                        <span class="n">Screenshot</span> <span class="n">will</span> <span class="n">based</span> <span class="n">on</span> <span class="n">the</span> <span class="n">position</span> <span class="n">of</span> <span class="n">main_gff</span> <span class="n">file</span>
                        <span class="n">to</span> <span class="n">generate</span> <span class="n">screenshot</span><span class="o">.</span>
  <span class="o">--</span><span class="n">side_gffs</span> <span class="n">SIDE_GFFS</span><span class="p">,</span> <span class="o">-</span><span class="n">sg</span> <span class="n">SIDE_GFFS</span>
                        <span class="n">If</span> <span class="n">comparing</span> <span class="n">between</span> <span class="n">other</span> <span class="n">features</span> <span class="ow">and</span> <span class="o">--</span><span class="n">main_gff</span> <span class="ow">is</span>
                        <span class="n">needed</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">other</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">here</span><span class="o">.</span> <span class="n">For</span>
                        <span class="n">multiple</span> <span class="n">gff</span> <span class="n">files</span><span class="p">,</span> <span class="n">just</span> <span class="n">use</span> <span class="n">comma</span> <span class="n">to</span> <span class="n">separate</span> <span class="n">them</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fasta</span> <span class="n">FASTA</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">fasta</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">frag_wig_folder</span> <span class="n">FRAG_WIG_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">fw</span> <span class="n">FRAG_WIG_FOLDER</span>
                        <span class="n">If</span> <span class="n">the</span> <span class="n">information</span> <span class="n">of</span> <span class="n">fragmented</span> <span class="n">wig</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">required</span><span class="p">,</span>
                        <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">tex_wig_folder</span> <span class="n">TEX_WIG_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">tw</span> <span class="n">TEX_WIG_FOLDER</span>
                        <span class="n">If</span> <span class="n">the</span> <span class="n">information</span> <span class="n">of</span> <span class="n">TEX</span><span class="o">+/-</span> <span class="n">wig</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">required</span><span class="p">,</span>
                        <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">folder</span><span class="o">.</span>
  <span class="o">--</span><span class="n">height</span> <span class="n">HEIGHT</span><span class="p">,</span> <span class="o">-</span><span class="n">he</span> <span class="n">HEIGHT</span>
                        <span class="n">The</span> <span class="n">height</span> <span class="n">of</span> <span class="n">screenshot</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">1500.</span>
  <span class="o">--</span><span class="n">tex_libs</span> <span class="n">TEX_LIBS</span><span class="p">,</span> <span class="o">-</span><span class="n">tl</span> <span class="n">TEX_LIBS</span>
                        <span class="n">If</span> <span class="n">the</span> <span class="n">TEX</span><span class="o">+/-</span> <span class="n">wig</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">required</span><span class="p">,</span> <span class="n">please</span> <span class="n">also</span> <span class="n">assign</span>
                        <span class="n">proper</span> <span class="nb">format</span> <span class="n">here</span><span class="o">.</span> <span class="n">The</span> <span class="nb">format</span> <span class="ow">is</span><span class="p">:</span>
                        <span class="n">wig_file_name</span><span class="p">:</span><span class="n">TEX</span><span class="o">+/-</span><span class="p">(</span><span class="n">tex</span> <span class="ow">or</span> <span class="n">notex</span><span class="p">):</span><span class="n">condition_id</span><span class="p">(</span><span class="n">intege</span>
                        <span class="n">r</span><span class="p">):</span><span class="n">replicate_id</span><span class="p">(</span><span class="n">alphabet</span><span class="p">):</span><span class="n">strand</span><span class="p">(</span><span class="o">+</span> <span class="ow">or</span> <span class="o">-</span><span class="p">)</span><span class="o">.</span> <span class="n">For</span> <span class="n">multiple</span>
                        <span class="n">wig</span> <span class="n">files</span><span class="p">,</span> <span class="n">please</span> <span class="n">use</span> <span class="n">comma</span> <span class="n">to</span> <span class="n">separate</span> <span class="n">the</span> <span class="n">wig</span> <span class="n">files</span><span class="o">.</span>
                        <span class="n">For</span> <span class="n">example</span><span class="p">,</span> <span class="n">wig1</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span><span class="p">,</span><span class="n">wig2</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-.</span>
  <span class="o">--</span><span class="n">frag_libs</span> <span class="n">FRAG_LIBS</span><span class="p">,</span> <span class="o">-</span><span class="n">fl</span> <span class="n">FRAG_LIBS</span>
                        <span class="n">If</span> <span class="n">the</span> <span class="n">fragmented</span> <span class="n">wig</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">required</span><span class="p">,</span> <span class="n">please</span> <span class="n">also</span>
                        <span class="n">assign</span> <span class="n">proper</span> <span class="nb">format</span> <span class="n">here</span><span class="o">.</span> <span class="n">The</span> <span class="nb">format</span> <span class="ow">is</span><span class="p">:</span> <span class="n">wig_file_nam</span>
                        <span class="n">e</span><span class="p">:</span><span class="n">fragmented</span><span class="p">(</span><span class="n">frag</span><span class="p">):</span><span class="n">condition_id</span><span class="p">(</span><span class="n">integer</span><span class="p">):</span><span class="n">replicate_id</span><span class="p">(</span>
                        <span class="n">alphabet</span><span class="p">):</span><span class="n">strand</span><span class="p">(</span><span class="o">+</span> <span class="ow">or</span> <span class="o">-</span><span class="p">)</span><span class="o">.</span> <span class="n">For</span> <span class="n">multiple</span> <span class="n">wig</span> <span class="n">files</span><span class="p">,</span>
                        <span class="n">please</span> <span class="n">use</span> <span class="n">comma</span> <span class="n">to</span> <span class="n">separate</span> <span class="n">the</span> <span class="n">wig</span> <span class="n">files</span><span class="o">.</span> <span class="n">For</span>
                        <span class="n">example</span><span class="p">,</span> <span class="n">wig1</span><span class="p">:</span><span class="n">frag</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span><span class="p">,</span><span class="n">wig2</span><span class="p">:</span><span class="n">frag</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-.</span>
  <span class="o">--</span><span class="n">present</span> <span class="n">PRESENT</span><span class="p">,</span> <span class="o">-</span><span class="n">p</span> <span class="n">PRESENT</span>
                        <span class="n">The</span> <span class="nb">type</span> <span class="n">of</span> <span class="n">presentation</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">screenshot</span><span class="o">.</span>
                        <span class="n">expand</span><span class="o">/</span><span class="n">collapse</span><span class="o">/</span><span class="n">squish</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="n">expand</span><span class="o">.</span>
  <span class="o">--</span><span class="n">output_folder</span> <span class="n">OUTPUT_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FOLDER</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">output</span> <span class="n">folder</span><span class="o">.</span> <span class="n">It</span> <span class="n">will</span> <span class="n">create</span> <span class="n">a</span> <span class="n">sub</span><span class="o">-</span>
                        <span class="n">folder</span> <span class="s2">&quot;screenshots&quot;</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">output</span> <span class="n">folder</span> <span class="n">to</span> <span class="n">store</span> <span class="n">the</span>
                        <span class="n">results</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>Based on the paths of <code class="docutils literal"><span class="pre">main_gff</span></code>, <code class="docutils literal"><span class="pre">screenshot</span></code> will generate a folder - <code class="docutils literal"><span class="pre">screenshots</span></code> under the
folder of <code class="docutils literal"><span class="pre">main_gff</span></code>. All output files will be generated in this folder.</p>
<p><code class="docutils literal"><span class="pre">forward.txt</span></code> is the batch file of forward strand for running on IGV.</p>
<p><code class="docutils literal"><span class="pre">reverse.txt</span></code> is the batch file of reverse strand for running on IGV.</p>
<p><code class="docutils literal"><span class="pre">forward</span></code> is the folder for storing screenshots of forward strand.</p>
<p><code class="docutils literal"><span class="pre">reverse</span></code> is the folder for storing screenshots of reverse strand.</p>
<p>When batch files are executed on IGV, the screenshots will automatically store in <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>.
The format of filenames will be <code class="docutils literal"><span class="pre">$STRAIN:$START-$END.png</span></code>. For example, <code class="docutils literal"><span class="pre">NC_007795:1051529-1051696.png</span></code>
means the strain is NC_007795. The feature&#8217;s start point is 1051529 and the end point is
1051696.</p>
</div>
<div class="section" id="color-png">
<h2>color_png<a class="headerlink" href="#color-png" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">color_png</span></code> is a following procedure of <code class="docutils literal"><span class="pre">screenshot</span></code>. If numerous samples are included in one figure,
it will be difficult to distinguish the tracks. <code class="docutils literal"><span class="pre">color_png</span></code> can color the tracks based on TEX +/- libraries
for assisting the checking process.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.imagemagick.org/script/index.php">ImageMagick</a>.</p>
<p>The screenshots which generated from <code class="docutils literal"><span class="pre">screenshot</span></code>. Please make sure the folders of <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>
exist in the folder of <code class="docutils literal"><span class="pre">screenshots</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">color_png</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">screenshot_folder</span> <span class="n">SCREENSHOT_FOLDER</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">track_number</span> <span class="n">TRACK_NUMBER</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">ImageMagick_covert_path</span> <span class="n">IMAGEMAGICK_COVERT_PATH</span><span class="p">]</span>
                           <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">screenshot_folder</span> <span class="n">SCREENSHOT_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">SCREENSHOT_FOLDER</span>
                        <span class="n">The</span> <span class="n">folder</span> <span class="n">which</span> <span class="n">stores</span> <span class="s2">&quot;screenshots&quot;</span> <span class="p">(</span><span class="n">a</span> <span class="n">folder</span>
                        <span class="n">generated</span> <span class="n">by</span> <span class="n">subcommand</span> <span class="s2">&quot;screenshot&quot;</span><span class="p">)</span><span class="o">.</span>
  <span class="o">--</span><span class="n">track_number</span> <span class="n">TRACK_NUMBER</span><span class="p">,</span> <span class="o">-</span><span class="n">t</span> <span class="n">TRACK_NUMBER</span>
                        <span class="n">The</span> <span class="n">number</span> <span class="n">of</span> <span class="n">tracks</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ImageMagick_covert_path</span> <span class="n">IMAGEMAGICK_COVERT_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">m</span> <span class="n">IMAGEMAGICK_COVERT_PATH</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="s2">&quot;convert&quot;</span> <span class="ow">in</span> <span class="n">ImageMagick</span>
                        <span class="n">package</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The new screenshots will replace the previous ones automatically.</p>
</div>
<div class="section" id="merge-features">
<h2>merge_features<a class="headerlink" href="#merge-features" title="Permalink to this headline">¶</a></h2>
<p>If merging multiple features to one gff file is needed, <code class="docutils literal"><span class="pre">merge_features</span></code> can achieve this purpose.
It will merge all features that user assigned and search the parent transcript to each feature.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">merge_features</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">]</span>
                                <span class="p">[</span><span class="o">--</span><span class="n">other_features_path</span> <span class="n">OTHER_FEATURES_PATH</span><span class="p">]</span>
                                <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_term</span> <span class="n">FUZZY_TERM</span><span class="p">]</span>
                                <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_TSS</span> <span class="n">FUZZY_TSS</span><span class="p">]</span>
                                <span class="p">[</span><span class="o">--</span><span class="n">strain_name</span> <span class="n">STRAIN_NAME</span><span class="p">]</span>
                                <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">transcript_path</span> <span class="n">TRANSCRIPT_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_PATH</span>
                        <span class="n">If</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">provided</span><span class="o">.</span> <span class="n">The</span> <span class="n">output</span> <span class="n">will</span> <span class="n">be</span>
                        <span class="n">generated</span> <span class="n">based</span> <span class="n">on</span> <span class="n">transcripts</span> <span class="ow">and</span> <span class="n">assign</span>
                        <span class="s2">&quot;Parent_tran&quot;</span> <span class="n">to</span> <span class="n">each</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">attributes</span> <span class="n">of</span> <span class="n">gff</span>
                        <span class="n">file</span><span class="o">.</span> <span class="n">If</span> <span class="n">there</span> <span class="ow">is</span> <span class="n">no</span> <span class="n">transcript</span><span class="p">,</span> <span class="n">the</span> <span class="n">output</span> <span class="n">will</span> <span class="n">just</span>
                        <span class="n">simply</span> <span class="n">combine</span> <span class="nb">all</span> <span class="nb">input</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">other_features_path</span> <span class="n">OTHER_FEATURES_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">of</span> <span class="n">OTHER_FEATURES_PATH</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">of</span> <span class="n">other</span> <span class="n">features</span> <span class="n">which</span>
                        <span class="n">you</span> <span class="n">want</span> <span class="n">to</span> <span class="n">merge</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fuzzy_term</span> <span class="n">FUZZY_TERM</span><span class="p">,</span> <span class="o">-</span><span class="n">fm</span> <span class="n">FUZZY_TERM</span>
                        <span class="n">For</span> <span class="n">merging</span> <span class="n">terminators</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">fuzzy</span>
                        <span class="n">between</span> <span class="n">transcript</span> <span class="ow">and</span> <span class="n">terminator</span><span class="o">.</span> <span class="n">ATTENTION</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">third</span> <span class="n">column</span> <span class="n">of</span> <span class="n">gff</span> <span class="n">file</span> <span class="n">of</span> <span class="n">terminator</span> <span class="n">should</span> <span class="n">be</span>
                        <span class="n">exactly</span> <span class="s2">&quot;terminator&quot;</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">30.</span>
  <span class="o">--</span><span class="n">fuzzy_TSS</span> <span class="n">FUZZY_TSS</span><span class="p">,</span> <span class="o">-</span><span class="n">ft</span> <span class="n">FUZZY_TSS</span>
                        <span class="n">For</span> <span class="n">merging</span> <span class="n">TSSs</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">fuzzy</span> <span class="n">between</span> <span class="n">TSS</span>
                        <span class="ow">and</span> <span class="n">transcript</span><span class="o">.</span> <span class="n">ATTENTION</span><span class="p">,</span> <span class="n">the</span> <span class="n">third</span> <span class="n">column</span> <span class="n">of</span> <span class="n">gff</span>
                        <span class="n">file</span> <span class="n">of</span> <span class="n">terminator</span> <span class="n">should</span> <span class="n">be</span> <span class="n">exactly</span> <span class="s2">&quot;TSS&quot;</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="mf">5.</span>
  <span class="o">--</span><span class="n">strain_name</span> <span class="n">STRAIN_NAME</span><span class="p">,</span> <span class="o">-</span><span class="n">s</span> <span class="n">STRAIN_NAME</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">strain</span> <span class="n">name</span> <span class="n">of</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">files</span><span class="o">.</span> <span class="n">It</span>
                        <span class="n">will</span> <span class="n">become</span> <span class="n">the</span> <span class="n">prefix</span> <span class="n">name</span> <span class="n">of</span> <span class="n">output</span> <span class="n">gff</span> <span class="n">file</span><span class="o">.</span>
</pre></div>
</div>
<p>-Output files</p>
<p>The gff file will be stored in <code class="docutils literal"><span class="pre">merge_all_features</span></code>. The tag - <code class="docutils literal"><span class="pre">Parent</span></code> in attributes of
gff file shows the parent transcript.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="test_case.html" title="Performing a test cast"
             >next</a></li>
        <li class="right" >
          <a href="docker.html" title="Dockfiles"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.1.3 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Sung-Huan Yu.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.3.
    </div>
  </body>
</html>