

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>ANNOgesicâ€™s subcommands &mdash; ANNOgesic 0.5.14 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="ANNOgesic 0.5.14 documentation" href="index.html"/>
        <link rel="next" title="Tutorial of ANNOgesic" href="tutorial.html"/>
        <link rel="prev" title="Docker image" href="docker.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="index.html" class="icon icon-home"> ANNOgesic
        

        
          
          <img src="_static/annogesic_logo_white.png" class="logo" />
        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">ANNOgesic - Transcriptome annotation pipeline</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#table-of-content">Table of content</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="required.html">Required tools or databases</a><ul>
<li class="toctree-l4"><a class="reference internal" href="required.html#tools">Tools</a></li>
<li class="toctree-l4"><a class="reference internal" href="required.html#databases">Databases</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="installation.html#github">Github</a></li>
<li class="toctree-l4"><a class="reference internal" href="installation.html#docker">Docker</a></li>
<li class="toctree-l4"><a class="reference internal" href="installation.html#pip3">pip3</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="docker.html">Docker image</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">ANNOgesic&#8217;s subcommands</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#the-format-of-filename">The format of filename</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic">The input format of libraries for running ANNOgesic</a></li>
<li class="toctree-l4"><a class="reference internal" href="#create-create-analysis-folder">create (create analysis folder)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#get-input-files-download-required-files">get_input_files (download required files)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#update-genome-fasta-update-reference-genome-fasta-file">update_genome_fasta (update reference genome fasta file)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#annotation-transfer-annotation-transfer">annotation_transfer (annotation transfer)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#snp-snp-calling">snp (SNP calling)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#tss-ps-tss-and-processing-site-prediction">tss_ps (TSS and processing site prediction)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#transcript-transcript-detection">transcript (transcript detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#terminator-terminator-detection">terminator (terminator detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#utr-utr-detection">utr (UTR detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#srna-srna-detection">srna (sRNA detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sorf-sorf-detection">sorf (sORF detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#promoter-promoter-motif-detection">promoter (Promoter motif detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#operon-operon-detection">operon (Operon detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#circrna-circular-rna-detection">circrna (circular RNA detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#go-term-go-term-retrieving">go_term (GO term retrieving)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#srna-target-srna-target-prediction">srna_target (sRNA target prediction)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ppi-network-protein-protein-interaction-network-detection">ppi_network (protein-protein interaction network detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#localization-subcellular-localization-prediction">localization (subcellular localization prediction)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#riboswitch-thermometer-riboswitch-and-rna-thermometer-detection">riboswitch_thermometer (riboswitch and RNA thermometer detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#crispr-crispr-detection">crispr (CRISPR detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimize-tss-ps-optimization-of-tss-and-processing-site-detection">optimize_tss_ps (optimization of TSS and processing site detection)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#screenshot-screenshot-generation">screenshot (screenshot generation)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#color-png-color-screenshot">color_png (color screenshot)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#merge-features-merge-all-annotation-features">merge_features (merge all annotation features)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tutorial.html">Tutorial of ANNOgesic</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#generating-a-project">Generating a project</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#retrieving-the-input-data">Retrieving the input data</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#putting-wig-and-reads-to-proper-location">Putting wig, and reads to proper location</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#improving-the-reference-genome">Improving the reference genome</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#generating-annotation-files">Generating annotation files</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#tss-and-processing-site-prediction-and-optimization">TSS and processing site prediction and optimization</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#performing-transcript-detection">Performing transcript detection</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#prediction-of-terminator">Prediction of terminator</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#generating-utr">Generating UTR</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#detecting-operon-and-suboperon">Detecting operon and suboperon</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#promoter-motif-detection">Promoter motif detection</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#prediction-of-srna-and-sorf">Prediction of sRNA and sORF</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#performing-srna-target-prediction">Performing sRNA target prediction</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#mapping-and-detecting-of-circular-rna">Mapping and detecting of circular RNA</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#snp-calling">SNP calling</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#mapping-gene-ontology">Mapping Gene ontology</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#prediction-of-subcellular-localization">Prediction of Subcellular localization</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#generating-protein-protein-interaction-network">Generating protein-protein interaction network</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#generating-riboswitch-and-rna-thermometer">Generating riboswitch and RNA thermometer</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#detection-of-crispr">Detection of CRISPR</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#merge-all-features-to-be-one-gff-file">Merge all features to be one gff file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#producing-the-screenshots">Producing the screenshots</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorial.html#coloring-the-screenshots">Coloring the screenshots</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a><ul>
<li class="toctree-l4"><a class="reference internal" href="troubleshooting.html#pickling-error">Pickling error</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="license.html">License</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#download">Download</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#source-code">Source code</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#license">License</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#cite">Cite</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#contact">Contact</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="required.html">Required tools or databases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="required.html#tools">Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="required.html#databases">Databases</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#github">Github</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#docker">Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#pip3">pip3</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="docker.html">Docker image</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">ANNOgesic&#8217;s subcommands</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#the-format-of-filename">The format of filename</a></li>
<li class="toctree-l2"><a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic">The input format of libraries for running ANNOgesic</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-create-analysis-folder">create (create analysis folder)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#get-input-files-download-required-files">get_input_files (download required files)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#update-genome-fasta-update-reference-genome-fasta-file">update_genome_fasta (update reference genome fasta file)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotation-transfer-annotation-transfer">annotation_transfer (annotation transfer)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#snp-snp-calling">snp (SNP calling)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tss-ps-tss-and-processing-site-prediction">tss_ps (TSS and processing site prediction)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#transcript-transcript-detection">transcript (transcript detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#terminator-terminator-detection">terminator (terminator detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#utr-utr-detection">utr (UTR detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#srna-srna-detection">srna (sRNA detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sorf-sorf-detection">sorf (sORF detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#promoter-promoter-motif-detection">promoter (Promoter motif detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#operon-operon-detection">operon (Operon detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#circrna-circular-rna-detection">circrna (circular RNA detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#go-term-go-term-retrieving">go_term (GO term retrieving)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#srna-target-srna-target-prediction">srna_target (sRNA target prediction)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ppi-network-protein-protein-interaction-network-detection">ppi_network (protein-protein interaction network detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#localization-subcellular-localization-prediction">localization (subcellular localization prediction)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#riboswitch-thermometer-riboswitch-and-rna-thermometer-detection">riboswitch_thermometer (riboswitch and RNA thermometer detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#crispr-crispr-detection">crispr (CRISPR detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#optimize-tss-ps-optimization-of-tss-and-processing-site-detection">optimize_tss_ps (optimization of TSS and processing site detection)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#screenshot-screenshot-generation">screenshot (screenshot generation)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#color-png-color-screenshot">color_png (color screenshot)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#merge-features-merge-all-annotation-features">merge_features (merge all annotation features)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial of ANNOgesic</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#generating-a-project">Generating a project</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#retrieving-the-input-data">Retrieving the input data</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#putting-wig-and-reads-to-proper-location">Putting wig, and reads to proper location</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#improving-the-reference-genome">Improving the reference genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#generating-annotation-files">Generating annotation files</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#tss-and-processing-site-prediction-and-optimization">TSS and processing site prediction and optimization</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#performing-transcript-detection">Performing transcript detection</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#prediction-of-terminator">Prediction of terminator</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#generating-utr">Generating UTR</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#detecting-operon-and-suboperon">Detecting operon and suboperon</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#promoter-motif-detection">Promoter motif detection</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#prediction-of-srna-and-sorf">Prediction of sRNA and sORF</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#performing-srna-target-prediction">Performing sRNA target prediction</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#mapping-and-detecting-of-circular-rna">Mapping and detecting of circular RNA</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#snp-calling">SNP calling</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#mapping-gene-ontology">Mapping Gene ontology</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#prediction-of-subcellular-localization">Prediction of Subcellular localization</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#generating-protein-protein-interaction-network">Generating protein-protein interaction network</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#generating-riboswitch-and-rna-thermometer">Generating riboswitch and RNA thermometer</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#detection-of-crispr">Detection of CRISPR</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#merge-all-features-to-be-one-gff-file">Merge all features to be one gff file</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#producing-the-screenshots">Producing the screenshots</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#coloring-the-screenshots">Coloring the screenshots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a><ul>
<li class="toctree-l2"><a class="reference internal" href="troubleshooting.html#pickling-error">Pickling error</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">ANNOgesic</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>ANNOgesic&#8217;s subcommands</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="annogesic-s-subcommands">
<span id="id1"></span><h1>ANNOgesic&#8217;s subcommands<a class="headerlink" href="#annogesic-s-subcommands" title="Permalink to this headline">Â¶</a></h1>
<p>In general, the subcommands need at least one argument - the analysis
folder. If it is not given, ANNOgesic assumes the current
folder as the analysis folder.</p>
<div class="section" id="the-format-of-filename">
<span id="id2"></span><h2>The format of filename<a class="headerlink" href="#the-format-of-filename" title="Permalink to this headline">Â¶</a></h2>
<p>In order to recognize file types as well as relation of genome name and features,
please use following principle of filename designation:</p>
<p>The genome filenames should be the same as the annotation file designation, i.e.
<code class="docutils literal"><span class="pre">NC_007795.fa,</span> <span class="pre">NC_007795.gff,</span> <span class="pre">NC_007795.ptt,</span> <span class="pre">NC_007795.rnt,</span> <span class="pre">NC_007705.gbk</span></code>.</p>
<p>For all input gff files for ANNOgesic please use following format:
<code class="docutils literal"><span class="pre">$GENOME_$FEATURE.gff</span></code>. For example, <code class="docutils literal"><span class="pre">NC_007795_TSS.gff,</span> <span class="pre">NC_007795_transcript.gff</span></code>.</p>
<p>Possible feature names are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="36%" />
<col width="64%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Feature name</td>
<td>meaning</td>
</tr>
<tr class="row-even"><td>TSS</td>
<td>transcription starting site</td>
</tr>
<tr class="row-odd"><td>processing</td>
<td>processing site</td>
</tr>
<tr class="row-even"><td>transcript</td>
<td>transcript</td>
</tr>
<tr class="row-odd"><td>sRNA</td>
<td>small RNA</td>
</tr>
<tr class="row-even"><td>sORF</td>
<td>small open reading frame</td>
</tr>
<tr class="row-odd"><td>term</td>
<td>terminator</td>
</tr>
<tr class="row-even"><td>5UTR</td>
<td>5&#8217;UTR</td>
</tr>
<tr class="row-odd"><td>3UTR</td>
<td>3&#8217;UTR</td>
</tr>
<tr class="row-even"><td>circRNA</td>
<td>circular RNA</td>
</tr>
<tr class="row-odd"><td>riboswitch</td>
<td>riboswitch</td>
</tr>
<tr class="row-even"><td>RNA_thermometer</td>
<td>RNA_thermometer</td>
</tr>
<tr class="row-odd"><td>CRISPR</td>
<td>CRISPR</td>
</tr>
</tbody>
</table>
<p>Please avoid <code class="docutils literal"><span class="pre">|</span></code> in the filename and genome name of Gff3 files or fasta file.</p>
</div>
<div class="section" id="the-input-format-of-libraries-for-running-annogesic">
<span id="id3"></span><h2>The input format of libraries for running ANNOgesic<a class="headerlink" href="#the-input-format-of-libraries-for-running-annogesic" title="Permalink to this headline">Â¶</a></h2>
<p>Some <code class="docutils literal"><span class="pre">ANNOgesic</span></code> modules require certain library information. Please use following format:</p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_FILENAME:$LIBRARY_TYPE:$CONDITION:$REPLICATE:$STRAND</span></code></p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_FILENAME</span></code> means the <code class="docutils literal"><span class="pre">.wig</span></code> file.</p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_TYPE</span></code> can be <code class="docutils literal"><span class="pre">tex</span></code> (TEX+) or <code class="docutils literal"><span class="pre">notex</span></code> (TEX-) or <code class="docutils literal"><span class="pre">frag</span></code> (fragmented/conventional).</p>
<p><code class="docutils literal"><span class="pre">$CONDITION</span></code> is the index of conditions. Please use 1, 2, 3, ... to represent different conditions.</p>
<p><code class="docutils literal"><span class="pre">$REPLICATE</span></code> is the index of replicates. Please use a, b, c, ... to represent different replicates.</p>
<p><code class="docutils literal"><span class="pre">$STRAND</span></code> is the strand of wiggle file. Please use + or -.</p>
<p>All libraries should be seperated by colons and listed in one line, like shown in the example above, i.e.
for TEX +/- treated libraries:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_TEX_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span> \
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_TEX_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span> \
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_TEX_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span> \
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_TEX_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">tex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span> \
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span> \
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span> \
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">2</span><span class="n">_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span> \
<span class="n">TSB_OD_0</span><span class="o">.</span><span class="mi">5</span><span class="n">_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">notex</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span>
</pre></div>
</div>
<p>or for fragmented libraries (RNA-Seq generated after transcript fragmentation):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">fragmented_forward</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">frag</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">+</span> <span class="n">fragmented_reverse</span><span class="o">.</span><span class="n">wig</span><span class="p">:</span><span class="n">frag</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="n">a</span><span class="p">:</span><span class="o">-</span>
</pre></div>
</div>
<p>If only conventional RNA-seq data without fragmentation or TEX treated can be provided,
it can still be assigned to fragmented libraries. However, it may influence the results.</p>
</div>
<div class="section" id="create-create-analysis-folder">
<span id="create"></span><h2>create (create analysis folder)<a class="headerlink" href="#create-create-analysis-folder" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">create</span></code> generates the folders for analysis. Once created, please move the required files
into the corresponding folders.</p>
<p>The folders are following:</p>
<p><strong>input:</strong> Stores all input files.</p>
<blockquote>
<div><p><strong>BAMs:</strong> For <code class="docutils literal"><span class="pre">.bam</span></code> files. <code class="docutils literal"><span class="pre">BAMs_map_closed_genomes</span></code>
is for the <code class="docutils literal"><span class="pre">.bam</span></code> files which are mapped on closed genomes of our real query genomes.
<code class="docutils literal"><span class="pre">BAMs_map_query_genomes</span></code> is for the <code class="docutils literal"><span class="pre">.bam</span></code> files which are mapped on our query genomes.</p>
<p><strong>databases:</strong> For all databases.</p>
<p><strong>manual_TSSs:</strong> If the manual detected transcription starting sites (TSSs) can be provided,
it can be stored here for running <code class="docutils literal"><span class="pre">TSS_optimization</span></code> or merging
the automatic predicted ones and manual detected ones. Please use gff3 format.</p>
<p><strong>manual_processing_sites:</strong> It is similar to <code class="docutils literal"><span class="pre">manual_TSS</span></code>, but for
processing sites.</p>
<p><strong>mutation_table:</strong> If the mutation table between the closed genomes and
query genomes is provided, please put the file here. Please check
the section of <a class="reference internal" href="#update-genome-fasta"><span class="std std-ref">update_genome_fasta (update reference genome fasta file)</span></a> for the format of
mutation table.</p>
<p><strong>reads:</strong> For running <code class="docutils literal"><span class="pre">circrna</span></code> with mapping reads by ANNOgesic,
please put the reads here. <code class="docutils literal"><span class="pre">.bzip2</span></code> and <code class="docutils literal"><span class="pre">.gzip</span></code> as input is accepted.</p>
<p><strong>references:</strong> For annotation files and fasta files.
If they can be downloaded from NCBI, the files can also be obtained via running <a class="reference internal" href="#get-input-files"><span class="std std-ref">get_input_files (download required files)</span></a>.</p>
<p><strong>riboswitch_ID_file:</strong> For storing the file which contains all the Rfam IDs of riboswitches.
For format details, please check the section of
<a class="reference internal" href="#riboswitch-thermometer"><span class="std std-ref">riboswitch_thermometer (riboswitch and RNA thermometer detection)</span></a>.</p>
<p><strong>RNA_thermometer_ID_file:</strong> For storing the file which contains all the Rfam IDs of RNA thermometer.
For format details, please check the section of
<a class="reference internal" href="#riboswitch-thermometer"><span class="std std-ref">riboswitch_thermometer (riboswitch and RNA thermometer detection)</span></a>.</p>
<p><strong>wigs:</strong> For wiggle files. Based on the methods of RNA-Seq, wiggle files can be stored in
<code class="docutils literal"><span class="pre">fragment</span></code> (fragmented/conventional libraries) or <code class="docutils literal"><span class="pre">tex_notex</span></code> (TEX +/- treated libraries).</p>
</div></blockquote>
<p><strong>output:</strong> Stores all output files.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">create</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="n">PROJECT_PATH</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="n">PROJECT_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">pj</span> <span class="n">PROJECT_PATH</span>
                        <span class="n">Name</span><span class="o">/</span><span class="n">path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span>
</pre></div>
</div>
</div>
<div class="section" id="get-input-files-download-required-files">
<span id="get-input-files"></span><h2>get_input_files (download required files)<a class="headerlink" href="#get-input-files-download-required-files" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">get_input_files</span></code> is the subcommand for downloading required files (fasta, annotation files) from NCBI.
Therefore, the web address of the reference genome in NCBI needs to be assigned. For example,
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000013425.1_ASM1342v1">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000013425.1_ASM1342v1</a>
Then, the user can assign the file type for download.</p>
<ul class="simple">
<li><strong>Reqired information</strong></li>
</ul>
<p><strong>FTP source:</strong> The IP of NCBI.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">get_input_files</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span>
                                 <span class="p">[</span><span class="o">--</span><span class="n">ftp_path</span> <span class="n">FTP_PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_fasta</span><span class="p">]</span>
                                 <span class="p">[</span><span class="o">--</span><span class="n">ref_gff</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_gbk</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ref_ptt</span><span class="p">]</span>
                                 <span class="p">[</span><span class="o">--</span><span class="n">ref_rnt</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">convert_embl</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ftp_path</span> <span class="n">FTP_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">F</span> <span class="n">FTP_PATH</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">NCBI</span> <span class="n">FTP</span> <span class="n">where</span> <span class="n">can</span> <span class="n">download</span> <span class="n">the</span> <span class="n">required</span>
                        <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_fasta</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span>       <span class="n">Download</span> <span class="n">fasta</span> <span class="n">files</span> <span class="n">of</span> <span class="n">the</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_gff</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span>         <span class="n">Download</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">of</span> <span class="n">the</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_gbk</span><span class="p">,</span> <span class="o">-</span><span class="n">k</span>         <span class="n">Download</span> <span class="n">genbank</span> <span class="n">files</span> <span class="n">of</span> <span class="n">the</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">False</span><span class="o">.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">ref_ptt</span><span class="p">,</span> <span class="o">-</span><span class="n">p</span>         <span class="n">Download</span> <span class="n">ptt</span> <span class="n">files</span> <span class="n">of</span> <span class="n">the</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">ref_rnt</span><span class="p">,</span> <span class="o">-</span><span class="n">r</span>         <span class="n">Download</span> <span class="n">rnt</span> <span class="n">files</span> <span class="n">of</span> <span class="n">the</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">convert_embl</span><span class="p">,</span> <span class="o">-</span><span class="n">e</span>    <span class="n">Convert</span> <span class="n">gbk</span> <span class="n">to</span> <span class="n">embl</span> <span class="n">files</span> <span class="n">of</span> <span class="n">the</span> <span class="n">reference</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/input/reference</span></code></p>
<p>Output folder names are following:</p>
<p><strong>fasta:</strong> Fasta files.</p>
<p><strong>annotation:</strong> Annotation files.</p>
</div>
<div class="section" id="update-genome-fasta-update-reference-genome-fasta-file">
<span id="update-genome-fasta"></span><h2>update_genome_fasta (update reference genome fasta file)<a class="headerlink" href="#update-genome-fasta-update-reference-genome-fasta-file" title="Permalink to this headline">Â¶</a></h2>
<p>If fasta files of the query genomes do not exist, <code class="docutils literal"><span class="pre">update_genome_fasta</span></code> can
update fasta files from the closed genomes of the real query genomes to our real query ones
via searching the mutations.
Therefore, a table of mutation information is required. For the format of the table, please check
<a class="reference external" href="https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/tutorial_data/mutation.csv">mutation table</a>.
Titles of the columns are presented on the top and they need to start with <code class="docutils literal"><span class="pre">#</span></code>.
Each column is separated by <code class="docutils literal"><span class="pre">tab</span></code>. If the mutation type is deletion or insertion,
the user can type <code class="docutils literal"><span class="pre">-</span></code> to represent them. The information of <code class="docutils literal"><span class="pre">Target_ids</span></code>
(the query genome names), <code class="docutils literal"><span class="pre">Reference_ids</span></code>, (the closed genome names)
<code class="docutils literal"><span class="pre">Reference_nts</span></code> (the nucleotides of the closed genomes), <code class="docutils literal"><span class="pre">Positions</span></code>, <code class="docutils literal"><span class="pre">Target_nts</span></code>
(the nucleotides of the query genomes) are required. The other columns can be blank.
Please use <code class="docutils literal"><span class="pre">tab</span></code> to separate all columns including blank ones.</p>
<p>If no mutation information is provided, <code class="docutils literal"><span class="pre">snp</span></code> can be used for detecting mutations.
(one module of <code class="docutils literal"><span class="pre">ANNOgesic</span></code>). Please check the section of <a class="reference internal" href="#snp"><span class="std std-ref">snp (SNP calling)</span></a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Fasta files of reference genome</strong></p>
<p><strong>Mutation table:</strong> Contains the information of mutations between closed and query genomes.
For example, please check <a class="reference external" href="https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/tutorial_data/mutation.csv">mutation table</a>.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic update_genome_fasta [-h] [--project_path [PROJECT_PATH]]
                                     --closed_fasta_files CLOSED_FASTA_FILES
                                     [CLOSED_FASTA_FILES ...] --mutation_table
                                     MUTATION_TABLE --output_format
                                     OUTPUT_FORMAT [OUTPUT_FORMAT ...]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --closed_fasta_files CLOSED_FASTA_FILES [CLOSED_FASTA_FILES ...], -c CLOSED_FASTA_FILES [CLOSED_FASTA_FILES ...]
                        Path of the closed fasta files.
  --mutation_table MUTATION_TABLE, -m MUTATION_TABLE
                        Path of the mutation table which stores the mutation
                        information between the query genome and closed genome
                        of the query one. For example, please check
                        https://github.com/Sung-
                        Huan/ANNOgesic/blob/master/tutorial_data/mutation.csv
  --output_format OUTPUT_FORMAT [OUTPUT_FORMAT ...], -o OUTPUT_FORMAT [OUTPUT_FORMAT ...]
                        Please assign the filename of output file and the
                        genome names (listed in Target_id of mutation table)
                        which should be included in the specific output file.
                        For example: $FILE_PATH1:genome1,genome2
                        $FILE_PATH2:genome3. FILE_PATH1 is a output fasta file
                        which includes the information of genome1 and genome2
                        (import multi-genomes to one file should be separated
                        by &quot;,&quot;.) And FILE_PATH2 is for genome3. The multiple
                        output files are splitted by spaces.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p><strong>Fasta files of updated genome</strong>: The updated fasta files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/updated_references/fasta_files</span></code>.</p>
</div>
<div class="section" id="annotation-transfer-annotation-transfer">
<span id="annotation-transfer"></span><h2>annotation_transfer (annotation transfer)<a class="headerlink" href="#annotation-transfer-annotation-transfer" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">annotation</span> <span class="pre">transfer</span></code> is the subcommand for transferring the annotation from the closed genomes
to the real query genomes. To achieve this, <a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">RATT</a>
is integrated in ANNOgesic. The higher similarity between closed genomes and query genomes are,
the more precise the performance is. Before running <code class="docutils literal"><span class="pre">annotation</span> <span class="pre">transfer</span></code>,
please run <code class="docutils literal"><span class="pre">source</span> <span class="pre">$PAGIT_HOME/sourceme.pagit</span></code> first. it will modify the path for executing RATT.
If you use Dockerfile to execute ANNOgesic, the path modification can be skipped.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">RATT</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Annotation files of the closed genomes</strong>: Genbank/embl files of the closed genomes.</p>
<p><strong>Fasta files of the closed genomes</strong></p>
<p><strong>Fasta files of the updated genomes</strong></p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">annotation_transfer</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span>
                                     <span class="o">--</span><span class="n">compare_pair</span> <span class="n">COMPARE_PAIR</span>
                                     <span class="p">[</span><span class="n">COMPARE_PAIR</span> <span class="o">...</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">closed_embl_files</span> <span class="n">CLOSED_EMBL_FILES</span> <span class="p">[</span><span class="n">CLOSED_EMBL_FILES</span> <span class="o">...</span><span class="p">]]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">closed_gbk_files</span> <span class="n">CLOSED_GBK_FILES</span> <span class="p">[</span><span class="n">CLOSED_GBK_FILES</span> <span class="o">...</span><span class="p">]]</span>
                                     <span class="o">--</span><span class="n">closed_fasta_files</span> <span class="n">CLOSED_FASTA_FILES</span>
                                     <span class="p">[</span><span class="n">CLOSED_FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                                     <span class="o">--</span><span class="n">updated_fasta_files</span> <span class="n">UPDATED_FASTA_FILES</span>
                                     <span class="p">[</span><span class="n">UPDATED_FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">ratt_path</span> <span class="n">RATT_PATH</span><span class="p">]</span> <span class="o">--</span><span class="n">element</span> <span class="n">ELEMENT</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">transfer_type</span> <span class="n">TRANSFER_TYPE</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">convert_to_gff_rnt_ptt</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">compare_pair</span> <span class="n">COMPARE_PAIR</span> <span class="p">[</span><span class="n">COMPARE_PAIR</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">p</span> <span class="n">COMPARE_PAIR</span> <span class="p">[</span><span class="n">COMPARE_PAIR</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">name</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">pairs</span><span class="p">,</span> <span class="n">ex</span><span class="o">.</span>
                        <span class="n">NC_007795</span><span class="p">:</span><span class="n">NEW_NC_007795</span><span class="o">.</span> <span class="n">The</span> <span class="n">closed</span> <span class="n">genome</span> <span class="ow">is</span>
                        <span class="n">NC_007795</span> <span class="ow">and</span> <span class="n">the</span> <span class="n">target</span> <span class="n">genome</span> <span class="ow">is</span> <span class="n">NEW_NC_007795</span><span class="o">.</span> <span class="n">The</span>
                        <span class="n">assigned</span> <span class="n">names</span> <span class="n">are</span> <span class="n">the</span> <span class="n">headers</span> <span class="n">of</span> <span class="n">the</span> <span class="n">fasta</span> <span class="n">file</span>
                        <span class="p">(</span><span class="n">start</span> <span class="k">with</span> <span class="s2">&quot;&gt;&quot;</span><span class="p">),</span> <span class="ow">not</span> <span class="n">the</span> <span class="n">filename</span> <span class="n">of</span> <span class="n">fasta</span> <span class="n">file</span><span class="o">.</span> <span class="n">If</span>
                        <span class="n">multiple</span> <span class="n">genomes</span> <span class="n">need</span> <span class="n">to</span> <span class="n">be</span> <span class="n">assigned</span><span class="p">,</span> <span class="n">please</span> <span class="n">use</span>
                        <span class="n">spaces</span> <span class="n">to</span> <span class="n">separate</span> <span class="n">the</span> <span class="n">genomes</span><span class="o">.</span>
  <span class="o">--</span><span class="n">closed_embl_files</span> <span class="n">CLOSED_EMBL_FILES</span> <span class="p">[</span><span class="n">CLOSED_EMBL_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">ce</span> <span class="n">CLOSED_EMBL_FILES</span> <span class="p">[</span><span class="n">CLOSED_EMBL_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">The</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">embl</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">closed_gbk_files</span> <span class="n">CLOSED_GBK_FILES</span> <span class="p">[</span><span class="n">CLOSED_GBK_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">cg</span> <span class="n">CLOSED_GBK_FILES</span> <span class="p">[</span><span class="n">CLOSED_GBK_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">If</span> <span class="n">you</span> <span class="n">have</span> <span class="n">no</span> <span class="n">embl</span> <span class="n">file</span><span class="p">,</span> <span class="n">you</span> <span class="n">can</span> <span class="n">assign</span> <span class="n">genbank</span>
                        <span class="n">files</span><span class="o">.</span> <span class="n">The</span> <span class="n">genbank</span> <span class="n">can</span> <span class="n">be</span> <span class="n">ended</span> <span class="n">by</span> <span class="o">.</span><span class="n">gbk</span><span class="p">,</span> <span class="o">.</span><span class="n">gbff</span> <span class="ow">or</span> <span class="o">.</span><span class="n">gb</span>
  <span class="o">--</span><span class="n">closed_fasta_files</span> <span class="n">CLOSED_FASTA_FILES</span> <span class="p">[</span><span class="n">CLOSED_FASTA_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">cf</span> <span class="n">CLOSED_FASTA_FILES</span> <span class="p">[</span><span class="n">CLOSED_FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">The</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">closed</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">updated_fasta_files</span> <span class="n">UPDATED_FASTA_FILES</span> <span class="p">[</span><span class="n">UPDATED_FASTA_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">uf</span> <span class="n">UPDATED_FASTA_FILES</span> <span class="p">[</span><span class="n">UPDATED_FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">The</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">updated</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">ratt_path</span> <span class="n">RATT_PATH</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">start</span><span class="o">.</span><span class="n">ratt</span><span class="o">.</span><span class="n">sh</span> <span class="n">file</span> <span class="n">of</span> <span class="n">RATT</span> <span class="n">folder</span><span class="o">.</span> <span class="n">Default</span>
                        <span class="ow">is</span> <span class="n">start</span><span class="o">.</span><span class="n">ratt</span><span class="o">.</span><span class="n">sh</span><span class="o">.</span>
  <span class="o">--</span><span class="n">element</span> <span class="n">ELEMENT</span><span class="p">,</span> <span class="o">-</span><span class="n">e</span> <span class="n">ELEMENT</span>
                        <span class="o">--</span><span class="n">element</span> <span class="n">will</span> <span class="n">become</span> <span class="n">the</span> <span class="n">prefix</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">output</span> <span class="n">file</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transfer_type</span> <span class="n">TRANSFER_TYPE</span><span class="p">,</span> <span class="o">-</span><span class="n">t</span> <span class="n">TRANSFER_TYPE</span>
                        <span class="n">The</span> <span class="n">transfer</span> <span class="nb">type</span> <span class="k">for</span> <span class="n">running</span> <span class="n">RATT</span><span class="o">.</span> <span class="p">(</span><span class="n">For</span> <span class="n">the</span> <span class="n">details</span><span class="p">,</span>
                        <span class="n">please</span> <span class="n">refer</span> <span class="n">to</span> <span class="n">the</span> <span class="n">manual</span> <span class="n">of</span> <span class="n">RATT</span><span class="o">.</span><span class="p">)</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="n">Strain</span><span class="o">.</span>
  <span class="o">--</span><span class="n">convert_to_gff_rnt_ptt</span><span class="p">,</span> <span class="o">-</span><span class="n">g</span>
                        <span class="n">Convert</span> <span class="n">the</span> <span class="n">annotation</span> <span class="n">to</span> <span class="n">gff</span><span class="p">,</span> <span class="n">rnt</span> <span class="ow">and</span> <span class="n">ptt</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files from <a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">RATT</a>
will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/annotation_transfer</span></code>.</p>
<p><strong>Annotation files</strong> (<code class="docutils literal"><span class="pre">.gff</span></code>, <code class="docutils literal"><span class="pre">.ptt</span></code>, <code class="docutils literal"><span class="pre">.rnt</span></code>) will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/updated_references/annotations</span></code>.</p>
</div>
<div class="section" id="snp-snp-calling">
<span id="snp"></span><h2>snp (SNP calling)<a class="headerlink" href="#snp-snp-calling" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">snp</span></code> can analyze the alignment files and fasta files to detect mutations by running
<a class="reference external" href="https://github.com/samtools">Samtools</a> and <a class="reference external" href="https://github.com/samtools">Bcftools</a>.
There are multiple programs which can be applied to detect mutations
(with BAQ, without BAQ and extend BAQ) and there are multiple flag options to set filters
(QUAL, DP, DP4, etc.). Moreover, <code class="docutils literal"><span class="pre">snp</span></code> can also be used for generating the fasta files of
query genomes if it is necessary.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="https://github.com/samtools">Samtools</a>.</p>
<p><a class="reference external" href="https://github.com/samtools">Bcftools</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>BAM files:</strong> BAM files from fragmented/conventional libraries or TEX +/- treated libraries both can be accepted.
For assigning the files, please follow the format &#8211; <code class="docutils literal"><span class="pre">$SET_NAME:$SAMPLE_NUM:$BAMFILE1,$BAMFILE2,...</span></code>.
For example, the user has four bam files and these files are for two different conditions. Then the input will be
<code class="docutils literal"><span class="pre">set1:2:sample1.bam,sample2.bam</span> <span class="pre">set2:2:sample3.bam,sample4.bam</span></code>. <code class="docutils literal"><span class="pre">$SAMPLE_NUM</span></code> in
this example is 2 which means there are two samples in set1 and set2.
<code class="docutils literal"><span class="pre">$SAMPLE_NUM</span></code> will influence <code class="docutils literal"><span class="pre">--dp4_cutoff</span></code>, <code class="docutils literal"><span class="pre">--indel_fraction</span></code>, and
<code class="docutils literal"><span class="pre">--read_depth_range</span></code> which will be used to filter the mutations.</p>
<p><strong>Fasta files of the closed genomes</strong> or <strong>Fasta files of the query genomes</strong></p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic snp [-h] [--project_path [PROJECT_PATH]] --bam_type BAM_TYPE
                     --program PROGRAM [PROGRAM ...] --fasta_files FASTA_FILES
                     [FASTA_FILES ...] --bam_files BAM_FILES [BAM_FILES ...]
                     [--samtools_path SAMTOOLS_PATH]
                     [--bcftools_path BCFTOOLS_PATH] [--quality QUALITY]
                     [--read_depth_range READ_DEPTH_RANGE] [--ploidy PLOIDY]
                     [--rg_tag] [--caller CALLER] [--dp4_cutoff DP4_CUTOFF]
                     [--indel_fraction INDEL_FRACTION]
                     [--filter_tag_info FILTER_TAG_INFO [FILTER_TAG_INFO ...]]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --bam_type BAM_TYPE, -t BAM_TYPE
                        Please assign the type of BAM. If the BAM files are
                        produced by mapping to the closed genome of the query
                        genome, please key in &quot;closed_genome&quot;. the mutations
                        between the closed genome and query genome can be
                        detected for generating sequence of the query genome.
                        If the BAM files are produced by mapping to exact
                        query genome, please key in &quot;query_genome&quot;. The exact
                        mutations of query genome can be detected.
  --program PROGRAM [PROGRAM ...], -p PROGRAM [PROGRAM ...]
                        Please assign the program for detecting SNP of
                        transcript: &quot;with_BAQ&quot;, &quot;without_BAQ&quot;, &quot;extend_BAQ&quot;.
                        Multi-programs can be executed at the same time
                        (separated by spaces). For example: with_BAQ
                        without_BAQ extend_BAQ.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        Paths of the genome fasta files.
  --bam_files BAM_FILES [BAM_FILES ...], -b BAM_FILES [BAM_FILES ...]
                        The input format is
                        $SET_NAME:$SAMPLE_NUM:$BAMFILE1,$BAMFILE2,...
                        $SAMPLE_NUM means the input bam files
                        (samples/replicates) should be considered as the given
                        number of samples. $SAMPLE_NUM also influences
                        --dp4_cutoff, --indel_fraction, and
                        --read_depth_range.For example, if sample1.bam,
                        sample2.bam, and sample3.bam are three
                        samples/replicates of one condition. You can assign
                        set1:2:sample1.bam,sample2.bam,sample3.bam. This means
                        these three bam files should be computed together. And
                        these bam files (samples/replicates) should be
                        considered as 2 samples. If multiple samples need to
                        be assigned, please use spaces to separate the input.

additional arguments:
  --samtools_path SAMTOOLS_PATH
                        If you want to assign the path of samtools, please
                        assign here.
  --bcftools_path BCFTOOLS_PATH
                        If you want to assign the path of bcftools, please
                        assign here.
  --quality QUALITY, -q QUALITY
                        The minimum quality of a real mutation. Default is 40.
  --read_depth_range READ_DEPTH_RANGE, -d READ_DEPTH_RANGE
                        Range of the read depth of a real mutation. The format
                        is $MIN,$MAX. This value can be assigned by different
                        types: 1. real number (&quot;r&quot;), 2. times of $SAMPLE_NUM
                        (assigned by --bam_files) (&quot;n&quot;) or 3. times of the
                        average read depth (&quot;a&quot;). For example, n_10,a_3 is
                        assigned, the average read depth is 70 and the
                        $SAMPLE_NUM is 4. Then, n_10 will be 40 (10 *
                        $SAMPLE_NUM) and a_3 will be 140 (average read depth *
                        3). Based on the same example, if this value is
                        r_10,a_3, the minimum read depth will become exact 10
                        reads. If you don&#39;t want to set maximum read depth,
                        you can assign &quot;none&quot;. Default is n_10,none.
  --ploidy PLOIDY, -pl PLOIDY
                        The query bacteria is haploid or diploid. Default is
                        haploid.
  --rg_tag, -R          This function is for one BAM file which includes multi
                        samples (opposite of --ignore-RG in samtools). Default
                        is False.
  --caller CALLER, -c CALLER
                        The types of caller - consensus-caller or
                        multiallelic-caller. For details, please check
                        bcftools. &quot;c&quot; represents consensus-caller. &quot;m&quot;
                        represents multiallelic-caller. Default is m.
  --dp4_cutoff DP4_CUTOFF, -D DP4_CUTOFF
                        The cutoff of DP4. DP4 is compose of four numbers: the
                        reads covering the reference forward bases (number 1),
                        reference reverse bases (number 2), alternate forward
                        bases (number 3) and alternate reverse bases (number
                        4). Two values need to be assigned, ex: n_10,0.8. The
                        first value is for (number 3 + number 4). This value
                        can be assigned based on 1. real number (&quot;r&quot;), 2.
                        times of $SAMPLE_NUM (assigned by --bam_files) (&quot;n&quot;)
                        or 3. times of average read depth (&quot;a&quot;). The second
                        value is for (number 3 + number 4) / (number 1 +
                        number 2 + number 3 + number 4). These two values are
                        split by commas. For example, n_10,0.8 is assigned and
                        the average read depth is 70 and $SAMPLE_NUM is 4. It
                        means that the sum of number 3 and number 4 should be
                        higher than 40 (10 * $SAMPLE_NUM), and the fraction --
                        (number 3 + number 4) / (number 1 + number 2 + number
                        3 + number 4) should be higher than 0.8. Based on the
                        same example, if r_10,0.8 is assigned, the sum of read
                        depth of number 3 and number 4 will become exact 10
                        reads. Default is n_10,0.8.
  --indel_fraction INDEL_FRACTION, -if INDEL_FRACTION
                        This value is the minimum IDV and IMF which supports
                        insertion of deletion. The minimum IDV can be assigned
                        by different types: 1. real number (&quot;r&quot;), 2. times of
                        $SAMPLE_NUM (assigned by --bam_files) (&quot;n&quot;) or 3.
                        times of the average read depth (&quot;a&quot;). For example,
                        n_10,0.8 is assigned, the average read depth is 70 and
                        $SAMPLE_NUM is 4. It means that IDV should be higher
                        than 40 (10 * $SAMPLE_NUM), and IMF should be higher
                        than 0.8. Based on the same example, if r_10,0.8 is
                        assigned, the minimum IDV will become exact 10 reads.
                        Default is n_10,0.8 and the two numbers are separated
                        by commas.
  --filter_tag_info FILTER_TAG_INFO [FILTER_TAG_INFO ...], -ft FILTER_TAG_INFO [FILTER_TAG_INFO ...]
                        This function can set more filters to improve the
                        results. Please assign 1. the tag, 2. bigger (&quot;b&quot;) or
                        smaller (&quot;s&quot;) and 3. value for filters. For example,
                        &quot;RPB_b0.1,MQ0F_s0&quot; means that RPB should be bigger
                        than 0.1 and MQ0F should be smaller than 0. Default is
                        RPB_b0.1,MQSB_b0.1,MQB_b0.1,BQB_b0.1.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>If <code class="docutils literal"><span class="pre">bam_type</span></code> is <code class="docutils literal"><span class="pre">closed_genome</span></code>,
the results will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/compare_closed_and_updated_references</span></code>.
If <code class="docutils literal"><span class="pre">bam_type</span></code> is <code class="docutils literal"><span class="pre">query_genome</span></code>, the results will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/mutations_of_query_genomes</span></code>.</p>
<p>The output folders and results are following:</p>
<p><strong>SNP_raw_output:</strong> Stores output tables which be only considered read depth and QUAL.</p>
<blockquote>
<div><strong>VCF Table (only consider read depth and QUAL):</strong> Filename is <code class="docutils literal"><span class="pre">$GENOME_$PROGRAM_$SAMPLE.vcf</span></code>.</div></blockquote>
<p><strong>SNP_table:</strong> Stores two types of output tables</p>
<blockquote>
<div><p><strong>VCF Table (consider all filters):</strong> Filename is <code class="docutils literal"><span class="pre">$GENOME_$PROGRAM_$SAMPLE_best.vcf</span></code>.</p>
<p><strong>Index of fasta files:</strong>: Filename is <code class="docutils literal"><span class="pre">$GENOME_$PROGRAM_$SAMPLE_seq_reference.csv</span></code>.
The meaning of this file is like following example:</p>
</div></blockquote>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Staphylococcus_aureus_HG003</span>     <span class="mi">1632629</span> <span class="o">.</span>       <span class="n">AaA</span>     <span class="n">AA</span>      <span class="mi">57</span>      <span class="o">.</span>
<span class="n">Staphylococcus_aureus_HG003</span>     <span class="mi">1632630</span> <span class="o">.</span>       <span class="n">aA</span>      <span class="n">a</span>       <span class="mi">57</span>      <span class="o">.</span>
<span class="n">Staphylococcus_aureus_HG003</span>     <span class="mi">1499572</span> <span class="o">.</span>       <span class="n">T</span>       <span class="n">TT</span><span class="p">,</span><span class="n">TTTTT</span>        <span class="mf">43.8525</span> <span class="o">.</span>
</pre></div>
</div>
<p>The example contains &#8220;position conflict&#8221; and &#8220;mutation conflict&#8221;.
As a result, the conflicts will affect the other mutation&#8217;s positions.
Therefore, it will generate four different fasta files. The first two lines are &#8220;position conflict&#8221;, and
the last line is &#8220;mutation conflict&#8221;.
<code class="docutils literal"><span class="pre">$GENOME_$PROGRAM_$SAMPLE_seq_reference.csv</span></code> is the index for these four fasta files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="mi">1</span>       <span class="mi">1632629</span> <span class="mi">1</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TT</span>      <span class="n">Staphylococcus_aureus_HG003</span>
<span class="mi">1</span>       <span class="mi">1632629</span> <span class="mi">2</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TTTTT</span>   <span class="n">Staphylococcus_aureus_HG003</span>
<span class="mi">2</span>       <span class="mi">1632630</span> <span class="mi">1</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TT</span>      <span class="n">Staphylococcus_aureus_HG003</span>
<span class="mi">2</span>       <span class="mi">1632630</span> <span class="mi">2</span>       <span class="mi">1499572</span><span class="p">:</span><span class="n">TTTTT</span>   <span class="n">Staphylococcus_aureus_HG003</span>
</pre></div>
</div>
<p>The first column is the index of the &#8220;position conflict&#8221;.
The second column is the selected position.
The third one is the index of the &#8220;mutations conflict&#8221;.
The fourth one is the selected position and nucleotides.
The last column is the genome name.</p>
<p><strong>Potential fasta files</strong>: Filename is <code class="docutils literal"><span class="pre">$FASTANAME_$SAMPLE_$STRIANNAME_$INDEXofPOSITIONCONNFLICT_$INDEXofMUTATIONCONFLICT.fa</span></code>,
and it is stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/seqs</span></code>.
Based on the example in <strong>Index of fasta files</strong>, <code class="docutils literal"><span class="pre">Staphylococcus_aureus_HG003_set1_Staphylococcus_aureus_HG003_1_1.fa</span></code>
will be generated based on the first line of <code class="docutils literal"><span class="pre">$GENOME_$PROGRAM_seq_reference.csv</span></code>.
<code class="docutils literal"><span class="pre">Staphylococcus_aureus_HG003_set1_Staphylococcus_aureus_HG003_1_2.fa</span></code> and will be generated based on the second line of
<code class="docutils literal"><span class="pre">$GENOME_$PROGRAM_seq_reference.csv</span></code> and so forth.</p>
<p><strong>statistics</strong>: Stores the statistic files and figures, ex: the distribution of SNPs based on QUAL.</p>
</div>
<div class="section" id="tss-ps-tss-and-processing-site-prediction">
<span id="tss-ps"></span><h2>tss_ps (TSS and processing site prediction)<a class="headerlink" href="#tss-ps-tss-and-processing-site-prediction" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">tss_ps</span></code> can generate the TSS and processing sites via running
<a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>. Since the parameters can affect the
results strongly, <code class="docutils literal"><span class="pre">optimize_tss_ps</span></code> can obtain the optimized parameters of
<a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>. Please check the section
<a class="reference internal" href="#optimize-tss-ps"><span class="std std-ref">optimize_tss_ps (optimization of TSS and processing site detection)</span></a> for details.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Wiggle files of TEX +/-:</strong> Please check the section <a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic"><span class="std std-ref">The input format of libraries for running ANNOgesic</span></a> for assigning correct format.</p>
<p><strong>Fasta files of the reference genomes</strong></p>
<p><strong>GFF files of the reference genomes</strong></p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the manual detected TSSs:</strong> If gff file of the manual detected TSSs can be provided, <code class="docutils literal"><span class="pre">tss_ps</span></code> can merge the manual detected TSSs
and TSSpredator predicted ones.</p>
<p><strong>Gff files of transcripts:</strong> If comparing TSSs with transcripts is required, gff files of the transcripts need to be assigned.
For the transcripts, please check the section <a class="reference internal" href="#transcript"><span class="std std-ref">transcript (transcript detection)</span></a>.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic tss_ps [-h] [--project_path [PROJECT_PATH]]
                        [--program PROGRAM] --fasta_files FASTA_FILES
                        [FASTA_FILES ...] --annotation_files ANNOTATION_FILES
                        [ANNOTATION_FILES ...] --tex_notex_libs TEX_NOTEX_LIBS
                        [TEX_NOTEX_LIBS ...]
                        [--replicate_tex REPLICATE_TEX [REPLICATE_TEX ...]]
                        --condition_names CONDITION_NAMES
                        [CONDITION_NAMES ...]
                        [--tsspredator_path TSSPREDATOR_PATH]
                        [--specify_genomes SPECIFY_GENOMES [SPECIFY_GENOMES ...]]
                        [--height HEIGHT [HEIGHT ...]]
                        [--height_reduction HEIGHT_REDUCTION [HEIGHT_REDUCTION ...]]
                        [--factor FACTOR [FACTOR ...]]
                        [--factor_reduction FACTOR_REDUCTION [FACTOR_REDUCTION ...]]
                        [--enrichment_factor ENRICHMENT_FACTOR [ENRICHMENT_FACTOR ...]]
                        [--processing_factor PROCESSING_FACTOR [PROCESSING_FACTOR ...]]
                        [--base_height BASE_HEIGHT [BASE_HEIGHT ...]]
                        [--utr_length UTR_LENGTH] [--fuzzy FUZZY]
                        [--cluster CLUSTER]
                        [--manual_files MANUAL_FILES [MANUAL_FILES ...]]
                        [--genome_lengths GENOME_LENGTHS [GENOME_LENGTHS ...]]
                        [--validate_gene]
                        [--compare_transcript_files COMPARE_TRANSCRIPT_FILES [COMPARE_TRANSCRIPT_FILES ...]]
                        [--re_check_orphan]
                        [--overlap_feature OVERLAP_FEATURE]
                        [--reference_gff_files REFERENCE_GFF_FILES [REFERENCE_GFF_FILES ...]]
                        [--remove_low_expression REMOVE_LOW_EXPRESSION]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --program PROGRAM, -p PROGRAM
                        Which feature you want to predict, please assign &quot;TSS&quot;
                        or &quot;processing_site&quot;. Default is TSS.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        Paths of the query genome fasta files.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the query genome gff files.
  --tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...], -tl TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]
                        The libraries of TEX+/- wig files. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --replicate_tex REPLICATE_TEX [REPLICATE_TEX ...], -rt REPLICATE_TEX [REPLICATE_TEX ...]
                        This value is the minimal number of replicates that a
                        TSS has to be detected. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATE. If different
                        --replicate_tex values need to be assigned to
                        different conditions, please use spaces to separate
                        them. For example, 1_2 2_2 3_3. It means that
                        --replicate_tex is 2 in number 1 and number 2
                        conditions. In number 3 condition, --replicate_tex is
                        3. For assigning the same --replicate_tex to all
                        conditions, just use like all_1 (--replicate_tex is 1
                        in all conditions). Default is all_1.
  --condition_names CONDITION_NAMES [CONDITION_NAMES ...], -cn CONDITION_NAMES [CONDITION_NAMES ...]
                        The output prefix of all conditions. If multiple
                        conditions need to be assigned, please use spaces to
                        separate them. For example, prefix_condition1
                        prefix_condition2.

additional arguments:
  --tsspredator_path TSSPREDATOR_PATH
                        If you want to assign the path of TSSpredator, please
                        assign here. Default is /usr/local/bin/TSSpredator.jar
  --specify_genomes SPECIFY_GENOMES [SPECIFY_GENOMES ...], -ss SPECIFY_GENOMES [SPECIFY_GENOMES ...]
                        If you want to assign different parameters to
                        different genomes, Please input the genome names that
                        you want to compute. The genome names should be
                        separated by spaces. Default is running all genomes
                        based on the same parameter.
  --height HEIGHT [HEIGHT ...], -he HEIGHT [HEIGHT ...]
                        This value relates to the minimal number of read
                        starts at a certain genomic position to be considered
                        as a TSS candidate. If --specify_genomes is assigned,
                        please input --height based on the order of
                        --specify_genomes. Ex: if --specify_genomes is s1 s2
                        and --height is 0.3 0.4, it means the height of s1 is
                        0.3 and the height of s2 is 0.4. Default is 0.3.
  --height_reduction HEIGHT_REDUCTION [HEIGHT_REDUCTION ...], -rh HEIGHT_REDUCTION [HEIGHT_REDUCTION ...]
                        When comparing different genomes/conditions and the
                        step height threshold is reached in at least one
                        genome/condition, the threshold is reduced for the
                        other genomes/conditions by the value set here. This
                        value must be smaller than the step height threshold.
                        If --specify_genomes is assigned, please input
                        --height_reduction based on the order of
                        --specify_genomes like the statement of --height.
                        Default is 0.2.
  --factor FACTOR [FACTOR ...], -fa FACTOR [FACTOR ...]
                        This is the minimal factor by which the TSS height has
                        to exceed the local expression background. If
                        --specify_genomes is assigned, please input --factor
                        based on the order of --specify_genomes like the
                        statement of --height. Default is 2.0.
  --factor_reduction FACTOR_REDUCTION [FACTOR_REDUCTION ...], -rf FACTOR_REDUCTION [FACTOR_REDUCTION ...]
                        When comparing different genomes/conditions and the
                        step factor threshold is reached in at least one
                        genome/condition, the threshold is reduced for the
                        other genomes/conditions by the value set here. This
                        value must be smaller than the step factor threshold.
                        If --specify_genomes is assigned, please input
                        --factor_reduction based on the order of
                        --specify_genomes like the statement of --height.
                        Default is 0.5.
  --enrichment_factor ENRICHMENT_FACTOR [ENRICHMENT_FACTOR ...], -ef ENRICHMENT_FACTOR [ENRICHMENT_FACTOR ...]
                        This is the minimal enrichment factor. If
                        --specify_genomes is assigned, please input
                        --enrichment_factor based on the order of
                        --specify_genomes like the statement of --height.
                        Default is 2.0.
  --processing_factor PROCESSING_FACTOR [PROCESSING_FACTOR ...], -pf PROCESSING_FACTOR [PROCESSING_FACTOR ...]
                        This is the minimal processing factor. If untreated
                        library is higher than the treated library and above
                        which the TSS candidate is considered as a processing
                        site and not annotated as detected. If
                        --specify_genomes is assigned, please input
                        --processing_factor based on the order of
                        --specify_genomes like the statement of --height.
                        Default is 1.5.
  --base_height BASE_HEIGHT [BASE_HEIGHT ...], -bh BASE_HEIGHT [BASE_HEIGHT ...]
                        This is the minimal number of reads should be mapped
                        on TSS. If --specify_genomes is assigned, please input
                        --base_height based on the order of --specify_genomes
                        like the statement of --height.Default is 0.0.
  --utr_length UTR_LENGTH, -u UTR_LENGTH
                        The length of UTR. It is for Primary and Secondary
                        TSSs. Default is 300.
  --fuzzy FUZZY, -fu FUZZY
                        If --compare_transcript_files is provided, please
                        assign the fuzzy for comparing TSS and transcript.
                        Default is 5.
  --cluster CLUSTER, -c CLUSTER
                        This value defines the maximal distance (nucleotides)
                        between TSS candidates have to be clustered together.
                        If the distance between these multiple TSSs is smaller
                        or equal to this value, only one of them will be
                        printed out. Default is 2.
  --manual_files MANUAL_FILES [MANUAL_FILES ...], -m MANUAL_FILES [MANUAL_FILES ...]
                        If gff files of the manual checked TSS are provided,
                        this function will merge manual checked ones and
                        TSSpredator predicted ones. Please assign the path of
                        manual-checked TSS gff files.
  --genome_lengths GENOME_LENGTHS [GENOME_LENGTHS ...], -le GENOME_LENGTHS [GENOME_LENGTHS ...]
                        If --manual_files is assigned, Please specify the
                        genome length of input genomes. If you want to compare
                        whole genome, please type &quot;all&quot;. The input format is
                        $GENOME:SLENGTH. Multiple genomes can be accepted,
                        please use spaces to separate them. For example,
                        test.gff contain two genomes (s1 and s2). s1 was
                        manual checked 100kb and s2 was checked whole genome.
                        The value of this argument will be s1:100000 s2:all.
                        Default setting will compute all genomes in manual-
                        detected TSS gff files with whole length of genome.
  --validate_gene, -v   Using TSS candidates to validate genes in annotation
                        file. it will be store in statistics folder. Default
                        is False.
  --compare_transcript_files COMPARE_TRANSCRIPT_FILES [COMPARE_TRANSCRIPT_FILES ...], -ta COMPARE_TRANSCRIPT_FILES [COMPARE_TRANSCRIPT_FILES ...]
                        If the paths of transcript gff files are provided,
                        this function will compare TSS and transcript to
                        obtain the overlap information. Default is False.
  --re_check_orphan, -ro
                        If there is no information of gene or locus_tag in
                        genome annotation gff file, all TSSs will be assigned
                        to orphan TSSs by TSSpredator. The function can
                        compare TSSs with CDSs to classify the TSS correctly.
                        Default is False.
  --overlap_feature OVERLAP_FEATURE, -of OVERLAP_FEATURE
                        If processing site and TSS are overlap, you can keep
                        &quot;TSS&quot; or &quot;processing_site&quot; or &quot;both&quot;. Default is both.
  --reference_gff_files REFERENCE_GFF_FILES [REFERENCE_GFF_FILES ...], -rg REFERENCE_GFF_FILES [REFERENCE_GFF_FILES ...]
                        If --overlap_feature is &quot;TSS&quot; or &quot;processing_site&quot;,
                        --reference_gff_files need to be assigned. For TSS,
                        please assign the folder of processing site. For
                        processing_site, please assign the folder of TSS. If
                        --overlap_feature is &quot;both&quot;, please don&#39;t use this
                        function (Default). Default is None (keep both).
  --remove_low_expression REMOVE_LOW_EXPRESSION, -rl REMOVE_LOW_EXPRESSION
                        If you want to remove low expressed TSS/processing
                        site, please assign the file of manual-checked gff
                        file here. This function will remove the low expressed
                        ones based on comparison of manual-checked ones and
                        predicted ones. BE CAREFUL: This function may remove
                        some True positives as sell. Please make sure you want
                        to do it.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>The results of TSS are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/TSSs</span></code>, and the results of processing site
are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/processing_sites</span></code>.</p>
<p>The output folders are following:</p>
<p><strong>MasterTables:</strong> MasterTable from <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p><strong>statistics:</strong> Statistic files.</p>
<blockquote>
<div><p><strong>Venn Figures of TSS types:</strong> Filename is <code class="docutils literal"><span class="pre">TSS_venn_$GENOME.png</span></code>.</p>
<p><strong>TSS types with corresponding amounts:</strong> Table is <code class="docutils literal"><span class="pre">stat_TSS_class_$GENOME.csv</span></code>, and Figure is <code class="docutils literal"><span class="pre">TSS_class_$GENOME.png</span></code>.</p>
<p><strong>Conditions with corresponding amounts:</strong> <code class="docutils literal"><span class="pre">stat_TSS_libs_$GENOME.csv</span></code> stores all combination of conditions with corresponding amounts.
<code class="docutils literal"><span class="pre">TSSstatistics.tsv</span></code> stores the number of TSS which can be detected or missing in each condition.</p>
<p><strong>Comparing TSSs with other features:</strong> <code class="docutils literal"><span class="pre">stat_compare_TSS_transcript_$GENOME.csv</span></code> is for comparing TSSs with transcripts.
<code class="docutils literal"><span class="pre">stat_gene_vali_$GENOME.csv</span></code> is for comparing TSS with genome annotations.</p>
<p><strong>Comparing manual detected TSSs and predicted TSSs:</strong> In <code class="docutils literal"><span class="pre">stat_compare_TSSpredator_manual_$GENOME.csv</span></code>, the accuracy of TSS prediction can be found.</p>
</div></blockquote>
<p><strong>configs</strong>: Configuration files for running TSSpredator.</p>
<p><strong>gffs</strong>: Output gff files of TSSs. Some useful information can be found in the tags of the attributes within the TSS gff files.
Based on this information, we can know the details of the specific TSS. The tags are as following:</p>
<blockquote>
<div><p><strong>method:</strong> Stores the information that this TSS is detected by manual detection or <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p><strong>type:</strong> TSS type of this TSS. It could be Primary, Secondary, Internal, Antisense or Orphan.</p>
<p><strong>utr_length:</strong> UTR length of this TSS.</p>
<p><strong>associated_gene</strong>: Which genes are associated with this TSS.</p>
<p><strong>Parent:</strong> Presents the parent transcripts of this TSS, if the user has compared TSS with the transcript.</p>
<p><strong>libs:</strong> Shows in which libraries the TSS can be detected.</p>
</div></blockquote>
</div>
<div class="section" id="transcript-transcript-detection">
<span id="transcript"></span><h2>transcript (transcript detection)<a class="headerlink" href="#transcript-transcript-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">transcript</span></code> can detect transcripts based on the coverage. Most of the transcript assembly tools are
focus on eukaryotic transcript. Due to this, we constructed a subcommand which is based on the nucleotide coverage data,
given gene annotations and several parameters that can be set by the user.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Wiggle files of fragmented/conventional libraries or TEX+/- treated libraries:</strong> For importing the information about libraries, please check the section
<a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic"><span class="std std-ref">The input format of libraries for running ANNOgesic</span></a>.</p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>TSS gff files:</strong> If the user wants to compare transcripts with TSSs, TSS gff files are required.</p>
<p><strong>Genome anntation gff files:</strong> If the user wants to compare transcripts with genome annotations or modify transcript by genome annotations,
genome annotation gff files are required. There are four options for modification of transcripts:</p>
<blockquote>
<div><p><strong>merge_overlap:</strong> If multiple transcripts overlap the same gene, they will be merged as one complete transcript.</p>
<p><strong>extend_3end:</strong> If the transcript starts at the upstream of the gene and ends within the gene,
the end point of the transcript will be extended to the end point of gene.</p>
<p><strong>extend_5end:</strong> If the transcript starts within the gene and ends at the downstream of gene,
the starting point of the transcript will be extended to the starting point of the gene.</p>
<p><strong>within_extend_ends:</strong> If the transcript is within the gene, the two ends of the transcript will be
extended to the two ends of gene.</p>
<p><strong>none:</strong> Transcripts will not be modified by the genome annotations</p>
</div></blockquote>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic transcript [-h] [--project_path [PROJECT_PATH]]
                            [--annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...]]
                            [--modify_transcript MODIFY_TRANSCRIPT [MODIFY_TRANSCRIPT ...]]
                            [--tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]]
                            [--frag_libs FRAG_LIBS [FRAG_LIBS ...]]
                            [--replicate_tex REPLICATE_TEX [REPLICATE_TEX ...]]
                            [--replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...]]
                            [--tex_notex TEX_NOTEX] [--length LENGTH]
                            [--height HEIGHT] [--width WIDTH]
                            [--tolerance TOLERANCE]
                            [--tolerance_coverage TOLERANCE_COVERAGE]
                            [--tss_files TSS_FILES [TSS_FILES ...]]
                            [--compare_feature_genome COMPARE_FEATURE_GENOME [COMPARE_FEATURE_GENOME ...]]
                            [--tss_fuzzy TSS_FUZZY] [--table_best]
                            [--terminator_files TERMINATOR_FILES [TERMINATOR_FILES ...]]
                            [--terminator_fuzzy TERMINATOR_FUZZY]
                            [--max_length_distribution MAX_LENGTH_DISTRIBUTION]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        If paths of the genome annotation gff files.
  --modify_transcript MODIFY_TRANSCRIPT [MODIFY_TRANSCRIPT ...], -mt MODIFY_TRANSCRIPT [MODIFY_TRANSCRIPT ...]
                        If --annotation_files is provided, you can assign how
                        to modify the transcripts by genome annotations. There
                        are five opetions. 1. &quot;merge_overlap&quot;: if multiple
                        transcripts overlap the same gene, they will be merged
                        as one complete transcript. 2. &quot;extend_3end&quot;: if the
                        transcript starts at the upstream of the gene and ends
                        within the gene, the end point of the transcript will
                        be extended to the end point of gene. 3.
                        &quot;extend_5end&quot;: if the transcript starts within the
                        gene and ends at the downstream of gene, the starting
                        point of the transcript will be extended to the
                        starting point of the gene. 4. &quot;within_extend_ends&quot;:
                        if the transcript is within the gene, the two ends of
                        the transcript will be extended to the two ends of
                        gene. 5. &quot;none&quot;: the transcript will not be modified
                        by the genome annotations. If you want to assign
                        mutliple modifications to this function, please use
                        spaces to separated them. Default is merge_overlapped.
  --tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...], -tl TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]
                        If the TEX+/- libraries can be provided, please assign
                        the name of TEX+/- library. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --frag_libs FRAG_LIBS [FRAG_LIBS ...], -fl FRAG_LIBS [FRAG_LIBS ...]
                        If the fragmented (or conventional) libraries can be
                        provided, please assign the name of fragmented
                        library. The format is: wig_file_path:fragmented(frag)
                        :condition_id(integer):replicate_id(alphabet):strand(+
                        or -). If multiple wig files need to be assigned,
                        please use spaces to separate the wig files. For
                        example, $WIG_PATH_1:frag:1:a:+
                        $WIG_PATH_2:frag:1:a:-.
  --replicate_tex REPLICATE_TEX [REPLICATE_TEX ...], -rt REPLICATE_TEX [REPLICATE_TEX ...]
                        This value (for TEX+/- libraries) is the minimal
                        number of replicates that a transcript has to be
                        detected. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATE. If different
                        --replicate_tex values need to be assigned to
                        different conditions, please use spaces to separate
                        them. For example, 1_2 2_2 3_3. It means that
                        --replicate_tex is 2 in number 1 and number 2
                        conditions. In number 3 condition, --replicate_tex is
                        3. For assigning the same --replicate_tex to all
                        conditions, just use like all_1 (--replicate_tex is 1
                        in all conditions). Default is all_1.
  --replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...], -rf REPLICATE_FRAG [REPLICATE_FRAG ...]
                        The meaning and input type is the same to
                        --replicates_tex. This value is for fragmented (or
                        conventional) libraries.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        If the libraries of TEX+/- need to be provided, please
                        assign this value as well. This value is that a
                        transcript should be detected in both (TEX+ and TEX-)
                        or can be detected in only one library (TEX+ or TEX-).
                        Please assign 1 or 2. Default is 1.

additional arguments:
  --length LENGTH, -l LENGTH
                        The minimum length of the transcript after modifying
                        by genome annotation. If --annotation_files is
                        assigned, this value will be for the final output.
                        Otherwise, --width will be the minimum length for the
                        final output. Default is 20.
  --height HEIGHT, -he HEIGHT
                        The minimum coverage of the transcript. If --tex_notex
                        is 1, coverage of TEX+ or TEX- libraries should higher
                        than this value. If --tex_notex is 2, the function
                        will compute the average of coverage of TEX+ and TEX-
                        libraries, and the average should higher than the
                        minimum coverage. The default is 10.
  --width WIDTH, -w WIDTH
                        The minimum length of the transcript without modifying
                        by genome annotation. This value will be for the final
                        output if --annotation_files is not provided.
                        Otherwise, --length would be the minimum length of the
                        transcript for the final output. The default is 20.
  --tolerance TOLERANCE, -t TOLERANCE
                        This value defines the number of nucleotides that
                        coverages drop below --height can be ignore in one
                        transcript. The default is 5.
  --tolerance_coverage TOLERANCE_COVERAGE, -tc TOLERANCE_COVERAGE
                        If the coverage is lower than tolerance_coverage, even
                        the length is within --tolerance, the algorithm will
                        still divide the current transcript to two parts.
                        Default is 0.
  --tss_files TSS_FILES [TSS_FILES ...], -ct TSS_FILES [TSS_FILES ...]
                        If the paths of TSS files are assigned here, this
                        function will compare transcripts with TSSs to detect
                        the overlap.
  --compare_feature_genome COMPARE_FEATURE_GENOME [COMPARE_FEATURE_GENOME ...], -cf COMPARE_FEATURE_GENOME [COMPARE_FEATURE_GENOME ...]
                        If --compare_genome_annotation is provided, please
                        assign the feature which you want to compare. Default
                        is None. If multiple features need to be assigned,
                        just insert spaces between each feature, such as gene
                        CDS.
  --tss_fuzzy TSS_FUZZY, -tf TSS_FUZZY
                        If --compare_TSS is assigned, please type the fuzzy
                        for comparing TSS with transcript here. Default is 5.
  --table_best, -tb     The output table only includes the information of the
                        highest expressed library. Default is False.
  --terminator_files TERMINATOR_FILES [TERMINATOR_FILES ...], -e TERMINATOR_FILES [TERMINATOR_FILES ...]
                        If the paths of terminator gff files are assigned
                        here, this function will compare transcripts with
                        terminators to detect the parent transcript of
                        terminator. Default is None.
  --terminator_fuzzy TERMINATOR_FUZZY, -ef TERMINATOR_FUZZY
                        If --terminator_files is assigned, please assign the
                        fuzzy here. Default is 30.
  --max_length_distribution MAX_LENGTH_DISTRIBUTION, -mb MAX_LENGTH_DISTRIBUTION
                        For generating the figure of distribution of
                        transcript length, please assign the maximum length
                        that you want to include. Default is 2000.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/transcripts</span></code>.</p>
<p>The generated output folders are as following:</p>
<p><strong>tables:</strong> Table of transcript with more details. The meanings of the columns in the table are following:</p>
<blockquote>
<div><p><strong>Genome:</strong> Genome name.</p>
<p><strong>Name:</strong> Transcript name in the gff file.</p>
<p><strong>Start:</strong> Starting point of this transcript.</p>
<p><strong>End:</strong> End point of this transcript.</p>
<p><strong>Strand:</strong> Strand of this transcript.</p>
<p><strong>Detect_lib_type:</strong> This transcript can be detected in fragmented/conventional or TEX+/- libraries.</p>
<p><strong>Associated_gene:</strong> Which genes are associated with this transcript.</p>
<p><strong>Associated_tss:</strong> Which TSSs are located on this transcript.</p>
<p><strong>Associated_term:</strong> Which terminators are associated with this transcript.</p>
<p><strong>Coverage_details:</strong> Stores the average coverage information of all libraries about this transcript.</p>
</div></blockquote>
<p><strong>statistics:</strong> Stores statistic files.</p>
<blockquote>
<div><p><strong>Comparing transcript with other features:</strong> <code class="docutils literal"><span class="pre">stat_compare_transcript_genome_$GENOMENAME.csv</span></code> is
for comparing transcript with genome annotation, <code class="docutils literal"><span class="pre">stat_compare_transcript_TSS_$GENOMENAME.csv</span></code> is for comparing
transcript with TSS, and <code class="docutils literal"><span class="pre">stat_compare_transcript_terminator_$GENOMENAME.csv</span></code> is for comparing
transcript with terminator.</p>
<p><strong>Figure of the distribution of transcript length:</strong> <code class="docutils literal"><span class="pre">$GENOME_length_all.png</span></code> is for analyzing of all transcript length.
<code class="docutils literal"><span class="pre">$GENOME_length_less_$LENGTH.png</span></code> is for the analyzing of the assigned length.</p>
</div></blockquote>
<p><strong>gffs:</strong> Stores gff files of transcripts. Some useful information can be found in the tags of the attributes within the transcript gff file.
Based on this information, we can know the details of the specific transcript. The tags are as following:</p>
<blockquote>
<div><p><strong>compare_$FEATURE:</strong> State of overlap between transcripts and features
(If <code class="docutils literal"><span class="pre">--compare_feature_genome</span></code> and <code class="docutils literal"><span class="pre">--annotation_files</span></code> are assigned). The value may be &#8220;cover&#8221;, &#8220;right_shift&#8221;, &#8220;left_shift&#8221;, &#8220;within&#8221; or &#8220;no_related&#8221;.</p>
<p><strong>associated_tss:</strong> Shows which TSSs are located on this transcript (If <code class="docutils literal"><span class="pre">--tss_files</span></code> is assigned).</p>
<p><strong>associated_term:</strong> Shows which terminators are located on this transcript (If <code class="docutils literal"><span class="pre">--terminator_files</span></code> is assigned).</p>
<p><strong>associated_$FEATURE:</strong> Shows that the features are located on this transcript
(If <code class="docutils literal"><span class="pre">--compare_feature_genome</span></code> and <code class="docutils literal"><span class="pre">--annotation_files</span></code> are assigned).</p>
<p><strong>detect_lib:</strong> This transcript is detected by Tex-treated libraries or fragmented/conventional libraries.</p>
<p><strong>best_avg_coverage:</strong> The average coverage of the highest expressed library within this transcript.</p>
</div></blockquote>
</div>
<div class="section" id="terminator-terminator-detection">
<span id="terminator"></span><h2>terminator (terminator detection)<a class="headerlink" href="#terminator-terminator-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">terminator</span></code> will predict the rho-independent terminators. <code class="docutils literal"><span class="pre">ANNOgesic</span></code> combines the results of
two methods in order to get more reliable candidates. The first method is using <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a>.
The other one detects the specific secondary structure between converging pairs
of transcripts and CDSs. <code class="docutils literal"><span class="pre">ANNOgesic</span></code> can check the coverages in order to generate the terminators
which have coverage significant decrease.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a></p>
<p><strong>RNAfold</strong> of <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Fasta files of the genome sequences</strong></p>
<p><strong>Wiggle files of TEX +/- treated libraries or fragmented/conventional libraries</strong></p>
<p><strong>Gff files of the transcripts</strong></p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic terminator [-h] [--project_path [PROJECT_PATH]] --fasta_files
                            FASTA_FILES [FASTA_FILES ...] --annotation_files
                            ANNOTATION_FILES [ANNOTATION_FILES ...]
                            --transcript_files TRANSCRIPT_FILES
                            [TRANSCRIPT_FILES ...]
                            [--tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]]
                            [--frag_libs FRAG_LIBS [FRAG_LIBS ...]]
                            [--tex_notex TEX_NOTEX]
                            [--replicate_tex REPLICATE_TEX [REPLICATE_TEX ...]]
                            [--replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...]]
                            [--transterm_path TRANSTERM_PATH]
                            [--expterm_path EXPTERM_PATH]
                            [--rnafold_path RNAFOLD_PATH]
                            [--srna_files SRNA_FILES [SRNA_FILES ...]]
                            [--decrease DECREASE]
                            [--fuzzy_detect_coverage FUZZY_DETECT_COVERAGE]
                            [--fuzzy_within_transcript FUZZY_WITHIN_TRANSCRIPT]
                            [--fuzzy_downstream_transcript FUZZY_DOWNSTREAM_TRANSCRIPT]
                            [--fuzzy_within_gene FUZZY_WITHIN_GENE]
                            [--fuzzy_downstream_gene FUZZY_DOWNSTREAM_GENE]
                            [--highest_coverage HIGHEST_COVERAGE]
                            [--table_best] [--window_size WINDOW_SIZE]
                            [--window_shift WINDOW_SHIFT]
                            [--min_loop_length MIN_LOOP_LENGTH]
                            [--max_loop_length MAX_LOOP_LENGTH]
                            [--min_stem_length MIN_STEM_LENGTH]
                            [--max_stem_length MAX_STEM_LENGTH]
                            [--miss_rate MISS_RATE]
                            [--min_u_tail_length MIN_U_TAIL_LENGTH]
                            [--range_u_tail RANGE_U_TAIL] [--keep_multi_term]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        Paths of the genome fasta files.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the genome annotation gff files.
  --transcript_files TRANSCRIPT_FILES [TRANSCRIPT_FILES ...], -a TRANSCRIPT_FILES [TRANSCRIPT_FILES ...]
                        Paths of the transcript gff files.
  --tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...], -tl TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]
                        If the libraries of TEX+/- can be provided, please
                        assign the name of TEX+/- library. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --frag_libs FRAG_LIBS [FRAG_LIBS ...], -fl FRAG_LIBS [FRAG_LIBS ...]
                        If the fragmented (or conventional) libraries can be
                        provided, please assign the name of fragmented
                        library. The format is: wig_file_path:fragmented(frag)
                        :condition_id(integer):replicate_id(alphabet):strand(+
                        or -). If multiple wig files need to be assigned,
                        please use spaces to separate the wig files. For
                        example, $WIG_PATH_1:frag:1:a:+
                        $WIG_PATH_2:frag:1:a:-.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        If the libraries of TEX+/- can be provided, please
                        assign this value as well. This value is that the
                        terminator should be detected in both (TEX+ and TEX-)
                        or can be detected in only one library (TEX+ or TEX-).
                        Please assign 1 or 2. Default is 1.
  --replicate_tex REPLICATE_TEX [REPLICATE_TEX ...], -rt REPLICATE_TEX [REPLICATE_TEX ...]
                        This value (for TEX+/- libraries) is the minimal
                        number of replicates that a terminator has to be
                        detected. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATE. If different
                        --replicate_tex values need to be assigned to
                        different conditions, please use spaces to separate
                        them. For example, 1_2 2_2 3_3. It means that
                        --replicate_tex is 2 in number 1 and number 2
                        conditions. In number 3 condition, --replicate_tex is
                        3. For assigning the same --replicate_tex to all
                        conditions, just use like all_1 (--replicate_tex is 1
                        in all conditions). Default is all_1.
  --replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...], -rf REPLICATE_FRAG [REPLICATE_FRAG ...]
                        The meaning and input type is the same as
                        --replicates_tex. This value is for fragmented (or
                        conventional) libraries.

additional arguments:
  --transterm_path TRANSTERM_PATH
                        Please assign the path of &quot;transterm&quot; in TransTermHP.
  --expterm_path EXPTERM_PATH
                        Please assign the path of expterm.dat for TransTermHP.
                        Default is /usr/local/bin/expterm.dat
  --rnafold_path RNAFOLD_PATH
                        If you want to assign the path of &quot;RNAfold&quot; of Vienna
                        package, please assign here.
  --srna_files SRNA_FILES [SRNA_FILES ...], -sr SRNA_FILES [SRNA_FILES ...]
                        If you want to include sRNA information to detect
                        terminator, please assign the paths of sRNA gff files.
  --decrease DECREASE, -d DECREASE
                        This value is maximum ratio -- (lowest coverage /
                        highest coverage) within (or nearby) the terminator.
                        If the ratio is smaller than --decrease, the candidate
                        will be considered as the terminator which has
                        coverage dramatic decreasing. Default is 0.5.
  --fuzzy_detect_coverage FUZZY_DETECT_COVERAGE, -fc FUZZY_DETECT_COVERAGE
                        This value is the extended region (nucleotides) of the
                        terminators for detecting coverage significant
                        decreasing. Ex: the location of terminator is 300-400,
                        and --fuzzy_detect_coverage is 30. If the coverage
                        decrease is detected within 270-430, this candidate
                        will be still considered as the terminator which have
                        coverage dramatic decrease. Default is 30.
  --fuzzy_within_transcript FUZZY_WITHIN_TRANSCRIPT, -fut FUZZY_WITHIN_TRANSCRIPT
                        If the candidates are within transcript and the
                        distance (nucleotides) between the end of
                        gene/transcript and terminator is within this value,
                        the candidate will be considered as a terminator.
                        Otherwise, it will be removed. Default is 30.
  --fuzzy_downstream_transcript FUZZY_DOWNSTREAM_TRANSCRIPT, -fdt FUZZY_DOWNSTREAM_TRANSCRIPT
                        The meaning is similar to --fuzzy_within_transcript.
                        This value is for the candidates which are downstream
                        of transcript. Default is 30.
  --fuzzy_within_gene FUZZY_WITHIN_GENE, -fuc FUZZY_WITHIN_GENE
                        The meaning is similar to --fuzzy_within_transcript.
                        This value is for gene in stead of transcript. Default
                        is 10.
  --fuzzy_downstream_gene FUZZY_DOWNSTREAM_GENE, -fdg FUZZY_DOWNSTREAM_GENE
                        The meaning is similar to
                        --fuzzy_downstream_transcript. This value is for gene
                        in stead of transcript. Default is 310.
  --highest_coverage HIGHEST_COVERAGE, -hc HIGHEST_COVERAGE
                        The highest coverage of terminator must be higher than
                        this value. The low expressed terminator will not be
                        included in &quot;best_candidates&quot;, but still in
                        &quot;all_candidates&quot;. Default is 10.
  --table_best, -tb     Output table only contains the information of the
                        library which has most significant coverage decrease.
                        Default is False.
  --window_size WINDOW_SIZE, -wz WINDOW_SIZE
                        Window size for searching secondary structure of
                        intergenic region. Default is 100 nts.
  --window_shift WINDOW_SHIFT, -ws WINDOW_SHIFT
                        The number of nucleotides for window shift. Default is
                        20 nts.
  --min_loop_length MIN_LOOP_LENGTH, -ml MIN_LOOP_LENGTH
                        The minimum loop length of terminator. Default is 3
                        nts.
  --max_loop_length MAX_LOOP_LENGTH, -Ml MAX_LOOP_LENGTH
                        The maximum loop length of terminator. Default is 10
                        nts.
  --min_stem_length MIN_STEM_LENGTH, -ms MIN_STEM_LENGTH
                        The minimum stem length of terminator. Default is 4
                        nts.
  --max_stem_length MAX_STEM_LENGTH, -Ms MAX_STEM_LENGTH
                        The maximum stem length of terminator. Default is 20
                        nts.
  --miss_rate MISS_RATE, -mr MISS_RATE
                        The percentage of nucleotides which can be no pair in
                        the stem. Default is 0.25.
  --min_u_tail_length MIN_U_TAIL_LENGTH, -mu MIN_U_TAIL_LENGTH
                        The minimum U-tail length of terminator. Default is 3
                        nts.
  --range_u_tail RANGE_U_TAIL, -ru RANGE_U_TAIL
                        The range (nucleotides) for detection of U-tail. For
                        example, if --range_u_tail is 6 and
                        --min_u_tail_length is 3, and there are 3 Us within 6
                        nts, This candidate will be assigned as the terminator
                        which has poly U-tail. Default is 6.
  --keep_multi_term, -kp
                        Sometimes, one gene is associated with more terminator
                        candidates. In default, it will only keep the high
                        confident one. This function can keep all terminators
                        which associated with the same gene. Default is False.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/terminators</span></code>.</p>
<p>The output folders are as following:</p>
<p><strong>statistics:</strong> Stores statistic files.</p>
<blockquote>
<div><p><strong>Terminator detection method with corresponding amounts:</strong> Filename is <code class="docutils literal"><span class="pre">stat_$GENOME.csv</span></code>.</p>
<p><strong>Comparing terminators with transcripts:</strong> Based on different types of terminators,
the files are <code class="docutils literal"><span class="pre">stat_compare_terminator_transcript_$GENOME_all_candidates.csv</span></code>,
<code class="docutils literal"><span class="pre">stat_comparison_terminator_transcript_$GENOME_best.csv</span></code> and <code class="docutils literal"><span class="pre">stat_comparison_terminator_transcript_$GENOME_express.csv</span></code></p>
</div></blockquote>
<p><strong>transtermhp_results:</strong> Store any output of <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a>.</p>
<p><strong>gffs:</strong> Store gff files of terminators.</p>
<p>There are four different sub-folders for storing different gff files.</p>
<blockquote>
<div><p><strong>all_candidates:</strong> Stores all terminators which <code class="docutils literal"><span class="pre">ANNOgesic</span></code> can detect.</p>
<p><strong>expressed_candidates:</strong> Stores the terminators revealing gene expression.</p>
<p><strong>best_candidates:</strong> Stores the terminators which reveal gene expression and show dramatic decrease of its coverage.</p>
<p><strong>non_expressed_candidates:</strong> Stores the terminators which has no gene expression.</p>
</div></blockquote>
<p>Some useful information can be found in the tags of the attributes within the terminator gff file.
Based on this information, we can know the details of the specific terminator. The tags are as following:</p>
<blockquote>
<div><p><strong>method:</strong> By which method the terminator is detected.</p>
<p><strong>coverage_decrease:</strong> The terminators coverage reveals dramatic decrease or not.</p>
<p><strong>express:</strong> The terminator reveals gene expression or not.</p>
<p><strong>diff_coverage:</strong> This value shows the library which reveals strongest coverage decreasing.</p>
<p><strong>associated_gene:</strong> Which genes are associated with this terminator.</p>
<p><strong>Parent:</strong> This tag presents the parent transcript of the terminator.</p>
</div></blockquote>
<p><strong>tables:</strong> Stores tables of terminators with more details.</p>
<p>There are four different sub-folders for storing different tables.</p>
<blockquote>
<div><p><strong>all_candidates:</strong> Stores all terminators which <code class="docutils literal"><span class="pre">ANNOgesic</span></code> can detect.</p>
<p><strong>express_candidates:</strong> Stores the terminators revealing gene expression.</p>
<p><strong>best_candidates:</strong> Stores the terminators which reveal gene expression and show dramatic decrease of its coverage.</p>
<p><strong>non_expressed_candidates:</strong> Stores the terminators which has no gene expression.</p>
</div></blockquote>
<p>The meanings of the columns are as following:</p>
<blockquote>
<div><p><strong>Genome:</strong> Genome name.</p>
<p><strong>Name:</strong> Name of this terminator in the gff file.</p>
<p><strong>Start:</strong> Staring point of this terminator.</p>
<p><strong>End:</strong> End point of this terminator.</p>
<p><strong>Strand:</strong> Strand of this terminator.</p>
<p><strong>Detect:</strong> This terminator is detected by which method.</p>
<p><strong>Associated_gene:</strong> Which genes are associated with this terminator.</p>
<p><strong>Associated_transcript:</strong> The parent transcript of this terminator.</p>
<p><strong>Coverage_decrease:</strong> This terminator shows dramatic decrease of its coverage or not.</p>
<p><strong>Coverage_detail:</strong> Shows the coverage information of the libraries about this terminator. &#8220;high&#8221; means the highest coverage of the libraries,
&#8220;low&#8221; means the lowest coverage of the libraries, and &#8220;diff&#8221; represents the difference between &#8220;high&#8221; and &#8220;low&#8221;. If &#8220;No_coverage_decreasing&#8221; is showed,
it means this terminator reveal gene expression but no coverage decrease. If &#8220;NA&#8221; is showed, it means that this terminator has no gene expression.</p>
</div></blockquote>
</div>
<div class="section" id="utr-utr-detection">
<span id="utr"></span><h2>utr (UTR detection)<a class="headerlink" href="#utr-utr-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">utr</span></code> can compare TSSs, CDSs/tRNAs/sRNAs, transcripts and terminators
to generate 5&#8217;UTR and 3&#8217;UTR. 5&#8217;UTRs are based on detecting the regions between TSSs and CDSs/tRNAs/sRNAs.
3&#8217;UTRs are based on detecting the
regions between the end of the transcripts and CDSs/tRNAs/sRNAs. If the input gff files of TSSs are not computed by
ANNOgesic, please use <code class="docutils literal"><span class="pre">--tss_source</span></code> to classify TSSs for the analysis.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Gff files of the TSSs</strong></p>
<p><strong>Gff files of the transcripts</strong></p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the terminators:</strong> If the information of terminators is needed, the gff files of terminators are required.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">utr</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span> <span class="o">--</span><span class="n">annotation_files</span>
                     <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">tss_files</span>
                     <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">transcript_files</span>
                     <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">terminator_files</span> <span class="n">TERMINATOR_FILES</span> <span class="p">[</span><span class="n">TERMINATOR_FILES</span> <span class="o">...</span><span class="p">]]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">tss_source</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">base_5utr</span> <span class="n">BASE_5UTR</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">utr_length</span> <span class="n">UTR_LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">base_3utr</span> <span class="n">BASE_3UTR</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">terminator_fuzzy</span> <span class="n">TERMINATOR_FUZZY</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_3utr</span> <span class="n">FUZZY_3UTR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_5utr</span> <span class="n">FUZZY_5UTR</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">tss_files</span> <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">t</span> <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">TSS</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">terminator_files</span> <span class="n">TERMINATOR_FILES</span> <span class="p">[</span><span class="n">TERMINATOR_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">e</span> <span class="n">TERMINATOR_FILES</span> <span class="p">[</span><span class="n">TERMINATOR_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">If</span> <span class="n">the</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">terminator</span> <span class="n">files</span> <span class="n">are</span> <span class="n">assigned</span> <span class="n">here</span><span class="p">,</span>
                        <span class="n">this</span> <span class="n">function</span> <span class="n">will</span> <span class="n">also</span> <span class="n">apply</span> <span class="n">terminator</span> <span class="n">to</span> <span class="n">detect</span>
                        <span class="mi">3</span><span class="s1">&#39;UTR.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">tss_source</span><span class="p">,</span> <span class="o">-</span><span class="n">s</span>      <span class="n">The</span> <span class="n">TSS</span> <span class="n">gff</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">generated</span> <span class="n">by</span> <span class="n">ANNOgesic</span> <span class="ow">or</span> <span class="ow">not</span><span class="o">.</span> <span class="n">If</span>
                        <span class="n">the</span> <span class="n">TSS</span> <span class="n">file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">generated</span> <span class="n">by</span> <span class="n">ANNOgesic</span><span class="p">,</span> <span class="n">this</span>
                        <span class="n">function</span> <span class="n">will</span> <span class="n">classify</span> <span class="n">the</span> <span class="n">TSSs</span> <span class="k">for</span> <span class="n">detecting</span> <span class="n">UTRs</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="kc">True</span> <span class="p">(</span><span class="kn">from</span> <span class="nn">ANNOgesic</span><span class="p">)</span><span class="o">.</span>
  <span class="o">--</span><span class="n">base_5utr</span> <span class="n">BASE_5UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">b5</span> <span class="n">BASE_5UTR</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">information</span> <span class="k">for</span> <span class="n">detection</span> <span class="n">of</span> <span class="mi">5</span><span class="s1">&#39;UTR.</span>
                        <span class="n">It</span> <span class="n">can</span> <span class="n">be</span> <span class="s2">&quot;TSS&quot;</span> <span class="ow">or</span> <span class="s2">&quot;transcript&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span>
                        <span class="n">both</span><span class="o">.</span>
  <span class="o">--</span><span class="n">utr_length</span> <span class="n">UTR_LENGTH</span><span class="p">,</span> <span class="o">-</span><span class="n">l</span> <span class="n">UTR_LENGTH</span>
                        <span class="n">The</span> <span class="n">maximum</span> <span class="n">UTR</span> <span class="n">length</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">300.</span>
  <span class="o">--</span><span class="n">base_3utr</span> <span class="n">BASE_3UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">b3</span> <span class="n">BASE_3UTR</span>
                        <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">information</span> <span class="k">for</span> <span class="n">detection</span> <span class="n">of</span> <span class="mi">3</span><span class="s1">&#39;UTR.</span>
                        <span class="n">It</span> <span class="n">can</span> <span class="n">be</span> <span class="s2">&quot;transcript&quot;</span> <span class="ow">or</span> <span class="s2">&quot;terminator&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="n">transcript</span><span class="o">.</span>
  <span class="o">--</span><span class="n">terminator_fuzzy</span> <span class="n">TERMINATOR_FUZZY</span><span class="p">,</span> <span class="o">-</span><span class="n">ef</span> <span class="n">TERMINATOR_FUZZY</span>
                        <span class="n">This</span> <span class="ow">is</span> <span class="n">only</span> <span class="k">for</span> <span class="o">--</span><span class="n">base_3utr</span> <span class="n">which</span> <span class="ow">is</span> <span class="n">assigned</span> <span class="n">by</span>
                        <span class="s2">&quot;transcript&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="ow">and</span> <span class="n">terminator</span> <span class="n">file</span> <span class="n">are</span>
                        <span class="n">provided</span><span class="o">.</span> <span class="n">If</span> <span class="n">the</span> <span class="n">distance</span> <span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span> <span class="n">between</span>
                        <span class="n">terminator</span> <span class="ow">and</span> <span class="n">the</span> <span class="n">end</span> <span class="n">of</span> <span class="n">transcript</span> <span class="ow">is</span> <span class="n">lower</span> <span class="n">than</span>
                        <span class="n">this</span> <span class="n">value</span><span class="p">,</span> <span class="n">the</span> <span class="n">terminator</span> <span class="ow">is</span> <span class="n">consider</span> <span class="n">to</span> <span class="n">be</span>
                        <span class="n">associated</span> <span class="k">with</span> <span class="n">the</span> <span class="mi">3</span><span class="s1">&#39;UTR. Default is 30.</span>
  <span class="o">--</span><span class="n">fuzzy_3utr</span> <span class="n">FUZZY_3UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">f3</span> <span class="n">FUZZY_3UTR</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">base_3utr</span> <span class="n">includes</span> <span class="n">transcript</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span>
                        <span class="n">fuzzy</span> <span class="n">of</span> <span class="mi">3</span><span class="s1">&#39;UTR. Default is 10 nucleotides.</span>
  <span class="o">--</span><span class="n">fuzzy_5utr</span> <span class="n">FUZZY_5UTR</span><span class="p">,</span> <span class="o">-</span><span class="n">f5</span> <span class="n">FUZZY_5UTR</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">base_5utr</span> <span class="n">includes</span> <span class="n">transcript</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span> <span class="n">the</span>
                        <span class="n">fuzzy</span> <span class="n">of</span> <span class="mi">5</span><span class="s1">&#39;UTR. Default is 5 nucleotides.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files of 5&#8217;UTRs are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/UTRs/5UTRs</span></code>.</p>
<p>Output files of 3&#8217;UTRs are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/UTRs/3UTRs</span></code>.</p>
<p>The output folders are as following:</p>
<p><strong>gffs:</strong> Stores gff files of the 5&#8217;UTR/3&#8217;UTR.
Some useful information can be found in the tags of the attributes within the UTR gff file.
Based on this information, we can know the details of the specific UTR. The tags are as following:</p>
<blockquote>
<div><p><strong>length:</strong> UTR length.</p>
<p><strong>associated_cds:</strong> Which CDSs/rRNAs/tRNAs are associated with this UTR.</p>
<p><strong>associated_gene:</strong> Which genes are associated with this UTR.</p>
<p><strong>Parent:</strong> Shows the parent transcript of this UTR.</p>
<p><strong>associated_tss:</strong> Which TSSs are associated with this 5&#8217;UTR.</p>
<p><strong>tss_type:</strong> What types of TSSs are associated with this 5&#8217;UTR.</p>
<p><strong>associated_term:</strong> Which terminators are associated with this 3&#8217;UTR.</p>
</div></blockquote>
<p><strong>statiatics:</strong> <code class="docutils literal"><span class="pre">$GFFNAME_$GENOME_$UTRTYPE_length.png</span></code> is the distribution of the UTR length.</p>
</div>
<div class="section" id="srna-srna-detection">
<span id="srna"></span><h2>srna (sRNA detection)<a class="headerlink" href="#srna-srna-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">srna</span></code> can predict different types of sRNAs. For intergenic and antisense sRNA, it
is detected via comparison of the transcripts and annotation profiles, as well as coverage files.
For UTR-derived sRNA, the detection is based on the TSSs, processing sites,
transcripts, genome annotations and coverage files. Further filters like folding free energy change,
BLAST to nr database and sRNA database can be set as well.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Gff files of the transcripts</strong></p>
<p><strong>Wiggle files of the fragmented/conventional or TEX+/- libraries:</strong> Please check the section
<a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic"><span class="std std-ref">The input format of libraries for running ANNOgesic</span></a>.</p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the TSSs:</strong> If you want to detect the UTR-derived sRNAs, it is necessary to input
TSS information. If you don&#8217;t want to detect UTR-derived sRNAs, TSS information still can be provided as a filter.
We strongly recommend input this file.</p>
<p><strong>Gff files of processing sites:</strong> For checking the sRNAs which end with processing sites. Moreover,
Some 3&#8217;UTR-derived and interCDS-derived sRNA candidates start
from processing sites not TSSs. If you don&#8217;t want to detect UTR-derived sRNAs,
This information still can be provided to increase the accuracy, especially for some
long non-coding regions. We strongly recommend input this file if you want to detect UTR-derived sRNAs.</p>
<p><strong>Promoter tables:</strong> Information of the promoter motifs can be used for prioritizing sRNA candidates via
promoters and sRNA coverage. The format should be as following:</p>
<table border="1" class="docutils">
<colgroup>
<col width="28%" />
<col width="30%" />
<col width="25%" />
<col width="18%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Genome</td>
<td>TSS_position</td>
<td>TSS_strand</td>
<td>Motif</td>
</tr>
<tr class="row-even"><td>NC_000915.1</td>
<td>237118</td>
<td>-</td>
<td>MOTIF_1</td>
</tr>
<tr class="row-odd"><td>NC_000915.1</td>
<td>729009</td>
<td>-</td>
<td>MOTIF_1</td>
</tr>
</tbody>
</table>
<p>First irow is header of the table, the last column is the name of promoter motif.
If subcommand <code class="docutils literal"><span class="pre">promoter</span></code> was implemented before, the table will be generated automatically.
Please refer to the section <a class="reference internal" href="#promoter"><span class="std std-ref">promoter (Promoter motif detection)</span></a>.</p>
<ul class="simple">
<li><strong>Filers with the corresponding input files and tools</strong></li>
</ul>
<p>There are some filters which can improve the prediction. The user can assign the information to remove false positive.
If the information is not assigned to be a filter, it still can input to the module. Then, the information will
be shown in the output files, but this information is not considered as a filter. For example, if terminator association
is not assigned to be a filter, the user still can specify the path of terminator gff files. The associated terminators
will be shown in output gff files and tables, but the sRNA candidates which are not associated with terminators will
still be included. Following is the filter names with the required files and tools.</p>
<p><strong>Secondary structure:</strong> Remove the false positive by checking the folding energy change of secondary structure.</p>
<blockquote>
<div><p><strong>Required tools:</strong></p>
<blockquote>
<div><p><a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a></p>
<p><a class="reference external" href="http://pages.cs.wisc.edu/~ghost/doc/AFPL/6.50/Ps2pdf.htm">Ps2pdf14</a></p>
</div></blockquote>
<p><strong>Required files:</strong></p>
<blockquote>
<div><strong>Fasta files of genome sequences</strong></div></blockquote>
</div></blockquote>
<p><strong>TSS:</strong> Remove the candidates which are not associated with TSSs.</p>
<blockquote>
<div><p><strong>Required files:</strong></p>
<blockquote>
<div><strong>Gff files of TSSs</strong></div></blockquote>
</div></blockquote>
<p><strong>Searching sRNA candidate in sRNA database:</strong> If homology of this sRNA candidate can be found in sRNA database,
this candidate will be included to the result without considering other filters.</p>
<blockquote>
<div><p><strong>Required tools:</strong></p>
<blockquote>
<div><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">Blast+</a></div></blockquote>
<p><strong>Required files:</strong></p>
<blockquote>
<div><strong>sRNA database:</strong> Such as <a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a>.
Format of the header should be <code class="docutils literal"><span class="pre">$ID|$GENOME|$SRNANAME</span></code>. For example,
<code class="docutils literal"><span class="pre">srn_4840|S._aureus_NCTC8325|RsaOV</span></code> The ID is srn_4840,
the strain of this sRNA is S._aureus_NCTC8325 and the name of sRNA is RsaOV.
If the format of the header is not correct, an error or non-sense results will occur.
If you want to use BSRD with proper headers, you can download it from our
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/database">Git repository</a> easily.</div></blockquote>
</div></blockquote>
<p><strong>Searching sRNA candidate in nr database:</strong> If homologs of this sRNA candidates can be found in nr database and the hit numbers are more than <code class="docutils literal"><span class="pre">--cutoff_nr_hit</span></code>,
this candidates will be removed.</p>
<blockquote>
<div><p><strong>Required tools:</strong></p>
<blockquote>
<div><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">Blast+</a></div></blockquote>
<p><strong>Required files:</strong></p>
<blockquote>
<div><strong>nr database:</strong> The file can be download from <a class="reference external" href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/">nr database</a>.</div></blockquote>
</div></blockquote>
<p><strong>Terminator:</strong> Remove the candidates which are not associated with terminators.</p>
<blockquote>
<div><p><strong>Required files:</strong></p>
<blockquote>
<div><strong>Gff files of the terminators</strong></div></blockquote>
</div></blockquote>
<p><strong>sORF:</strong> Remove the candidates which overlap sORF.</p>
<blockquote>
<div><p><strong>Required files:</strong></p>
<blockquote>
<div><strong>Gff files of the sORFs</strong></div></blockquote>
</div></blockquote>
<p><strong>Promoter:</strong> Remove the candidates which are not associated with promoter motif.</p>
<blockquote>
<div><p><strong>Required files:</strong></p>
<blockquote>
<div><strong>Tables of the promoters:</strong> Please check the Promoter Tables of this section.</div></blockquote>
</div></blockquote>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic srna [-h] [--project_path [PROJECT_PATH]]
                      [--utr_derived_srna]
                      [--filter_info FILTER_INFO [FILTER_INFO ...]]
                      --transcript_files TRANSCRIPT_FILES
                      [TRANSCRIPT_FILES ...] --annotation_files
                      ANNOTATION_FILES [ANNOTATION_FILES ...]
                      [--tss_files TSS_FILES [TSS_FILES ...]]
                      [--processing_site_files PROCESSING_SITE_FILES [PROCESSING_SITE_FILES ...]]
                      [--terminator_files TERMINATOR_FILES [TERMINATOR_FILES ...]]
                      [--fasta_files FASTA_FILES [FASTA_FILES ...]]
                      [--compute_sec_structures]
                      [--promoter_tables PROMOTER_TABLES [PROMOTER_TABLES ...]]
                      [--promoter_names PROMOTER_NAMES [PROMOTER_NAMES ...]]
                      [--sorf_files SORF_FILES [SORF_FILES ...]]
                      [--srna_database_path SRNA_DATABASE_PATH]
                      [--nr_database_path NR_DATABASE_PATH]
                      [--tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]]
                      [--frag_libs FRAG_LIBS [FRAG_LIBS ...]]
                      [--tex_notex TEX_NOTEX]
                      [--replicate_tex REPLICATE_TEX [REPLICATE_TEX ...]]
                      [--replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...]]
                      [--rnafold_path RNAFOLD_PATH]
                      [--relplot_path RELPLOT_PATH]
                      [--mountain_path MOUNTAIN_PATH]
                      [--blastn_path BLASTN_PATH] [--blastx_path BLASTX_PATH]
                      [--makeblastdb_path MAKEBLASTDB_PATH]
                      [--ps2pdf14_path PS2PDF14_PATH]
                      [--parallel_blast PARALLEL_BLAST] [--tss_source]
                      [--tss_intergenic_fuzzy TSS_INTERGENIC_FUZZY]
                      [--tss_5utr_fuzzy TSS_5UTR_FUZZY]
                      [--tss_3utr_fuzzy TSS_3UTR_FUZZY]
                      [--tss_intercds_fuzzy TSS_INTERCDS_FUZZY]
                      [--terminator_fuzzy_in_srna TERMINATOR_FUZZY_IN_SRNA]
                      [--terminator_fuzzy_out_srna TERMINATOR_FUZZY_OUT_SRNA]
                      [--min_length MIN_LENGTH] [--max_length MAX_LENGTH]
                      [--run_intergenic_tex_coverage RUN_INTERGENIC_TEX_COVERAGE]
                      [--run_intergenic_notex_coverage RUN_INTERGENIC_NOTEX_COVERAGE]
                      [--run_intergenic_fragmented_coverage RUN_INTERGENIC_FRAGMENTED_COVERAGE]
                      [--run_break_transcript RUN_BREAK_TRANSCRIPT]
                      [--run_antisense_tex_coverage RUN_ANTISENSE_TEX_COVERAGE]
                      [--run_antisense_notex_coverage RUN_ANTISENSE_NOTEX_COVERAGE]
                      [--run_antisense_fragmented_coverage RUN_ANTISENSE_FRAGMENTED_COVERAGE]
                      [--run_utr_tex_coverage RUN_UTR_TEX_COVERAGE]
                      [--run_utr_notex_coverage RUN_UTR_NOTEX_COVERAGE]
                      [--run_utr_fragmented_coverage RUN_UTR_FRAGMENTED_COVERAGE]
                      [--min_utr_coverage MIN_UTR_COVERAGE]
                      [--cutoff_energy CUTOFF_ENERGY] [--mountain_plot]
                      [--nr_format] [--srna_format] [--table_best]
                      [--decrease_intergenic_antisense DECREASE_INTERGENIC_ANTISENSE]
                      [--decrease_utr DECREASE_UTR]
                      [--fuzzy_intergenic_antisense FUZZY_INTERGENIC_ANTISENSE]
                      [--fuzzy_utr FUZZY_UTR] [--cutoff_nr_hit CUTOFF_NR_HIT]
                      [--blast_e_nr BLAST_E_NR] [--blast_e_srna BLAST_E_SRNA]
                      [--detect_srna_in_cds]
                      [--overlap_percent_cds OVERLAP_PERCENT_CDS]
                      [--ignore_hypothetical_protein]
                      [--ranking_time_promoter RANKING_TIME_PROMOTER]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --utr_derived_srna, -u
                        The function is for detecting UTR-derived sRNA.
                        Default is False.
  --filter_info FILTER_INFO [FILTER_INFO ...], -d FILTER_INFO [FILTER_INFO ...]
                        There are several filters that you can use to improve
                        sRNA detection: 1. tss (sRNA has to start with TSS),
                        2. sec_str (free energy change of secondary structure
                        (normalized by length) has to be smaller than
                        --cutoff_energy), 3. blast_nr (the number of the
                        homology can not be found more than --cutoff_nr_hit in
                        the non-redundant database), 4. blast_srna (as long as
                        the homology can be found in sRNA database, the
                        candidates will be included to best candidtes without
                        considering other filters), 5. sorf (sRNA can not
                        overlap sORF), 6. term (sRNA has to be associated with
                        a terminator), 7. promoter (sRNA has to be associated
                        with a promoter motif). ATTENTION: without importing
                        any information, the results may include many false
                        positives. If multiple filters needs to be assigned,
                        please use spaces to separated them. ex: tss sec_str
                        blast_nr - means it used 1. TSS, 2. free energy change
                        of secondary structure and 3. blast to nr database to
                        detect sRNA. If you want to use blast_srna as a
                        filter, please follow the format:
                        $ID|$GENOME|$SRNANAME. &quot;tss sec_str blast_nr
                        blast_srna&quot; is recommended to be assigned. If you
                        don&#39;t want to use any filters, please assign &quot;none&quot;.
                        Default is tss sec_str blast_nr blast_srna.
  --transcript_files TRANSCRIPT_FILES [TRANSCRIPT_FILES ...], -a TRANSCRIPT_FILES [TRANSCRIPT_FILES ...]
                        Paths of the transcript files.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the genome annotation gff files.
  --tss_files TSS_FILES [TSS_FILES ...], -t TSS_FILES [TSS_FILES ...]
                        If the paths of TSS gff files are assigned here, TSS
                        information will be used for detecting sRNA. If you
                        want to detect UTR-derived sRNA or &quot;tss&quot; in
                        --filter_info, TSS gff files MUST be provided.
  --processing_site_files PROCESSING_SITE_FILES [PROCESSING_SITE_FILES ...], -p PROCESSING_SITE_FILES [PROCESSING_SITE_FILES ...]
                        If the paths of processing site gff files are assigned
                        here, processing site information will be used for
                        detecting sRNA. For detection of UTR-derived sRNA,
                        processing site information can improve the results.
  --terminator_files TERMINATOR_FILES [TERMINATOR_FILES ...], -e TERMINATOR_FILES [TERMINATOR_FILES ...]
                        If terminator information can be provided, please
                        assign the paths of gff files of terminators. If
                        &quot;term&quot; in --filter_info, terminator gff files MUST be
                        provided.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        If &quot;sec_str&quot; or &quot;blast_nr&quot; or &quot;blast_srna&quot; is assigned
                        to --filter_info, fasta files MUST be assigned here.
  --compute_sec_structures, -cs
                        Compute secondary structures of sRNAs. Default is
                        False.
  --promoter_tables PROMOTER_TABLES [PROMOTER_TABLES ...], -pt PROMOTER_TABLES [PROMOTER_TABLES ...]
                        If the paths of promoter tables can be provided,
                        please assign the paths of promoter tables here. If
                        &quot;promoter&quot; in --filter_info, the promoter tables MUST
                        be assigned. The format of table is $GENOME
                        $TSS_POSITION $TSS_STRAND $PROMOTER_NAME. TSS
                        information is also required.
  --promoter_names PROMOTER_NAMES [PROMOTER_NAMES ...], -pn PROMOTER_NAMES [PROMOTER_NAMES ...]
                        If --promoter_tables is provided, please assign the
                        promoter name (the last column of promoter table)
                        which you want to compare. If multiple promoters need
                        to be assigned, please put spaces between the
                        promoters. Default is None.
  --sorf_files SORF_FILES [SORF_FILES ...], -O SORF_FILES [SORF_FILES ...]
                        If the paths of sORF gff files can be provided, please
                        assign the paths of sORF gff files here. If &quot;sorf&quot; in
                        --filter_info, sORF gff files MUST is assigned.
  --srna_database_path SRNA_DATABASE_PATH, -sd SRNA_DATABASE_PATH
                        If sRNA database can be provided, please assign here.
                        If &quot;blast_srna&quot; is included in --filter_info, please
                        assign the path of sRNA database with proper header.
                        Format of the header should be $ID|$GENOME|$NAME For
                        example of the proper header, please check
                        https://github.com/Sung-Huan/ANNOgesic/blob/master/dat
                        abase/sRNA_database_BSRD.fa
  --nr_database_path NR_DATABASE_PATH, -nd NR_DATABASE_PATH
                        If nr database can be provided, please assign here. If
                        &quot;blast_nr&quot; is included in --filter_info, please assign
                        the path of nr database.
  --tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...], -tl TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]
                        If TEX+/- libraries can be provided, please assign the
                        name of TEX+/- libraries here. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --frag_libs FRAG_LIBS [FRAG_LIBS ...], -fl FRAG_LIBS [FRAG_LIBS ...]
                        If fragmented (or conventional) libraries can be
                        provided, please assign the name of fragmented
                        libraries here. The format is: wig_file_path:fragmente
                        d(frag):condition_id(integer):replicate_id(alphabet):s
                        trand(+ or -). If multiple wig files need to be
                        assigned, please use spaces to separate the wig files.
                        For example, $WIG_PATH_1:frag:1:a:+
                        $WIG_PATH_2:frag:1:a:-.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        If TEX+/- libraries is assigned, this value is that a
                        sRNA should be detected in both (TEX+ and TEX-) or can
                        be detected in only one library (TEX+ or TEX-). Please
                        assign 1 or 2. Default is 2.
  --replicate_tex REPLICATE_TEX [REPLICATE_TEX ...], -rt REPLICATE_TEX [REPLICATE_TEX ...]
                        This value (for TEX+/- libraries) is the minimal
                        number of replicates that a sRNA has to be detected.
                        The format is $NUMBERofCONDITION_$NUMBERofREPLICATE.
                        If different --replicate_tex values need to be
                        assigned to different conditions, please use spaces to
                        separate them. For example, 1_2 2_2 3_3. It means that
                        --replicate_tex is 2 in number 1 and number 2
                        conditions. In number 3 condition, --replicate_tex is
                        3. For assigning the same --replicate_tex to all
                        conditions, just use like all_1 (--replicate_tex is 1
                        in all conditions). Default is all_1.
  --replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...], -rf REPLICATE_FRAG [REPLICATE_FRAG ...]
                        The meaning and input type is the same as
                        --replicates_tex. This value is for fragmented (or
                        conventional) libraries.

additional arguments:
  --rnafold_path RNAFOLD_PATH
                        Please assign RNAfold path.
  --relplot_path RELPLOT_PATH
                        Please assign the path of relplot.pl in Vienna
                        package.
  --mountain_path MOUNTAIN_PATH
                        Please assign the path of mountain.pl in Vienna
                        package.
  --blastn_path BLASTN_PATH
                        Please assign the path of blastn in blast+ package.
  --blastx_path BLASTX_PATH
                        Please assign the path of blastx in blast+ package.
  --makeblastdb_path MAKEBLASTDB_PATH
                        Please assign the path of makeblastdb in blast+
                        package.
  --ps2pdf14_path PS2PDF14_PATH
                        Please assign the path of ps2pdf14.
  --parallel_blast PARALLEL_BLAST, -pb PARALLEL_BLAST
                        How many parallels that you want to use for blast.
                        Default is 10.
  --tss_source, -ts     If the TSS gff file is not generated by ANNOgesic,
                        please use this function to classify TSSs and generate
                        the proper format for sRNA prediction. Default is True
                        (from ANNOgesic).
  --tss_intergenic_fuzzy TSS_INTERGENIC_FUZZY, -ft TSS_INTERGENIC_FUZZY
                        If --tss_files is provided, please assign the fuzzy
                        for comparing TSS with transcript. It is for
                        intergenic sRNA. Default is 3.
  --tss_5utr_fuzzy TSS_5UTR_FUZZY, -f5 TSS_5UTR_FUZZY
                        If --tss_files is provided, please assign the fuzzy
                        for comparing TSS with transcript. It is for 5&#39;UTR-
                        derived sRNA.The input type can be percentage (&quot;p&quot;) or
                        the real amount of reads (&quot;n&quot;). Ex: p_0.05 means the
                        fuzzy is 5 percent of the length of 5&#39;UTR. n_10 means
                        the fuzzy is 10 base pair. Default is n_3.
  --tss_3utr_fuzzy TSS_3UTR_FUZZY, -f3 TSS_3UTR_FUZZY
                        The meaning is similar to --tss_5utr_fuzzy. This value
                        is for 3&#39;UTR-derived sRNA instead of 5&#39;UTR-derived
                        sRNA. Default is p_0.04.
  --tss_intercds_fuzzy TSS_INTERCDS_FUZZY, -fc TSS_INTERCDS_FUZZY
                        The meaning is similar to --tss_5utr_fuzzy. This value
                        is for interCDS-derived sRNA instead of 5&#39;UTR-derived
                        sRNA. Default is p_0.04.
  --terminator_fuzzy_in_srna TERMINATOR_FUZZY_IN_SRNA, -efi TERMINATOR_FUZZY_IN_SRNA
                        If --terminator_files is provided, please assign the
                        fuzzy for comparing terminator with transcript. This
                        value is for the terminator which is within sRNA.
                        Default is 30.
  --terminator_fuzzy_out_srna TERMINATOR_FUZZY_OUT_SRNA, -efo TERMINATOR_FUZZY_OUT_SRNA
                        The meaning is the same as --terminator_fuzzy_in_sRNA.
                        This value is for the terminator which is outside of
                        sRNA. Default is 30.
  --min_length MIN_LENGTH, -lm MIN_LENGTH
                        Please assign the minimum sRNA length. Default is 30.
  --max_length MAX_LENGTH, -lM MAX_LENGTH
                        Please assign the maximum sRNA length. Default is 500.
  --run_intergenic_tex_coverage RUN_INTERGENIC_TEX_COVERAGE, -it RUN_INTERGENIC_TEX_COVERAGE
                        The minimum average coverage of intergenic sRNA
                        candidates in TEX+ libraries. This value is based on
                        different types of TSSs. The order of numbers is
                        &quot;Primary,Secondary,Internal,Antisense,Orphan&quot;. Ex: The
                        input is 0,0,0,50,10. It means that antisense TSS
                        (minimum coverage is 50) and orphan TSS (minimum
                        coverage is 10) are used for sRNA prediction. The
                        other types of TSSs will not be used for sRNA
                        detection (assign to 0). If TSS information is not
                        provided, it will choose the lowest one as a general
                        cutoff for prediction. Ex: if the input is 0,0,0,50,10
                        and --tss_files is not provided, 10 will be the
                        general cutoff for prediction. Default is 0,0,0,40,20
                        and each number is separated by commas.
  --run_intergenic_notex_coverage RUN_INTERGENIC_NOTEX_COVERAGE, -in RUN_INTERGENIC_NOTEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_tex_coverage. This value is for TEX-
                        libraries. Default is 0,0,0,30,10 and each number is
                        separated by commas.
  --run_intergenic_fragmented_coverage RUN_INTERGENIC_FRAGMENTED_COVERAGE, -if RUN_INTERGENIC_FRAGMENTED_COVERAGE
                        The meaning is the same as
                        --run_intergenic_tex_coverage. This value is for
                        fragmented (or conventional) libraries. Default is
                        400,200,0,50,20 and each number is separated by
                        commas.
  --run_break_transcript RUN_BREAK_TRANSCRIPT, -ib RUN_BREAK_TRANSCRIPT
                        Several primary/secondary TSSs are associated with
                        transcripts which has no CDSs/tRNA/rRNA inside
                        (perhaps associated with ncRNA). In order to detect
                        the sRNA candidates in these transcripts, please
                        assign the minimum average coverage of the sRNA
                        candidates. The format is $TEX,$NOTEX,$FRAG, ex:
                        200,100,100 is means that the minimum average coverage
                        is 200 for TEX+ libraries, 100 for TEX- and fragmented
                        (or conventional) libraries. Default is 30,20,30 and
                        each number is separated by commas.
  --run_antisense_tex_coverage RUN_ANTISENSE_TEX_COVERAGE, -at RUN_ANTISENSE_TEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_tex_coverage. This value is for
                        antisense in stead of intergenic. Default is
                        0,0,0,40,20 and each number is separated by commas.
  --run_antisense_notex_coverage RUN_ANTISENSE_NOTEX_COVERAGE, -an RUN_ANTISENSE_NOTEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_notex_coverage. This value is for
                        antisense in stead of intergenic. Default is
                        0,0,0,30,10 and each number is separated by commas.
  --run_antisense_fragmented_coverage RUN_ANTISENSE_FRAGMENTED_COVERAGE, -af RUN_ANTISENSE_FRAGMENTED_COVERAGE
                        The meaning is the same as
                        --run_intergenic_fragmented_coverage. This value is
                        for antisense in stead of intergenic. Default is
                        400,200,0,50,20 and each number is separated by
                        commas.
  --run_utr_tex_coverage RUN_UTR_TEX_COVERAGE, -ut RUN_UTR_TEX_COVERAGE
                        The minimum average coverage of UTR-derived sRNA
                        candidates in TEX+ libraries. The input can be
                        assigned by the percentile (&quot;p&quot;) or real number of
                        coverage (&quot;n&quot;). The order of numbers is
                        &quot;5&#39;UTR,3&#39;UTR,interCDS&quot;. Ex: if the input is
                        &quot;p_0.7,p_0.5,p_0.5&quot;, it will use 70 percentile of
                        5&#39;UTR coverage as minimum coverage for 5&#39;UTR-derived
                        sRNA, median of 3&#39;UTR and interCDS coverage as minimum
                        coverage for 3&#39;UTR and interCDS-derived sRNA. Ex: if
                        the input is &quot;n_30,n_10,n_20 &quot; it will use 30 as
                        minimum coverage for 5&#39;UTR-derived sRNA, 10 as minimum
                        coverage for 3&#39;UTR-derived sRNA and 20 as minimum
                        coverage for interCDS-derived sRNA. Default is
                        p_0.8,p_0.6,p_0.7 and each number is separated by
                        commas.
  --run_utr_notex_coverage RUN_UTR_NOTEX_COVERAGE, -un RUN_UTR_NOTEX_COVERAGE
                        The meaning is the same as --run_utr_tex_coverage.
                        This value is for TEX- libraries. Default is
                        p_0.7,p_0.5,p_0.6 and each number is separated by
                        commas.
  --run_utr_fragmented_coverage RUN_UTR_FRAGMENTED_COVERAGE, -uf RUN_UTR_FRAGMENTED_COVERAGE
                        The meaning is the same as --run_utr_tex_coverage.
                        This value is for fragmented (or conventional)
                        libraries. Default is p_0.7,p_0.5,p_0.6 and each
                        number is separated by commas.
  --min_utr_coverage MIN_UTR_COVERAGE, -mu MIN_UTR_COVERAGE
                        The minimum general coverage of UTR-derived sRNA. The
                        coverage of UTR-derived sRNA should not only fit the
                        --run_utr_TEX_coverage, --run_utr_noTEX_coverage and
                        --run_utr_fragmented_coverage, but also this value.
                        Default is 50.
  --cutoff_energy CUTOFF_ENERGY, -ce CUTOFF_ENERGY
                        If &quot;sec_str&quot; is included in --filter_info, please
                        assign the maximum folding energy change (normalized
                        by length of gene). Default is -0.05.
  --mountain_plot, -m   This function will generate mountain plot of sRNA
                        candidate. Default is False.
  --nr_format, -nf      This function will format nr database. If the nr
                        database has formatted, the step can be skip. Default
                        is False.
  --srna_format, -sf    The meaning is the same as --nr_format. It is for sRNA
                        database in stead of nr database. Default is False.
  --table_best, -tb     The output table of sRNA candidates only contains the
                        information of the highest expressed library. Default
                        is False.
  --decrease_intergenic_antisense DECREASE_INTERGENIC_ANTISENSE, -di DECREASE_INTERGENIC_ANTISENSE
                        This value is for detecting the coverage decrease in
                        intergenic/antisense transcript. If the length of
                        intergenic transcript is longer than the max_length,
                        it will check the sRNA candidates based on coverage.
                        If the ratio -- (the lowest coverage / the highest
                        coverage) of the sRNA region is smaller than this
                        value, the spot of lowest coverage will be considered
                        as the end point of the sRNA. If the new sRNA length
                        is suitable for a sRNA candidate, this candidate will
                        be included in output. Default is 0.1.
  --decrease_utr DECREASE_UTR, -du DECREASE_UTR
                        This value is for detecting the coverage decrease in
                        UTR region. The end point of UTR-derived sRNA is
                        defined by processing site or the end of transcript
                        (3&#39;UTR-derived sRNA). If there is no processing sites
                        in 5&#39;UTR or interCDS, the algorithm will check the
                        coverage to detect the end point of sRNA. If the ratio
                        -- (the lowest coverage / the highest coverage) of the
                        sRNA region is smaller than this value, the spot of
                        lowest coverage will be considered as the end point of
                        the sRNA. If the new sRNA length is suitable for a
                        sRNA candidate, this candidate will be included in
                        output. Default is 0.05.
  --fuzzy_intergenic_antisense FUZZY_INTERGENIC_ANTISENSE, -fi FUZZY_INTERGENIC_ANTISENSE
                        If the situation is like
                        --decrease_intergenic_antisense mentioned, This value
                        is the fuzzy nucleotides for detecting the coverage
                        decrease. Ex: the location of intergenic sRNA is
                        300-400, and --fuzzy_intergenic_antisense is 30. The
                        algorithm will search the coverage decrease within
                        270-430. Default is 10.
  --fuzzy_utr FUZZY_UTR, -fu FUZZY_UTR
                        It is similar with --fuzzy_intergenic_antisense. This
                        is for UTR-derived sRNAs. Default is 10.
  --cutoff_nr_hit CUTOFF_NR_HIT, -cn CUTOFF_NR_HIT
                        The maximum hits number in nr database. Default is 0.
  --blast_e_nr BLAST_E_NR, -en BLAST_E_NR
                        The maximum e value for blast in nr database. Default
                        is 0.0001.
  --blast_e_srna BLAST_E_SRNA, -es BLAST_E_SRNA
                        The maximum e value for blast in sRNA database.
                        Default is 0.0001.
  --detect_srna_in_cds, -ds
                        This function will search sRNA in CDS (ex: the genome
                        annotation is not correct). More sRNA candidates which
                        overlap with CDS will be detected. Default is False.
  --overlap_percent_cds OVERLAP_PERCENT_CDS, -oc OVERLAP_PERCENT_CDS
                        If --detect_srna_in_cds is True, please assign the
                        maximum ratio of overlap between CDS and sRNA
                        candidates. Default is 0.5
  --ignore_hypothetical_protein, -ih
                        This function is for ignoring the hypothetical
                        proteins in genome annotation file. Default is False.
  --ranking_time_promoter RANKING_TIME_PROMOTER, -rp RANKING_TIME_PROMOTER
                        If --promoter_tables is provided, the information of
                        promoter can be use for ranking sRNA candidates as
                        well. The ranking score is --ranking_time_promoter *
                        average coverage. For example, a sRNA candidate which
                        is associated with promoter and its average coverage
                        is 10. If --ranking_time_promoter is 2, the ranking
                        score will be 20 (2*10). For the candidate which are
                        not associated with promoter, the
                        --ranking_time_promoter will be 1. Therefore,
                        --ranking_time_promoter can not be smaller than 1.
                        Default is 2.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sRNAs</span></code>. the output folders and files are following:</p>
<p><strong>sRNA_2d_$GENOME:</strong> The secondary structures of all sRNA candidates.</p>
<p><strong>sRNA_seq_$GENOME:</strong> The sequences of all sRNA candidates.</p>
<p><strong>blast_results_and_misc:</strong> Stores the results of blast.</p>
<blockquote>
<div><p><strong>nr_blast_$GENOME.txt:</strong> output of BLAST for the nr database.</p>
<p><strong>sRNA_blast_$GENOME.txt:</strong> output of BLAST for the sRNA database.</p>
</div></blockquote>
<p><strong>figs:</strong> Stores the figures about secondary structures of sRNAs.</p>
<blockquote>
<div><p><strong>mountain_plots:</strong> Stores mountain plots of the sRNA candidates. Filename is as <code class="docutils literal"><span class="pre">srna10_NC_009839.1_335339_335435_+_mountain.pdf</span></code>.
&#8220;srna10&#8221;, &#8220;NC_009839.1&#8221;, &#8220;335339&#8221;, &#8220;335435&#8221;, &#8220;+&#8221; are ID of sRNA gff file, genome name, starting point, end point and strand, respectively.</p>
<p><strong>sec_plots:</strong> Stores the secondary structure plots of sRNA candidates.
Filename of is as <code class="docutils literal"><span class="pre">srna10_NC_009839.1_335339_335435_+_rss.pdf</span></code>.
&#8220;srna10&#8221;, &#8220;NC_009839.1&#8221;, &#8220;335339&#8221;, &#8220;335435&#8221;, &#8220;+&#8221; are ID of sRNA gff file, genome name, starting point, end point and strand, respectively.</p>
<p><strong>dot_plots:</strong> Stores the dot plots of sRNA candidates.
Filename of dot plot is as <code class="docutils literal"><span class="pre">srna10_NC_009839.1_335339_335435_+_dp.pdf</span></code>.
&#8220;srna10&#8221;, &#8220;NC_009839.1&#8221;, &#8220;335339&#8221;, &#8220;335435&#8221;, &#8220;+&#8221; are ID of sRNA gff file, genome name, starting point, end point and strand, respectively.</p>
</div></blockquote>
<p><strong>statistics:</strong> Stores statistics files. <code class="docutils literal"><span class="pre">stat_$GENOME_sRNA_blast.csv</span></code> is the analysis result of BLAST for sRNA databases.
<code class="docutils literal"><span class="pre">stat_sRNA_class_Staphylococcus_aureus_HG003.csv</span></code> is the classification of sRNA candidates.</p>
<p><strong>TSS_classes:</strong> If the TSSs are not computed by ANNOgesic, <code class="docutils literal"><span class="pre">TSS_classes</span></code> will be generated for classification of TSS.
TSS gff files with TSS types will be stored here.</p>
<p><strong>tables:</strong> Stores sRNA tables with more details. There are also some sub-folders:</p>
<blockquote>
<div><p><strong>for_classes:</strong> Stores the results based on different sRNA classes.</p>
<p><strong>best_candidates:</strong> Stores the best results of sRNAs after filtering.</p>
<p><strong>all_candidates:</strong> Stores all candidates without filtering.</p>
</div></blockquote>
<p>The meanings of the columns are as following:</p>
<blockquote>
<div><p><strong>Rank:</strong> Ranking number of this sRNA.</p>
<p><strong>Genome:</strong> Genome name.</p>
<p><strong>Name:</strong> sRNA Name which is shown in gff file.</p>
<p><strong>Start:</strong> Starting point of this sRNA.</p>
<p><strong>End:</strong> End point of this sRNA.</p>
<p><strong>Strand:</strong> Strand of this sRNA.</p>
<p><strong>Start_with_TSS/Cleavage_site:</strong> This sRNA starts with which TSS or cleavage site.</p>
<p><strong>End_with_cleavage:</strong> If the sRNA ends with a cleavage site, the information of this cleavage site will be showed here.</p>
<p><strong>Candidates:</strong> Position of this sRNA.</p>
<p><strong>Lib_type:</strong> This sRNA is detected by TEX+/- or fragmented/conventional library.</p>
<p><strong>Best_avg_coverage:</strong> Based on coverage of all libraries, The best average coverage of this sRNA will be showed here.</p>
<p><strong>Best_highest_coverage:</strong> Based on coverage of all libraries, The highest average coverage of this sRNA will be showed here.</p>
<p><strong>Best_lowest_coverage:</strong> Based on coverage of all libraries, The lowest average coverage of this sRNA will be showed here.</p>
<p><strong>Track/Coverage:</strong> Shows the coverage information of the libraries about this sRNA. &#8220;high&#8221; means the highest coverage of the libraries,
&#8220;low&#8221; means the lowest coverage of the libraries, and &#8220;avg&#8221; represents the average coverage of this sRNA.</p>
<p><strong>Normalized_secondary_energy_change(by_length):</strong> Secondary folding energy change (normalized by length) of this sRNA.</p>
<p><strong>sRNA_types:</strong> Shows the sRNA type.</p>
<p><strong>Conflict_sORF:</strong> If this sRNA overlaps sORF, the overlapped sORF will be showed here.</p>
<p><strong>nr_hit_number:</strong> The hit numbers of this sRNA in nr database.</p>
<p><strong>sRNA_hit_number:</strong> The hit numbers of this sRNA in sRNA database.</p>
<p><strong>nr_hit_top3|ID|e-value:</strong> The top 3 hits of this sRNA in nr database will be showed here. The information includes protein name, ID and e-value.</p>
<p><strong>sRNA_hit|e-value:</strong> If the homology of this sRNA can be found in sRNA database, the information will be showed here.</p>
<p><strong>Overlap_CDS:</strong> If the sRNA overlaps CDS, the information of the overlapped CDS will be showed here.</p>
<p><strong>Overlap_percent:</strong> If the sRNA overlaps CDS, the percentage of overlap between sRNA and CDS will be showed here.</p>
<p><strong>End_with_terminator:</strong> The terminator which is associated with this sRNA.</p>
<p><strong>Associated_promoter:</strong> The promoter which is associated with this sRNA.</p>
<p><strong>sRNA_length:</strong> sRNA length.</p>
</div></blockquote>
<p><strong>gffs:</strong> Stores gff files of the sRNA. There are also some sub-folders:</p>
<blockquote>
<div><p><strong>for_classes:</strong> Stores the results based on different sRNA classes.</p>
<p><strong>best_candidates:</strong> Stores the best results of sRNAs after filtering.</p>
<p><strong>all_candidates:</strong> Stores all candidates without filtering.</p>
</div></blockquote>
<p>Some useful information can be found in the tags of the attributes within the sRNA gff file.
Based on this information, we can know the details of the specific sRNA. The tags are as following:</p>
<blockquote>
<div><p><strong>sRNA_type:</strong> This sRNA is from 5&#8217;UTR, 3&#8217;UTR, interCDS, intergenic, antisense or within CDS.</p>
<p><strong>with_TSS:</strong> Which TSSs are related to this sRNA.</p>
<p><strong>sORF:</strong> Which sORFs overlap this sRNA.</p>
<p><strong>sRNA_hit:</strong> Blast hit numbers of this sRNA in sRNA database.</p>
<p><strong>nr_hit:</strong> Blast hit numbers of this sRNA in nr database.</p>
<p><strong>2d_energy</strong>: Normalized (by the length of sRNA) folding energy change of the sRNA secondary structure.</p>
<p><strong>with_term:</strong> Terminators which are associated with the sRNA candidate.</p>
<p><strong>end_cleavage:</strong> If this sRNA ends with a cleavage site, information of the cleavage site will be showed here.</p>
<p><strong>overlap_cds:</strong> This sRNA overlaps CDS or not.</p>
<p><strong>overlap_percentage:</strong> If this sRNA overlap CDS. The percentage of the overlap between CDS and sRNA will be showed here.</p>
<p><strong>promoter:</strong> Promoters which are associated with the sRNA.</p>
</div></blockquote>
</div>
<div class="section" id="sorf-sorf-detection">
<span id="sorf"></span><h2>sorf (sORF detection)<a class="headerlink" href="#sorf-sorf-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">sorf</span></code> can detect sORF based on searching ribosome binding sites (Shine-Dalgarno sequence),
start codons and stop codons within the non-coding expressed regions.
Since these regions may be sRNAs or sORFs, Comparison between sORFs and sRNAs can be done by this subcommand as well.
If multiple sORFs overlap with each other, this subcommand will merge them to be one sORF. Therefore, one region may contain more than one sORF.
Position of the start codon which listed in output table is assigned by the first nucleotide. The position of stop codon is assigned by the last nucleotide.
Moreover, one region may contain different frame shifts.
Ex: (200, 202, 203) are the positions of three start codons and (241, 243) are two stop codons in
a small transcript. There are three sORF candidates (200-241, 203-241 and 202-243).</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of the genome annotations</strong>
<strong>Gff files of the transcripts</strong></p>
<p><strong>Wiggle files of TEX+/- or fragmented/conventional libraries:</strong> Please refer to the section <a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic"><span class="std std-ref">The input format of libraries for running ANNOgesic</span></a>.</p>
<p><strong>fasta files of the genome sequences</strong></p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the TSSs:</strong> For checking the sORFs start from TSS or not. We strongly recommend to input this file.</p>
<p><strong>Gff files of sRNAs:</strong> For checking the overlap of sRNAs and sORFs.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic sorf [-h] [--project_path [PROJECT_PATH]]
                      [--utr_derived_sorf] --fasta_files FASTA_FILES
                      [FASTA_FILES ...] --transcript_files TRANSCRIPT_FILES
                      [TRANSCRIPT_FILES ...] --annotation_files
                      ANNOTATION_FILES [ANNOTATION_FILES ...]
                      [--tss_files TSS_FILES [TSS_FILES ...]]
                      [--srna_files SRNA_FILES [SRNA_FILES ...]]
                      [--tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]]
                      [--frag_libs FRAG_LIBS [FRAG_LIBS ...]]
                      [--tex_notex TEX_NOTEX]
                      [--replicate_tex REPLICATE_TEX [REPLICATE_TEX ...]]
                      [--replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...]]
                      [--utr_length UTR_LENGTH] [--min_length MIN_LENGTH]
                      [--max_length MAX_LENGTH]
                      [--cutoff_intergenic_coverage CUTOFF_INTERGENIC_COVERAGE]
                      [--cutoff_antisense_coverage CUTOFF_ANTISENSE_COVERAGE]
                      [--cutoff_5utr_coverage CUTOFF_5UTR_COVERAGE]
                      [--cutoff_3utr_coverage CUTOFF_3UTR_COVERAGE]
                      [--cutoff_intercds_coverage CUTOFF_INTERCDS_COVERAGE]
                      [--cutoff_background CUTOFF_BACKGROUND] [--table_best]
                      [--start_codon START_CODON [START_CODON ...]]
                      [--stop_codon STOP_CODON [STOP_CODON ...]]
                      [--min_rbs_distance MIN_RBS_DISTANCE]
                      [--max_rbs_distance MAX_RBS_DISTANCE]
                      [--rbs_not_after_tss] [--fuzzy_rbs FUZZY_RBS]
                      [--print_all_combination] [--best_no_srna]
                      [--best_no_tss]
                      [--ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --utr_derived_sorf, -u
                        This function will detect UTR-derived sORF. Default is
                        False.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        The paths of genome fasta files.
  --transcript_files TRANSCRIPT_FILES [TRANSCRIPT_FILES ...], -a TRANSCRIPT_FILES [TRANSCRIPT_FILES ...]
                        Paths of the transcript gff files.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the genome annotation gff files.
  --tss_files TSS_FILES [TSS_FILES ...], -t TSS_FILES [TSS_FILES ...]
                        If the paths of TSS gff files are assigned here, this
                        function will use the TSS information to detect sORF.
                        If running without TSS, the results may be influenced.
  --srna_files SRNA_FILES [SRNA_FILES ...], -s SRNA_FILES [SRNA_FILES ...]
                        If the paths of sRNA gff files can be provided, this
                        function will compare sORF and sRNA to detect the
                        overlap.
  --tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...], -tl TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]
                        If the TEX+/- libraries can be provided, please assign
                        the name of TEX+/- library here. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --frag_libs FRAG_LIBS [FRAG_LIBS ...], -fl FRAG_LIBS [FRAG_LIBS ...]
                        If the fragmented (or conventional) libraries can be
                        provided, please assign the name of fragmented library
                        here. The format is: wig_file_path:fragmented(frag):co
                        ndition_id(integer):replicate_id(alphabet):strand(+ or
                        -). If multiple wig files need to be assigned, please
                        use spaces to separate the wig files. For example,
                        $WIG_PATH_1:frag:1:a:+ $WIG_PATH_2:frag:1:a:-.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        If the TEX+/- libraries is provided, this value is
                        that a sORF should be detected in both (TEX+ and TEX-)
                        or can be detected in only one library (TEX+ or TEX-).
                        Please assign 1 or 2. Default is 2.
  --replicate_tex REPLICATE_TEX [REPLICATE_TEX ...], -rt REPLICATE_TEX [REPLICATE_TEX ...]
                        This value (for TEX+/- libraries) is the minimal
                        number of replicates that a sORF has to be detected.
                        The format is $NUMBERofCONDITION_$NUMBERofREPLICATE.
                        If different --replicate_tex values need to be
                        assigned to different conditions, please use spaces to
                        separate them. For example, 1_2 2_2 3_3. It means that
                        --replicate_tex is 2 in number 1 and number 2
                        conditions. In number 3 condition, --replicate_tex is
                        3. For assigning the same --replicate_tex to all
                        conditions, just use like all_1 (--replicate_tex is 1
                        in all conditions). Default is all_1.
  --replicate_frag REPLICATE_FRAG [REPLICATE_FRAG ...], -rf REPLICATE_FRAG [REPLICATE_FRAG ...]
                        The meaning and input type is the same as
                        --replicates_tex. This value is for fragmented (or
                        conventional) libraries.

additional arguments:
  --utr_length UTR_LENGTH, -ul UTR_LENGTH
                        If --tss_files is provided, please assign the utr
                        length for comparing TSS with sORF. The default number
                        is 300.
  --min_length MIN_LENGTH, -lm MIN_LENGTH
                        Please assign the minimum residue length of sORF.
                        Default is 30.
  --max_length MAX_LENGTH, -lM MAX_LENGTH
                        Please assign the maximum residue length of sORF.
                        Default is 150.
  --cutoff_intergenic_coverage CUTOFF_INTERGENIC_COVERAGE, -ci CUTOFF_INTERGENIC_COVERAGE
                        The minimum coverage of intergenic sORF candidates.
  --cutoff_antisense_coverage CUTOFF_ANTISENSE_COVERAGE, -ai CUTOFF_ANTISENSE_COVERAGE
                        The minimum coverage of antisense sORF candidates.
  --cutoff_5utr_coverage CUTOFF_5UTR_COVERAGE, -cu5 CUTOFF_5UTR_COVERAGE
                        The minimum coverage of 5&#39;UTR derived sORF candidates.
                        This value can be assigned by percentage (&quot;p&quot;) or the
                        amount of reads (&quot;n&quot;). Ex: p_0.05 means that the
                        coverage of sORF candidates should be higher than 5
                        percentile of all 5&#39;UTR transcripts. n_10 means that
                        the coverage of sORF candidates should be higher than
                        10. Default is p_0.5.
  --cutoff_3utr_coverage CUTOFF_3UTR_COVERAGE, -cu3 CUTOFF_3UTR_COVERAGE
                        The meaning is the same as --cutoff_5utr_coverage.
                        This value is for 3&#39;UTR. Default is p_0.5.
  --cutoff_intercds_coverage CUTOFF_INTERCDS_COVERAGE, -cuf CUTOFF_INTERCDS_COVERAGE
                        The meaning is the same as --cutoff_5utr_coverage.
                        This value is for interCDS. Default is p_0.5.
  --cutoff_background CUTOFF_BACKGROUND, -cub CUTOFF_BACKGROUND
                        The general minimum coverage of all sORF candidates.
                        All candidates should fit this condition as well.
                        Default is 10.
  --table_best, -tb     The output table of sORF candidates only includes
                        information of the highest expressed library. Default
                        is False.
  --start_codon START_CODON [START_CODON ...], -ac START_CODON [START_CODON ...]
                        The types of start coden. If multiple types of start
                        codon need to be assigned, please use spaces to
                        separate them. Default is ATG.
  --stop_codon STOP_CODON [STOP_CODON ...], -oc STOP_CODON [STOP_CODON ...]
                        The types of stop codon. If multiple types of stop
                        codon need to be assigned, please use spaces to
                        separate them. Default is TTA TAG TGA.
  --min_rbs_distance MIN_RBS_DISTANCE, -mr MIN_RBS_DISTANCE
                        The minimum distance (nucleotides) between the
                        ribosome binding site (Shine-Dalgarno sequence) and
                        start codon. Default is 3.
  --max_rbs_distance MAX_RBS_DISTANCE, -Mr MAX_RBS_DISTANCE
                        The maximum distance (nucleotides) between the
                        ribosome binding site (Shine-Dalgarno sequence) and
                        start codon. Default is 15.
  --rbs_not_after_tss, -at
                        For generating best results, if the ribosome binding
                        site (Shine-Dalgarno sequence) of sORF is not
                        associated with TSS, this function will include this
                        candidate as well. Default is False.
  --fuzzy_rbs FUZZY_RBS, -zr FUZZY_RBS
                        The number of nucleotides of Shine-Dalgarno sequence
                        allow be different with AGGAGG. Default is 2.
  --print_all_combination, -pa
                        Non-annotated transcript may has many start codons and
                        stop codons. This function can print all combinations
                        of start codons and stop codons. Default is False.
  --best_no_srna, -bs   For generating best results, this function can exclude
                        the sORFs which overlap with sRNA. Default is False.
  --best_no_tss, -bt    For generating best results, this function can include
                        the sORFs which are not associated with TSS. Default
                        is False.
  --ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN, -ih IGNORE_HYPOTHETICAL_PROTEIN
                        This function is for ignoring hypothetical protein in
                        genome annotation file. Default is False.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sORFs</span></code>.</p>
<p><strong>statistics</strong>: Stores statistic files.</p>
<p><strong>tables:</strong> Stores tables of the sORFs with more details. There are also some sub-folders:</p>
<blockquote>
<div><p><strong>best_candidates:</strong> Stores the best results of sORFs after filtering.</p>
<p><strong>all_candidates:</strong> Stores all candidates without filtering.</p>
</div></blockquote>
<p>The meaning of each column is as following:</p>
<blockquote>
<div><p><strong>Genome:</strong> Genome name.</p>
<p><strong>Name:</strong> the sORF name which is also shown in gff file.</p>
<p><strong>Start:</strong> Starting point of this sORF.</p>
<p><strong>End:</strong> End point of this sORF.</p>
<p><strong>Strand:</strong> Strand of this sORF.</p>
<p><strong>Type:</strong> sORF type.</p>
<p><strong>TSS:</strong> TSSs which are associated with this sORF.</p>
<p><strong>Ribosome_binding_site:</strong> Ribosome binding site (Shine-Dalgarno sequence) of this sORF.</p>
<p><strong>All_start_points:</strong> Positions of all start codons which can be found in the region of this sORF.</p>
<p><strong>All_stop_points:</strong> Positions of all stop codons which can be found in the region of this sORF.</p>
<p><strong>sRNA_conflict:</strong> If this sORF overlaps sRNA, the overlapped sRNA will be showed here.</p>
<p><strong>Frame_shift:</strong> If there are sORF candidates which can be found by frame shift,
the number of frame shift will be showed here. &#8220;1&#8221; means there
are some candidates can be found by frame shift once. &#8220;2&#8221; means there are some candidates can be found by frame shift twice.</p>
<p><strong>Lib_type:</strong> This sORF can be detected in TEX+/- or fragmented/conventional libraries.</p>
<p><strong>Best_avg_coverage:</strong> Based on coverage of all libraries, The best average coverage of this sORF will be showed here.</p>
<p><strong>Best_highest_coverage:</strong> Based on coverage of all libraries, The highest average coverage of this sORF will be showed here.</p>
<p><strong>Best_lowest_coverage:</strong> Based on coverage of all libraries, The lowest average coverage of this sORF will be showed here.</p>
<p><strong>Track_detail:</strong> Shows the coverage information of the libraries about this sORF. &#8220;high&#8221; means the highest coverage of the libraries,
&#8220;low&#8221; means the lowest coverage of the libraries, and &#8220;avg&#8221; represents the average coverage of this sORF.</p>
<p><strong>Seq:</strong> Sequence of this sORF.</p>
</div></blockquote>
<p><strong>gffs:</strong> Stores gff files of the sORFs. There are also some sub-folders:</p>
<blockquote>
<div><p><strong>best_candidates:</strong> Stores the best results of sORFs after filtering.</p>
<p><strong>all_candidates:</strong> Stores all candidates without filtering.</p>
</div></blockquote>
<p>Some useful information can be found in the tags of the attributes within the sORF gff file.
Based on this information, we can know the details of the specific sORF. The tags are as following:</p>
<blockquote>
<div><p><strong>start_TSS:</strong> Shows this sORF starts with which TSS.</p>
<p><strong>with_TSS:</strong> Which TSSs are associated with this sORF.</p>
<p><strong>sORF_type:</strong> Type of the sORF (5&#8217;UTR, 3&#8217;UTR, interCDS, intergenic, antisense or within CDS).</p>
<p><strong>sRNA:</strong> Which sRNAs are overlap with this sORF.</p>
<p><strong>rbs:</strong> Ribosome binding sites (Shine-Dalgarno sequences) of this sORF.</p>
<p><strong>frame_shift:</strong> The number of frame shifts in the regions.</p>
</div></blockquote>
</div>
<div class="section" id="promoter-promoter-motif-detection">
<span id="promoter"></span><h2>promoter (Promoter motif detection)<a class="headerlink" href="#promoter-promoter-motif-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">promoter</span></code> can scan the upstream of TSSs to discover the promoter motifs.
We integrated <a class="reference external" href="http://meme-suite.org/tools/meme">MEME</a> (for ungapped motifs) and
<a class="reference external" href="http://meme-suite.org/tools/glam2">GLAM2</a> (for gapped motifs) to predict the promoters.
Based on the tool, HTML files can be generated for visualization. If the input TSS gff file
is not computed by ANNOgesic, please add <code class="docutils literal"><span class="pre">--tss_source</span></code> to classify TSSs for predicting
promoter motifs.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://meme-suite.org/tools/meme">MEME</a>.</p>
<p><a class="reference external" href="http://meme-suite.org/tools/glam2">GLAM2</a></p>
<p><a class="reference external" href="https://http://www.mpich.org/">MPICH</a> (if parallel runs are required)</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Fasta files of the genome sequences</strong></p>
<p><strong>Gff files of the TSSs:</strong> If the input TSS gff file is not generated by ANNOgesic,
the libraries and wiggle files are necessary.
Please refer to the <a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic"><span class="std std-ref">The input format of libraries for running ANNOgesic</span></a>
in order to assign the correct format.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic promoter [-h] [--program PROGRAM]
                          [--project_path [PROJECT_PATH]] --fasta_files
                          FASTA_FILES [FASTA_FILES ...] --tss_files TSS_FILES
                          [TSS_FILES ...]
                          [--motif_width MOTIF_WIDTH [MOTIF_WIDTH ...]]
                          [--num_motifs NUM_MOTIFS]
                          [--nt_before_tss NT_BEFORE_TSS]
                          [--meme_path MEME_PATH] [--glam2_path GLAM2_PATH]
                          [--e_value E_VALUE] [--end_run END_RUN]
                          [--parallels PARALLELS] [--tss_source]
                          [--tex_libs TEX_LIBS [TEX_LIBS ...]]
                          [--annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...]]
                          [--combine_all]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --program PROGRAM, -p PROGRAM
                        Please assign the program -- meme or glam2 or both.
                        Default is both
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        Paths of genome fasta files.
  --tss_files TSS_FILES [TSS_FILES ...], -t TSS_FILES [TSS_FILES ...]
                        Paths of TSS gff files.
  --motif_width MOTIF_WIDTH [MOTIF_WIDTH ...], -w MOTIF_WIDTH [MOTIF_WIDTH ...]
                        The length for motif detection. For detecting a range
                        of length, please insert &quot;-&quot; between two values.
                        Moreover, if multiple motif length need to be
                        assigned, please use spaces to separate them. For
                        example, 50 2-10. It means that the length of motif
                        for detection is 50 and within 2 to 10. The number
                        should be less or equal than --nt_before_TSS. Default
                        is 50.
  --num_motifs NUM_MOTIFS, -n NUM_MOTIFS
                        The number of motifs that you want to detect. Default
                        is 10.
  --nt_before_tss NT_BEFORE_TSS, -b NT_BEFORE_TSS
                        The number of upstream nucleotides of TSS for promoter
                        prediction. Default is 50.

additional arguments:
  --meme_path MEME_PATH
                        path of MEME.
  --glam2_path GLAM2_PATH
                        path of GLAM2.
  --e_value E_VALUE, -e E_VALUE
                        The maximum e value for running MEME. Default is 0.05.
  --end_run END_RUN, -er END_RUN
                        If the result of GLAM2 is not improved after running
                        this number of iteration, it will be ended. Default is
                        10000.
  --parallels PARALLELS, -pl PARALLELS
                        This function can run MEME by parallel. Please input
                        the number of parallel runs.
  --tss_source, -s      If the TSS gff file is not generated by ANNOgesic,
                        this function can classify TSS and generate the proper
                        format for promoter detection. Default is True (from
                        ANNOgesic)
  --tex_libs TEX_LIBS [TEX_LIBS ...], -tl TEX_LIBS [TEX_LIBS ...]
                        If --tss_source is False, please assign the name of
                        TEX+/- library. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        If --tss_source is False, please assign the paths of
                        genome annotation gff files.
  --combine_all, -c     This function will combine all TSS files in
                        &quot;tss_files&quot; to generate global promoter motifs.
                        Default is False.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/promoters</span></code>. The output folders are following:</p>
<p><strong>allfasta:</strong> If <code class="docutils literal"><span class="pre">--combine_all</span></code> is True, it will combine all TSS files in <code class="docutils literal"><span class="pre">--tss_files</span></code>
to generate promoter motifs. The results will be stored in this folder.</p>
<p><strong>fasta_classes:</strong> The fasta files of different TSS types.</p>
<p><strong>$GENOME_NAME:</strong> Stores output of MEME and GLAM2 based on different TSS types. Two sub-folders are under this folder.</p>
<blockquote>
<div><p><strong>MEME</strong>: Stores the results of MEME.</p>
<p><strong>GLAM2</strong>: Stores the results of GLAM2.</p>
<p>The format of folders which under these two folders is <code class="docutils literal"><span class="pre">promoter_motifs_$FILENAME_$GENOME_$TSSTYPE_$PROMOTERLEGNTH</span></code>,
ex: <code class="docutils literal"><span class="pre">promoter_motifs_NC_000915.1_allgenome_internal_45_nt</span></code>.
&#8220;NC_000915.1&#8221;, &#8220;allgenome&#8221;, &#8220;primary&#8221; and &#8220;45_nt&#8221; are gff filename, genome name, TSS type and upstream nucleotides of TSS, respectively.
If genome name is &#8220;allgenome&#8221;, this means that the result is generated by the information of all genomes of gff files.
If there is only one genome in the gff file, the genome name will be assigned as &#8220;allgenome&#8221; as well. Several files are stored in the sub-folder:</p>
<blockquote>
<div><p><strong>Figures of the promoter motifs:</strong> Contains EPS and PNG files.</p>
<p><strong>Details of the promoter motifs:</strong> Contains HTML file, XML file and TXT file. These files include the TSS information.</p>
<p><strong>Promoter tables:</strong> <code class="docutils literal"><span class="pre">meme.csv</span></code> or <code class="docutils literal"><span class="pre">glam2.csv</span></code> is the promoter table which also includes the TSS information.
Moreover, it can used as an input for sRNA detection (<code class="docutils literal"><span class="pre">srna</span></code>). Please check the section <code class="docutils literal"><span class="pre">srna</span></code>.</p>
</div></blockquote>
</div></blockquote>
<p><strong>TSS_classes:</strong> If the TSSs are not computed by ANNOgesic, <code class="docutils literal"><span class="pre">TSS_classes</span></code> will be generated for classification of TSS.
TSS gff files with TSS types will be stored here.</p>
</div>
<div class="section" id="operon-operon-detection">
<span id="operon"></span><h2>operon (Operon detection)<a class="headerlink" href="#operon-operon-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">operon</span></code> will group TSSs, genes/CDSs/tRNAs/rRNAs, transcripts, terminators and UTRs to operons and
sub-operons.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of the TSSs</strong></p>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Gff files of the transcripts</strong></p>
<p><strong>Gff files of the 5&#8217;UTRs</strong></p>
<p><strong>Gff files of the 3&#8217;UTRs</strong></p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the terminators</strong></p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">operon</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span> <span class="o">--</span><span class="n">tss_files</span>
                        <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">annotation_files</span>
                        <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span>
                        <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">utr5_files</span> <span class="n">UTR5_FILES</span>
                        <span class="p">[</span><span class="n">UTR5_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">utr3_files</span> <span class="n">UTR3_FILES</span>
                        <span class="p">[</span><span class="n">UTR3_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">terminator_files</span> <span class="n">TERMINATOR_FILES</span> <span class="p">[</span><span class="n">TERMINATOR_FILES</span> <span class="o">...</span><span class="p">]]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">tss_fuzzy</span> <span class="n">TSS_FUZZY</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">terminator_fuzzy</span> <span class="n">TERMINATOR_FUZZY</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">min_length</span> <span class="n">MIN_LENGTH</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">tss_files</span> <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">t</span> <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">TSS</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">utr5_files</span> <span class="n">UTR5_FILES</span> <span class="p">[</span><span class="n">UTR5_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">u5</span> <span class="n">UTR5_FILES</span> <span class="p">[</span><span class="n">UTR5_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="mi">5</span><span class="s1">&#39;UTR gff files.</span>
  <span class="o">--</span><span class="n">utr3_files</span> <span class="n">UTR3_FILES</span> <span class="p">[</span><span class="n">UTR3_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">u3</span> <span class="n">UTR3_FILES</span> <span class="p">[</span><span class="n">UTR3_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="mi">3</span><span class="s1">&#39;UTR gff files.</span>
  <span class="o">--</span><span class="n">terminator_files</span> <span class="n">TERMINATOR_FILES</span> <span class="p">[</span><span class="n">TERMINATOR_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">e</span> <span class="n">TERMINATOR_FILES</span> <span class="p">[</span><span class="n">TERMINATOR_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">If</span> <span class="n">terminator</span> <span class="n">information</span> <span class="n">can</span> <span class="n">be</span> <span class="n">provided</span><span class="p">,</span> <span class="n">please</span>
                        <span class="n">assign</span> <span class="n">the</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">terminator</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">here</span><span class="o">.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">tss_fuzzy</span> <span class="n">TSS_FUZZY</span><span class="p">,</span> <span class="o">-</span><span class="n">tf</span> <span class="n">TSS_FUZZY</span>
                        <span class="n">The</span> <span class="n">fuzzy</span> <span class="k">for</span> <span class="n">comparing</span> <span class="n">between</span> <span class="n">TSS</span> <span class="ow">and</span> <span class="n">transcript</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="mf">5.</span>
  <span class="o">--</span><span class="n">terminator_fuzzy</span> <span class="n">TERMINATOR_FUZZY</span><span class="p">,</span> <span class="o">-</span><span class="n">ef</span> <span class="n">TERMINATOR_FUZZY</span>
                        <span class="n">The</span> <span class="n">fuzzy</span> <span class="k">for</span> <span class="n">comparing</span> <span class="n">between</span> <span class="n">terminator</span> <span class="ow">and</span>
                        <span class="n">transcript</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">30.</span>
  <span class="o">--</span><span class="n">min_length</span> <span class="n">MIN_LENGTH</span><span class="p">,</span> <span class="o">-</span><span class="n">l</span> <span class="n">MIN_LENGTH</span>
                        <span class="n">The</span> <span class="n">minimum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">operon</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">20.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/operons</span></code>. The output folders are as following:</p>
<p><strong>gffs:</strong> Stores gff files which are integrated the information of TSSs, annotations,
transcripts, 5&#8217;UTRs, and 3&#8217;UTRs and assign parent transcript to all features (presented by
<strong>Parent</strong> in attributes of gff files).</p>
<p><strong>tables:</strong> The tables of operons which store all information of operons and sub-operons.</p>
<p>The meaning of each column is as following:</p>
<blockquote>
<div><p><strong>Operon_ID:</strong> Operon ID.</p>
<p><strong>Genome:</strong> Genome name.</p>
<p><strong>Operon_position:</strong> Starting point and end point of the operon.</p>
<p><strong>Strand:</strong> Strand of the operon.</p>
<p><strong>Number_of_suboperon:</strong> The amount of sub-operons in this operon region.</p>
<p><strong>Position_of_suboperon:</strong> Starting point and end point of the sub-operons.</p>
<p><strong>Start_with_TSS:</strong> This operon starts with TSS or not.</p>
<p><strong>Number_of_TSS:</strong> The number of the TSSs which are located on this operon.</p>
<p><strong>Terminated_with_terminator:</strong> This operon ends with TSS or not.</p>
<p><strong>Number_of_terminator:</strong> The number of the terminators which are associated with this operon.</p>
<p><strong>Number_of_gene_associated_suboperon:</strong> The number of the genes which are associated with the sub-operon.</p>
<p><strong>Number_of_gene_associated_operon:</strong> The number of the genes which are associated with the operon.</p>
<p><strong>Associated_genes_with_suboperon:</strong> Locus tag of the genes which are associated with the sub-operon.</p>
<p><strong>Associated_genes_with_whole_operon:</strong> Locus tag of the genes which are associated with the operon.</p>
</div></blockquote>
<p><strong>statistics:</strong> Stores statistic file which includes the number of sub-operons, monocistronic operon, polycistronic operon, etc.</p>
</div>
<div class="section" id="circrna-circular-rna-detection">
<span id="circrna"></span><h2>circrna (circular RNA detection)<a class="headerlink" href="#circrna-circular-rna-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">circrna</span></code> can detect the potential circular RNAs via <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>.
Moreover, the false positive can be removed by checking genome annotation files and quality of splicing site detection.
The user can assign reads for detecting circular RNAs or assign BAM files to skip mapping.
BE CAREFUL, BAM files must be mapped by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>
with <code class="docutils literal"><span class="pre">--splits</span></code> or <code class="docutils literal"><span class="pre">circrna</span></code> can&#8217;t find the proper candidates.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">segemehl.x and testrealign.x in Segemehl</a>.
For generating testrealign.x, please refer to <a class="reference internal" href="required.html#required-tools-or-databases"><span class="std std-ref">Required tools or databases</span></a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Fasta files of reads or BAM files:</strong> If you want to use BAM files directly, they should be
mapped by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> with <code class="docutils literal"><span class="pre">--splits</span></code>.
The input format is <code class="docutils literal"><span class="pre">$SET_NAME:$READ1,$READ2,...</span></code> or <code class="docutils literal"><span class="pre">$SET_NAME:$BAM1,$BAM2,...</span></code>.
For example, <code class="docutils literal"><span class="pre">set1:read1.fa,read2.fa</span></code> means these two fasta files need to be computed together.
If your BAM files are generated by mapping reads on multiple reference genomes,
<a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">testrealign.x</a> may not be able to
handle them.</p>
<p><strong>Fasta files of the genome annotations</strong></p>
<p><strong>Gff files of the genome annotations</strong></p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic circrna [-h] [--project_path [PROJECT_PATH]]
                         [--read_files READ_FILES [READ_FILES ...]]
                         [--bam_files BAM_FILES [BAM_FILES ...]] --fasta_files
                         FASTA_FILES [FASTA_FILES ...] --annotation_files
                         ANNOTATION_FILES [ANNOTATION_FILES ...]
                         [--segemehl_path SEGEMEHL_PATH]
                         [--testrealign_path TESTREALIGN_PATH]
                         [--samtools_path SAMTOOLS_PATH]
                         [--parallels PARALLELS]
                         [--support_reads SUPPORT_READS]
                         [--start_ratio START_RATIO] [--end_ratio END_RATIO]
                         [--ignore_hypothetical_proteins IGNORE_HYPOTHETICAL_PROTEINS]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --read_files READ_FILES [READ_FILES ...], -rp READ_FILES [READ_FILES ...]
                        If this argument is given, ANNOgesic will map the
                        reads via segemehl (with -S). BE CAREFUL, this
                        function only use default parameters of segemehl to
                        map the reads. If some specific functions of segemehl
                        need to be implemented, please directly run segemehl
                        (MUST run with -S). please assign the fasta files. The
                        input format is $SET_NAME:$READ1,$READ2,... For
                        example, set1:read1.fa,read2.fa means these two fasta
                        files need to be compute together. .bz2 and .gz
                        compressed fasta files can be accepted as well.
  --bam_files BAM_FILES [BAM_FILES ...], -b BAM_FILES [BAM_FILES ...]
                        You can want assign the paths of Bam files to skip
                        mapping. The input format is $SET_NAME:$BAM1,$BAM2,...
                        For example, set1:bam1.bam,bam2.bam means these two
                        bam files need to be compute together. BE CAREFUL, the
                        bam files must be generated via segemehl with -S.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        Paths of the genome fasta files.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the genome annotation gff files.

additional arguments:
  --segemehl_path SEGEMEHL_PATH
                        Please assign the path of segemehl.x in segemehl
                        package.
  --testrealign_path TESTREALIGN_PATH
                        Please assign the path of testrealign.x in segemehl
                        package.
  --samtools_path SAMTOOLS_PATH
                        Please assign the path of samtools.
  --parallels PARALLELS, -p PARALLELS
                        The number of parallels processes for --align. Default
                        is 10.
  --support_reads SUPPORT_READS, -s SUPPORT_READS
                        The minimum supported reads of circular RNA. Default
                        is 10.
  --start_ratio START_RATIO, -sr START_RATIO
                        The minimum ratio -- (supported reads of circRNA / all
                        reads) at starting point of candidate. Default is 0.5.
  --end_ratio END_RATIO, -er END_RATIO
                        The meaning is similar to --start_ratio. This value is
                        for end point of candidate. Default is 0.5.
  --ignore_hypothetical_proteins IGNORE_HYPOTHETICAL_PROTEINS, -ih IGNORE_HYPOTHETICAL_PROTEINS
                        This function is for ignoring hypothetical protein in
                        genome annotation file. Default is False.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/circRNAs</span></code>. The output folders are following:</p>
<p><strong>segemehl_alignment_files:</strong> If read files are assigned, the folder is for results of mapping.</p>
<p><strong>segemehl_splice_results:</strong> The results of splicing detection. For understanding the splicing tables, please
refer to <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>.</p>
<p><strong>gffs:</strong> Stores gff files of the circular RNAs. <code class="docutils literal"><span class="pre">$GENOME_$SET_circRNA_best.gff</span></code> is gff files for the best results after checking
genome annotation and quality of splicing. <code class="docutils literal"><span class="pre">$GENOME_$SET_circRNA_all.gff</span></code> is for all candidates without filtering.
Some useful information can be found in the tags of the attributes within the circular RNA gff file.
Based on this information, we can know the details of the specific circular RNA. The tags are as following:</p>
<blockquote>
<div><p><strong>support_reads:</strong> The number of reads which support the circular RNA.</p>
<p><strong>read_at_start:</strong> (supported reads / total reads) at starting point of the circular RNA.</p>
<p><strong>read_at_end:</strong> (supported reads / total reads) at end point of the circular RNA.</p>
<p><strong>conflict:</strong> The circular RNA overlap genome annotation or not.</p>
<p><strong>method:</strong> The circular RNA is detected by which method.</p>
</div></blockquote>
<p><strong>circRNA_tables:</strong> Stores tables of the circular RNAs with more details. The meaning of each column is as following:</p>
<blockquote>
<div><p><strong>ID:</strong> ID of this circular RNA.</p>
<p><strong>Genome:</strong> Genome name.</p>
<p><strong>Strand:</strong> Strand of the circular RNA.</p>
<p><strong>Start:</strong> Starting point of the circular RNA.</p>
<p><strong>End:</strong> End point of the circular RNA.</p>
<p><strong>Annotation_overlap:</strong> If there is genome annotation which overlap this circular RNA, the overlapped feature will be showed here.</p>
<p><strong>Supported_reads:</strong> The number of reads which support the circular RNA.</p>
<p><strong>Supported_reads/Reads_at_start:</strong> (supported reads / total reads) at starting point of the circular RNA.</p>
<p><strong>Supported_reads/Reads_at_end:</strong> (supported reads / total reads) at end point of the circular RNA.</p>
</div></blockquote>
</div>
<div class="section" id="go-term-go-term-retrieving">
<span id="go-term"></span><h2>go_term (GO term retrieving)<a class="headerlink" href="#go-term-go-term-retrieving" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">go_term</span></code> can retrieve the information of Gene Ontology from Uniprot.
Some analyses of Go terms can be done as well.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Uniprot mapping table:</strong> <a class="reference external" href="http://www.uniprot.org/downloads">idmapping_selected.tab from Uniprot</a>.</p>
<p><strong>GOslim file:</strong> <a class="reference external" href="http://geneontology.org/page/go-slim-and-subset-guide">goslim.obo</a>.</p>
<p><strong>GO file:</strong> <a class="reference external" href="http://geneontology.org/page/download-ontology">go.obo</a>.</p>
<p><strong>Gff files of the genome annotations</strong></p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the transcripts:</strong> For detecting the GO terms only based on expressed CDSs.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">go_term</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span>
                         <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span>
                         <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]]</span>
                         <span class="o">--</span><span class="n">uniprot_id</span> <span class="n">UNIPROT_ID</span> <span class="o">--</span><span class="n">go_obo</span> <span class="n">GO_OBO</span> <span class="o">--</span><span class="n">goslim_obo</span>
                         <span class="n">GOSLIM_OBO</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">If</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">are</span> <span class="n">provided</span><span class="p">,</span> <span class="n">GO</span>
                        <span class="n">term</span> <span class="n">can</span> <span class="n">be</span> <span class="n">retrieved</span> <span class="n">based</span> <span class="n">on</span> <span class="n">expressed</span> <span class="n">CDS</span> <span class="ow">and</span> <span class="nb">all</span>
                        <span class="n">CDS</span><span class="o">.</span>
  <span class="o">--</span><span class="n">uniprot_id</span> <span class="n">UNIPROT_ID</span><span class="p">,</span> <span class="o">-</span><span class="n">u</span> <span class="n">UNIPROT_ID</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">UniProt</span> <span class="n">ID</span> <span class="n">mapping</span> <span class="n">database</span><span class="o">.</span>
  <span class="o">--</span><span class="n">go_obo</span> <span class="n">GO_OBO</span><span class="p">,</span> <span class="o">-</span><span class="n">go</span> <span class="n">GO_OBO</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">go</span><span class="o">.</span><span class="n">obo</span><span class="o">.</span>
  <span class="o">--</span><span class="n">goslim_obo</span> <span class="n">GOSLIM_OBO</span><span class="p">,</span> <span class="o">-</span><span class="n">gs</span> <span class="n">GOSLIM_OBO</span>
                        <span class="n">The</span> <span class="n">path</span> <span class="n">of</span> <span class="n">goslim</span><span class="o">.</span><span class="n">obo</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">ANNOgesic/output/Go_terms</span></code>. If gff files of the transcript are assigned,
two sub-folders will be generated. Results of the expressed CDSs are stored in <code class="docutils literal"><span class="pre">ANNOgesic/output/Go_term/expressed_CDSs</span></code> and
results of all CDSs are stored in <code class="docutils literal"><span class="pre">ANNOgesic/output/Go_term/all_CDSs</span></code>.</p>
<p><strong>Go_term_results:</strong> Stores tables of the Go terms information. The meaning of each column is as following:</p>
<blockquote>
<div><p><strong>Genome:</strong> Genome name.</p>
<p><strong>Strand:</strong> Strand of the CDS.</p>
<p><strong>Start:</strong> Starting point of this CDS.</p>
<p><strong>End:</strong> End point of this CDS.</p>
<p><strong>Protein_id</strong> Protein ID of this CDS.</p>
<p><strong>Go_term</strong> Go term of This CDS.</p>
</div></blockquote>
<p><strong>statistics:</strong> Stores statistic files and figures.</p>
<blockquote>
<div><p><strong>GO term with corresponding amount:</strong> Format of the filename is <code class="docutils literal"><span class="pre">stat_$GENOME.csv</span></code>.</p>
<p><strong>Figures of the three GO term classes:</strong> All figures are stored in the sub-folder - <code class="docutils literal"><span class="pre">figs</span></code>.
<code class="docutils literal"><span class="pre">$GENOME_biological_process.png</span></code>, <code class="docutils literal"><span class="pre">$GENOME_cellular_component.png</span></code>,
<code class="docutils literal"><span class="pre">$GENOME_molecular_function.png</span></code> and <code class="docutils literal"><span class="pre">$GENOME_three_roots.png</span></code> are the figures of
biological process, cellular component, molecular_function and three roots of GO term classes, respectively.</p>
</div></blockquote>
</div>
<div class="section" id="srna-target-srna-target-prediction">
<span id="srna-target"></span><h2>srna_target (sRNA target prediction)<a class="headerlink" href="#srna-target-srna-target-prediction" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">srna_target</span></code> can search potential targets of the sRNA via different programs
(RNAup or RNAplex or both). We recommend running with both
programs. <code class="docutils literal"><span class="pre">srna_target</span></code> can also compare the results of both programs and provide the best ones.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a> .</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Gff files of the sRNAs</strong></p>
<p><strong>Fasta files of the genomes</strong></p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic srna_target [-h] [--project_path [PROJECT_PATH]]
                             --annotation_files ANNOTATION_FILES
                             [ANNOTATION_FILES ...] --fasta_files FASTA_FILES
                             [FASTA_FILES ...] --srna_files SRNA_FILES
                             [SRNA_FILES ...]
                             [--query_srnas QUERY_SRNAS [QUERY_SRNAS ...]]
                             [--program PROGRAM] [--top TOP]
                             [--rnaplfold_path RNAPLFOLD_PATH]
                             [--rnaplex_path RNAPLEX_PATH]
                             [--rnaup_path RNAUP_PATH]
                             [--interaction_length INTERACTION_LENGTH]
                             [--window_size_target WINDOW_SIZE_TARGET]
                             [--span_target SPAN_TARGET]
                             [--window_size_srna WINDOW_SIZE_SRNA]
                             [--span_srna SPAN_SRNA]
                             [--unstructured_region_rnaplex_target UNSTRUCTURED_REGION_RNAPLEX_TARGET]
                             [--unstructured_region_rnaplex_srna UNSTRUCTURED_REGION_RNAPLEX_SRNA]
                             [--unstructured_region_rnaup UNSTRUCTURED_REGION_RNAUP]
                             [--energy_threshold ENERGY_THRESHOLD]
                             [--duplex_distance DUPLEX_DISTANCE]
                             [--parallels_rnaplex PARALLELS_RNAPLEX]
                             [--parallels_rnaup PARALLELS_RNAUP]
                             [--continue_rnaup]
                             [--potential_target_start POTENTIAL_TARGET_START]
                             [--potential_target_end POTENTIAL_TARGET_END]
                             [--target_feature TARGET_FEATURE [TARGET_FEATURE ...]]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the genome annotation gff files.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        Paths of the genome fasta files.
  --srna_files SRNA_FILES [SRNA_FILES ...], -s SRNA_FILES [SRNA_FILES ...]
                        Paths of the sRNA gff files.
  --query_srnas QUERY_SRNAS [QUERY_SRNAS ...], -q QUERY_SRNAS [QUERY_SRNAS ...]
                        Please assign the query sRNA. The input format is
                        $GENOME:$START:$END:$STRAND. If multiple sRNAs need to
                        be assigned, please use spaces to separate them. For
                        example, NC_007795.1:200:534:+
                        NC_007795.1:6767:6900:-. If you want to detect all
                        sRNAs in gff file, please assign &quot;all&quot;. Default is
                        all.
  --program PROGRAM, -p PROGRAM
                        The program for detecting sRNA-mRNA interaction.
                        Please assign &quot;RNAplex&quot; or &quot;RNAup&quot; or &quot;both&quot;. Default
                        is both.
  --top TOP, -t TOP     The minimum ranking number of targets which will be
                        included to final output. The ranking is based on the
                        binding energy. Default is 20.

additional arguments:
  --rnaplfold_path RNAPLFOLD_PATH
                        Please assign the path of RNAplfold in Vienna package.
  --rnaplex_path RNAPLEX_PATH
                        Please assign the path of RNAplex in Vienna package.
  --rnaup_path RNAUP_PATH
                        Please assign the path of RNAup in Vienna package.
  --interaction_length INTERACTION_LENGTH, -i INTERACTION_LENGTH
                        Maximum length of an interaction. Default is 30.
  --window_size_target WINDOW_SIZE_TARGET, -wt WINDOW_SIZE_TARGET
                        This value is the average of the pair probabilities
                        over windows for mRNA target. This function only works
                        for &quot;RNAplex&quot;. Default is 240.
  --span_target SPAN_TARGET, -st SPAN_TARGET
                        This value is the maximum allowed separation of a base
                        pair to span for mRNA target. This function only works
                        for &quot;RNAplex&quot;. Default is 160.
  --window_size_srna WINDOW_SIZE_SRNA, -ws WINDOW_SIZE_SRNA
                        The meaning is similar to --window_size_target. This
                        value is for sRNA. Default is 30.
  --span_srna SPAN_SRNA, -ss SPAN_SRNA
                        The meaning is similar to --span_target. This value is
                        for sRNA. Default is 30.
  --unstructured_region_rnaplex_target UNSTRUCTURED_REGION_RNAPLEX_TARGET, -ut UNSTRUCTURED_REGION_RNAPLEX_TARGET
                        Calculate the mean probability that from length 1 to
                        this value are unpaired for mRNA target. This function
                        only works for &quot;RNAplex&quot;. Default is 30.
  --unstructured_region_rnaplex_srna UNSTRUCTURED_REGION_RNAPLEX_SRNA, -us UNSTRUCTURED_REGION_RNAPLEX_SRNA
                        The meaning is similar to
                        --unstructured_region_rnaplex_target. This value is
                        for sRNA. Default is 30.
  --unstructured_region_rnaup UNSTRUCTURED_REGION_RNAUP, -uu UNSTRUCTURED_REGION_RNAUP
                        Compute the mean probability that from length 1 to
                        this value are unpaired. This function only works for
                        &quot;RNAup&quot;. Default is 30.
  --energy_threshold ENERGY_THRESHOLD, -e ENERGY_THRESHOLD
                        The minimum energy for a duplex. This function only
                        works for &quot;RNAplex&quot;. Default is -8.
  --duplex_distance DUPLEX_DISTANCE, -d DUPLEX_DISTANCE
                        Distance between target 3&#39; ends of two consecutive
                        duplexes. This function only works for &quot;RNAplex&quot;.
                        Default is 20.
  --parallels_rnaplex PARALLELS_RNAPLEX, -pp PARALLELS_RNAPLEX
                        The number of parallels for running RNAplex. Default
                        is 5.
  --parallels_rnaup PARALLELS_RNAUP, -pu PARALLELS_RNAUP
                        The number of parallel processes for running RNAup.
                        Default is 20.
  --continue_rnaup, -cr
                        The running time of RNAup is long if numerous sRNAs
                        are assigned. This function can continue to run RNAup
                        based on the previous intermediate results if the
                        previous process was crushed. Default is False.
  --potential_target_start POTENTIAL_TARGET_START, -ps POTENTIAL_TARGET_START
                        --potential_target_start and --potential_target_end
                        will be applied to extract the potential target. This
                        value indicates the number of nucleotides at the
                        upstream of --target_feature starting point which will
                        be extracted as part of the potential target. Default
                        is 200.
  --potential_target_end POTENTIAL_TARGET_END, -pe POTENTIAL_TARGET_END
                        --potential_target_start and --potential_target_end
                        will be applied to extract the potential target. This
                        value indicates the number of nucleotides at the
                        downstream of --target_feature starting point which
                        will be extracted as part of the potential target.
                        Default is 150.
  --target_feature TARGET_FEATURE [TARGET_FEATURE ...], -tf TARGET_FEATURE [TARGET_FEATURE ...]
                        This is the feature name for extracting as potential
                        targets. If multiple features need to be assigned,
                        please use spaces to separate them. Ex: CDS tRNA.
                        Default is CDS.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sRNA_targets</span></code>.</p>
<p><strong>RNAplex_results:</strong> Stores all results of RNAplex. <code class="docutils literal"><span class="pre">$GENOME_RNAplex.txt</span></code> is raw results of RNAplex.
<code class="docutils literal"><span class="pre">$GENOME_RNAplex_rank.csv</span></code> is the tables with details, and the targets are sorted by binding energy.
The meaning of each column in <code class="docutils literal"><span class="pre">$GENOME_RNAplex_rank.csv</span></code> is as following:</p>
<blockquote>
<div><p><strong>sRNA:</strong> sRNA name which is shown in sRNA gff file.</p>
<p><strong>Genome:</strong> Genome name.</p>
<p><strong>sRNA_position:</strong> Starting point and end point of this sRNA.</p>
<p><strong>sRNA_interacted_position_RNAplex:</strong> The interaction region of this sRNA.</p>
<p><strong>sRNA_strand:</strong> Strand of this sRNA.</p>
<p><strong>Target:</strong> Locus tag or gene name of the target mRNA.</p>
<p><strong>Target_position:</strong> Starting point and end point of this mRNA.</p>
<p><strong>Target_interacted_position_RNAplex:</strong> The interaction region of this mRNA.</p>
<p><strong>Target_strand:</strong> Strand of this target mRNA.</p>
<p><strong>Energy_RNAplex:</strong> Interaction energy change of this interaction.</p>
<p><strong>Rank_RNAplex:</strong> Ranking of the interaction (the ranking is based on the binding energy).</p>
</div></blockquote>
<p><strong>RNAup_results:</strong> Stored all results of RNAup. <code class="docutils literal"><span class="pre">$GENOME_RNAup.txt</span></code> is raw results of RNAup.
<code class="docutils literal"><span class="pre">$GENOME_RNAup_rank.csv</span></code> is the tables with details, and the targets are
sorted by binding energy. The meaning of each column is similar to the table of RNAplex.</p>
<p><strong>merged_results:</strong> Store the results which are merged by the results of <code class="docutils literal"><span class="pre">RNAplex_results</span></code> and <code class="docutils literal"><span class="pre">RNAup_results</span></code>.
<code class="docutils literal"><span class="pre">$GENOME_merge.csv</span></code> contains all candidates of the both programs.
<code class="docutils literal"><span class="pre">$GENOME_overlap.csv</span></code> contains the results which are top 20 (default) in the both methods.
The meaning of each column is similar to the table of RNAplex.</p>
<p><strong>sRNA_seqs:</strong> Stores fasta sequences of the sRNAs.</p>
<p><strong>target_seqs:</strong> Stores fasta sequences of the potential targets.</p>
</div>
<div class="section" id="ppi-network-protein-protein-interaction-network-detection">
<span id="ppi-network"></span><h2>ppi_network (protein-protein interaction network detection)<a class="headerlink" href="#ppi-network-protein-protein-interaction-network-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">ppi_network</span></code> can retrieve the PPI data from <a class="reference external" href="http://string-db.org/">STRING</a>.
Then using <a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a> to search
the supported literature of the protein-protein interaction networks.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Species table of STRING:</strong> <a class="reference external" href="http://string-db.org/cgi/download.pl">species.v${VERSION}.txt from STRING</a>.</p>
<p><strong>Gff files of the genome annotations</strong></p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic ppi_network [-h] [--project_path [PROJECT_PATH]]
                             --annotation_files ANNOTATION_FILES
                             [ANNOTATION_FILES ...] --species_string
                             SPECIES_STRING --query_strains QUERY_STRAINS
                             [QUERY_STRAINS ...] [--query QUERY [QUERY ...]]
                             [--without_strain_pubmed] [--score SCORE]
                             [--node_size NODE_SIZE]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the genome annotation gff files. BE CAREFUL:
                        The gff files MUST have proper locus_tag item in the
                        attributes. The locus_tag items can be assigned by a
                        real locus tag like locus_tag=SAOUHSC_00003, or a gene
                        name, such as, locus_tag=dnaA.
  --species_string SPECIES_STRING, -d SPECIES_STRING
                        Please assign path of the species table of STRING
                        (species.$VERSION.txt).
  --query_strains QUERY_STRAINS [QUERY_STRAINS ...], -s QUERY_STRAINS [QUERY_STRAINS ...]
                        This is for query strain and filename. In order to
                        retrieve the data from STRING and Pubmed, The strain
                        name must be assigned. If the query strain is not
                        available in database, you can assign the close strain
                        of your query strain. For example, the query strain is
                        Staphylococcus aureus HG003, but there is no
                        Staphylococcus aureus HG003 in STRING database.
                        Therefore, Staphylococcus aureus NCTC 8325 (close
                        strain) can be used as reference. The input format can
                        be: Staphylococcus_aureus_HG003.gff:Staphylococcus_aur
                        eus_HG003:&quot;Staphylococcus aureus 8325&quot;:&quot;Staphylococcus
                        aureus&quot;. or Staphylococcus_aureus_HG003.gff:Staphyloco
                        ccus_aureus_HG003:&quot;93061&quot;:&quot;Staphylococcus aureus&quot;. or
                        Staphylococcus_aureus_HG003.gff:Staphylococcus_aureus_
                        HG003:&quot;Staphylococcus aureus NCTC
                        8325&quot;:&quot;Staphylococcus aureus&quot;. (gff_filename:gff_strai
                        n_name:STRING_name:Pubmed_name). First one is the gff
                        file name. Second one is the strain name in gff files.
                        Third one is from STRING (please check the
                        species.vXXX.txt of STRING, taxon_id,
                        STRING_name_compact or official_name_NCBI can be
                        assigned here), and the fourth one is for searching
                        Pubmed. If you want to detect for multiple strains,
                        just put spaces between these strains. Before running
                        it, please check the species table
                        (species.v$VERSION.txt) which should be downloaded. If
                        the file was not downloaded before, please download
                        it. BE CAREFUL, if the assigned name is with spaces,
                        please put &quot;&quot; at two ends. For searching Pubmen, if a
                        specific name is assigned, it may not be able to find
                        the related literature.
  --query QUERY [QUERY ...], -q QUERY [QUERY ...]
                        Please assign the query protein here. The format is
                        $GENOME_NAME_OF_GFF:$START_POINT:$END_POINT:$STRAND.
                        If multiple proteins need to be assigned, please use
                        spaces to separate them. For example,
                        Staphylococcus_aureus_HG003:345:456:+
                        Staphylococcus_aureus_HG003:2000:3211:-. For computing
                        all protein, just type &quot;all&quot;. Default is all.
  --without_strain_pubmed, -n
                        This function will retrieve the literature from Pubmed
                        without assigning strains as well. Default is False.

additional arguments:
  --score SCORE, -ps SCORE
                        Please assign minimum PIE score for searching
                        literature. The value is from -1 (worst) to 1 (best).
                        Default is 0.
  --node_size NODE_SIZE, -ns NODE_SIZE
                        Please assign the size of nodes in figure, default is
                        4000.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/PPI_networks</span></code>. The output folders are as following:</p>
<p><strong>best_results:</strong> Stores the results which the scores of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a>
for supported literature are higher than <code class="docutils literal"><span class="pre">--score</span></code>.</p>
<p><strong>all_results:</strong> Stores the results of all protein-protein interactions
(including the low score(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a>) literature).</p>
<p>Under &#8220;best_results&#8221; and &#8220;all_results&#8221;, several files and folders are generated:</p>
<blockquote>
<div><p><strong>Results of searching literatures without assigning a specific strain:</strong> <code class="docutils literal"><span class="pre">$STRAIN_without_strain.csv</span></code>.</p>
<p><strong>Results of searching literatures with assigning a specific strain:</strong> <code class="docutils literal"><span class="pre">$STRAIN_with_strain.csv</span></code>.</p>
<p><strong>without_strain:</strong> Stores all interaction information which is searched without assigning a specific strain.</p>
<p><strong>with_strain:</strong> Stores all interaction information which is searched with assigning a specific strain.</p>
</div></blockquote>
<p><strong>figures:</strong> Stores the protein-protein networks of the query proteins.
Thickness represents how many literature can be found for the interactions.
Solid line means that strong supported literature can be found. Dash-dot line
means that the supported literature are very weak. Dot line means that no supported literature can be found.
Color is the best score of the supported literature of the interactions.</p>
</div>
<div class="section" id="localization-subcellular-localization-prediction">
<span id="subcellular-localization"></span><h2>localization (subcellular localization prediction)<a class="headerlink" href="#localization-subcellular-localization-prediction" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">localization</span></code> can predict the subcellular localization of CDSs.
Some statistics and visualization files are provided as well.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://www.psort.org/psortb/">Psortb</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Fasta files of the genome sequences</strong></p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the transcripts:</strong> For detecting subcellular localization only based on expressed CDSs.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">localization</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span>
                              <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span>
                              <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">fasta_files</span> <span class="n">FASTA_FILES</span>
                              <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]]</span>
                              <span class="o">--</span><span class="n">bacteria_type</span> <span class="n">BACTERIA_TYPE</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">psortb_path</span> <span class="n">PSORTB_PATH</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">difference_multi</span> <span class="n">DIFFERENCE_MULTI</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">merge_to_gff</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fasta_files</span> <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">genome</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">If</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">are</span> <span class="n">provided</span><span class="p">,</span> <span class="n">it</span>
                        <span class="n">can</span> <span class="n">get</span> <span class="n">the</span> <span class="n">results</span> <span class="n">based</span> <span class="n">on</span> <span class="n">expressed</span> <span class="n">CDS</span> <span class="ow">and</span> <span class="nb">all</span>
                        <span class="n">CDS</span><span class="o">.</span>
  <span class="o">--</span><span class="n">bacteria_type</span> <span class="n">BACTERIA_TYPE</span><span class="p">,</span> <span class="o">-</span><span class="n">b</span> <span class="n">BACTERIA_TYPE</span>
                        <span class="n">The</span> <span class="nb">type</span> <span class="n">of</span> <span class="n">bacteria</span> <span class="p">(</span><span class="n">Gram</span><span class="o">-</span><span class="n">positive</span> <span class="ow">or</span> <span class="n">Gram</span><span class="o">-</span><span class="n">negative</span><span class="p">)</span><span class="o">.</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="s1">&#39;positive&#39;</span> <span class="ow">or</span> <span class="s1">&#39;negative&#39;</span><span class="o">.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">psortb_path</span> <span class="n">PSORTB_PATH</span>
                        <span class="n">If</span> <span class="n">you</span> <span class="n">want</span> <span class="n">to</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">Psortb</span><span class="p">,</span> <span class="n">please</span>
                        <span class="n">assign</span> <span class="n">here</span><span class="o">.</span>
  <span class="o">--</span><span class="n">difference_multi</span> <span class="n">DIFFERENCE_MULTI</span><span class="p">,</span> <span class="o">-</span><span class="n">d</span> <span class="n">DIFFERENCE_MULTI</span>
                        <span class="n">For</span> <span class="n">the</span> <span class="n">protein</span> <span class="n">which</span> <span class="n">have</span> <span class="n">multiple</span> <span class="n">locations</span><span class="p">,</span> <span class="k">if</span> <span class="n">the</span>
                        <span class="n">difference</span> <span class="n">of</span> <span class="n">psorb</span> <span class="n">scores</span> <span class="ow">is</span> <span class="n">smaller</span> <span class="n">than</span>
                        <span class="o">--</span><span class="n">difference_multi</span><span class="p">,</span> <span class="n">it</span> <span class="n">will</span> <span class="n">be</span> <span class="n">printed</span> <span class="n">out</span> <span class="k">as</span> <span class="n">well</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="mf">0.5</span><span class="o">.</span> <span class="n">The</span> <span class="n">maximum</span> <span class="n">value</span> <span class="ow">is</span> <span class="mf">10.</span>
  <span class="o">--</span><span class="n">merge_to_gff</span><span class="p">,</span> <span class="o">-</span><span class="n">m</span>    <span class="n">Merging</span> <span class="n">the</span> <span class="n">information</span> <span class="n">to</span> <span class="n">genome</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">file</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/subcellular_localization</span></code>. If gff files of the transcripts are assigned,
two sub-folders will be generated. Results of the expressed CDSs are stored in
<code class="docutils literal"><span class="pre">ANNOgesic/output/subcellular_localization/expressed_CDSs</span></code> and
results of all CDS are stored in <code class="docutils literal"><span class="pre">ANNOgesic/output/subcellular_localization/all_CDSs</span></code>.</p>
<p><strong>psortb_results</strong>: Stores the results of Psortb:</p>
<blockquote>
<div><p><strong>Raw output data of Psortb:</strong> Format of the filename is <code class="docutils literal"><span class="pre">$GENOME_raw.txt</span></code>.</p>
<p><strong>Table of subcellular localization:</strong> Format of the filename is <code class="docutils literal"><span class="pre">$GENOME_table.csv</span></code>.
The meaning of each column is as following:</p>
<blockquote>
<div><p><strong>Genome:</strong> Genome name.</p>
<p><strong>Protein:</strong> Protein ID.</p>
<p><strong>Strand:</strong> Strand of this protein.</p>
<p><strong>Start:</strong> Starting point of this protein.</p>
<p><strong>End:</strong> End point of this protein.</p>
<p><strong>Location:</strong> Predicted subcellular localization of this protein.</p>
<p><strong>Score:</strong> Psortb score.</p>
</div></blockquote>
</div></blockquote>
<p><strong>statistics:</strong> Stores statistic files and figures.</p>
<blockquote>
<div><p><strong>Subcellular localization with corresponding amounts:</strong> Format of the filename is <code class="docutils literal"><span class="pre">stat_$GENOME_sublocal.csv</span></code>.</p>
<p><strong>Figure of Subcellular localization with corresponding amounts:</strong> Format of the filename is <code class="docutils literal"><span class="pre">$FILENAME_$GENOME_sublocal.png</span></code>.</p>
</div></blockquote>
</div>
<div class="section" id="riboswitch-thermometer-riboswitch-and-rna-thermometer-detection">
<span id="riboswitch-thermometer"></span><h2>riboswitch_thermometer (riboswitch and RNA thermometer detection)<a class="headerlink" href="#riboswitch-thermometer-riboswitch-and-rna-thermometer-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">riboswitch_thermometer</span></code> can search riboswitches and RNA thermometers between
TSSsï¼ˆthe starting point of transcript was assigned if no TSS was detected)
and its downstream CDSs, as well as associated ribosome binding sites
(Shine-Dalgarno sequence). Then using <a class="reference external" href="http://infernal.janelia.org/">Infernal</a> to scan
the region in <a class="reference external" href="http://rfam.xfam.org/">Rfam</a>.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://infernal.janelia.org/">Infernal</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><a class="reference external" href="http://rfam.xfam.org/">Rfam</a>.</p>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Gff files of the transcripts</strong></p>
<p><strong>Gff files of the TSSs</strong></p>
<p><strong>Fasta files of the genome sequences</strong></p>
<p><strong>Rfam ID files of the riboswitch or RNA thermometer:</strong>
The file should contain Rfam IDs,
name and description of riboswitches or RNA thermometers as following.</p>
<table border="1" class="docutils">
<colgroup>
<col width="21%" />
<col width="11%" />
<col width="68%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>#Rfam_ID</td>
<td>Name</td>
<td>Description</td>
</tr>
<tr class="row-even"><td>RF00162</td>
<td>SAM</td>
<td>SAM riboswitch box leader</td>
</tr>
<tr class="row-odd"><td>RF00059</td>
<td>TPP</td>
<td>TPP riboswitch THI element</td>
</tr>
</tbody>
</table>
<p>All columns are separated by <code class="docutils literal"><span class="pre">tab</span></code>. You can also download
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/database">riboswitch and RNA thermometer data</a>
from our Git repository.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">riboswitch_thermometer</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">program</span> <span class="n">PROGRAM</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">riboswitch_id_file</span> <span class="n">RIBOSWITCH_ID_FILE</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">rna_thermometer_id_file</span> <span class="n">RNA_THERMOMETER_ID_FILE</span><span class="p">]</span>
                                        <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span>
                                        <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">tss_files</span>
                                        <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">]</span>
                                        <span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span>
                                        <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">fasta_files</span>
                                        <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                                        <span class="o">--</span><span class="n">rfam_path</span> <span class="n">RFAM_PATH</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">cmscan_path</span> <span class="n">CMSCAN_PATH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">cmpress_path</span> <span class="n">CMPRESS_PATH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">utr_length</span> <span class="n">UTR_LENGTH</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">e_value</span> <span class="n">E_VALUE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output_all</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">fuzzy</span> <span class="n">FUZZY</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">fuzzy_rbs</span> <span class="n">FUZZY_RBS</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">start_codon</span> <span class="n">START_CODON</span> <span class="p">[</span><span class="n">START_CODON</span> <span class="o">...</span><span class="p">]]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">max_dist_rbs</span> <span class="n">MAX_DIST_RBS</span><span class="p">]</span>
                                        <span class="p">[</span><span class="o">--</span><span class="n">min_dist_rbs</span> <span class="n">MIN_DIST_RBS</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">program</span> <span class="n">PROGRAM</span><span class="p">,</span> <span class="o">-</span><span class="n">p</span> <span class="n">PROGRAM</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">feature</span> <span class="n">that</span> <span class="n">you</span> <span class="n">want</span> <span class="n">to</span> <span class="n">detect</span><span class="o">.</span> <span class="n">The</span>
                        <span class="n">options</span> <span class="n">can</span> <span class="n">be</span> <span class="s2">&quot;riboswitch&quot;</span><span class="p">,</span> <span class="s2">&quot;thermometer&quot;</span><span class="p">,</span> <span class="s2">&quot;both&quot;</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="n">both</span><span class="o">.</span>
  <span class="o">--</span><span class="n">riboswitch_id_file</span> <span class="n">RIBOSWITCH_ID_FILE</span><span class="p">,</span> <span class="o">-</span><span class="n">ri</span> <span class="n">RIBOSWITCH_ID_FILE</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">program</span> <span class="ow">is</span> <span class="s2">&quot;riboswitch&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span>
                        <span class="n">the</span> <span class="n">file</span> <span class="n">path</span> <span class="n">of</span> <span class="n">riboswitch</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">Rfam</span><span class="o">.</span> <span class="n">The</span> <span class="n">file</span>
                        <span class="n">should</span> <span class="n">include</span> <span class="n">the</span> <span class="n">Rfam</span> <span class="n">ID</span><span class="p">,</span> <span class="n">riboswitch</span> <span class="n">name</span><span class="p">,</span> <span class="ow">and</span>
                        <span class="n">Description</span> <span class="n">of</span> <span class="n">the</span> <span class="n">riboswitch</span><span class="o">.</span> <span class="n">For</span> <span class="n">example</span><span class="p">,</span> <span class="n">please</span>
                        <span class="n">check</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">Sung</span><span class="o">-</span><span class="n">Huan</span><span class="o">/</span><span class="n">ANNOgesic</span><span class="o">/</span><span class="n">blob</span><span class="o">/</span><span class="n">mast</span>
                        <span class="n">er</span><span class="o">/</span><span class="n">database</span><span class="o">/</span><span class="n">Rfam_riboswitch_ID</span><span class="o">.</span><span class="n">csv</span>
  <span class="o">--</span><span class="n">rna_thermometer_id_file</span> <span class="n">RNA_THERMOMETER_ID_FILE</span><span class="p">,</span> <span class="o">-</span><span class="n">ti</span> <span class="n">RNA_THERMOMETER_ID_FILE</span>
                        <span class="n">If</span> <span class="o">--</span><span class="n">program</span> <span class="ow">is</span> <span class="s2">&quot;thermometer&quot;</span> <span class="ow">or</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="n">please</span> <span class="n">assign</span>
                        <span class="n">the</span> <span class="n">file</span> <span class="n">path</span> <span class="n">of</span> <span class="n">RNA</span> <span class="n">thermometer</span> <span class="n">ID</span> <span class="n">of</span> <span class="n">Rfam</span><span class="o">.</span> <span class="n">The</span> <span class="n">file</span>
                        <span class="n">should</span> <span class="n">include</span> <span class="n">the</span> <span class="n">Rfam</span> <span class="n">ID</span><span class="p">,</span> <span class="n">RNA</span> <span class="n">thermometer</span> <span class="n">name</span> <span class="ow">and</span>
                        <span class="n">Description</span> <span class="n">of</span> <span class="n">the</span> <span class="n">RNA</span> <span class="n">thermometer</span><span class="o">.</span> <span class="n">For</span> <span class="n">example</span><span class="p">,</span>
                        <span class="n">please</span> <span class="n">check</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">Sung</span><span class="o">-</span><span class="n">Huan</span><span class="o">/</span><span class="n">ANNOgesic</span><span class="o">/</span><span class="n">bl</span>
                        <span class="n">ob</span><span class="o">/</span><span class="n">master</span><span class="o">/</span><span class="n">database</span><span class="o">/</span><span class="n">Rfam_RNA_thermometer_ID</span><span class="o">.</span><span class="n">csv</span>
  <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">annotation</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">tss_files</span> <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">t</span> <span class="n">TSS_FILES</span> <span class="p">[</span><span class="n">TSS_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">TSS</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">transcript_files</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">a</span> <span class="n">TRANSCRIPT_FILES</span> <span class="p">[</span><span class="n">TRANSCRIPT_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">transcript</span> <span class="n">gff</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fasta_files</span> <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">genome</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">rfam_path</span> <span class="n">RFAM_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">R</span> <span class="n">RFAM_PATH</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">Rfam</span> <span class="n">CM</span> <span class="n">database</span><span class="o">.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">cmscan_path</span> <span class="n">CMSCAN_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">cs</span> <span class="n">CMSCAN_PATH</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">cmscan</span> <span class="ow">in</span> <span class="n">Infernal</span> <span class="n">package</span><span class="o">.</span>
  <span class="o">--</span><span class="n">cmpress_path</span> <span class="n">CMPRESS_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">cp</span> <span class="n">CMPRESS_PATH</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">cmpress</span> <span class="ow">in</span> <span class="n">Infernal</span> <span class="n">package</span><span class="o">.</span>
  <span class="o">--</span><span class="n">utr_length</span> <span class="n">UTR_LENGTH</span><span class="p">,</span> <span class="o">-</span><span class="n">u</span> <span class="n">UTR_LENGTH</span>
                        <span class="n">The</span> <span class="n">UTR</span> <span class="n">length</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">300.</span>
  <span class="o">--</span><span class="n">e_value</span> <span class="n">E_VALUE</span><span class="p">,</span> <span class="o">-</span><span class="n">e</span> <span class="n">E_VALUE</span>
                        <span class="n">The</span> <span class="n">maximum</span> <span class="n">e</span> <span class="n">value</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">0.001</span><span class="o">.</span>
  <span class="o">--</span><span class="n">output_all</span><span class="p">,</span> <span class="o">-</span><span class="n">o</span>      <span class="n">One</span> <span class="n">query</span> <span class="n">sequence</span> <span class="n">may</span> <span class="n">fit</span> <span class="n">multiple</span> <span class="n">riboswitches</span> <span class="ow">or</span>
                        <span class="n">RNA</span> <span class="n">thermometers</span><span class="o">.</span> <span class="n">If</span> <span class="o">--</span><span class="n">output_all</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">,</span> <span class="nb">all</span> <span class="n">results</span>
                        <span class="n">will</span> <span class="n">be</span> <span class="n">printed</span> <span class="n">out</span><span class="o">.</span> <span class="n">Otherwise</span><span class="p">,</span> <span class="n">only</span> <span class="n">the</span> <span class="n">best</span> <span class="n">one</span> <span class="n">will</span>
                        <span class="n">be</span> <span class="n">printed</span> <span class="n">out</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fuzzy</span> <span class="n">FUZZY</span><span class="p">,</span> <span class="o">-</span><span class="n">z</span> <span class="n">FUZZY</span>
                        <span class="n">The</span> <span class="n">fuzzy</span> <span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span> <span class="k">for</span> <span class="n">extracting</span> <span class="n">the</span> <span class="n">sequences</span>
                        <span class="n">of</span> <span class="n">potential</span> <span class="n">riboswitches</span> <span class="ow">or</span> <span class="n">RNA</span> <span class="n">thermometers</span><span class="o">.</span> <span class="n">Default</span>
                        <span class="ow">is</span> <span class="mf">10.</span>
  <span class="o">--</span><span class="n">fuzzy_rbs</span> <span class="n">FUZZY_RBS</span><span class="p">,</span> <span class="o">-</span><span class="n">zr</span> <span class="n">FUZZY_RBS</span>
                        <span class="n">The</span> <span class="n">number</span> <span class="n">of</span> <span class="n">nucleotides</span> <span class="n">of</span> <span class="n">Shine</span><span class="o">-</span><span class="n">Dalgarno</span> <span class="n">sequence</span>
                        <span class="n">allow</span> <span class="n">to</span> <span class="n">be</span> <span class="n">different</span> <span class="k">with</span> <span class="n">AGGAGG</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">2.</span>
  <span class="o">--</span><span class="n">start_codon</span> <span class="n">START_CODON</span> <span class="p">[</span><span class="n">START_CODON</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">ac</span> <span class="n">START_CODON</span> <span class="p">[</span><span class="n">START_CODON</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">The</span> <span class="n">types</span> <span class="n">of</span> <span class="n">start</span> <span class="n">codon</span><span class="o">.</span> <span class="n">If</span> <span class="n">multiple</span> <span class="n">types</span> <span class="n">need</span> <span class="n">to</span> <span class="n">be</span>
                        <span class="n">assigned</span><span class="p">,</span> <span class="n">please</span> <span class="n">use</span> <span class="n">spaces</span> <span class="n">to</span> <span class="n">separate</span> <span class="n">them</span><span class="o">.</span> <span class="n">Default</span>
                        <span class="ow">is</span> <span class="n">ATG</span><span class="o">.</span>
  <span class="o">--</span><span class="n">max_dist_rbs</span> <span class="n">MAX_DIST_RBS</span><span class="p">,</span> <span class="o">-</span><span class="n">Mr</span> <span class="n">MAX_DIST_RBS</span>
                        <span class="n">The</span> <span class="n">maximum</span> <span class="n">distance</span> <span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span> <span class="n">between</span> <span class="n">ribosome</span>
                        <span class="n">binding</span> <span class="n">site</span> <span class="p">(</span><span class="n">Shine</span><span class="o">-</span><span class="n">Dalgarno</span> <span class="n">sequence</span><span class="p">)</span> <span class="ow">and</span> <span class="n">start</span>
                        <span class="n">codon</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">14.</span>
  <span class="o">--</span><span class="n">min_dist_rbs</span> <span class="n">MIN_DIST_RBS</span><span class="p">,</span> <span class="o">-</span><span class="n">mr</span> <span class="n">MIN_DIST_RBS</span>
                        <span class="n">The</span> <span class="n">minimum</span> <span class="n">distance</span> <span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span> <span class="n">between</span> <span class="n">ribosome</span>
                        <span class="n">binding</span> <span class="n">site</span> <span class="p">(</span><span class="n">Shine</span><span class="o">-</span><span class="n">Dalgarno</span> <span class="n">sequence</span><span class="p">)</span> <span class="ow">and</span> <span class="n">start</span>
                        <span class="n">codon</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">5.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files of the riboswitches are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/riboswitches</span></code> and
output files of the RNA thermometers are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/RNA_thermometers</span></code>.</p>
<p>Names of the output folders are as following:</p>
<p><strong>scan_Rfam_results:</strong> Stores the results of searching to Rfam with <code class="docutils literal"><span class="pre">cmscan</span></code> (<a class="reference external" href="http://infernal.janelia.org/">Infernal</a>).</p>
<p><strong>gffs:</strong> Stores gff files of riboswitches/RNA_thermometers.
Some useful information can be found in the tags of the attributes within the gff file.
Based on this information, we can know the details of the specific riboswitches/RNA_thermometers.
The tags are as following:</p>
<blockquote>
<div><p><strong>rfam_id:</strong> Rfam ID of this riboswitch/RNA_thermometer.</p>
<p><strong>e-value:</strong> E-value of searching this riboswitch/RNA_thermometer to Rfam.</p>
<p><strong>method:</strong> This riboswitch/RNA_thermometer is detected by which method.</p>
</div></blockquote>
<p><strong>tables:</strong> Stores tables of riboswiches/RNA_thermometers with more details. The meaning of each column in this table is as following:</p>
<blockquote>
<div><p><strong>ID:</strong> Riboswich/RNA_thermometer ID.</p>
<p><strong>Genome:</strong> Genome name.</p>
<p><strong>Strand:</strong> Strand of the riboswich/RNA_thermometer.</p>
<p><strong>Associated_CDS:</strong> Downstream CDS of the riboswich/RNA_thermometer.</p>
<p><strong>Start_genome:</strong> This riboswich/RNA_thermometer starts from which position of the genome.</p>
<p><strong>End_genome:</strong> This riboswich/RNA_thermometer ends to which position of the genome.</p>
<p><strong>Rfam:</strong> Rfam ID of this riboswich/RNA_thermometer.</p>
<p><strong>E_value:</strong> E-value of searching this riboswitch/RNA_thermometer to Rfam.</p>
<p><strong>Start_align:</strong> Position of this riboswich/RNA_thermometer can be aligned to the genome.</p>
<p><strong>End_align:</strong> Position this riboswich/RNA_thermometer can be aligned to the genome.</p>
</div></blockquote>
<p><strong>statistics:</strong> Stores the file which contains the riboswich/RNA_thermometer with corresponding amount.</p>
</div>
<div class="section" id="crispr-crispr-detection">
<span id="crispr"></span><h2>crispr (CRISPR detection)<a class="headerlink" href="#crispr-crispr-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">crispr</span></code> integrates CRISPR Recognition Tool (<a class="reference external" href="http://www.room220.com/crt/">CRT</a>) which can detect the repeat
units and spacers of CRISPR. Moreover, the false positive can be removed by comparing candidates with genome annotation.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://www.room220.com/crt/">CRT</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Fasta files of the genome sequences</strong></p>
<ul class="simple">
<li><strong>Optional input files</strong></li>
</ul>
<p><strong>Gff files of the genome annotations:</strong> This file can be used for removing false positive.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">crispr</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span> <span class="o">--</span><span class="n">fasta_files</span>
                        <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">crt_path</span> <span class="n">CRT_PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">window_size</span> <span class="n">WINDOW_SIZE</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">min_number_repeats</span> <span class="n">MIN_NUMBER_REPEATS</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">min_length_repeat</span> <span class="n">MIN_LENGTH_REPEAT</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">Max_length_repeat</span> <span class="n">MAX_LENGTH_REPEAT</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">min_length_spacer</span> <span class="n">MIN_LENGTH_SPACER</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">Max_length_spacer</span> <span class="n">MAX_LENGTH_SPACER</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">ignore_hypothetical_protein</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">fasta_files</span> <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">f</span> <span class="n">FASTA_FILES</span> <span class="p">[</span><span class="n">FASTA_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">genome</span> <span class="n">fasta</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">--</span><span class="n">annotation_files</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">],</span> <span class="o">-</span><span class="n">g</span> <span class="n">ANNOTATION_FILES</span> <span class="p">[</span><span class="n">ANNOTATION_FILES</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">If</span> <span class="n">paths</span> <span class="n">of</span> <span class="n">the</span> <span class="n">genome</span> <span class="n">gff</span> <span class="n">files</span> <span class="n">are</span> <span class="n">provided</span><span class="p">,</span> <span class="n">this</span>
                        <span class="n">function</span> <span class="n">will</span> <span class="n">compare</span> <span class="n">CRISPRs</span> <span class="k">with</span> <span class="n">the</span> <span class="n">genome</span>
                        <span class="n">annotation</span> <span class="k">for</span> <span class="n">removing</span> <span class="n">the</span> <span class="n">false</span> <span class="n">positives</span><span class="o">.</span> <span class="n">Default</span>
                        <span class="ow">is</span> <span class="kc">None</span><span class="o">.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">crt_path</span> <span class="n">CRT_PATH</span>   <span class="n">If</span> <span class="n">you</span> <span class="n">want</span> <span class="n">to</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="n">CRT</span><span class="o">.</span><span class="n">jar</span><span class="p">,</span> <span class="n">please</span>
                        <span class="n">assign</span> <span class="n">here</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">CRT</span><span class="o">.</span><span class="n">jar</span>
  <span class="o">--</span><span class="n">window_size</span> <span class="n">WINDOW_SIZE</span><span class="p">,</span> <span class="o">-</span><span class="n">w</span> <span class="n">WINDOW_SIZE</span>
                        <span class="n">Length</span> <span class="n">of</span> <span class="n">the</span> <span class="n">window</span> <span class="n">size</span> <span class="k">for</span> <span class="n">searching</span> <span class="n">CRISPR</span> <span class="p">(</span><span class="nb">range</span><span class="p">:</span>
                        <span class="mi">6</span><span class="o">-</span><span class="mi">9</span><span class="p">)</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">8.</span>
  <span class="o">--</span><span class="n">min_number_repeats</span> <span class="n">MIN_NUMBER_REPEATS</span><span class="p">,</span> <span class="o">-</span><span class="n">mn</span> <span class="n">MIN_NUMBER_REPEATS</span>
                        <span class="n">Minimum</span> <span class="n">number</span> <span class="n">of</span> <span class="n">repeats</span> <span class="n">that</span> <span class="n">a</span> <span class="n">CRISPR</span> <span class="n">must</span> <span class="n">contain</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="ow">is</span> <span class="mf">3.</span>
  <span class="o">--</span><span class="n">min_length_repeat</span> <span class="n">MIN_LENGTH_REPEAT</span><span class="p">,</span> <span class="o">-</span><span class="n">ml</span> <span class="n">MIN_LENGTH_REPEAT</span>
                        <span class="n">Minimum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">repeats</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">23.</span>
  <span class="o">--</span><span class="n">Max_length_repeat</span> <span class="n">MAX_LENGTH_REPEAT</span><span class="p">,</span> <span class="o">-</span><span class="n">Ml</span> <span class="n">MAX_LENGTH_REPEAT</span>
                        <span class="n">Maximum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">repeats</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">47.</span>
  <span class="o">--</span><span class="n">min_length_spacer</span> <span class="n">MIN_LENGTH_SPACER</span><span class="p">,</span> <span class="o">-</span><span class="n">ms</span> <span class="n">MIN_LENGTH_SPACER</span>
                        <span class="n">Minimum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">spacers</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">26.</span>
  <span class="o">--</span><span class="n">Max_length_spacer</span> <span class="n">MAX_LENGTH_SPACER</span><span class="p">,</span> <span class="o">-</span><span class="n">Ms</span> <span class="n">MAX_LENGTH_SPACER</span>
                        <span class="n">Maximum</span> <span class="n">length</span> <span class="n">of</span> <span class="n">CRISPR</span> <span class="n">spacers</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="mf">50.</span>
  <span class="o">--</span><span class="n">ignore_hypothetical_protein</span><span class="p">,</span> <span class="o">-</span><span class="ow">in</span>
                        <span class="n">This</span> <span class="n">function</span> <span class="ow">is</span> <span class="k">for</span> <span class="n">ignoring</span> <span class="n">the</span> <span class="n">hypothetical</span>
                        <span class="n">proteins</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="kc">False</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/crisprs</span></code>. The folders which are generated by the subcommand are as following:</p>
<p><strong>CRT_results:</strong> Stores the output of <a class="reference external" href="http://www.room220.com/crt/">CRT</a>.</p>
<p><strong>gffs:</strong> Stores CRSIPR gff files. Two sub-folders are under this folder:</p>
<blockquote>
<div><p><strong>all_candidates:</strong> Stores gff files which contains all CRISPRs.</p>
<p><strong>best_candidates:</strong> Stores gff files which contains the CRISPRs without overlapping genome annotation.</p>
</div></blockquote>
<p><strong>statistics:</strong> Stores statistic files.</p>
</div>
<div class="section" id="optimize-tss-ps-optimization-of-tss-and-processing-site-detection">
<span id="optimize-tss-ps"></span><h2>optimize_tss_ps (optimization of TSS and processing site detection)<a class="headerlink" href="#optimize-tss-ps-optimization-of-tss-and-processing-site-detection" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">optimize_tss_ps</span></code> can adapt the parameter set of <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.
For running it, please manual detect TSSs around 200kb and find at least 50 TSSs (using gff format).
If there are less than 50 TSSs within 200kb, please continue checking until 50 TSSs are detected.
Then <code class="docutils literal"><span class="pre">optimize_tss_ps</span></code> can scan whole genome based on the manual detected set to get optimized parameters.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Wiggle files of TEX +/-:</strong> Please check the section <a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic"><span class="std std-ref">The input format of libraries for running ANNOgesic</span></a>.</p>
<p><strong>Fasta files of the genome sequences</strong></p>
<p><strong>Gff files of the genome annotations</strong></p>
<p><strong>Gff files of the manual-detected TSSs</strong></p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic optimize_tss_ps [-h] [--project_path [PROJECT_PATH]]
                                 [--program PROGRAM] --fasta_files FASTA_FILES
                                 [FASTA_FILES ...] --annotation_files
                                 ANNOTATION_FILES [ANNOTATION_FILES ...]
                                 --manual_files MANUAL_FILES
                                 [MANUAL_FILES ...]
                                 [--genome_lengths GENOME_LENGTHS [GENOME_LENGTHS ...]]
                                 --tex_notex_libs TEX_NOTEX_LIBS
                                 [TEX_NOTEX_LIBS ...]
                                 [--replicate_tex REPLICATE_TEX [REPLICATE_TEX ...]]
                                 --condition_names CONDITION_NAMES
                                 [CONDITION_NAMES ...]
                                 [--tsspredator_path TSSPREDATOR_PATH]
                                 [--max_height MAX_HEIGHT]
                                 [--max_height_reduction MAX_HEIGHT_REDUCTION]
                                 [--max_factor MAX_FACTOR]
                                 [--max_factor_reduction MAX_FACTOR_REDUCTION]
                                 [--max_base_height MAX_BASE_HEIGHT]
                                 [--max_enrichment_factor MAX_ENRICHMENT_FACTOR]
                                 [--max_processing_factor MAX_PROCESSING_FACTOR]
                                 [--utr_length UTR_LENGTH] [--cluster CLUSTER]
                                 [--parallels PARALLELS] [--steps STEPS]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --program PROGRAM, -p PROGRAM
                        The feature for optimization. Please assign &quot;TSS&quot; or
                        &quot;Processing_site&quot;. Default is TSS.
  --fasta_files FASTA_FILES [FASTA_FILES ...], -f FASTA_FILES [FASTA_FILES ...]
                        Paths of the fasta file that you want to optimize.
  --annotation_files ANNOTATION_FILES [ANNOTATION_FILES ...], -g ANNOTATION_FILES [ANNOTATION_FILES ...]
                        Paths of the query genome annotation gff file.
  --manual_files MANUAL_FILES [MANUAL_FILES ...], -m MANUAL_FILES [MANUAL_FILES ...]
                        Paths of the manual-checked gff files. It is the
                        benchmark for training. Please detect the
                        TSSs/processing sites at least 200kb or 50 true
                        TSSs/processing sites.
  --genome_lengths GENOME_LENGTHS [GENOME_LENGTHS ...], -le GENOME_LENGTHS [GENOME_LENGTHS ...]
                        Please specify the genome length of input genomes in
                        the manual-detected TSS gff files. If the genome was
                        manual detected for compare whole genome, please type
                        &quot;all&quot;. The input format is $GENOME:SLENGTH. Multiple
                        genomes can be accepted, please use spaces to separate
                        them. For example, test.gff contain two genomes (s1
                        and s2). s1 was manual checked 100kb and s2 was
                        checked whole genome. The value of this argument will
                        be s1:100000 s2:all. Default setting will compute all
                        genomes in manual-detected TSS gff files with whole
                        length of genome.
  --tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...], -tl TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]
                        TEX+/- wig files for TSSpredator. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --replicate_tex REPLICATE_TEX [REPLICATE_TEX ...], -rt REPLICATE_TEX [REPLICATE_TEX ...]
                        This value is the minimal number of replicates that a
                        TSS has to be detected. The format is
                        $NUMBERofCONDITION_$NUMBERofREPLICATE. If different
                        --replicate_tex values need to be assigned to
                        different conditions, please use spaces to separate
                        them. For example, 1_2 2_ 3_3. It means that
                        --replicate_tex is 2 in number 1 and number 2
                        conditions. In number 3 condition, --replicate_tex is
                        3. For assigning the same --replicate_tex to all
                        conditions, just use like all_1 (--replicate_tex is 1
                        in all conditions). Default is all_1.
  --condition_names CONDITION_NAMES [CONDITION_NAMES ...], -cn CONDITION_NAMES [CONDITION_NAMES ...]
                        The output prefixes of all conditions. If multiple
                        conditions need to be assigned, please use spaces to
                        separate them. For example, prefix_condition1
                        prefix_condition2.

additional arguments:
  --tsspredator_path TSSPREDATOR_PATH
                        If you want to assign the path of TSSpredator, please
                        assign here. Default is /usr/local/bin/TSSpredator.jar
  --max_height MAX_HEIGHT, -he MAX_HEIGHT
                        This value relates to the minimum number of read
                        starts at a certain genomic position to be considered
                        as a TSS candidate. During optimization, --max_height
                        will be never larger than this value. Default is 2.5.
  --max_height_reduction MAX_HEIGHT_REDUCTION, -rh MAX_HEIGHT_REDUCTION
                        When comparing different genomes/conditions and the
                        step height threshold is reached in at least one
                        genome/condition, the threshold is reduced for the
                        other genomes/conditions by the value set here. This
                        value must be smaller than the step height threshold.
                        During optimization, --max_height_reduction will be
                        never larger than this value. Default is 2.4.
  --max_factor MAX_FACTOR, -fa MAX_FACTOR
                        This is the minimum factor by which the TSS height has
                        to exceed the local expression background. During
                        optimization, --max_factor will be never larger than
                        this value. Default is 10.
  --max_factor_reduction MAX_FACTOR_REDUCTION, -rf MAX_FACTOR_REDUCTION
                        When comparing different genomes/conditions and the
                        step factor threshold is reached in at least one
                        genome/condition, the threshold is reduced for the
                        other genomes/conditions by the value set here. This
                        value must be smaller than the step factor threshold.
                        During optimization, --max_factor_reduction will be
                        never larger than this value. Default is 9.9.
  --max_base_height MAX_BASE_HEIGHT, -bh MAX_BASE_HEIGHT
                        This is the minimum number of reads should be mapped
                        on TSS. During optimization, --max_base_height will be
                        never larger than this value. Default is 0.06.
  --max_enrichment_factor MAX_ENRICHMENT_FACTOR, -ef MAX_ENRICHMENT_FACTOR
                        This is the minimum enrichment factor. During
                        optimization, --max_enrichment_factor will be never
                        larger than this value. Default is 6.0.
  --max_processing_factor MAX_PROCESSING_FACTOR, -pf MAX_PROCESSING_FACTOR
                        This is the minimum processing factor. If untreated
                        library is higher than the treated library and above
                        which the TSS candidate is considered as a processing
                        site and not annotated as detected. During
                        optimization, --max_processing_factor will be never
                        larger than this value. Default is 6.0
  --utr_length UTR_LENGTH, -u UTR_LENGTH
                        The length of UTR. It is for Primary and Secondary
                        TSSs. Default is 300.
  --cluster CLUSTER, -cu CLUSTER
                        This value defines the maximal distance (nucleotides)
                        between TSS candidates have to be clustered together.
                        If the distance between these multiple TSSs is smaller
                        or equal to this value, only one of them will be
                        printed out. Default is 2.
  --parallels PARALLELS, -c PARALLELS
                        Parallel runs for optimization. Default is 4.
  --steps STEPS, -s STEPS
                        The total runs for optimization. Default is 4000 runs.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Based on the programs (TSS/processing site), Output files are stored in
<code class="docutils literal"><span class="pre">$ANNOgesic/output/TSSs/optimized_TSSpredator</span></code> or <code class="docutils literal"><span class="pre">$ANNOgesic/output/processing_sites/optimized_TSSpredator</span></code>.
Two output files are as following:</p>
<p><strong>stat_$GENOME.csv:</strong> Stores the information of every run. The first column is the number of run.
The second column is the parameter set. <code class="docutils literal"><span class="pre">he</span></code> represents height; <code class="docutils literal"><span class="pre">rh</span></code> represents
height reduction; <code class="docutils literal"><span class="pre">fa</span></code> means factor; <code class="docutils literal"><span class="pre">rf</span></code> means factor reduction; <code class="docutils literal"><span class="pre">bh</span></code> indicates
base height; <code class="docutils literal"><span class="pre">ef</span></code> indicates enrichment factor; <code class="docutils literal"><span class="pre">pf</span></code> means processing factor. About the details
of parameters, please refer to <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.
For example, <code class="docutils literal"><span class="pre">he_2.0_rh_1.8_fa_4.4_rf_2.8_bh_0.08_ef_3.0_pf_2.6</span></code> means that height is 2.0,
height reduction is 1.8, factor is 4.4, factor reduction is 2.8, base height is 0.08,
enrichment factor is 3.0 and processing factor is 2.6. The third column is
the number of true positive. The fourth column is true positive rate. The fifth
columns is the number of false positive. The sixth is false positive rate.
The seventh column is the number of false negatives. The eighth
column is missing rate.</p>
<p><strong>best_$GENOME.csv:</strong> Stores the best parameter set. The meanings of all columns are the same as <code class="docutils literal"><span class="pre">stat_$GENOME.csv</span></code>.</p>
</div>
<div class="section" id="screenshot-screenshot-generation">
<span id="screenshot"></span><h2>screenshot (screenshot generation)<a class="headerlink" href="#screenshot-screenshot-generation" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">screenshot</span></code> can generate batch files for producing screenshot of <a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>.
Generating screenshots can reduce the time for checking the results in genome browser.
When the batch files is produced, the user just needs to open <a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>, then presses <code class="docutils literal"><span class="pre">tools</span></code>
on the top tags and choose <code class="docutils literal"><span class="pre">run</span> <span class="pre">batch</span> <span class="pre">script</span></code>. The program will automatically produce screenshots.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files that the user wants to produce screenshots:</strong> All screenshots will be produced based on the positions of <code class="docutils literal"><span class="pre">--main_gff</span></code>.
If comparing <code class="docutils literal"><span class="pre">--main_gff</span></code> with other features is required, please assign gff files of other features to <code class="docutils literal"><span class="pre">--side_gffs</span></code>.</p>
<p><strong>Fasta files of the genomes</strong></p>
<p><strong>Wiggle files of TEX+/- or fragmented/conventional libraries:</strong> Please check the section <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic screenshot [-h] [--project_path [PROJECT_PATH]] --fasta_file
                            FASTA_FILE --main_gff MAIN_GFF
                            [--side_gffs SIDE_GFFS [SIDE_GFFS ...]]
                            [--tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]]
                            [--frag_libs FRAG_LIBS [FRAG_LIBS ...]]
                            --output_folder OUTPUT_FOLDER [--height HEIGHT]
                            [--present PRESENT]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --fasta_file FASTA_FILE, -f FASTA_FILE
                        Path of the genome fasta file.
  --main_gff MAIN_GFF, -mg MAIN_GFF
                        Screenshot will be generated based on the position of
                        --main_gff file.
  --side_gffs SIDE_GFFS [SIDE_GFFS ...], -sg SIDE_GFFS [SIDE_GFFS ...]
                        If you want to compare other features with the
                        --main_gff, please assign the path of other gff files
                        here. If multiple gff files need to be assigned, just
                        use spaces to separate them.
  --tex_notex_libs TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...], -tl TEX_NOTEX_LIBS [TEX_NOTEX_LIBS ...]
                        If TEX+/- wig file is required, please also assign the
                        proper format here. The format is:
                        wig_file_path:TEX+/-(tex or notex):condition_id(intege
                        r):replicate_id(alphabet):strand(+ or -). If multiple
                        wig files need to be assigned, please use spaces to
                        separate the wig files. For example,
                        $WIG_PATH_1:tex:1:a:+ $WIG_PATH_2:tex:1:a:-.
  --frag_libs FRAG_LIBS [FRAG_LIBS ...], -fl FRAG_LIBS [FRAG_LIBS ...]
                        If the fragmented (or conventional) wig file is
                        required, please also assign the proper format here.
                        The format is: wig_file_path:fragmented(frag):conditio
                        n_id(integer):replicate_id(alphabet):strand(+ or -).
                        If multiple wig files need to be assigned, please use
                        spaces to separate the wig files. For example,
                        $WIG_PATH_1:frag:1:a:+ $WIG_PATH_2:frag:1:a:-.
  --output_folder OUTPUT_FOLDER, -o OUTPUT_FOLDER
                        Please assign the output folder. It will create a sub-
                        folder &quot;screenshots&quot; in --output_folder to store the
                        results.

additional arguments:
  --height HEIGHT, -he HEIGHT
                        Please assign the height of the screenshot. Default is
                        1500.
  --present PRESENT, -p PRESENT
                        The presentation type of the features in the
                        screenshot. expand/collapse/squish. Default is expand.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Based on the paths of <code class="docutils literal"><span class="pre">--main_gff</span></code>, <code class="docutils literal"><span class="pre">screenshot</span></code> will generate a folder - <code class="docutils literal"><span class="pre">screenshots</span></code> under the
folder of <code class="docutils literal"><span class="pre">--main_gff</span></code>. Output files will be stored in this folder. the output files and folders are
as following:</p>
<p><strong>forward.txt:</strong> Batch file of the forward strand for running on IGV.</p>
<p><strong>reverse.txt:</strong> Batch file of reverse strand for running on IGV.</p>
<p><strong>forward:</strong> Folder for storing screenshots of the forward strand.</p>
<p><strong>reverse:</strong> Folder for storing screenshots of the reverse strand.</p>
<p>When batch files are executed on IGV, the screenshots will be automatically stored in the folder called <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>.
Format of the filenames will be <code class="docutils literal"><span class="pre">$GENOME:$START-$END.png</span></code>. For example, <code class="docutils literal"><span class="pre">NC_007795:1051529-1051696.png</span></code>
means the genome is NC_007795, the feature&#8217;s start point is 1051529 and end point is
1051696.</p>
</div>
<div class="section" id="color-png-color-screenshot">
<span id="color-png"></span><h2>color_png (color screenshot)<a class="headerlink" href="#color-png-color-screenshot" title="Permalink to this headline">Â¶</a></h2>
<p><code class="docutils literal"><span class="pre">color_png</span></code> is a following procedure of <code class="docutils literal"><span class="pre">screenshot</span></code>. If numerous samples are included in one figure,
Tracks will be difficult to check. <code class="docutils literal"><span class="pre">color_png</span></code> can color the tracks based on TEX +/- libraries
for improving the checking process.</p>
<ul class="simple">
<li><strong>Required tools</strong></li>
</ul>
<p><a class="reference external" href="http://www.imagemagick.org/script/index.php">ImageMagick</a>.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>The screenshots folder:</strong> Please make sure the folders of <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>
exist in the folder of <code class="docutils literal"><span class="pre">screenshots</span></code>.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">annogesic</span> <span class="n">color_png</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]]</span>
                           <span class="o">--</span><span class="n">screenshot_folder</span> <span class="n">SCREENSHOT_FOLDER</span>
                           <span class="o">--</span><span class="n">track_number</span> <span class="n">TRACK_NUMBER</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">imagemagick_covert_path</span> <span class="n">IMAGEMAGICK_COVERT_PATH</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">basic</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">project_path</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">],</span> <span class="o">-</span><span class="n">pj</span> <span class="p">[</span><span class="n">PROJECT_PATH</span><span class="p">]</span>
                        <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>
  <span class="o">--</span><span class="n">screenshot_folder</span> <span class="n">SCREENSHOT_FOLDER</span><span class="p">,</span> <span class="o">-</span><span class="n">f</span> <span class="n">SCREENSHOT_FOLDER</span>
                        <span class="n">The</span> <span class="n">folder</span> <span class="n">which</span> <span class="n">stores</span> <span class="s2">&quot;screenshots&quot;</span> <span class="p">(</span><span class="n">a</span> <span class="n">folder</span>
                        <span class="n">generated</span> <span class="n">by</span> <span class="n">subcommand</span> <span class="s2">&quot;screenshot&quot;</span><span class="p">)</span><span class="o">.</span>
  <span class="o">--</span><span class="n">track_number</span> <span class="n">TRACK_NUMBER</span><span class="p">,</span> <span class="o">-</span><span class="n">t</span> <span class="n">TRACK_NUMBER</span>
                        <span class="n">The</span> <span class="n">number</span> <span class="n">of</span> <span class="n">tracks</span><span class="o">.</span>

<span class="n">additional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">--</span><span class="n">imagemagick_covert_path</span> <span class="n">IMAGEMAGICK_COVERT_PATH</span><span class="p">,</span> <span class="o">-</span><span class="n">m</span> <span class="n">IMAGEMAGICK_COVERT_PATH</span>
                        <span class="n">Please</span> <span class="n">assign</span> <span class="n">the</span> <span class="n">path</span> <span class="n">of</span> <span class="s2">&quot;convert&quot;</span> <span class="ow">in</span> <span class="n">ImageMagick</span>
                        <span class="n">package</span><span class="o">.</span>
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>The new screenshots will replace the previous ones automatically.</p>
</div>
<div class="section" id="merge-features-merge-all-annotation-features">
<span id="merge-feature"></span><h2>merge_features (merge all annotation features)<a class="headerlink" href="#merge-features-merge-all-annotation-features" title="Permalink to this headline">Â¶</a></h2>
<p>If merging multiple features of the annotation to one gff file is needed, <code class="docutils literal"><span class="pre">merge_features</span></code> can achieve this purpose.
<code class="docutils literal"><span class="pre">merge_features</span></code> can merge all features (the user assigned) to one gff file, and search the parent transcript to each feature.</p>
<ul class="simple">
<li><strong>Required files</strong></li>
</ul>
<p><strong>Gff files of features that the user wants to merge:</strong>
If transcript gff files can be provided, this module will search the parent transcripts to other input features.</p>
<ul class="simple">
<li><strong>Arguments</strong></li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: annogesic merge_features [-h] [--project_path [PROJECT_PATH]]
                                --output_prefix OUTPUT_PREFIX
                                [--transcript_file TRANSCRIPT_FILE]
                                [--other_features_files OTHER_FEATURES_FILES [OTHER_FEATURES_FILES ...]]
                                [--terminator_fuzzy TERMINATOR_FUZZY]
                                [--tss_fuzzy TSS_FUZZY]

optional arguments:
  -h, --help            show this help message and exit

basic arguments:
  --project_path [PROJECT_PATH], -pj [PROJECT_PATH]
                        Path of the project folder. If none is given, the
                        current directory is used.
  --output_prefix OUTPUT_PREFIX, -op OUTPUT_PREFIX
                        Please assign the prefix name of output gff file. The
                        filename will be $OUTPUT_PREFIX_merge_features.gff.
  --transcript_file TRANSCRIPT_FILE, -a TRANSCRIPT_FILE
                        If transcript gff file is provided. The parent
                        transcripts (&quot;Parent&quot; in gff attributes) of all
                        features will be generated. If there is no transcript,
                        this function will just simply combine all input gff
                        files.
  --other_features_files OTHER_FEATURES_FILES [OTHER_FEATURES_FILES ...], -of OTHER_FEATURES_FILES [OTHER_FEATURES_FILES ...]
                        Please assign the gff files (besides transcript gff
                        file) which you want to merge. You can use spaces to
                        separate multiple gff files.

additional arguments:
  --terminator_fuzzy TERMINATOR_FUZZY, -ef TERMINATOR_FUZZY
                        For merging terminators, please assign the fuzzy
                        nucleotides between transcript and terminator.
                        ATTENTION, the third column of gff file of terminator
                        should be exact &quot;terminator&quot;. Default is 30.
  --tss_fuzzy TSS_FUZZY, -tf TSS_FUZZY
                        For merging TSSs, please assign the fuzzy between TSS
                        and transcript. ATTENTION, the third column of gff
                        file of terminator should be exact &quot;TSS&quot;. Default is
                        5.
</pre></div>
</div>
<ul class="simple">
<li><strong>Output files</strong></li>
</ul>
<p>Output gff files are stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/merge_all_features</span></code>. The tag - <code class="docutils literal"><span class="pre">Parent</span></code> in the attributes of
gff file shows the parent transcript.</p>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="tutorial.html" class="btn btn-neutral float-right" title="Tutorial of ANNOgesic" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="docker.html" class="btn btn-neutral" title="Docker image" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Sung-Huan Yu.
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.5.14',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>