<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>ANNOgesic’s subcommands &mdash; ANNOgesic 0.1.3 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ANNOgesic 0.1.3 documentation" href="index.html" />
    <link rel="next" title="Performing a test cast" href="test_case.html" />
    <link rel="prev" title="Dockfiles" href="docker.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="test_case.html" title="Performing a test cast"
             accesskey="N">next</a></li>
        <li class="right" >
          <a href="docker.html" title="Dockfiles"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.1.3 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo_annogesic.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">ANNOgesic&#8217;s subcommands</a><ul>
<li><a class="reference internal" href="#the-format-of-filename">The format of filename</a></li>
<li><a class="reference internal" href="#the-input-format-of-libraries-for-running-annogesic">The input format of libraries for running ANNOgesic</a></li>
<li><a class="reference internal" href="#the-format-of-srna-database">The format of sRNA database</a></li>
<li><a class="reference internal" href="#definition-of-reference-strain-and-target-strain">Definition of reference strain and target strain</a></li>
<li><a class="reference internal" href="#create">create</a></li>
<li><a class="reference internal" href="#get-input-files">get_input_files</a></li>
<li><a class="reference internal" href="#get-target-fasta">get_target_fasta</a></li>
<li><a class="reference internal" href="#annotation-transfer">annotation_transfer</a></li>
<li><a class="reference internal" href="#snp">snp</a></li>
<li><a class="reference internal" href="#tsspredator-tss-and-processing-site-prediction">tsspredator(TSS and processing site prediction)</a></li>
<li><a class="reference internal" href="#transcript-assembly">transcript_assembly</a></li>
<li><a class="reference internal" href="#terminator">terminator</a></li>
<li><a class="reference internal" href="#utr">utr</a></li>
<li><a class="reference internal" href="#srna">srna</a></li>
<li><a class="reference internal" href="#sorf">sorf</a></li>
<li><a class="reference internal" href="#promoter">promoter</a></li>
<li><a class="reference internal" href="#operon">operon</a></li>
<li><a class="reference internal" href="#circrna">circrna</a></li>
<li><a class="reference internal" href="#go-term">go_term</a></li>
<li><a class="reference internal" href="#srna-target">srna_target</a></li>
<li><a class="reference internal" href="#ppi-network">ppi_network</a></li>
<li><a class="reference internal" href="#subcellular-localization">subcellular_localization</a></li>
<li><a class="reference internal" href="#riboswitch">riboswitch</a></li>
<li><a class="reference internal" href="#optimize-tsspredator">optimize_tsspredator</a></li>
<li><a class="reference internal" href="#screenshot">screenshot</a></li>
<li><a class="reference internal" href="#color-png">color_png</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="docker.html"
                        title="previous chapter">Dockfiles</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="test_case.html"
                        title="next chapter">Performing a test cast</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="annogesic-s-subcommands">
<h1>ANNOgesic&#8217;s subcommands<a class="headerlink" href="#annogesic-s-subcommands" title="Permalink to this headline">¶</a></h1>
<p>In general, the subcommands need at least one argument - the analysis
folder. If this is not given ANNOgesic assumes that the current
folder is the analysis folder.</p>
<div class="section" id="the-format-of-filename">
<h2>The format of filename<a class="headerlink" href="#the-format-of-filename" title="Permalink to this headline">¶</a></h2>
<p>In order to recognize the file types and the relationship between different files,
please follow the principle of the format of filename.</p>
<p>Keep the filename of genome fasta file the same as annotation file. For example,
<code class="docutils literal"><span class="pre">NC_007795.fa,</span> <span class="pre">NC_007795.gff,</span> <span class="pre">NC_007795.ptt,</span> <span class="pre">NC_007795.rnt,</span> <span class="pre">NC_007705.gbk</span></code>.</p>
<p>All the <code class="docutils literal"><span class="pre">.gff</span></code> files which are import into <code class="docutils literal"><span class="pre">ANNOgesic</span></code>, please follow the format.
<code class="docutils literal"><span class="pre">$STRAINNAME_$FEATURE.gff</span></code>. For example, <code class="docutils literal"><span class="pre">NC_007795_TSS.gff,</span> <span class="pre">NC_007795_transcript.gff</span></code>.</p>
<p>The names of features are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="31%" />
<col width="69%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Feature name</td>
<td>meaning</td>
</tr>
<tr class="row-even"><td>TSS</td>
<td>transcription starting site</td>
</tr>
<tr class="row-odd"><td>processing</td>
<td>processing site</td>
</tr>
<tr class="row-even"><td>transcript</td>
<td>transcript assembly</td>
</tr>
<tr class="row-odd"><td>sRNA</td>
<td>small RNA</td>
</tr>
<tr class="row-even"><td>sORF</td>
<td>small open reading frame</td>
</tr>
<tr class="row-odd"><td>term</td>
<td>terminator</td>
</tr>
<tr class="row-even"><td>5UTR</td>
<td>5&#8217;UTR</td>
</tr>
<tr class="row-odd"><td>3UTR</td>
<td>3&#8217;UTR</td>
</tr>
<tr class="row-even"><td>circRNA</td>
<td>circular RNA</td>
</tr>
<tr class="row-odd"><td>riboswitch</td>
<td>riboswitch</td>
</tr>
</tbody>
</table>
<p>Please also notice the name of strain or genome in files. It should be avoid to use <code class="docutils literal"><span class="pre">|</span></code> in the name.
Because <code class="docutils literal"><span class="pre">|</span></code> is one character which already used in Unix system. However,
if your fasta files or gff files are download from NCBI through ANNOgesic, ANNOgesic will automatically modify the names.</p>
</div>
<div class="section" id="the-input-format-of-libraries-for-running-annogesic">
<h2>The input format of libraries for running ANNOgesic<a class="headerlink" href="#the-input-format-of-libraries-for-running-annogesic" title="Permalink to this headline">¶</a></h2>
<p>Some modules of <code class="docutils literal"><span class="pre">ANNOgesic</span></code> need to import the information of libraries.
For import the information of libraries, please follow this format:</p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_FILENAME:$LIBRARY_TYPE:$CONDITION:$REPLICATE:$STRAND</span></code></p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_FILENAME</span></code> means the <code class="docutils literal"><span class="pre">.wig</span></code> file.</p>
<p><code class="docutils literal"><span class="pre">$LIBRARY_TYPE</span></code> can be <code class="docutils literal"><span class="pre">tex</span></code> (TEX+) or <code class="docutils literal"><span class="pre">notex</span></code> (TEX-) or <code class="docutils literal"><span class="pre">frag</span></code> (fragmented, not for <code class="docutils literal"><span class="pre">tsspredator</span></code>).</p>
<p><code class="docutils literal"><span class="pre">$CONDITION</span></code> is the index of conditions. Please use 1, 2, 3, ... to represent different conditions.</p>
<p><code class="docutils literal"><span class="pre">$REPLICATE</span></code> is the index of replicates. Please use a, b, c, ... to represent different replicates.</p>
<p><code class="docutils literal"><span class="pre">$STRAND</span></code> is the strand of wiggle file. Please use + or -.</p>
<p>For example,</p>
<p>for TEX +/- treated libraries:</p>
<div class="highlight-python"><div class="highlight"><pre>TSB_OD_0.2_TEX_reverse.wig:tex:1:a:-,
TSB_OD_0.5_TEX_reverse.wig:tex:2:a:-,
TSB_OD_0.2_TEX_forward.wig:tex:1:a:+,
TSB_OD_0.5_TEX_forward.wig:tex:2:a:+,
TSB_OD_0.2_reverse.wig:notex:1:a:-,
TSB_OD_0.5_reverse.wig:notex:2:a:-,
TSB_OD_0.2_forward.wig:notex:1:a:+,
TSB_OD_0.5_forward.wig:notex:2:a:+,
</pre></div>
</div>
<p>All the libraries should be separated by comma like the above.
The libraries should be combined to one line.</p>
<p>for fragmented libraries:</p>
<div class="highlight-python"><div class="highlight"><pre>fragmented_forward.wig:frag:1:a:+,fragmented_reverse.wig:frag:1:a:-
</pre></div>
</div>
</div>
<div class="section" id="the-format-of-srna-database">
<h2>The format of sRNA database<a class="headerlink" href="#the-format-of-srna-database" title="Permalink to this headline">¶</a></h2>
<p>If you want to import sRNA database to sRNA detection, please follow the format. Or the output of sRNA name will become
nonsense. The format is</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ID|$STRAIN|$SRNANAME</dd>
</dl>
<p>for example</p>
<dl class="docutils">
<dt>::</dt>
<dd>srn_4840|S._aureus_NCTC8325|RsaOV</dd>
</dl>
</div>
<div class="section" id="definition-of-reference-strain-and-target-strain">
<h2>Definition of reference strain and target strain<a class="headerlink" href="#definition-of-reference-strain-and-target-strain" title="Permalink to this headline">¶</a></h2>
<p>We call the strain that user want to compute - &#8220;target strain&#8221;.
If user have no fasta file or genome annotation files of &#8220;target strain&#8221;,
ANNOgesic can generate them. It requires a strain which is close to &#8220;target strain&#8221;.
We call this required strain - &#8220;reference strain&#8221;.</p>
</div>
<div class="section" id="create">
<h2>create<a class="headerlink" href="#create" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">create</span></code> generates the required folders for input and
output files. Once these folders are created, the input files have to
be placed into the correct locations. Therefore, after creating the folders,
please put the required files in the proper folders.</p>
<p>BAMs: For <code class="docutils literal"><span class="pre">.bam</span></code> files. It has two sub-folders. <code class="docutils literal"><span class="pre">BAMs_map_reference</span></code>
is for the <code class="docutils literal"><span class="pre">.bam</span></code> files which mapped on &#8220;reference strain&#8221;.
<code class="docutils literal"><span class="pre">BAMs_map_target</span></code> is for the <code class="docutils literal"><span class="pre">.bam</span></code> files which mapped on &#8220;target strain&#8221;.</p>
<p>database: For all databases</p>
<p>manual_TSS: If you detected transcription starting sites(TSSs) manually,
you can put the results here. When you compute TSSs, ANNOgesic
can merge manual and predicted TSSs together.
If you want to run <code class="docutils literal"><span class="pre">TSS_optimization</span></code>  for TSSs, it is also necessary.
Please use gff3 format.</p>
<p>manual_processing_site: It is similar with <code class="docutils literal"><span class="pre">manual_TSS</span></code>, it is for
processing sites.</p>
<p>mutation_table: If you detected the mutations between reference genome and
target genome manually, please put the file here. When
you run <code class="docutils literal"><span class="pre">get_target_fasta</span></code>, it is required. Please refer
to the section of <code class="docutils literal"><span class="pre">get_target_fasta</span></code> for the format of
mutation table.</p>
<p>reads: If you want to run <code class="docutils literal"><span class="pre">circrna</span></code> and you also want to align read data by ANNOgesic,
please put the read data here. It can also deal with <code class="docutils literal"><span class="pre">.bzip2</span></code> and <code class="docutils literal"><span class="pre">.gzip</span></code>.</p>
<p>reference: For annotation files and fasta files. If the information of
the reference strain can be download from NCBI, you can also get
the files through running <code class="docutils literal"><span class="pre">get_input_files</span></code>.</p>
<p>riboswitch_ID: For storing the file which contains all the Rfam ID of riboswitch.
For the details of format, please refer to the section of
<code class="docutils literal"><span class="pre">riboswicth</span></code>.</p>
<p>wigs: For wiggle files. Based on the methods of RNA-Seq, you can put them to
<code class="docutils literal"><span class="pre">fragment</span></code> (fragmented libraries) or <code class="docutils literal"><span class="pre">tex_notex</span></code> (TEX +/- treated libraries).</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: ANNOgesic.py create [-h] project_path

positional arguments:
  project_path  Name/path of the project.

optional arguments:
  -h, --help    show this help message and exit
</pre></div>
</div>
</div>
<div class="section" id="get-input-files">
<h2>get_input_files<a class="headerlink" href="#get-input-files" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">get_input_files</span></code> is the subcommand for downloading required files (fasta, annotation files) from NCBI.
Therefore, user needs to assign the IP of the reference genome in NCBI. For example,
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000013425.1_ASM1342v1">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000013425.1_ASM1342v1</a>
Then, user can choose which kinds of files he/she wants to download.</p>
<ul class="simple">
<li>Pre-reqired information</li>
</ul>
<p><code class="docutils literal"><span class="pre">FTP</span> <span class="pre">source</span></code>: The IP of NCBI.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic get_input_files [-h] [--FTP_path FTP_PATH] [--ref_fasta]
                                 [--ref_gff] [--ref_ptt] [--ref_rnt]
                                 [--ref_gbk] [--convert_embl] [--for_target]
                                 [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --FTP_path FTP_PATH, -F FTP_PATH
                        Path of website where can download the required files.
  --ref_fasta, -f       Download fasta files of reference. Default is False.
  --ref_gff, -g         Download gff files of reference. Default is False.
  --ref_ptt, -p         Download ptt files of reference. Default is False.
  --ref_rnt, -r         Download rnt files of reference. Default is False.
  --ref_gbk, -k         Download genbank files of reference. Default is False.
  --convert_embl, -e    Convert gbk to embl files of reference. Default is
                        False.
  --for_target, -t      If the genome which you download from NCBI is your
                        query sequence(you won&#39;t modify the genome), you can
                        assign the download files to store in target folder in
                        stead of reference folder.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will store in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/input/reference</span></code> if <code class="docutils literal"><span class="pre">--for_target</span></code> is False.
The output files will store in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/target</span></code> if <code class="docutils literal"><span class="pre">--for_target</span></code> is True.</p>
<p><code class="docutils literal"><span class="pre">fasta</span></code>: Fasta files.</p>
<p><code class="docutils literal"><span class="pre">annotation</span></code>: Annotation files.</p>
</div>
<div class="section" id="get-target-fasta">
<h2>get_target_fasta<a class="headerlink" href="#get-target-fasta" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">get_target_fasta</span></code> is the subcommand for generating fasta files of &#8220;target strain&#8221; from
&#8220;reference strain&#8221;. The example of format of mutation table is following:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="12%" />
<col width="12%" />
<col width="8%" />
<col width="9%" />
<col width="20%" />
<col width="13%" />
<col width="4%" />
<col width="12%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>#target_id</td>
<td>reference_id</td>
<td>reference_nt</td>
<td>position</td>
<td>target_nt</td>
<td>impact of correction</td>
<td>locus tag</td>
<td>gene</td>
<td>Description</td>
</tr>
<tr class="row-even"><td>HG003</td>
<td>NC_007795.1</td>
<td>a</td>
<td>333</td>
<td>c</td>
<td>&nbsp;</td>
<td>SAOUHSC_00002</td>
<td>dnaA</td>
<td>XXXXXX</td>
</tr>
<tr class="row-odd"><td>HG003</td>
<td>NC_007795.1</td>
<td>t</td>
<td>543</td>
<td>-</td>
<td>deletion</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>YYYYYY</td>
</tr>
<tr class="row-even"><td>HG003</td>
<td>NC_007795.1</td>
<td>-</td>
<td>600</td>
<td>g</td>
<td>insertion</td>
<td>SAOUHSC_00132</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
<p>If user wants to put the titles of columns in the top, it needs to start from <code class="docutils literal"><span class="pre">#</span></code>.
Each column is separated by <code class="docutils literal"><span class="pre">tab</span></code>. If the mutation type is deletion or insertion,
user can put <code class="docutils literal"><span class="pre">-</span></code> to represent them. The information of <code class="docutils literal"><span class="pre">target_id</span></code>, <code class="docutils literal"><span class="pre">reference_id</span></code>,
<code class="docutils literal"><span class="pre">reference_nt</span></code>, <code class="docutils literal"><span class="pre">position</span></code>, <code class="docutils literal"><span class="pre">target_nt</span></code> is required. The others can be blank.
However, please still use <code class="docutils literal"><span class="pre">tab</span></code> to separate all blank columns.</p>
<p>If user has no mutation information, user can also use <code class="docutils literal"><span class="pre">SNP_calling</span></code>
(one module of <code class="docutils literal"><span class="pre">ANNOgesic</span></code>) to compute it. Please refer to the section of <code class="docutils literal"><span class="pre">SNP_calling</span></code>.</p>
<ul class="simple">
<li>Pre-required files</li>
</ul>
<p>Fasta files of reference genome.</p>
<p>Mutation table which indicates the information of mutations between reference and target strain.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic get_target_fasta [-h] [--ref_fasta_folder REF_FASTA_FOLDER]
                                  [--mutation_table MUTATION_TABLE]
                                  [--output_format OUTPUT_FORMAT]
                                  [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --ref_fasta_folder REF_FASTA_FOLDER, -r REF_FASTA_FOLDER
                        The path of the folder of fasta files.
  --mutation_table MUTATION_TABLE, -m MUTATION_TABLE
                        The path of mutation table.
  --output_format OUTPUT_FORMAT, -o OUTPUT_FORMAT
                        Please assign the output filename and which strain
                        should be included in it. For example:
                        FILE1:strain1_and_strain2,FILE2:strain3. FILE1 is a
                        output fasta file which include the information of
                        strain1 and strain2 (import multi-strains to one file
                        should be separated by &quot;_and_&quot;.) And FILE2 is for
                        strain3. Comma is for split the files.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>Fasta files of target genome will store in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/target/fasta</span></code>.</p>
</div>
<div class="section" id="annotation-transfer">
<h2>annotation_transfer<a class="headerlink" href="#annotation-transfer" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">annotation</span> <span class="pre">transfer</span></code> is the subcommand for transfering the annotation from &#8220;reference strain&#8221;
to target &#8220;strain&#8221;. In this subcommand, we use <a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">PAGIT and RATT</a>
to achieve it. The similarity of &#8220;reference strain&#8221; and &#8220;target strain&#8221; should be closed.
Or it will influence the final results.
Please be attation, before you start to run RATT(annotation transfer),
run <code class="docutils literal"><span class="pre">source</span> <span class="pre">$PAGIT_HOME/sourceme.pagit</span></code> first. it will modify the path for execute RATT.
If you use Docker to execute ANNOgesic, you can skip the path modification.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">PAGIT and RATT</a>.</p>
<p>The genbank files of reference genome.</p>
<p>The fasta files of reference genome.</p>
<p>The fasta files of target genome.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic annotation_transfer [-h] [--RATT_path RATT_PATH]
                                     [--compare_pair COMPARE_PAIR]
                                     [--element ELEMENT]
                                     [--transfer_type TRANSFER_TYPE]
                                     [--ref_embl_gbk REF_EMBL_GBK]
                                     [--ref_fasta REF_FASTA]
                                     [--target_fasta TARGET_FASTA]
                                     [--convert_to_gff_rnt_ptt]
                                     [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --RATT_path RATT_PATH
                        Path of the start.ratt.sh file of RATT folder. Default
                        is start.ratt.sh.
  --compare_pair COMPARE_PAIR, -p COMPARE_PAIR
                        Please assign the name of strain pairs. ex.
                        NC_007795:NEW_NC_007795. The reference strain is
                        NC_007795 and the target strain is NEW_NC_007795.
                        Please assign the names of strain, not filenames of
                        fasta files. If you want to assign multiple strains,
                        please use comma to separate the strains.
  --element ELEMENT, -e ELEMENT
                        It will become the prefix of all output file.
  --transfer_type TRANSFER_TYPE, -t TRANSFER_TYPE
                        The transfer type for running RATT.(details can refer
                        to the manual of RATT.) Default is Strain.
  --ref_embl_gbk REF_EMBL_GBK, -re REF_EMBL_GBK
                        The folder which stores every reference embl
                        folders.If you have no embl folder, you can assign the
                        folder of genbank.
  --ref_fasta REF_FASTA, -rf REF_FASTA
                        The folder of reference fasta files.
  --target_fasta TARGET_FASTA, -tf TARGET_FASTA
                        The folder which stores target fasta files.
  --convert_to_gff_rnt_ptt, -g
                        Do you want to convert to gff, rnt and ptt? Default is
                        False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All the output files from <a class="reference external" href="http://www.sanger.ac.uk/resources/software/pagit/">PAGIT and RATT</a>
will store in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/annotation_transfer</span></code>.</p>
<p>All annotation files(<code class="docutils literal"><span class="pre">.gff</span></code>, <code class="docutils literal"><span class="pre">.ptt</span></code>, <code class="docutils literal"><span class="pre">.rnt</span></code>) will store in <code class="docutils literal"><span class="pre">$ANNOgesic_folder/output/target/annotation</span></code>.</p>
</div>
<div class="section" id="snp">
<h2>snp<a class="headerlink" href="#snp" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">snp</span></code> can detect the potential mutations by comparing the results of alignment and fasta files.
<a class="reference external" href="https://github.com/samtools">Samtools</a>, <a class="reference external" href="https://github.com/samtools">Bcftools</a> are the main tools
for detection of mutations. User can choose programs (with BAQ, without BAQ and extend BAQ) to run <code class="docutils literal"><span class="pre">snp</span></code>.
User can also set the quality, read depth and the fraction of maximum read depth which support for indel.
If you have no fasta file of &#8220;target strain&#8221; and want to generate it, <code class="docutils literal"><span class="pre">snp</span></code> can get it from
the alignment results of &#8220;reference strain&#8221;.</p>
<ul class="simple">
<li>Pre-required files and tools:</li>
</ul>
<p><a class="reference external" href="https://github.com/samtools">Samtools</a>.</p>
<p><a class="reference external" href="https://github.com/samtools">Bcftools</a>.</p>
<p>BAM files for fragmented libraries or TEX +/- treated libraries.</p>
<p>Reference or target genome fasta files.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic snp [-h] [--samtools_path SAMTOOLS_PATH]
                     [--bcftools_path BCFTOOLS_PATH] [--bam_type BAM_TYPE]
                     [--program PROGRAM] [--fasta_path FASTA_PATH]
                     [--tex_bam_path TEX_BAM_PATH]
                     [--frag_bam_path FRAG_BAM_PATH] [--quality QUALITY]
                     [--read_depth READ_DEPTH]
                     [--indel_fraction INDEL_FRACTION] [--ploidy PLOIDY]
                     [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --samtools_path SAMTOOLS_PATH
                        If you want to assign the path of samtools, please
                        assign here.
  --bcftools_path BCFTOOLS_PATH
                        If you want to assign the path of bcftools, please
                        assign here.
  --bam_type BAM_TYPE, -t BAM_TYPE
                        Please assign the type of BAM. If your BAM file are
                        mapped to reference strain and you want to know the
                        mutations between refenece strain and target strain,
                        plase keyin &#39;reference&#39;. If your BAM file are mapped
                        to target strain and you want to check the mutations
                        of genome sequence, please keyin &#39;target&#39;.
  --program PROGRAM, -p PROGRAM
                        Please assign the program for detecting SNP of
                        transcript: 1: calculate with BAQ, 2: calculate
                        without BAQ, 3: calculate with extend BAQ. You can
                        assign more than 1 program. For example: 1,2,3.
                        Default is 1,2,3.
  --fasta_path FASTA_PATH, -f FASTA_PATH
                        The path of genome fasta folder.
  --tex_bam_path TEX_BAM_PATH, -tw TEX_BAM_PATH
                        The path of tex+/- wig folder.
  --frag_bam_path FRAG_BAM_PATH, -fw FRAG_BAM_PATH
                        The path of fragmented wig folder.
  --quality QUALITY, -q QUALITY
                        The minimum quality which considers a real snp.
                        Default is 40.
  --read_depth READ_DEPTH, -d READ_DEPTH
                        The minimum read depth. Default is 100.
  --indel_fraction INDEL_FRACTION, -imf INDEL_FRACTION
                        The fraction of maximum read depth, which supports
                        insertion of deletion. Default is 0.5.
  --ploidy PLOIDY, -pl PLOIDY
                        haploid or diploid. Default is haploid.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>If <code class="docutils literal"><span class="pre">bam_type</span></code> is <code class="docutils literal"><span class="pre">reference</span></code>,
the results will store in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/compare_reference</span></code>.
If it is <code class="docutils literal"><span class="pre">target</span></code>, the results will store in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/validate_target</span></code>.</p>
<p>The raw data from <a class="reference external" href="https://github.com/samtools">Samtools</a> and <a class="reference external" href="https://github.com/samtools">Bcftools</a>
will store in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/SNP_raw_outputs</span></code>.</p>
<p>The results will store in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/SNP_table</span></code>.</p>
<p>The meaning of file names are:</p>
<p><code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_depth_only.vcf</span></code> means the results only match the condition of read depth.</p>
<p><code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_depth_quality.vcf</span></code> means the results match the condition of read depth and quality.</p>
<p><code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_seq_reference.csv</span></code> is the index of fasta files which generated by <code class="docutils literal"><span class="pre">snp</span></code>.</p>
<p>For example,</p>
<div class="highlight-python"><div class="highlight"><pre>Staphylococcus_aureus_HG003     1632629 .       AaA     AA      57      .
Staphylococcus_aureus_HG003     1632630 .       aA      a       57      .
Staphylococcus_aureus_HG003     1499572 .       T       TT,TTTTT        43.8525 .
</pre></div>
</div>
<p>These mutations will cause conflict. Then the conflict will effect the positions of other mutations.
Therefore, it will generate four different fasta files.
<code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_seq_reference.csv</span></code> is the index for these four fasta files.</p>
<div class="highlight-python"><div class="highlight"><pre>1       1632629 1       1499572:TT      Staphylococcus_aureus_HG003
1       1632629 2       1499572:TTTTT   Staphylococcus_aureus_HG003
2       1632630 1       1499572:TT      Staphylococcus_aureus_HG003
2       1632630 2       1499572:TTTTT   Staphylococcus_aureus_HG003
</pre></div>
</div>
<p>The first column is the index of conflict. The second column is the positions which be selected.
The third one is the index of two potential mutations in the same position. The fourth one is
the position and nucleotides of mutations. The last column is the name of strain.
If you refer to <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/seqs</span></code>, the filename of fasta is like
<code class="docutils literal"><span class="pre">$FILENAME_$STRIANNAME_$INDEXofCONFLICT_$INDEXofTWOMUTATION.fa</span></code>. Therefore, the first line of
<code class="docutils literal"><span class="pre">$STRAIN_$PROGRAM_seq_reference.csv</span></code> will generate
<code class="docutils literal"><span class="pre">Staphylococcus_aureus_HG003_Staphylococcus_aureus_HG003_1_1.fa</span></code>
(if the file name of genome is Staphylococcus_aureus_HG003). The second line will generate
<code class="docutils literal"><span class="pre">Staphylococcus_aureus_HG003_Staphylococcus_aureus_HG003_1_2.fa</span></code> and so forth.</p>
<p>The statistics files will store in <code class="docutils literal"><span class="pre">$ANNOgesic/output/SNP_calling/$BAM_TYPE/statistics</span></code>.</p>
</div>
<div class="section" id="tsspredator-tss-and-processing-site-prediction">
<h2>tsspredator(TSS and processing site prediction)<a class="headerlink" href="#tsspredator-tss-and-processing-site-prediction" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">tsspredator</span></code> can generate the candidates of TSSs and processing sites. The main tool is
<a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>. We can easily switch the
TEX+ libraries and TEX- libraries to detect processing sites. User can assign the parameters
of <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>. If User want to get the
optimized parameters of <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>,
there is <code class="docutils literal"><span class="pre">optimize_tsspredator</span></code> for this purpose. Please refer to the section of <code class="docutils literal"><span class="pre">optimize_tsspredator</span></code>.</p>
<p>For the information of libraries, please refer to the section
<code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p>The libraries and wiggle files of TEX +/-. Please refer to <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>Fasta file of genome sequence.</p>
<p>Gff file of genome annotation.</p>
<p>If user has gff file of manual detected TSSs, <code class="docutils literal"><span class="pre">tsspredator</span></code> can merge the manual one
and predicted one.</p>
<p>If user wants to compare TSSs with transcripts, it also need the gff files of transcripts.
For the transcripts, please refer to the section of <code class="docutils literal"><span class="pre">transcript_assembly</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic tsspredator [-h] [--TSSpredator_path TSSPREDATOR_PATH]
                             [--fasta_folder FASTA_FOLDER]
                             [--annotation_folder ANNOTATION_FOLDER]
                             [--wig_folder WIG_FOLDER] [--height HEIGHT]
                             [--height_reduction HEIGHT_REDUCTION]
                             [--factor FACTOR]
                             [--factor_reduction FACTOR_REDUCTION]
                             [--enrichment_factor ENRICHMENT_FACTOR]
                             [--processing_factor PROCESSING_FACTOR]
                             [--base_height BASE_HEIGHT]
                             [--replicate_match REPLICATE_MATCH]
                             [--utr_length UTR_LENGTH] [--lib LIB]
                             [--output_prefix OUTPUT_PREFIX]
                             [--merge_manual MERGE_MANUAL] [--statistics]
                             [--validate_gene]
                             [--compute_program COMPUTE_PROGRAM]
                             [--compare_transcript_assembly COMPARE_TRANSCRIPT_ASSEMBLY]
                             [--fuzzy FUZZY] [--cluster CLUSTER]
                             [--length LENGTH] [--re_check_orphan]
                             [--overlap_feature OVERLAP_FEATURE]
                             [--reference_gff_folder REFERENCE_GFF_FOLDER]
                             [--remove_low_expression REMOVE_LOW_EXPRESSION]
                             [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TSSpredator_path TSSPREDATOR_PATH
                        If you want to assign the path of TSSpredator, please
                        assign here. Default is /usr/local/bin/TSSpredator.jar
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        Path of the target genome fasta folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        Path of the target genome annotation gff folder.
  --wig_folder WIG_FOLDER, -w WIG_FOLDER
                        The folder of the wig folder.
  --height HEIGHT, -he HEIGHT
                        This value relates to the minimal number of read
                        starts at a certain genomic position to be considered
                        as a TSS candidate. Default is 0.3.
  --height_reduction HEIGHT_REDUCTION, -rh HEIGHT_REDUCTION
                        When comparing different strains/conditions and the
                        step height threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step height threshold.
                        Default is 0.2.
  --factor FACTOR, -fa FACTOR
                        This is the minimal factor by which the TSS height has
                        to exceed the local expression background. Default is
                        2.0.
  --factor_reduction FACTOR_REDUCTION, -rf FACTOR_REDUCTION
                        When comparing different strains/conditions and the
                        step factor threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step factor threshold.
                        Default is 0.5.
  --enrichment_factor ENRICHMENT_FACTOR, -ef ENRICHMENT_FACTOR
                        This is the minimal enrichment factor. During
                        optimization will never larger than this value.
                        Default is 2.0.
  --processing_factor PROCESSING_FACTOR, -pf PROCESSING_FACTOR
                        This is the minimal processing factor. If untreated
                        library is higher than the treated library and above
                        which the TSS candidate is considered as a processing
                        site and not annotated as detected. During
                        optimization will never larger than this value.
                        Default is 1.5.
  --base_height BASE_HEIGHT, -bh BASE_HEIGHT
                        This is the minimal number of reads should be mapped
                        on TSS. Default is 0.0.
  --replicate_match REPLICATE_MATCH, -rm REPLICATE_MATCH
                        The TSS candidates should match to how many number of
                        the replicates. Default is 1.
  --utr_length UTR_LENGTH, -u UTR_LENGTH
                        The length of UTR. It is for Primary and Secondary
                        definition. Default is 300.
  --lib LIB, -l LIB     The libraries of wig files for TSSpredator. The format
                        is: wig_file_name:tex_treat_or_not(tex or notex):condi
                        tion_id(integer):replicate_id(alphabet):strand(+ or
                        -). If you have multiple wig files, please use comma
                        to separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --output_prefix OUTPUT_PREFIX, -p OUTPUT_PREFIX
                        The output prefix of all conditions. If you have
                        multiple conditions, please use comma to separate
                        them. For example,
                        prefix_condition1,prefix_condition2.
  --merge_manual MERGE_MANUAL, -m MERGE_MANUAL
                        If you have gff file of manual checked TSS, you can
                        use this function to merge manual checked ones and
                        predicted ones. please assign the path of gff file
                        of manual checked TSS.
  --statistics, -s      Doing statistics for TSS candidates. it will store in
                        statistics folder. Default is False.
  --validate_gene, -v   Using TSS candidates to validate genes in annotation
                        file. it will store in statistics folder. Default is
                        False.
  --compute_program COMPUTE_PROGRAM, -t COMPUTE_PROGRAM
                        Which program do you want to predict. (TSS or
                        processing_site) Default is TSS.
  --compare_transcript_assembly COMPARE_TRANSCRIPT_ASSEMBLY, -ta COMPARE_TRANSCRIPT_ASSEMBLY
                        If you want to compare with transcriptome assembly,
                        please assign the folder of gff file of transcript
                        assembly. Default is False.
  --fuzzy FUZZY, -fu FUZZY
                        The fuzzy for comparing TSS and transcript assembly.
                        Default is 5.
  --cluster CLUSTER, -c CLUSTER
                        This number is for compare manual detected TSS and
                        prediced one. If the position between manual checked
                        one and predicted one is smaller or equal than this
                        value, It will only print one of them. Default is 2.
  --length LENGTH, -le LENGTH
                        The length of genome that you want to compare between
                        predicted one and manual checked one for statistics.
                        If you want to compare whole genome, please don&#39;t turn
                        it on. The default is comparing whole genome.
  --re_check_orphan, -ro
                        If your annotation file lack information of gene or
                        locus_tag, you can turn it on. It will try to compare
                        with CDS. Default is False.
  --overlap_feature OVERLAP_FEATURE, -of OVERLAP_FEATURE
                        If processing site and TSS are overlap, you can keep
                        &quot;TSS&quot; or &quot;processing_site&quot; or &quot;both&quot;. Default is both.
  --reference_gff_folder REFERENCE_GFF_FOLDER, -rg REFERENCE_GFF_FOLDER
                        For --overlap_feature, if you want to only keep &quot;TSS&quot;
                        or &quot;processing_site&quot;, you need to assign the
                        --reference_gff_folder. If you are running TSS, please
                        assign the folder of processing site. If you are
                        running processing_site, please assign the folder of
                        TSS. If you want to keep &quot;both&quot; at overlap position,
                        please don&#39;t turn it on. Default is None(for keep
                        both).
  --remove_low_expression REMOVE_LOW_EXPRESSION, -rl REMOVE_LOW_EXPRESSION
                        If you want to remove low expressed TSS/processing
                        site, please assign the file of manual checked gff
                        file here. It will remove the low expressed ones
                        based on comparisone of manual checked ones.
                        Please Be ATTENTION: this parameter may
                        remove some True positive, too. So, please make sure
                        you want to do it.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/TSS</span></code>.</p>
<p><code class="docutils literal"><span class="pre">MasterTables</span></code>: The MasterTable from <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p>The output files of processing sites are similar. Just replace <code class="docutils literal"><span class="pre">TSS</span></code> to <code class="docutils literal"><span class="pre">processing_site</span></code>
like <code class="docutils literal"><span class="pre">$ANNOgesic/output/processing_site</span></code>.</p>
<p><code class="docutils literal"><span class="pre">configs</span></code>: The configuration files for running TSSpredator.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files of TSSs.</p>
<p>There are some useful tags in the attributes of gff files:</p>
<p><code class="docutils literal"><span class="pre">Method</span></code>: The TSSs are from manual detection or <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p><code class="docutils literal"><span class="pre">type</span></code>: The type of TSSs. It could be Primary, Secondary, Internal, Antisense or Orphan.</p>
<p><code class="docutils literal"><span class="pre">UTR_length</span></code>: The length of UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_gene</span></code>: Which genes are associated with this TSS.</p>
<p><code class="docutils literal"><span class="pre">Parent_tran</span></code>: Which transcript are associated with this TSS, if user has compared with transcript.</p>
<p>If user has compared with genome annotation files, the tag - <code class="docutils literal"><span class="pre">start_TSS</span></code> will appear in the gff files
of genome annotation. It represents the TSSs which associates with the CDS/tRNA/rRNA.</p>
<p>If user has compared with transcripts, the tag - <code class="docutils literal"><span class="pre">associated_tss</span></code> will appear in the gff files
of transcript. It will show the associated TSSs which is in the transcript.</p>
</div>
<div class="section" id="transcript-assembly">
<h2>transcript_assembly<a class="headerlink" href="#transcript-assembly" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">transcript_assembly</span></code> will detect transcripts based on the coverage.</p>
<p>For importing the information of libraries, please refer to the section of
<code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p>Wiggle files of fragmented libraries or TEX +/- treated libraries. We don&#8217;t recommend that only
use TEX +/- treated libraries to generate transcripts. It will lose some information
in 3&#8217;end.</p>
<p>If user wants to compare transcripts with TSSs, it requires <code class="docutils literal"><span class="pre">.gff</span></code> files of TSSs.
If user wants to compare transcripts with genome anntation, it requires <code class="docutils literal"><span class="pre">.gff</span></code> files of genomes.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic transcript_assembly [-h]
                                     [--annotation_folder ANNOTATION_FOLDER]
                                     [--length LENGTH]
                                     [--tex_wig_path TEX_WIG_PATH]
                                     [--frag_wig_path FRAG_WIG_PATH]
                                     [--height HEIGHT] [--width WIDTH]
                                     [--tolerance TOLERANCE]
                                     [--tolerance_coverage TOLERANCE_COVERAGE]
                                     [--replicates_tex REPLICATES_TEX]
                                     [--replicates_frag REPLICATES_FRAG]
                                     [--tex_notex TEX_NOTEX]
                                     [--compare_TSS COMPARE_TSS]
                                     [--compare_genome_annotation COMPARE_GENOME_ANNOTATION]
                                     [--compare_feature_genome COMPARE_FEATURE_GENOME]
                                     [--TSS_fuzzy TSS_FUZZY]
                                     [--Tex_treated_libs TEX_TREATED_LIBS]
                                     [--fragmented_libs FRAGMENTED_LIBS]
                                     [--table_best]
                                     [--terminator_folder TERMINATOR_FOLDER]
                                     [--fuzzy_term FUZZY_TERM]
                                     [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        It is for comparing transcript assembly and genome
                        annotation gff file. It can use annotation gff file
                        as reference and modify transcript assembly file.
                        If you want to do it, please assign the annotation
                        gff folder. Otherwise, don&#39;t turn it on.
  --length LENGTH, -l LENGTH
                        The minimum width of transcript. It is for comparing
                        to annotation file(--annotation_folder). If you want
                        to compare with annotation files, it will be the final
                        output. If you don&#39;t want to compare with annotation
                        files, --width would be minimum length for the final
                        output. The default is 20.
  --tex_wig_path TEX_WIG_PATH, -tw TEX_WIG_PATH
                        The path of TEX+/- wig folder.
  --frag_wig_path FRAG_WIG_PATH, -fw FRAG_WIG_PATH
                        The path of fragment wig folder.
  --height HEIGHT, -he HEIGHT
                        The minimum height of coverage to be a transcript. The
                        default is 10.
  --width WIDTH, -w WIDTH
                        The minimum width of transcript. It is for not
                        comparing to annotation file(--annotation_folder). If
                        you don&#39;t want to compare with annotation files, it
                        will be the final output. Otherwise, --length would be
                        the minimum length of transcript for the final output.
                        The default is 20.
  --tolerance TOLERANCE, -t TOLERANCE
                        This number indicates how many nucleotides which
                        coverages drop below --height can be ignore. The
                        default is 5.
  --tolerance_coverage TOLERANCE_COVERAGE, -tc TOLERANCE_COVERAGE
                        If the coverage is lower than tolerance_coverage, even
                        the range is within --tolerance, it will terminate the
                        current transcript. The default is 0.
  --replicates_tex REPLICATES_TEX, -rt REPLICATES_TEX
                        The position is included in the current transcript if
                        the supported replicates are more than --replicates_tex.
                        (for tex +/- library)
  --replicates_frag REPLICATES_FRAG, -rf REPLICATES_FRAG
                        The position is included in the current transcript if
                        the supported replicates are more than --replicates_frag.
                        (for fragmented library)
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        If you use tex +/- libraries to run transcript
                        assembly, the transcripts should be detected by both
                        or just one. (1 or 2). Default is 1.
  --compare_TSS COMPARE_TSS, -ct COMPARE_TSS
                        If you want to compare with TSS, please assign TSS
                        folder.
  --compare_genome_annotation COMPARE_GENOME_ANNOTATION, -cg COMPARE_GENOME_ANNOTATION
                        If you want to compare with genome annotation file,
                        please assign genome annotation folder.
  --compare_feature_genome COMPARE_FEATURE_GENOME, -cf COMPARE_FEATURE_GENOME
                        If you want to compare with genome annotation file,
                        please assign the feature which you want to compare.
                        Default is gene. If you want to compare more than one
                        feature, just insert comma between each feature,
                        such as gene,CDS.
  --TSS_fuzzy TSS_FUZZY, -fu TSS_FUZZY
                        The fuzzy for comparing TSS and transcript assembly.
                        Default is 5.
  --Tex_treated_libs TEX_TREATED_LIBS, -tl TEX_TREATED_LIBS
                        Tex +/- library. The format is:
                        wig_file_name:tex_treat_or_not(tex or notex):condition
                        _id(integer):replicate_id(alphabet):strand(+ or -). If
                        you have multiple wig files, please use comma to
                        separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --fragmented_libs FRAGMENTED_LIBS, -fl FRAGMENTED_LIBS
                        Fragmented library. The format is: wig_file_name:fragm
                        ented(frag):condition_id(integer):replicate_id(alphabe
                        t):strand(+ or -). If you have multiple wig files,
                        please use comma to separate the wig files. For
                        example, wig1:frag:1:a:+,wig2:frag:1:a:-.
  --table_best, -tb     The output table only includes the best library.
                        Default is False.
  --terminator_folder TERMINATOR_FOLDER, -tr TERMINATOR_FOLDER
                        If you want to compare between transcripts and
                        terminators, you can assign the folder of gff files of
                        terminator here. Default is None.
  --fuzzy_term FUZZY_TERM, -fz FUZZY_TERM
                        If you want to compare between transcripts and
                        terminators, please assign the fuzzy here. Default is
                        30.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/transcriptome_assembly</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files of transcript.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: For more details of transcripts.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p>There are some useful tags in gff files.</p>
<p><code class="docutils literal"><span class="pre">compare_FEATURE</span></code>: The situation of overlap between transcripts and features(&#8211;compare_feature_genome)
(If user has compared transcripts with genome annotation.)</p>
<p><code class="docutils literal"><span class="pre">associated_tss</span></code>: Which TSSs are located in this transcripts.
(If user has compared transcripts with TSSs.)</p>
<p><code class="docutils literal"><span class="pre">associated_$FEATURE</span></code>: It shows the feature(&#8211;compare_feature_genome) which are located in this transcripts.
(If user has compared transcripts with genome annotations.)</p>
<p><code class="docutils literal"><span class="pre">detect_lib</span></code>: The transcript is detected by tex-treated libraries or fragmented libraries.</p>
<p><code class="docutils literal"><span class="pre">best_avg_coverage</span></code>: The average coverage of highest expressed library.</p>
<p>If user has compared transcripts with genome annotations. The tag - <code class="docutils literal"><span class="pre">Parent_tran</span></code> will appear
in the gff files of genome annotations. It will show which transcirpt that CDSs/tRNAs/rRNAs are located.</p>
<p>If user has compared transcripts with TSSs. The tag - <code class="docutils literal"><span class="pre">Parent_tran</span></code> will appear
in the gff files of TSSs. It will show which transcripts that TSSs are located.</p>
</div>
<div class="section" id="terminator">
<h2>terminator<a class="headerlink" href="#terminator" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">terminator</span></code> will predict the rho-independent terminators. <code class="docutils literal"><span class="pre">ANNOgesic</span></code> combine the results of
two methods in order to get more reliable candidates. First one is using <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a>.
The other one is detect the specific secondary structure between converging pairs
of transcripts and CDSs. <code class="docutils literal"><span class="pre">ANNOgesic</span></code> can also compare with coverages in order to generate the terminators
which has coverage significant decreasing.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a></p>
<p>RNAfold of <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a>.</p>
<p>Gff files target genome annotation.</p>
<p>Fasta files of target genome sequence.</p>
<p>Wiggle files of TEX +/- treated libraries or fragmented libraries. we don&#8217;t
recommand only use TEX +/- treated libraries. Because it will lose some information in 3&#8217;end.</p>
<p>Gff files of transcript.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic terminator [-h] [--TransTermHP_path TRANSTERMHP_PATH]
                            [--expterm_path EXPTERM_PATH]
                            [--RNAfold_path RNAFOLD_PATH]
                            [--fasta_folder FASTA_FOLDER]
                            [--annotation_folder ANNOTATION_FOLDER]
                            [--transcript_folder TRANSCRIPT_FOLDER]
                            [--sRNA SRNA] [--statistics]
                            [--tex_wig_folder TEX_WIG_FOLDER]
                            [--frag_wig_folder FRAG_WIG_FOLDER]
                            [--decrease DECREASE]
                            [--fuzzy_detect_coverage FUZZY_DETECT_COVERAGE]
                            [--fuzzy_within_transcript FUZZY_WITHIN_TRANSCRIPT]
                            [--fuzzy_downstream_transcript FUZZY_DOWNSTREAM_TRANSCRIPT]
                            [--fuzzy_within_gene FUZZY_WITHIN_GENE]
                            [--fuzzy_downstream_gene FUZZY_DOWNSTREAM_GENE]
                            [--highest_coverage HIGHEST_COVERAGE]
                            [-tl TEX_NOTEX_LIBS] [-fl FRAG_LIBS]
                            [-te TEX_NOTEX] [-rt REPLICATES_TEX]
                            [-rf REPLICATES_FRAG] [-tb] [-ml MIN_LOOP_LENGTH]
                            [-Ml MAX_LOOP_LENGTH] [-ms MIN_STEM_LENGTH]
                            [-Ms MAX_STEM_LENGTH] [-mr MISS_RATE]
                            [-mu MIN_U_TAIL_LENGTH] [-ru RANGE_U_TAIL]
                            [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TransTermHP_path TRANSTERMHP_PATH
                        Please assign the path of &quot;transterm&quot; in TransTermHP.
  --expterm_path EXPTERM_PATH
                        Please assign the path of your expterm.dat for
                        TransTermHP. Default is /usr/local/bin/expterm.dat
  --RNAfold_path RNAFOLD_PATH
                        If you want to assign the path of &quot;RNAfold&quot; of Vienna
                        package, please assign here.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        The path of genome fasta folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --transcript_folder TRANSCRIPT_FOLDER, -a TRANSCRIPT_FOLDER
                        The folder which stores gff files of transcript
                        assembly.
  --sRNA SRNA, -sr SRNA
                        If you want to include sRNA information, please assign
                        the folder of sRNA gff files.
  --statistics, -s      Doing statistics for TransTermHP. The name of
                        statistics file is - stat_terminator_$STRAIN_NAME.csv.
                        Default is False.
  --tex_wig_folder TEX_WIG_FOLDER, -tw TEX_WIG_FOLDER
                        If you want to use tex +/- libraries, please assign
                        tex +/- wig folder.
  --frag_wig_folder FRAG_WIG_FOLDER, -fw FRAG_WIG_FOLDER
                        If you want to use fragmented libraries, please assign
                        fragmented wig folder.
  --decrease DECREASE, -d DECREASE
                        If the (lowest coverage / highest coverage) in the
                        terminator is smaller than this number, it will
                        consider this terminator have coverage dramatic
                        decreasing in it. Default is 0.5.
  --fuzzy_detect_coverage FUZZY_DETECT_COVERAGE, -fc FUZZY_DETECT_COVERAGE
                        It will elongate the number of nucleotides(you assign
                        here) from both terminal site. If it can found the
                        coverage dramatic decreasing within this range, it
                        will still consider the terminator have coverage
                        dramatic decrease in it. Default is 30.
  --fuzzy_within_transcript FUZZY_WITHIN_TRANSCRIPT, -fut FUZZY_WITHIN_TRANSCRIPT
                        If the candidates are within transcript and the
                        distance between the end of gene/transcript and
                        terminator candidate is within this number, it will be
                        consider as terminator. Default is 30.
  --fuzzy_downstream_transcript FUZZY_DOWNSTREAM_TRANSCRIPT, -fdt FUZZY_DOWNSTREAM_TRANSCRIPT
                        If the candidates are downstream of transcript and the
                        distance between the end of gene/transcript and
                        terminator candidate is within this number, it will be
                        consider as terminator. Default is 30.
  --fuzzy_within_gene FUZZY_WITHIN_GENE, -fuc FUZZY_WITHIN_GENE
                        If the candidates are upstream of gene and the
                        distance between the end of gene and terminator
                        candidate is within this number, it will be consider
                        as terminator. Default is 10.
  --fuzzy_downstream_gene FUZZY_DOWNSTREAM_GENE, -fdg FUZZY_DOWNSTREAM_GENE
                        If the candidates are downstream of gene and the
                        distance between the end of gene and terminator
                        candidate is within this number, it will be consider
                        as terminator. Default is 310.
  --highest_coverage HIGHEST_COVERAGE, -hc HIGHEST_COVERAGE
                        If the highest coverage of terminator is below to this
                        number, the terminator will be classify to non-detect,
                        but still included in &quot;all_candidates&quot;. Default is 10.
  -tl TEX_NOTEX_LIBS, --tex_notex_libs TEX_NOTEX_LIBS
                        Library name of tex and notex library. The format is:
                        wig_file_name:tex_treat_or_not(tex or notex):condition
                        _id(integer):replicate_id(alphabet):strand(+ or -). If
                        you have multiple wig files, please use comma to
                        separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  -fl FRAG_LIBS, --frag_libs FRAG_LIBS
                        Library name of fragmented library. The format is: wig
                        _file_name:fragmented(frag):condition_id(integer):repl
                        icate_id(alphabet):strand(+ or -). If you have
                        multiple wig files, please use comma to separate the
                        wig files. For example,
                        wig1:frag:1:a:+,wig2:frag:1:a:-.
  -te TEX_NOTEX, --tex_notex TEX_NOTEX
                        For tex +/- library, terminators should be detected by
                        both or just one.(1/2) Default is 1.
  -rt REPLICATES_TEX, --replicates_tex REPLICATES_TEX
                        The terminator of tex +/- library should be detected by
                        more than this number of replicates.
  -rf REPLICATES_FRAG, --replicates_frag REPLICATES_FRAG
                        The terminator of fragmented library should be
                        detected more by than this number of replicates.
  -tb, --table_best     Output table only contains most decreasing track.
                        Default is False.
  -ml MIN_LOOP_LENGTH, --min_loop_length MIN_LOOP_LENGTH
                        The minimum length of loop for terminator. Default is
                        3 nts.
  -Ml MAX_LOOP_LENGTH, --max_loop_length MAX_LOOP_LENGTH
                        The maximum length of loop for terminator. Default is
                        10 nts.
  -ms MIN_STEM_LENGTH, --min_stem_length MIN_STEM_LENGTH
                        The minimum length of stem for terminator. Default is
                        4 nts.
  -Ms MAX_STEM_LENGTH, --max_stem_length MAX_STEM_LENGTH
                        The maximum length of stem for terminator. Default is
                        20 nts.
  -mr MISS_RATE, --miss_rate MISS_RATE
                        How many percentage of nucleotides which has no base
                        pair in the stem. Default is 0.25.
  -mu MIN_U_TAIL_LENGTH, --min_U_tail_length MIN_U_TAIL_LENGTH
                        The minimum length of U tail for terminator. Default
                        is 3 nts.
  -ru RANGE_U_TAIL, --range_U_tail RANGE_U_TAIL
                        How long of nucleotides that you want to detect for U
                        tail. For example, if --range_U_tail is 6 and
                        --min_U_tail_length is 3, and there are 3 U within 6
                        nts, it will be assigned to detecting U tail
                        successfully. Default is 6.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The output files will stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/terminator</span></code>.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p><code class="docutils literal"><span class="pre">transtermhp</span></code>: All output of <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of terminator.
There are four different sub-folders to store terminators.</p>
<p><code class="docutils literal"><span class="pre">all_candidates</span></code> will store all terminators which <code class="docutils literal"><span class="pre">ANNOgesic</span></code> can detect.</p>
<p><code class="docutils literal"><span class="pre">express</span></code> will store the terminators which has gene expression.</p>
<p><code class="docutils literal"><span class="pre">best</span></code> will store the terminators which not only has gene expression but also
has coverage dramatic decreasing.</p>
<p><code class="docutils literal"><span class="pre">non_express</span></code> will store the terminators which has no gene expression.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of terminators which store more details.</p>
<p>The tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">Method</span></code>: The method that this terminator be detected.</p>
<p><code class="docutils literal"><span class="pre">coverage_decrease</span></code>: The coverage of the terminator has dramatic decreasing or not.</p>
<p><code class="docutils literal"><span class="pre">express</span></code>: The terminator has gene expression or not.</p>
<p><code class="docutils literal"><span class="pre">diff_coverage</span></code>: The highest coverage and lowest coverage of the library which expresses highest.
The numbers in parens are highest coverage and lowest coverage.</p>
</div>
<div class="section" id="utr">
<h2>utr<a class="headerlink" href="#utr" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">utr</span></code> can compare with the information of TSSs, CDSs/tRNAs/sRNAs, transcripts and terminators
to generate proper UTRs. 5&#8217;UTRs are based on detecting the regions between TSSs (which
are located in transcripts) and CDSs/tRNAs/sRNAs. 3&#8217;UTRs are based on detecting the
regions between the end of transcripts and CDSs/tRNAs/sRNAs. If the gff files of TSSs are not computed by
ANNOgesic, please use &#8211;TSS_source. <code class="docutils literal"><span class="pre">utr</span></code> would classify TSSs for the analysis.</p>
<ul class="simple">
<li>Pre-required files</li>
</ul>
<p>Gff files of genome annotations, TSSs and transcripts.</p>
<p>If user wants to combine the information of terminators, it also need the gff files of terminator.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic utr [-h] [--annotation_folder ANNOTATION_FOLDER]
                     [--TSS_folder TSS_FOLDER]
                     [--transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER]
                     [--terminator_folder TERMINATOR_FOLDER] [--TSS_source]
                     [--base_5UTR BASE_5UTR] [--UTR_length UTR_LENGTH]
                     [--base_3UTR BASE_3UTR]
                     [--terminator_fuzzy TERMINATOR_FUZZY]
                     [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        The path of TSS folder.
  --transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER, -a TRANSCRIPT_ASSEMBLY_FOLDER
                        The path of transcriptome assembly folder.
  --terminator_folder TERMINATOR_FOLDER, -e TERMINATOR_FOLDER
                        If you want to add the information of terminator, you
                        can assign the path of terminator folder here.
  --TSS_source, -s      If you generate TSS which is not from ANNOgesic,
                        please turn it on. Default is True(ANNOgesic).
  --base_5UTR BASE_5UTR, -b5 BASE_5UTR
                        Which information that you want to use for generating
                        5&#39;UTR. TSS/transcript/both. Default is both.
  --UTR_length UTR_LENGTH, -l UTR_LENGTH
                        The maximum length of UTR. Default is 300.
  --base_3UTR BASE_3UTR, -b3 BASE_3UTR
                        Which information that you want to use for generating
                        3&#39;UTR. transcript/terminantor/both. Default is
                        transcript.
  --terminator_fuzzy TERMINATOR_FUZZY, -f TERMINATOR_FUZZY
                        This is only for --base_3UTR which assigned by
                        &quot;transcript&quot; or &quot;both&quot;. If the distance(nucleotides)
                        between terminator and the end of transcript lower
                        than this value, it will assign the terminator
                        associated with the 3&#39;UTR. Default is 30.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output of 5&#8217;UTRs will store in <code class="docutils literal"><span class="pre">$ANNOgesic/output/UTR/5UTR</span></code>.</p>
<p>All output of 3&#8217;UTRs will store in <code class="docutils literal"><span class="pre">$ANNOgesic/output/UTR/3UTR</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of 5&#8217;UTR/3&#8217;UTR</p>
<p>The tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">length</span></code>: UTR length.</p>
<p><code class="docutils literal"><span class="pre">associated_cds</span></code>: Which CDSs/rRNAs/tRNAs are associated with this UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_gene</span></code>: Which genes are associated with this UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_tran</span></code>: Which transcript is associated with this UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_tss</span></code>: Which TSSs are associated with this 5&#8217;UTR.</p>
<p><code class="docutils literal"><span class="pre">TSS_type</span></code>: What types of TSSs are associated with this 5&#8217;UTR.</p>
<p><code class="docutils literal"><span class="pre">associated_term</span></code>: Which terminators are associated with this 3&#8217;UTR.</p>
</div>
<div class="section" id="srna">
<h2>srna<a class="headerlink" href="#srna" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">srna</span></code> can predict sRNAs through comparing the transcripts and annotation profile.
The transcripts in intergenic region might be sRNA candidates. Moreover, based on
the information of TSSs and processing sites, we can also predict UTR-derived sRNA candidates.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a>.</p>
<p><a class="reference external" href="http://pages.cs.wisc.edu/~ghost/doc/AFPL/6.50/Ps2pdf.htm">Ps2pdf14</a>.</p>
<p><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">Blast+</a>.</p>
<p><a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a>.</p>
<p><a class="reference external" href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/">nr database</a>.</p>
<p>Gff files of genome annotation and Transcript data.</p>
<p>It can also import more information to improve the accuracy of prediction.</p>
<p>wiggle files: The libraries and wiggle files, Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>If user want to detect the UTR-derived sRNAs, it will be necessary to import
more information.</p>
<p>TSSs: UTR-derived sRNAs must to generated from TSSs. If you don&#8217;t want to detect UTR-derived sRNAs,
you can also import this information for statistics.</p>
<p>processing sites: It is for check the sRNAs which end with processing sites. Moreover,
Some 3&#8217;UTR-derived and interCDS-derived sRNA candidates start
from processing sites not TSSs. If you don&#8217;t want to detect UTR-derived sRNAs,
you can also import this information for statistics.</p>
<p>Optional input file:</p>
<p>Fasta files of genome sequence.</p>
<p>sRNA database: It could be used for searching the known sRNA. The format of header should be
<code class="docutils literal"><span class="pre">$ID|$STRAIN|$SRNANAME</span></code>.
For example, <code class="docutils literal"><span class="pre">&gt;saci2813.1|Acinetobacter</span> <span class="pre">sp.</span> <span class="pre">ADP1|Aar</span></code>.
The ID is saci403.1; the strain of this sRNA is Acinetobacter sp. ADP1 and the name of sRNA is Aar.
If the user doesn&#8217;t follow the format, it will occur error when the user runs with <code class="docutils literal"><span class="pre">--sRNA_blast_stat,</span> <span class="pre">-sb</span></code>.
Or the results will be meaningless.
Of course, it can run <code class="docutils literal"><span class="pre">srna</span></code> without <code class="docutils literal"><span class="pre">--sRNA_blast_stat,</span> <span class="pre">-sb</span></code>.</p>
<p>nr database: It could be used for searching the known proteins in order to exclude false positive.</p>
<p>Gff files of sORFs: It could compare sORFs and sRNAs. User can refer these information to find the best sRNA candidates.</p>
<p>Gff files of terminator: It could compare terminators and sRNAs. User can refer these information
to find the best sRNA candidates.</p>
<p>Table of promoters: It could compare promoters and sRNAs. User can refer these information
to find the best sRNA candidates.
The format should be</p>
<table border="1" class="docutils">
<colgroup>
<col width="28%" />
<col width="30%" />
<col width="25%" />
<col width="18%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>strain</td>
<td>TSS_position</td>
<td>TSS_strand</td>
<td>Motif</td>
</tr>
<tr class="row-even"><td>NC_000915.1</td>
<td>237118</td>
<td>-</td>
<td>MOTIF_1</td>
</tr>
<tr class="row-odd"><td>NC_000915.1</td>
<td>729009</td>
<td>-</td>
<td>MOTIF_1</td>
</tr>
</tbody>
</table>
<p>First row is the header of table, the last column is the name of motif/promoter.
If you want apply ANNOgesci to compute the promoters, the table will be generated automatically.
Please refer to the section of <code class="docutils literal"><span class="pre">promoter</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic srna [-h] [--Vienna_folder VIENNA_FOLDER]
                      [--Vienna_utils VIENNA_UTILS]
                      [--blast_plus_folder BLAST_PLUS_FOLDER]
                      [--ps2pdf14_path PS2PDF14_PATH] [--UTR_derived_sRNA]
                      [--import_info IMPORT_INFO]
                      [--transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER]
                      [--annotation_folder ANNOTATION_FOLDER]
                      [--TSS_folder TSS_FOLDER]
                      [--processing_site_folder PROCESSING_SITE_FOLDER]
                      [--promoter_table PROMOTER_TABLE]
                      [--promoter_name PROMOTER_NAME] [--TSS_source]
                      [--TSS_intergenic_fuzzy TSS_INTERGENIC_FUZZY]
                      [--TSS_5UTR_fuzzy TSS_5UTR_FUZZY]
                      [--TSS_3UTR_fuzzy TSS_3UTR_FUZZY]
                      [--TSS_interCDS_fuzzy TSS_INTERCDS_FUZZY]
                      [--terminator_folder TERMINATOR_FOLDER]
                      [--terminator_fuzzy_in_CDS TERMINATOR_FUZZY_IN_CDS]
                      [--terminator_fuzzy_out_CDS TERMINATOR_FUZZY_OUT_CDS]
                      [--min_length MIN_LENGTH] [--max_length MAX_LENGTH]
                      [--tex_wig_folder TEX_WIG_FOLDER]
                      [--frag_wig_folder FRAG_WIG_FOLDER]
                      [--run_intergenic_TEX_coverage RUN_INTERGENIC_TEX_COVERAGE]
                      [--run_intergenic_noTEX_coverage RUN_INTERGENIC_NOTEX_COVERAGE]
                      [--run_intergenic_fragmented_coverage RUN_INTERGENIC_FRAGMENTED_COVERAGE]
                      [--run_antisense_TEX_coverage RUN_ANTISENSE_TEX_COVERAGE]
                      [--run_antisense_noTEX_coverage RUN_ANTISENSE_NOTEX_COVERAGE]
                      [--run_antisense_fragmented_coverage RUN_ANTISENSE_FRAGMENTED_COVERAGE]
                      [--intergenic_tolerance INTERGENIC_TOLERANCE]
                      [--run_utr_TEX_coverage RUN_UTR_TEX_COVERAGE]
                      [--run_utr_noTEX_coverage RUN_UTR_NOTEX_COVERAGE]
                      [--run_utr_fragmented_coverage RUN_UTR_FRAGMENTED_COVERAGE]
                      [--min_utr_coverage MIN_UTR_COVERAGE]
                      [--fasta_folder FASTA_FOLDER]
                      [--cutoff_energy CUTOFF_ENERGY] [--mountain_plot]
                      [--nr_format] [--srna_format]
                      [--sRNA_database_path SRNA_DATABASE_PATH]
                      [--nr_database_path NR_DATABASE_PATH]
                      [--tex_notex_libs TEX_NOTEX_LIBS]
                      [--frag_libs FRAG_LIBS] [--tex_notex TEX_NOTEX]
                      [--replicates_tex REPLICATES_TEX]
                      [--replicates_frag REPLICATES_FRAG] [--table_best]
                      [--decrease_intergenic DECREASE_INTERGENIC]
                      [--decrease_utr DECREASE_UTR]
                      [--fuzzy_intergenic FUZZY_INTERGENIC]
                      [--fuzzy_utr FUZZY_UTR] [--cutoff_nr_hit CUTOFF_NR_HIT]
                      [--blast_e_nr BLAST_E_NR] [--blast_e_srna BLAST_E_SRNA]
                      [--sORF SORF] [--best_with_all_sRNAhit]
                      [--best_without_sORF_candidate] [--best_with_terminator]
                      [--best_with_promoter] [--detect_sRNA_in_CDS]
                      [--overlap_percent_CDS OVERLAP_PERCENT_CDS]
                      [--ignore_hypothetical_protein]
                      [--ranking_time_promoter RANKING_TIME_PROMOTER]
                      [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --Vienna_folder VIENNA_FOLDER
                        Please assign the folder of Vienna package. It should
                        include RNAfold.
  --Vienna_utils VIENNA_UTILS
                        Please assign the folder of Utils of Vienna package.
                        It should include relplot.pl and mountain.pl.
  --blast_plus_folder BLAST_PLUS_FOLDER
                        Please assign the folder of blast+ which include
                        blastn, blastx, makeblastdb.
  --ps2pdf14_path PS2PDF14_PATH
                        Please assign the path of ps2pdf14.
  --UTR_derived_sRNA, -u
                        If you want to detect UTR-derived sRNA, please turn it
                        on. Default is False.
  --import_info IMPORT_INFO, -d IMPORT_INFO
                        There are several types of information that you can
                        import to detect and filter sRNAs: tss(the sRNA should
                        start from a TSS), sec_str(free energy change of
                        secondary structure(normalized by length)),
                        blast_nr(blast to nr), blast_srna(blast to sRNA),
                        sorf(compare with sORF), term(compare with
                        terminator), promoter(compare with promoter motif).
                        ATTENTION: without filters, the results may be not
                        well. Please assign the information you want to import
                        (comma for separate the filters), i.e.
                        tss,sec_str,blast_nr - means it used TSS, energy and
                        blast result to detect sRNA. Besides these
                        information, it will also consider the sequence length
                        of sRNA. ATTENTION: if you want to import sRNA
                        database, please follow the format that we define
                        $ID|$STRAIN|$SRNANAME. Default is
                        tss,sec_str,blast_nr,blast_srna.
  --transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER, -a TRANSCRIPT_ASSEMBLY_FOLDER
                        The path of transcriptome assembly folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        If you want to import TSS information, please assign
                        the path of gff folder of TSS. If you want to detect
                        UTR-derived sRNA, you MUST assign the folder of TSS.
  --processing_site_folder PROCESSING_SITE_FOLDER, -p PROCESSING_SITE_FOLDER
                        If you want to import processing site information,
                        please assign the path of gff folder of processing
                        site.If you want to detect UTR-derived sRNA, you MUST
                        assign the folder of processing site.
  --promoter_table PROMOTER_TABLE, -pt PROMOTER_TABLE
                        If you want to import promoter information, please
                        assign the path of promoter table. The format of table
                        is $STRAIN $TSS_POSITION $TSS_STRAND $PROMOTER_NAME.
                        If you want to import promoter information, the
                        associated tss imformation is required.
  --promoter_name PROMOTER_NAME, -pn PROMOTER_NAME
                        If you want to import promoter information, please
                        assign the promoter name (the last column of promoter
                        table) which you want to compare. If you want to
                        import multiple promoters, please put comma between
                        the promoters. Default is None.
  --TSS_source, -ts     If your gff file of TSS is not generated by ANNOgesic,
                        please you turn it on. It will classify TSSs and the
                        proper format for sRNA prediction. Default is True.
  --TSS_intergenic_fuzzy TSS_INTERGENIC_FUZZY, -ft TSS_INTERGENIC_FUZZY
                        If you want to import TSS information, you need to
                        assign the fuzzy for comparing TSS and transcript
                        assembly/CDS. It is for intergenic.Default is 3.
  --TSS_5UTR_fuzzy TSS_5UTR_FUZZY, -f5 TSS_5UTR_FUZZY
                        If you want to import TSS information, you need to
                        assign the fuzzy for comparing TSS and transcript
                        assembly. It is for 5&#39;UTR of UTR derived sRNA.You can
                        use percentage or the amount of reads. p_0.05 means
                        the fuzzy is 5 percent of the length of CDS which
                        overlap with transcript. n_10 means the fuzzy is 10
                        base pair. Default is n_3.
  --TSS_3UTR_fuzzy TSS_3UTR_FUZZY, -f3 TSS_3UTR_FUZZY
                        If you want to import TSS information, you need to
                        assign the fuzzy for comparing TSS and transcript
                        assembly/CDS. It is for 3&#39;UTR of UTR derived sRNA.You
                        can use percentage or the amount of reads. p_0.05
                        means the fuzzy is 5 percent of the length of CDS
                        which overlap with transcript. n_10 means the fuzzy is
                        10 base pair. Default is p_0.04.
  --TSS_interCDS_fuzzy TSS_INTERCDS_FUZZY, -fc TSS_INTERCDS_FUZZY
                        If you want to import TSS information, you need to
                        assign the fuzzy for comparing TSS and transcript
                        assembly. It is for interCDS derived sRNA.You can use
                        percentage or the amount of reads. p_0.05 means the
                        fuzzy is 5 percent of the length of CDS which overlap
                        with transcript. n_10 means the fuzzy is 10 base pair.
                        Default is p_0.04.
  --terminator_folder TERMINATOR_FOLDER, -tf TERMINATOR_FOLDER
                        If you want to import terminator information, please
                        assign the path of gff folder of terminator.
  --terminator_fuzzy_in_CDS TERMINATOR_FUZZY_IN_CDS, -tfi TERMINATOR_FUZZY_IN_CDS
                        If you want to import terminator information, you need
                        to assign the fuzzy for comparing terminator and
                        transcript assembly. It is the fuzzy for the
                        terminator which is within CDS. Default is 30.
  --terminator_fuzzy_out_CDS TERMINATOR_FUZZY_OUT_CDS, -tfo TERMINATOR_FUZZY_OUT_CDS
                        If you want to import terminator information, you need
                        to assign the fuzzy for comparing terminator and
                        transcript assembly. It is the fuzzy for the
                        terminator which is out of CDS. Default is 30.
  --min_length MIN_LENGTH, -lm MIN_LENGTH
                        Please assign the minimum length of sRNA. Default
                        is 30.
  --max_length MAX_LENGTH, -lM MAX_LENGTH
                        Please assign the maximum length of sRNA. Default
                        is 500.
  --tex_wig_folder TEX_WIG_FOLDER, -tw TEX_WIG_FOLDER
                        The path of tex+/- wig folder.
  --frag_wig_folder FRAG_WIG_FOLDER, -fw FRAG_WIG_FOLDER
                        The path of fragment wig folder.
  --run_intergenic_TEX_coverage RUN_INTERGENIC_TEX_COVERAGE, -it RUN_INTERGENIC_TEX_COVERAGE
                        The minimum average coverage of intergenic sRNA
                        candidates for TEX +. You can choose what kinds of TSS
                        that you want to use for sRNA prediction and assign
                        the cutoff of coverage to TSS. The order of numbers is
                        &quot;Primary Secondary Internal Antisense Orphan&quot;
                        (separated by comma). Ex: if you assign 0,0,0,50,10,
                        it means you want to use Antisense(cutoff coverage is
                        50) and Orphan(cutoff coverage is 10) for sRNA
                        prediction. 0 means you don&#39;t want to use it for
                        prediction. If you don&#39;t use TSS information for
                        prediction, it will choose the lowest one as a cutoff.
                        Ex: if you assign 0,0,0,50,10 and you have no TSS
                        information, it will use 10 as a general cutoff for
                        prediction. Default is 0,0,0,40,20.
  --run_intergenic_noTEX_coverage RUN_INTERGENIC_NOTEX_COVERAGE, -in RUN_INTERGENIC_NOTEX_COVERAGE
                        The minimum average coverage of intergenic sRNA
                        candidates for TEX -. You can choose what kinds of TSS
                        that you want to use for sRNA prediction and assign
                        the cutoff of coverage to TSS. The order of numbers is
                        &quot;Primary Secondary Internal Antisense Orphan&quot;
                        (separated by comma). Ex: if you assign 0,0,0,50,10,
                        it means you want to use Antisense(cutoff coverage is
                        50) and Orphan(cutoff coverage is 10) for sRNA
                        prediction. 0 means you don&#39;t want to use it for
                        prediction. If you don&#39;t use TSS information for
                        prediction, it will choose the lowest one as a cutoff.
                        Ex: if you assign 0,0,0,50,10 and you have no TSS
                        information, it will use 10 as a general cutoff for
                        prediction. Default is 0,0,0,30,10.
  --run_intergenic_fragmented_coverage RUN_INTERGENIC_FRAGMENTED_COVERAGE, -if RUN_INTERGENIC_FRAGMENTED_COVERAGE
                        The minimum average coverage of intergenic sRNA
                        candidates for fragmented library. You can choose what
                        kinds of TSS that you want to use for sRNA prediction
                        and assign the cutoff of coverage to TSS. The order of
                        numbers is &quot;Primary Secondary Internal Antisense
                        Orphan&quot; (separated by comma). Ex: if you assign
                        0,0,0,50,10, it means you want to use Antisense(cutoff
                        coverage is 50) and Orphan(cutoff coverage is 10) for
                        sRNA prediction. 0 means you don&#39;t want to use it for
                        prediction. If you don&#39;t use TSS information for
                        prediction, it will choose the lowest one as a cutoff.
                        Ex: if you assign 0,0,0,50,10 and you have no TSS
                        information, it will use 10 as a general cutoff for
                        prediction. Default is 400,200,0,50,20.
  --run_antisense_TEX_coverage RUN_ANTISENSE_TEX_COVERAGE, -at RUN_ANTISENSE_TEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_TEX_coverage. Just apply to
                        antisense. Default is 0,0,0,40,20.
  --run_antisense_noTEX_coverage RUN_ANTISENSE_NOTEX_COVERAGE, -an RUN_ANTISENSE_NOTEX_COVERAGE
                        The meaning is the same as
                        --run_intergenic_noTEX_coverage. Just apply to
                        antisense. Default is 0,0,0,30,10.
  --run_antisense_fragmented_coverage RUN_ANTISENSE_FRAGMENTED_COVERAGE, -af RUN_ANTISENSE_FRAGMENTED_COVERAGE
                        The meaning is the same as
                        --run_intergenic_fragmented_coverage. Just apply to
                        antisense. Default is 400,200,0,50,20.
  --intergenic_tolerance INTERGENIC_TOLERANCE, -ti INTERGENIC_TOLERANCE
                        This number indicates the tolerance of temporary drop
                        below cutoff of coverage. Default is 5.
  --run_utr_TEX_coverage RUN_UTR_TEX_COVERAGE, -ut RUN_UTR_TEX_COVERAGE
                        The minimum average coverage of UTR-derived sRNA
                        candidates for TEX +. You can assign the percentile or
                        real number of coverage for 5&#39;UTR, 3&#39;UTR and interCDS.
                        The order of numbers are &quot;5&#39;UTR, 3&#39;UTR and interCDS&quot;
                        (separated by comma). Ex: if you assign
                        &quot;p_0.7,p_0.5,p_0.5&quot;, it will use 70 percentile of
                        coverage as cutoff for 5&#39;UTR, median of coverage as
                        cutoff for 3&#39;UTR and interCDS. Ex: if you assign
                        &quot;n_30,n_10,n_20 &quot; it will use 30 as cutoff for 5&#39;UTR
                        and 10 as cutoff for 3&#39;UTR and 20 for interCDS.
                        Default is p_0.8,p_0.6,p_0.7.
  --run_utr_noTEX_coverage RUN_UTR_NOTEX_COVERAGE, -un RUN_UTR_NOTEX_COVERAGE
                        The minimum average coverage of UTR-derived sRNA
                        candidates for TEX -. You can assign the percentile or
                        real number of coverage for 5&#39;UTR, 3&#39;UTR and interCDS.
                        The order of numbers are &quot;5&#39;UTR, 3&#39;UTR and interCDS&quot;
                        (separated by comma). Ex: if you assign
                        &quot;p_0.7,p_0.5,p_0.5&quot;, it will use 70 percentile of
                        coverage as cutoff for 5&#39;UTR, median of coverage as
                        cutoff for 3&#39;UTR and interCDS. Ex: if you assign
                        &quot;n_30,n_10,n_20 &quot; it will use 30 as cutoff for 5&#39;UTR
                        and 10 as cutoff for 3&#39;UTR and 20 for interCDS.
                        Default is p_0.7,p_0.5,p_0.6.
  --run_utr_fragmented_coverage RUN_UTR_FRAGMENTED_COVERAGE, -uf RUN_UTR_FRAGMENTED_COVERAGE
                        The minimum average coverage of UTR-derived sRNA
                        candidates for fragmented library. You can assign the
                        percentile or real number of coverage for 5&#39;UTR, 3&#39;UTR
                        and interCDS. The order of numbers are &quot;5&#39;UTR, 3&#39;UTR
                        and interCDS&quot; (separated by comma). Ex: if you assign
                        &quot;p_0.7,p_0.5,p_0.5&quot;, it will use 70 percentile of
                        coverage as cutoff for 5&#39;UTR, median of coverage as
                        cutoff for 3&#39;UTR and interCDS. Ex: if you assign
                        &quot;n_30,n_10,n_20 &quot; it will use 30 as cutoff for 5&#39;UTR
                        and 10 as cutoff for 3&#39;UTR and 20 for interCDS.
                        Default is p_0.7,p_0.5,p_0.6.
  --min_utr_coverage MIN_UTR_COVERAGE, -mu MIN_UTR_COVERAGE
                        The minimum coverage of UTR-derived sRNA. The coverage
                        should not only fit the --run_utr_TEX_coverage,
                        --run_utr_noTEX_coverage and
                        --run_utr_fragmented_coverage, but also this value.
                        Defaul is 50.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        If you want to import sec_str, blast_nr, blast_srna
                        (--import_info), please assign the path of genome
                        fasta folder.
  --cutoff_energy CUTOFF_ENERGY, -e CUTOFF_ENERGY
                        If you want to import secondary structure information,
                        please assign the cutoff of folding energy change
                        (normalized by length of gene). Default is -0.05.
  --mountain_plot, -m   If you want to generate mountain plots of sRNA
                        candidates, please turn it on. Default is False.
  --nr_format, -nf      It is for formating nr database. If you already format
                        nr database, you don&#39;t need to turn it on. Default is
                        False.
  --srna_format, -sf    It is for formating sRNA database. If you already
                        format sRNA database, you don&#39;t need to turn it on.
                        Default is False.
  --sRNA_database_path SRNA_DATABASE_PATH, -sd SRNA_DATABASE_PATH
                        If you want to import blast results of sRNA, please
                        assign the path of sRNA database.
  --nr_database_path NR_DATABASE_PATH, -nd NR_DATABASE_PATH
                        If you want to import blast results of nr, please
                        assign the path of nr database.
  --tex_notex_libs TEX_NOTEX_LIBS, -tl TEX_NOTEX_LIBS
                        library name of tex and notex libraries. The format
                        is: wig_file_name:tex_treat_or_not(tex or notex):condi
                        tion_id(integer):replicate_id(alphabet):strand(+ or
                        -). If you have multiple wig files, please use comma
                        to separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --frag_libs FRAG_LIBS, -fl FRAG_LIBS
                        library name of fragmented libraries. The format is: w
                        ig_file_name:fragmented(frag):condition_id(integer):re
                        plicate_id(alphabet):strand(+ or -). If you have
                        multiple wig files, please use comma to separate the
                        wig files. For example,
                        wig1:frag:1:a:+,wig2:frag:1:a:-.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        For tex +/- libraries, sRNA candidates should be
                        detected by both or just one.(1/2) Default is 2.
  --replicates_tex REPLICATES_TEX, -rt REPLICATES_TEX
                        The sRNA of tex +/- libraries should be detected by
                        more than this number of replicates.
  --replicates_frag REPLICATES_FRAG, -rf REPLICATES_FRAG
                        The sRNA of fragmented libraries should be detected by
                        more than this number of replicates.
  --table_best, -tb     The output table of sRNA candidates only print the
                        best track of libraries. Default is False.
  --decrease_intergenic DECREASE_INTERGENIC, -di DECREASE_INTERGENIC
                        If the intergenic region is longer than the
                        max_length, it will based on coverage to check the
                        sRNA candidates. If (lowest coverage / the highest
                        coverage) of intergenic region is smaller than this
                        number, it will consider the the spot of lowest
                        coverage as end point. If the length of from start to
                        the end point is proper for sRNA candidate, it also
                        consider it as a sRNA candidate. Default is 0.1.
  --decrease_utr DECREASE_UTR, -du DECREASE_UTR
                        It is similar with --decrease_intergenic. It is for
                        UTR-derived sRNAs. Default is 0.05.
  --fuzzy_intergenic FUZZY_INTERGENIC, -fi FUZZY_INTERGENIC
                        If the situation is like --decrease_intergenic
                        mentioned, it is fuzzy value between the end of sRNA.
                        Default is 10.
  --fuzzy_utr FUZZY_UTR, -fu FUZZY_UTR
                        It is simliar with --fuzzy_intergenic. It is for UTR-
                        derived sRNAs. Default is 10.
  --cutoff_nr_hit CUTOFF_NR_HIT, -cn CUTOFF_NR_HIT
                        The cutoff of number of hits in nr database. If the
                        number of nr hits more than this cutoff, it will be
                        excluded. Default is 0.
  --blast_e_nr BLAST_E_NR, -en BLAST_E_NR
                        The cutoff of blast e value for nr alignment. Default
                        is 0.0001.
  --blast_e_srna BLAST_E_SRNA, -es BLAST_E_SRNA
                        The cutoff of blast e value for sRNA alignment.
                        Default is 0.0001.
  --sORF SORF, -O SORF  If you want to compare sORF and sRNA, please assign
                        the path of sORF gff folder.
  --best_with_all_sRNAhit, -ba
                        When you want to include the sRNA candidates which can
                        find the homology from blast sRNA database without
                        considering other information(ex. TSS, blast in nr...)
                        to the best results. Please turn it on. Or it will
                        just select the best candidates based on all filter
                        conditions. Default is False.
  --best_without_sORF_candidate, -bs
                        If you want to generate the best sRNA candidates which
                        excluded the overlap with sORFs.Please turn it on. Or
                        it will select the best candidates without considering
                        the overlap with sORF. Default is False.
  --best_with_terminator, -bt
                        If you want to generate the best sRNA candidates which
                        must be associated with terminator. Please turn it on.
                        Or it will select the best candidates without
                        considering the terminator. If the sRNA candidate ends
                        with processing site, it will include to best results,
                        too. Default is False.
  --best_with_promoter, -bp
                        If you want to generate the best sRNA candidates which
                        is must be associated with promoter.Please turn it on.
                        Default is False.
  --detect_sRNA_in_CDS, -ds
                        If you assume there are some wrong annotation in your
                        genome annotation file and you want to search the
                        small transcript in CDS, you can turn it on. It may
                        find more sRNA candidates which overlap with CDS.
                        Default is False.
  --overlap_percent_CDS OVERLAP_PERCENT_CDS, -oc OVERLAP_PERCENT_CDS
                        If you want to execute --detect_sRNA_in_CDS, you can
                        assign the cutoff. If the ratio of overlap between CDS
                        and sRNA candidates is lower than the cutoff, It may
                        be sRNA candidates. Default is 0.5
  --ignore_hypothetical_protein, -ih
                        If you want to ignore hypothetical protein in genome
                        annotation file, you can turn it on. Default is False.
  --ranking_time_promoter RANKING_TIME_PROMOTER, -rp RANKING_TIME_PROMOTER
                        If you imported promoter information, you can also use
                        it to rank sRNA candidates. The ranking will base on
                        --ranking_time_promoter * average coverage. For
                        example, one candidates which average coverage is 10,
                        associated with promoter and --ranking_time_promoter
                        is 2, the score for ranking will be 20 (2*10). The
                        candidates which are not associated with promoters,
                        the --ranking_time_promoter is 1. Default is 2. This
                        number can not be smaller than 1.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sRNA</span></code>.</p>
<p><code class="docutils literal"><span class="pre">sRNA_2d_$STRAIN_NAME</span></code>: The secondary structure of all sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">sRNA_seq_$STRAIN_NAME</span></code>: The sequence of all sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">blast_result_and_misc</span></code>: The results of blast.</p>
<p><code class="docutils literal"><span class="pre">mountain_plot</span></code>: The mountain plots of sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">sec_structure</span></code>: The dot plots and secondary structure plots of sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files. <code class="docutils literal"><span class="pre">stat_$STRAIN_NAME_sRNA_blast.csv</span></code> is the results of analysis of blast sRNA databases.
<code class="docutils literal"><span class="pre">stat_sRNA_class_Staphylococcus_aureus_HG003.csv</span></code> is the results of classification of sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: sRNA tables for more details. It also includes the ranking of sRNA candidates.
<code class="docutils literal"><span class="pre">for</span> <span class="pre">class</span></code> is for different classes of sRNAs.
<code class="docutils literal"><span class="pre">best</span></code> is the best results of sRNAs after filtering. <code class="docutils literal"><span class="pre">all_candidates</span></code> is for all candidates without filtering.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of sRNAs. The meanings of <code class="docutils literal"><span class="pre">for</span> <span class="pre">class</span></code>, <code class="docutils literal"><span class="pre">best</span></code>, <code class="docutils literal"><span class="pre">all_candidates</span></code> are the same as <code class="docutils literal"><span class="pre">tables</span></code>.</p>
<p>The useful tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">sRNA_type</span></code>: The sRNA is from 5&#8217;UTR or 3&#8217;UTR or interCDS or intergenic or antisense or in_CDS.</p>
<p><code class="docutils literal"><span class="pre">with_TSS</span></code>: Which TSSs are related to this sRNA. NA means the sRNA is not related to any TSSs.</p>
<p><code class="docutils literal"><span class="pre">sORF</span></code>: Which sORFs overlap with this sRNA.</p>
<p><code class="docutils literal"><span class="pre">sRNA_hit</span></code>: The blast hit of sRNA database.</p>
<p><code class="docutils literal"><span class="pre">nr_hit</span></code>: The blast hit of nr database.</p>
<p><code class="docutils literal"><span class="pre">2d_energy</span></code>: The normalized(by the length of sRNA) free energy change of secondary structure of sRNA candidate.</p>
<p><code class="docutils literal"><span class="pre">with_term</span></code>: The terminators which associated with the sRNA candidate.</p>
<p>If you assigned <code class="docutils literal"><span class="pre">--TSS_source</span></code> for sRNA prediction, <code class="docutils literal"><span class="pre">TSS_class</span></code> will be generated and store the gff files of TSSs.</p>
<p><code class="docutils literal"><span class="pre">promoter</span></code>: The promoters which associated with this sRNA candidates.</p>
<p><code class="docutils literal"><span class="pre">overlap_cds</span></code>: The CDSs which overlap with sRNA.</p>
<p><code class="docutils literal"><span class="pre">overlap_percent</span></code>: If there are CDSs overlap with sRNA, it will shows the percentage of overlap.</p>
</div>
<div class="section" id="sorf">
<h2>sorf<a class="headerlink" href="#sorf" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">sorf</span></code> can detect sORF based on ribosomal binding sites, start codons and stop codons within the intergenic transcripts.
Because non-coding region may be sRNAs or sORFs, it also provide the function to compare sORFs and sRNAs.
If there are some sORFs overlaped, it will merge them together. Therefore, one region may contain more than one sORF.
BE CAREFUL, The position of start codon is the first nucleotide. The position of stop codon is the last nucleotide.
As we know that three nucleotides form one amino acid. Therefore, one small transcript may contain
three open reading frames for one strand. Since <code class="docutils literal"><span class="pre">sorf</span></code> provides the region which covers all possible sORFs,
the start point and end point may not be the multiple of 3. The region may contain different frame shift.
Ex: (200, 202, 203) are the positions of three start codons and (241, 243) are two stop codons in
a small transcript. We can see that there are three possible ORFs(200-241, 203-241 and 202-243).
for 200-241 and 203-241, the frame has no shift, because the difference of positions for these two
ORFs are the multiple of 3. However, for 202-243, there is a frame shift. Please be aware this point for
using the results.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p>The gff files of CDSs/tRNAs/rRNAs and transcripts.</p>
<p>The libraries and wiggle files, Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>The fasta files of genome sequence for detection of ribosomal binding sites, start codons and stop codons.</p>
<p>User can also import some useful information to improve the prediction:</p>
<p>gff files of TSSs for checking the sORFs start from TSS or not.</p>
<p>gff files of sRNAs for checking the conflict of sRNAs and sORFs.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic sorf [-h] [--UTR_derived_sORF]
                      [--transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER]
                      [--annotation_folder ANNOTATION_FOLDER]
                      [--TSS_folder TSS_FOLDER] [--utr_length UTR_LENGTH]
                      [--min_length MIN_LENGTH] [--max_length MAX_LENGTH]
                      [--tex_wig_folder TEX_WIG_FOLDER]
                      [--frag_wig_folder FRAG_WIG_FOLDER]
                      [--cutoff_intergenic_coverage CUTOFF_INTERGENIC_COVERAGE]
                      [--cutoff_antisense_coverage CUTOFF_ANTISENSE_COVERAGE]
                      [--cutoff_5utr_coverage CUTOFF_5UTR_COVERAGE]
                      [--cutoff_3utr_coverage CUTOFF_3UTR_COVERAGE]
                      [--cutoff_interCDS_coverage CUTOFF_INTERCDS_COVERAGE]
                      [--cutoff_background CUTOFF_BACKGROUND]
                      [--fasta_folder FASTA_FOLDER]
                      [--tex_notex_libs TEX_NOTEX_LIBS]
                      [--frag_libs FRAG_LIBS] [--tex_notex TEX_NOTEX]
                      [--replicates_tex REPLICATES_TEX]
                      [--replicates_frag REPLICATES_FRAG] [--table_best]
                      [--sRNA_folder SRNA_FOLDER] [--start_codon START_CODON]
                      [--stop_codon STOP_CODON]
                      [--min_rbs_distance MIN_RBS_DISTANCE]
                      [--max_rbs_distance MAX_RBS_DISTANCE]
                      [--rbs_not_after_TSS] [--fuzzy_rbs FUZZY_RBS]
                      [--print_all_combination] [--best_no_sRNA]
                      [--best_no_TSS]
                      [--ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN]
                      [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --UTR_derived_sORF, -u
                        If you want to detect UTR-derived sORF, please turn it
                        on. Default is False.
  --transcript_assembly_folder TRANSCRIPT_ASSEMBLY_FOLDER, -a TRANSCRIPT_ASSEMBLY_FOLDER
                        The path of transcriptome assembly folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        If you want to import TSS information, please assign
                        the path of gff folder of TSS.
  --utr_length UTR_LENGTH, -ul UTR_LENGTH
                        If you want to import TSS information, please assign
                        the utr length for comparing TSS and sORF. The default
                        number is 300.
  --min_length MIN_LENGTH, -lm MIN_LENGTH
                        Please assign the minimum length of sORF. It will
                        predict sORF candidates based on the value. Default is
                        30.
  --max_length MAX_LENGTH, -lM MAX_LENGTH
                        Please assign the maximum length of sORF. It will
                        predict sORF candidates based on the value. Default is
                        150.
  --tex_wig_folder TEX_WIG_FOLDER, -tw TEX_WIG_FOLDER
                        The path of tex+/- wig folder.
  --frag_wig_folder FRAG_WIG_FOLDER, -fw FRAG_WIG_FOLDER
                        The path of fragment wig folder.
  --cutoff_intergenic_coverage CUTOFF_INTERGENIC_COVERAGE, -ci CUTOFF_INTERGENIC_COVERAGE
                        The cutoff of minimum coverage of intergenic sORF
                        candidates.
  --cutoff_antisense_coverage CUTOFF_ANTISENSE_COVERAGE, -ai CUTOFF_ANTISENSE_COVERAGE
                        The cutoff of minimum coverage of antisense sORF
                        candidates.
  --cutoff_5utr_coverage CUTOFF_5UTR_COVERAGE, -cu5 CUTOFF_5UTR_COVERAGE
                        The cutoff of minimum coverage of 5&#39;UTR derived sORF
                        candidates. You can use percentage or the amount of
                        reads. p_0.05 means the coverage of sORF candidates
                        should higher than 5 percentile of all 5&#39;UTR
                        transcripts. n_10 means the coverage of sORF
                        candidates should be 10. Default is p_0.5.
  --cutoff_3utr_coverage CUTOFF_3UTR_COVERAGE, -cu3 CUTOFF_3UTR_COVERAGE
                        The cutoff of minimum coverage of 3&#39;UTR derived sORF
                        candidates. You can use percentage or the amount of
                        reads. p_0.05 means the coverage of sORF candidates
                        should higher than 5 percentile of all 3&#39;UTR
                        transcripts. n_10 means the coverage of sORF
                        candidates should be 10. Default is p_0.5.
  --cutoff_interCDS_coverage CUTOFF_INTERCDS_COVERAGE, -cuf CUTOFF_INTERCDS_COVERAGE
                        The cutoff of minimum coverage of interCDS derived
                        sORF candidates. You can use percentage or the amount
                        of reads. p_0.05 means the coverage of sORF candidates
                        should higher than 5 percentile of all interCDS
                        transcripts. n_10 means the coverage of sORF
                        candidates should be 10. Default is p_0.5.
  --cutoff_background CUTOFF_BACKGROUND, -cub CUTOFF_BACKGROUND
                        The cutoff of minimum coverage of all sORF candidates.
                        Default is 10.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        The folder of genome fasta file.
  --tex_notex_libs TEX_NOTEX_LIBS, -tl TEX_NOTEX_LIBS
                        Library name of tex+/- library. The format is:
                        wig_file_name:tex_treat_or_not(tex or notex):condition
                        _id(integer):replicate_id(alphabet):strand(+ or -). If
                        you have multiple wig files, please use comma to
                        separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --frag_libs FRAG_LIBS, -fl FRAG_LIBS
                        Library name of fragmented library The format is: wig_
                        file_name:fragmented(frag):condition_id(integer):repli
                        cate_id(alphabet):strand(+ or -). If you have multiple
                        wig files, please use comma to separate the wig files.
                        For example, wig1:frag:1:a:+,wig2:frag:1:a:-.
  --tex_notex TEX_NOTEX, -te TEX_NOTEX
                        For tex +/- library, sORF candidates should be
                        detected by both or just one.(1/2) Default is 2.
  --replicates_tex REPLICATES_TEX, -rt REPLICATES_TEX
                        The sORF of tex +/- library should be detected by more
                        than this number of replicates.
  --replicates_frag REPLICATES_FRAG, -rf REPLICATES_FRAG
                        The sORF of fragmented library should be detected by
                        more than this number of replicates.
  --table_best, -tb     The output table of sORF candidates only print the
                        best track. Default is False.
  --sRNA_folder SRNA_FOLDER, -s SRNA_FOLDER
                        If you want to compare sORF and sRNA, please assign
                        the path of sORF gff folder.
  --start_codon START_CODON, -ac START_CODON
                        What kinds of start coden you want to use. If you want
                        to assign more than one type of start codon, please
                        use comma to separate them. Default is ATG.
  --stop_codon STOP_CODON, -oc STOP_CODON
                        What kinds of stop coden you want to use. If you want
                        to assign more than one type of stop codon, please use
                        comma to separate them. Default is TTA,TAG,TGA.
  --min_rbs_distance MIN_RBS_DISTANCE, -mr MIN_RBS_DISTANCE
                        The minimum distance between the ribosomal binding
                        site and start codon. Default is 3.
  --max_rbs_distance MAX_RBS_DISTANCE, -Mr MAX_RBS_DISTANCE
                        The maximum distance between the ribosomal binding
                        site and start codon. Default is 15.
  --rbs_not_after_TSS, -at
                        If you want to generate best gff file which include
                        ribosomal binding site not after TSS, please turn it
                        on. Default is False.
  --fuzzy_rbs FUZZY_RBS, -zr FUZZY_RBS
                        How many nucleotides of ribosomal binding site is
                        different with AGGAGG? Default is 2.
  --print_all_combination, -pa
                        Every expressed transcript of sORF has many start
                        codons and stop codons. If you want to print all
                        combinations of start codons and stop codons, please
                        turn it on. Default is False.
  --best_no_sRNA, -bs   If you want to generate best gff file without
                        overlaping with sRNA, please turn it on. Default is
                        False.
  --best_no_TSS, -bt    If you want to generate best gff file which not refere
                        to TSS, please turn it on. Default is False.
  --ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN, -ih IGNORE_HYPOTHETICAL_PROTEIN
                        If you want to ignore hypothetical protein in genome
                        annotation file, you can turn it on. Default is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sORF</span></code>.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of sORFs for more details. <code class="docutils literal"><span class="pre">all_candidates</span></code> is for all sORF candidates without filtering.
<code class="docutils literal"><span class="pre">best</span></code> is for the best sORF candidates with filtering.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of sORFs. The meanings of <code class="docutils literal"><span class="pre">all_candidates</span></code> and <code class="docutils literal"><span class="pre">best</span></code> are the same as <code class="docutils literal"><span class="pre">tables</span></code>.</p>
<p>The tags of gff files:</p>
<p><code class="docutils literal"><span class="pre">start_TSS</span></code>: The starting TSS of this sORF.</p>
<p><code class="docutils literal"><span class="pre">with_TSS</span></code>: Which TSSs are associated with this sORFs.</p>
<p><code class="docutils literal"><span class="pre">sORF_type</span></code>: The type of the sORF (5&#8217;UTR, 3&#8217;UTR, interCDS, intergenic, antisense or in_CDS).</p>
<p><code class="docutils literal"><span class="pre">sRNA</span></code>: Which sRNAs overlap with the the region of sORFs.</p>
<p><code class="docutils literal"><span class="pre">RBS</span></code>: The ribosomal binding sites of the region of sORFs.</p>
<p><code class="docutils literal"><span class="pre">frame_shift</span></code>: How many frame shifts in the regions of sORFs.</p>
</div>
<div class="section" id="promoter">
<h2>promoter<a class="headerlink" href="#promoter" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">promoter</span></code> can scan the upstream of TSSs to discover the promoter motifs.
User can assign the regions of upstream TSSs. We integrate MEME to compute the promoters.
User can view the results very easily. If the gff files of TSSs is not computed by
ANNOgesic, please use &#8211;TSS_source. <code class="docutils literal"><span class="pre">promoter</span></code> will classify the TSSs for computing
promoter motifs.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://meme-suite.org/tools/meme">MEME</a>.</p>
<p>Fasta files of genome sequence.</p>
<p>Gff files of genome annotation.</p>
<p>Gff files of TSSs.</p>
<p>If the gff files of TSSs is not computed by ANNOgesic, the libraries and wiggle files are necessary.
Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code> in order to assign the correct format.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic promoter [-h] [--MEME_path MEME_PATH]
                          [--fasta_folder FASTA_FOLDER]
                          [--TSS_folder TSS_FOLDER] [--num_motif NUM_MOTIF]
                          [--nt_before_TSS NT_BEFORE_TSS] [--e_value E_VALUE]
                          [--motif_width MOTIF_WIDTH] [--TSS_source]
                          [--tex_libs TEX_LIBS] [--tex_wig_path TEX_WIG_PATH]
                          [--annotation_folder ANNOTATION_FOLDER]
                          [--combine_all]
                          [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --MEME_path MEME_PATH
                        path of MEME.
  --fasta_folder FASTA_FOLDER, -f FASTA_FOLDER
                        Please assign the folder of genome fasta file.
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        The folder of TSS gff file.
  --num_motif NUM_MOTIF, -n NUM_MOTIF
                        How many of motifs you want to produce? Default is 10.
  --nt_before_TSS NT_BEFORE_TSS, -b NT_BEFORE_TSS
                        How many nucleotides that you want to extract before
                        TSS for promoter prediction? Default is 50.
  --e_value E_VALUE, -e E_VALUE
                        The cutoff of e value. Default is 0.05.
  --motif_width MOTIF_WIDTH, -w MOTIF_WIDTH
                        Motif length - it will refer the value to find the
                        motif. If you want to detect a range of width, you can
                        insert &quot;-&quot; between two values. Moreover, If you want
                        to compute more than one motif lenght, please use
                        comma to separate them. for example, 50,2-10. It means
                        the range of width which you want to detect is 50 and
                        within 2 to 10. The number should be less or equal
                        than --nt_before_TSS. Default is 50.
  --TSS_source, -s      If you generate TSS from other method, please turn it
                        on. Default is True(from ANNOgesic)
  --tex_libs TEX_LIBS, -tl TEX_LIBS
                        Library name of tex+/- library. If your TSS is not
                        from ANNOgesic, please assign it.The format is:
                        wig_file_name:tex_treat_or_not(tex or notex):condition
                        _id(integer):replicate_id(alphabet):strand(+ or -). If
                        you have multiple wig files, please use comma to
                        separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --tex_wig_path TEX_WIG_PATH, -tw TEX_WIG_PATH
                        The path of tex+/- wig folder. If your TSS is not from
                        ANNOgesic, please assign the folder of tex+/- wig
                        folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder. If your TSS
                        is not from ANNOgesic, please assign the annotation
                        gff path.
  --combine_all, -c     If you want to combine all TSS files in &quot;TSS_folder&quot;
                        to generate global promoter motifs, please turn it on.
                        Default is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/promoter_analysis</span></code>.</p>
<p><code class="docutils literal"><span class="pre">allfasta</span></code>: If you turn <code class="docutils literal"><span class="pre">--combine_all</span></code> on, it will combine all TSSs in <code class="docutils literal"><span class="pre">--TSS_folder</span></code>
to generate promoter motifs. The results will store in this folder.</p>
<p><code class="docutils literal"><span class="pre">fasta_class</span></code>: The fasta files of the classes of TSSs.</p>
<p>Every strain will generate one folder for storing the information of promoter motifs.
The sub-folder is <code class="docutils literal"><span class="pre">promoter_motifs_$FILENAME_$STRAINNAME_$TSSTYPE_$PROMOTERLEGNTH</span></code>.
Each sub-folder stores PNG files and HTML files of promoters. It also contains a table(CSV file) of
prometer which can be used to sRNA detection (please refer to <code class="docutils literal"><span class="pre">srna</span></code>).</p>
<p>If the TSSs are not computed by ANNOgesic, <code class="docutils literal"><span class="pre">TSS_class</span></code> will be generated. It will classify the
TSSs and store as gff files in it.</p>
</div>
<div class="section" id="operon">
<h2>operon<a class="headerlink" href="#operon" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">operon</span></code> will group TSSs, genes/CDSs/tRNAs/rRNAs, transcripts, terminators and UTRs to operons and
suboperons.</p>
<ul class="simple">
<li>Pre-required tools or files</li>
</ul>
<p>Gff files of TSSs, annotations, transcripts, 5&#8217;UTRs, and 3&#8217;UTRs.</p>
<p>If user wants to import the information of terminators, <code class="docutils literal"><span class="pre">operon</span></code> can integrate terminators, too.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic operon [-h] [--TSS_folder TSS_FOLDER]
                        [--annotation_folder ANNOTATION_FOLDER]
                        [--transcript_folder TRANSCRIPT_FOLDER]
                        [--UTR5_folder UTR5_FOLDER]
                        [--UTR3_folder UTR3_FOLDER]
                        [--term_folder TERM_FOLDER] [--TSS_fuzzy TSS_FUZZY]
                        [--term_fuzzy TERM_FUZZY] [--min_length MIN_LENGTH]
                        [--statistics] [--combine_gff]
                        [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TSS_folder TSS_FOLDER, -t TSS_FOLDER
                        The path of TSS gff folder.
  --annotation_folder ANNOTATION_FOLDER, -g ANNOTATION_FOLDER
                        The path of genome annotation gff folder.
  --transcript_folder TRANSCRIPT_FOLDER, -a TRANSCRIPT_FOLDER
                        The path of transcript gff folder.
  --UTR5_folder UTR5_FOLDER, -u5 UTR5_FOLDER
                        The path of 5&#39;UTR gff folder.
  --UTR3_folder UTR3_FOLDER, -u3 UTR3_FOLDER
                        The path of 3&#39;UTR gff folder.
  --term_folder TERM_FOLDER, -e TERM_FOLDER
                        If you want to import the information of terminator,
                        please assign the path of terminator gff folder.
  --TSS_fuzzy TSS_FUZZY, -tf TSS_FUZZY
                        The fuzzy for comparing TSS and transcript assembly.
                        Default is 5.
  --term_fuzzy TERM_FUZZY, -ef TERM_FUZZY
                        The fuzzy for comparing terminator and transcript
                        assembly. Default is 30.
  --min_length MIN_LENGTH, -l MIN_LENGTH
                        The minimum length of operon. Default is 20.
  --statistics, -s      Doing statistics for operon analysis. Default is
                        False. The name of statistics file is -
                        stat_operon_$STRAIN_NAME.csv.
  --combine_gff, -c     Convert the operon and all features you assigned to
                        one gff file. Default is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/operon</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files which integrate the information of TSSs, annotations,
transcripts, 5&#8217;UTRs, and 3&#8217;UTRs.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of operons which store all information of operons and suboperons.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
</div>
<div class="section" id="circrna">
<h2>circrna<a class="headerlink" href="#circrna" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">circrna</span></code> can detect the potential circular RNAs. It applies <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>
to detect circular RNAs. Then it will checks genome annotation files and quality of splicing site detection to
exclude false positive. User can assign reads for mapping and detecting circular RNAs or assign alignment files to skip mapping.
But BE CAREFUL, If user uses alignment files, they should be mapped by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>
and with <code class="docutils literal"><span class="pre">-S</span></code>. Or <code class="docutils literal"><span class="pre">circrna</span></code> can&#8217;t find the proper candidates.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>.</p>
<p>Fasta files of reads or alignment files. If you want to use alignment files directly, they should be
mapped by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> and with <code class="docutils literal"><span class="pre">-S</span></code>.</p>
<p>Fasta files and gff files of genome.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic circrna [-h] [--segemehl_folder SEGEMEHL_FOLDER]
                         [--samtools_path SAMTOOLS_PATH] [--align]
                         [--tex_bam_path TEX_BAM_PATH]
                         [--fragmented_bam_path FRAGMENTED_BAM_PATH]
                         [--process PROCESS] [--fasta_path FASTA_PATH]
                         [--annotation_path ANNOTATION_PATH]
                         [--support_reads SUPPORT_READS]
                         [--start_ratio START_RATIO] [--end_ratio END_RATIO]
                         [--ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN]
                         [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --segemehl_folder SEGEMEHL_FOLDER, -sg SEGEMEHL_FOLDER
                        Please assign the folder of segemehl.
  --samtools_path SAMTOOLS_PATH, -st SAMTOOLS_PATH
                        Please assign the path of samtools.
  --align, -a           Using segemehl to map reads (included splice
                        detection). If you already usd segemehl with -S to map
                        your reads, you can skip this step, don&#39;t need to turn
                        it on. Please be attention, it only use default
                        parameters of segemehl to map your reads. Moreover, it
                        will map all read files in ANNOgesic/input/reads. If
                        you want to run some specific functions of segemehl,
                        please directly run segemehl by yourself. Default is
                        False.
  --tex_bam_path TEX_BAM_PATH, -tb TEX_BAM_PATH
                        If you already has Bam files, Please assign the TEX+/-
                        Bam path.
  --fragmented_bam_path FRAGMENTED_BAM_PATH, -fb FRAGMENTED_BAM_PATH
                        If you already has Bam files, Please assign the
                        fragmented Bam path.
  --process PROCESS, -p PROCESS
                        The number of parallels processes for --align. Default
                        is 10.
  --fasta_path FASTA_PATH, -f FASTA_PATH
                        The folder of genome fasta.
  --annotation_path ANNOTATION_PATH, -g ANNOTATION_PATH
                        The folder of genome annotation gff files.
  --support_reads SUPPORT_READS, -s SUPPORT_READS
                        The cut off of supported reads. Default is 10.
  --start_ratio START_RATIO, -sr START_RATIO
                        The ratio of (read support circ / all read) at
                        starting point. The ratio of candidates should higher
                        than this cutoff. Default is 0.5.
  --end_ratio END_RATIO, -er END_RATIO
                        The ratio of (read support circ / all read) at end
                        point. The ratio of candidates should higher than this
                        cutoff. Default is 0.5.
  --ignore_hypothetical_protein IGNORE_HYPOTHETICAL_PROTEIN, -ih IGNORE_HYPOTHETICAL_PROTEIN
                        If you want to ignore hypothetical protein in genome
                        annotation file, you can turn it on. Default is False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All the output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/circRNA</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: Gff files of circular RNAs. <code class="docutils literal"><span class="pre">$STRAINNAME_best.gff</span></code> is the gff files for best result after comparing
with genome annotation and quality of splicing. <code class="docutils literal"><span class="pre">$STRAINNAME_all.gff</span></code> is for all candidates without filering.</p>
<p><code class="docutils literal"><span class="pre">circRNA_tables</span></code>: The tables for circular RNAs with more details.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files.</p>
<p><code class="docutils literal"><span class="pre">segemehl_align</span></code>: If <code class="docutils literal"><span class="pre">circrna</span></code> starts from read mapping, the folder is for results of mapping.</p>
<p><code class="docutils literal"><span class="pre">segemehl_splice</span></code>: The results of splicing detection. The information of the splicing tables, please
refer to <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>.</p>
</div>
<div class="section" id="go-term">
<h2>go_term<a class="headerlink" href="#go-term" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">go_term</span></code> can compare the genome annotation files and Uniprot to retreive the information of Gene Ontology.
It also provides some analyses of Go terms.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.uniprot.org/downloads">idmapping_selected.tab from Uniprot</a>.</p>
<p><a class="reference external" href="http://geneontology.org/page/go-slim-and-subset-guide">goslim.obo</a>.</p>
<p><a class="reference external" href="http://geneontology.org/page/download-ontology">go.obo</a>.</p>
<p>Gff files of genome annotation.</p>
<p>If you want to compute the expressed CDSs only, you need to assign the transcript gff files.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic go_term [-h] [--annotation_path ANNOTATION_PATH]
                         [--transcript_path TRANSCRIPT_PATH]
                         [--UniProt_id UNIPROT_ID] [--go_obo GO_OBO]
                         [--goslim_obo GOSLIM_OBO]
                         [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --annotation_path ANNOTATION_PATH, -g ANNOTATION_PATH
                        The path of genome annotation gff folder.
  --transcript_path TRANSCRIPT_PATH, -a TRANSCRIPT_PATH
                        If you assign the path of transcript gff folder, it
                        can only extract the expressed CDS to retrieve GO
                        term.
  --UniProt_id UNIPROT_ID, -u UNIPROT_ID
                        The path of UniProt ID mapping database. Default is
                        ANNOgesic/input/database/idmapping_selected.tab.
  --go_obo GO_OBO, -go GO_OBO
                        The path of go.obo. Default is
                        ANNOgesic/input/database/go.obo.
  --goslim_obo GOSLIM_OBO, -gs GOSLIM_OBO
                        The path of goslim.obo. Default is
                        ANNOgesic/input/database/goslim_generic.obo.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">ANNOgesic/output/Go_term</span></code>.</p>
<p><code class="docutils literal"><span class="pre">Go_term_results</span></code>: The tables of Go terms information.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files and figures.</p>
</div>
<div class="section" id="srna-target">
<h2>srna_target<a class="headerlink" href="#srna-target" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">srna_target</span></code> will search the potential targets of sRNAs. User can assign the
programs for running (RNAup or RNAplex or both). We recommand user to run with both
programs. <code class="docutils literal"><span class="pre">srna_target</span></code> can compare the both results and provide the best ones.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">ViennaRNA</a> .</p>
<p>Gff files of genome annotation.</p>
<p>Gff files of sRNAs.</p>
<p>Fasta files of genome.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic srna_target [-h] [--Vienna_folder VIENNA_FOLDER]
                             [--annotation_path ANNOTATION_PATH]
                             [--fasta_path FASTA_PATH] [--sRNA_path SRNA_PATH]
                             [--query_sRNA QUERY_SRNA] [--program PROGRAM]
                             [--interaction_length INTERACTION_LENGTH]
                             [--window_size_target WINDOW_SIZE_TARGET]
                             [--span_target SPAN_TARGET]
                             [--window_size_srna WINDOW_SIZE_SRNA]
                             [--span_srna SPAN_SRNA]
                             [--unstructured_region_RNAplex_target UNSTRUCTURED_REGION_RNAPLEX_TARGET]
                             [--unstructured_region_RNAplex_srna UNSTRUCTURED_REGION_RNAPLEX_SRNA]
                             [--unstructured_region_RNAup UNSTRUCTURED_REGION_RNAUP]
                             [--energy_threshold ENERGY_THRESHOLD]
                             [--duplex_distance DUPLEX_DISTANCE] [--top TOP]
                             [--process_rnaplex PROCESS_RNAPLEX]
                             [--process_rnaup PROCESS_RNAUP]
                             [--continue_rnaup]
                             [--potential_target_start POTENTIAL_TARGET_START]
                             [--potential_target_end POTENTIAL_TARGET_END]
                             [--target_feature TARGET_FEATURE]
                             [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --Vienna_folder VIENNA_FOLDER
                        Please assign the folder of Vienna package. It should
                        include RNAplfold, RNAup and RNAplex.
  --annotation_path ANNOTATION_PATH, -g ANNOTATION_PATH
                        The path of genome annotation gff folder.
  --fasta_path FASTA_PATH, -f FASTA_PATH
                        The path of genome fasta folder.
  --sRNA_path SRNA_PATH, -r SRNA_PATH
                        The path of sRNA gff folder.
  --query_sRNA QUERY_SRNA, -q QUERY_SRNA
                        Please assign the query sRNA. If you want to compute
                        all sRNA in gff file, please keyin &#39;all&#39;.the input
                        format should be like, $STRAIN:$STRAND:$START:$END. If
                        you want to compute more than one sRNA, please use
                        comma to separate them.For example,
                        NC_007795.1:+:200:534,NC_007795.1:-:6767:6900. Default
                        is all.
  --program PROGRAM, -p PROGRAM
                        Using RNAplex, RNAup or both. Default is both.
  --interaction_length INTERACTION_LENGTH, -i INTERACTION_LENGTH
                        Maximum length of an interaction. Default is 30.
  --window_size_target WINDOW_SIZE_TARGET, -wt WINDOW_SIZE_TARGET
                        Only works when --program is RNAplex or both. Average
                        the pair probabilities over windows of given size for
                        RNAplex. Default is 240.
  --span_target SPAN_TARGET, -st SPAN_TARGET
                        Only works when --program is RNAplex or both. Set the
                        maximum allowed separation of a base pair to span for
                        RNAplex. Default is 160.
  --window_size_srna WINDOW_SIZE_SRNA, -ws WINDOW_SIZE_SRNA
                        Only works when --program is RNAplex or both. Average
                        the pair probabilities over windows of given size for
                        RNAplex. Default is 30.
  --span_srna SPAN_SRNA, -ss SPAN_SRNA
                        Only works when --program is RNAplex or both. Set the
                        maximum allowed separation of a base pair to span for
                        RNAplex. Default is 30.
  --unstructured_region_RNAplex_target UNSTRUCTURED_REGION_RNAPLEX_TARGET, -ut UNSTRUCTURED_REGION_RNAPLEX_TARGET
                        Only works when --program is RNAplex or both. Compute
                        the mean probability that regions of length 1 to a
                        given length are unpaired for RNAplex. Default is 30.
  --unstructured_region_RNAplex_srna UNSTRUCTURED_REGION_RNAPLEX_SRNA, -us UNSTRUCTURED_REGION_RNAPLEX_SRNA
                        Only works when --program is RNAplex or both. Compute
                        the mean probability that regions of length 1 to a
                        given length are unpaired for RNAplex. Default is 30.
  --unstructured_region_RNAup UNSTRUCTURED_REGION_RNAUP, -uu UNSTRUCTURED_REGION_RNAUP
                        Only works when --program is RNAup or both. Compute
                        the mean probability that regions of length 1 to a
                        given length are unpaired for RNAplex. Default is 30.
  --energy_threshold ENERGY_THRESHOLD, -e ENERGY_THRESHOLD
                        Only works when --program is RNAplex or both. Minimum
                        energy for a duplex to be returned for RNAplex.
                        Default is -8.
  --duplex_distance DUPLEX_DISTANCE, -d DUPLEX_DISTANCE
                        Only works when --program is RNAplex or both. Distance
                        between target 3&#39; ends of two consecutive duplexes for
                        RNAplex. Default is 20.
  --top TOP, -t TOP     The output file only includes --top. Default is 20.
  --process_rnaplex PROCESS_RNAPLEX, -pp PROCESS_RNAPLEX
                        The amount of parallel processes for running RNAplex
                        prediction. Default is 5.
  --process_rnaup PROCESS_RNAUP, -pu PROCESS_RNAUP
                        The amount of parallel processes for running RNAup
                        prediction. Default is 20.
  --continue_rnaup, -cr
                        RNAup will take a long time for running if you want to
                        compute a lot of sRNA. If the process crush, you can
                        turn it on. This flag will continue running RNAup
                        based on your previous running. Default is False.
  --potential_target_start POTENTIAL_TARGET_START, -ps POTENTIAL_TARGET_START
                        How many upstream nucleotides of the start point of
                        CDS that you want to extract as potential target.
                        Default is 200.
  --potential_target_end POTENTIAL_TARGET_END, -pe POTENTIAL_TARGET_END
                        How many downstream nucleotides of the start point of
                        CDS that you want to extract as potential target.
                        Default is 150.
  --target_feature TARGET_FEATURE, -tf TARGET_FEATURE
                        What features you want to extract as potential
                        targets. If you want to use multi-features, just use
                        comma to separate them. For example, CDS,gene. Default
                        is CDS.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/sRNA_targets</span></code>.</p>
<p><code class="docutils literal"><span class="pre">RNAplex</span></code>: All results of RNAplex. <code class="docutils literal"><span class="pre">$STRAIN_RNAplex.txt</span></code> is raw results of RNAplex.
It includes the information of binding situations. <code class="docutils literal"><span class="pre">$STRAIN_RNAplex_rank.csv</span></code> is the results
that sorted by binding energy.</p>
<p><code class="docutils literal"><span class="pre">RNAup</span></code>: All results of RNAup. <code class="docutils literal"><span class="pre">$STRAIN_RNAup.txt</span></code> is raw results of RNAup.
It includes the information of binding situations. <code class="docutils literal"><span class="pre">$STRAIN_RNAup_rank.csv</span></code> is the results
that sorted by binding energy.</p>
<p><code class="docutils literal"><span class="pre">merge</span></code>: The results which merge <code class="docutils literal"><span class="pre">RNAplex</span></code> and <code class="docutils literal"><span class="pre">RNAup</span></code>. <code class="docutils literal"><span class="pre">$STRAIN_merge.csv</span></code> merges all candidates of both programs.
<code class="docutils literal"><span class="pre">$STRAIN_overlap.csv</span></code> lists the results which be top 20(default) in both methods.</p>
<p><code class="docutils literal"><span class="pre">sRNA_seqs</span></code>: The fasta sequences of sRNAs.</p>
<p><code class="docutils literal"><span class="pre">target_seqs</span></code>: The fasta sequences of potential targets.</p>
</div>
<div class="section" id="ppi-network">
<h2>ppi_network<a class="headerlink" href="#ppi-network" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">ppi_network</span></code> will retrieve the data from <a class="reference external" href="http://string-db.org/">STRING</a>.
Then use <a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a> to search
the literatures to support the protein-protein interaction networks. Therefore,
<code class="docutils literal"><span class="pre">ppi_network</span></code> can generate the protein-protein interaction networks with supported literatures.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://string-db.org/cgi/download.pl">species.vXXXX.txt from STRING</a>.</p>
<p>Ptt files of genome annotation.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic ppi_network [-h] [--gff_path GFF_PATH]
                             [--proteinID_strains PROTEINID_STRAINS]
                             [--without_strain_pubmed]
                             [--species_STRING SPECIES_STRING] [--score SCORE]
                             [--node_size NODE_SIZE] [--query QUERY]
                             [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --gff_path GFF_PATH, -g GFF_PATH
                        The path of genome annotation gff folder. Please
                        confirm your gff files which have proper locus_tag
                        item in attributes. The locus_tag items can be real
                        locus tag like locus_tag=SAOUHSC_00003, or they can be
                        assigned by gene name, such as, locus_tag=dnaA.
  --proteinID_strains PROTEINID_STRAINS, -s PROTEINID_STRAINS
                        This is for query strain and filename. In order to
                        retrieve the data from STRING and Pubmed, you also
                        have to assign the similar reference. For example, the
                        query strain is Staphylococcus aureus HG003, but there
                        is no Staphylococcus aureus HG003 in STRING database.
                        However, there is Staphylococcus aureus 8325 which is
                        highly similar with Staphylococcus aureus HG003.
                        Therefore, we can use it as reference. you can assign
                        it like Staphylococcus_aureus_HG003.gff:Staphylococcus
                        _aureus_HG003:&quot;Staphylococcus aureus
                        8325&quot;:&quot;Staphylococcus aureus&quot;. or Staphylococcus_aureu
                        s_HG003.gff:Staphylococcus_aureus_HG003:&quot;93061&quot;:&quot;Staph
                        ylococcus aureus&quot;. or Staphylococcus_aureus_HG003.gff:
                        Staphylococcus_aureus_HG003:&quot;Staphylococcus aureus
                        NCTC 8325&quot;:&quot;Staphylococcus aureus&quot;. (gff_filename:gff_
                        strain_name:STRING_name:Pubmed_name). First one is the
                        gff file name. Second one is the strain name of gff
                        files. Third one is for STRING database, and the
                        fourth one is for Pubmed. Of course, you can run the
                        script for several strains at the same time, just put
                        comma between these strains. Before running it, please
                        check the species table which should located in
                        ANNOgesic/input/database .If you didn&#39;t download the
                        file, please download it. You can use taxon_id,
                        STRING_name_compact or official_name_NCBI to represent
                        STRING_name.BE CAREFUL, if the name which you assigned
                        has spaces, please put &quot;&quot; at two ends. For the name of
                        Pubmed, you can assign the name not so specific. If
                        you assign a specific name, it may not be able to find
                        the related literatures.
  --without_strain_pubmed, -n
                        If you want to retrieve pubmed without assign any
                        strains, please turn it on. Default is False.
  --species_STRING SPECIES_STRING, -d SPECIES_STRING
                        Please assign the path of species table of STRING.
  --score SCORE, -ps SCORE
                        Please assign the cutoff of text-mining score. The
                        value is from -1 to 1. Default is 0.
  --node_size NODE_SIZE, -ns NODE_SIZE
                        Please assign the size of nodes in figure, default is
                        4000.
  --query QUERY, -q QUERY
                        Please assign the query protein here. The format is
                        $STRAINOFGFF:$START_POINT:$END_POINT:$STRAND.If you
                        want to compute more than one protein, please use
                        comma to separate them. For example, Staphylococcus_au
                        reus_HG003:345:456:+.Staphylococcus_aureus_HG003:2000:
                        3211:-. If you want to compute all protein, just type
                        all. Default is all.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All the output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/PPI</span></code>.</p>
<p><code class="docutils literal"><span class="pre">best_results</span></code>: The results which the score of supported literatures(the text-mining
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a>) is high enough).
<code class="docutils literal"><span class="pre">$STRAIN_without_strain.csv</span></code> is the results of searching literatures without specific strain.
<code class="docutils literal"><span class="pre">$STRAIN_with_strain.csv</span></code> is the results of searching literatures with specific strain.
For example, Staphylococcus_aureus_8325_without_strain.csv is search with Staphylococcus aureus;
Staphylococcus_aureus_8325_without_strain.csv is search without Staphylococcus aureus.
<code class="docutils literal"><span class="pre">without_strain</span></code> stores all interaction information which search without specific strain.
<code class="docutils literal"><span class="pre">with_strain</span></code> stores all interaction information which search with specific strain.</p>
<p><code class="docutils literal"><span class="pre">all_results</span></code>: The results of all protein-protein interactions.
(Even the text-mining(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a>) score is too low)</p>
<p><code class="docutils literal"><span class="pre">figures</span></code>: The thickness represents how many literatures can be found for the interactions.
The solid line means there is some literatures which strongly support the interactions. The dash-dot line
means the supported literatures are very weak. The dot line means there is no literatures which can support the
interactions. The color is the best score of the literatures of the interactions.</p>
</div>
<div class="section" id="subcellular-localization">
<h2>subcellular_localization<a class="headerlink" href="#subcellular-localization" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">subcellular</span> <span class="pre">localization</span></code> can predict the subcellular localization of CDSs. It also provide some
statistics and visualization files.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.psort.org/psortb/">Psortb</a>.</p>
<p>Gff files of genome annotation.</p>
<p>Fasta files of genome sequence.</p>
<p>If you want to compute the expressed CDSs only, you need to assign the transcript gff files.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic subcellular_localization [-h] [--Psortb_path PSORTB_PATH]
                                          [--gff_path GFF_PATH]
                                          [--fasta_path FASTA_PATH]
                                          [--transcript_path TRANSCRIPT_PATH]
                                          [--bacteria_type BACTERIA_TYPE]
                                          [--difference_multi DIFFERENCE_MULTI]
                                          [--merge_to_gff]
                                          [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --Psortb_path PSORTB_PATH
                        If you want to assign the path of Psortb, please
                        assign here.
  --gff_path GFF_PATH, -g GFF_PATH
                        The path of genome annotation gff folder.
  --fasta_path FASTA_PATH, -f FASTA_PATH
                        The path of genome fasta folder.
  --transcript_path TRANSCRIPT_PATH, -a TRANSCRIPT_PATH
                        If you assign the path of transcript gff folder, it
                        can get the results which only include expressed CDS.
  --bacteria_type BACTERIA_TYPE, -b BACTERIA_TYPE
                        Is Gram-positive or Gram-negative. Please assign
                        &#39;positive&#39; or &#39;negative&#39;.
  --difference_multi DIFFERENCE_MULTI, -d DIFFERENCE_MULTI
                        If the protein may have multiple locations, it will
                        calculate the difference of scores(psortb) between
                        best one and others. If the difference is within this
                        value, it will also print it out. Default is 0.5. The
                        maximum value is 10.
  --merge_to_gff, -m    If you want to merge the information to genome
                        annotation gff file, please turn it on. Default is
                        False.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/subcellular_localization</span></code>.</p>
<p><code class="docutils literal"><span class="pre">psortb_results</span></code>: The results of Psortb.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: Statistics files and figures.</p>
</div>
<div class="section" id="riboswitch">
<h2>riboswitch<a class="headerlink" href="#riboswitch" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">riboswitch</span></code> will search the pre-CDS region which have ribosome binding sites.
Basically, <code class="docutils literal"><span class="pre">riboswitch</span></code> use <a class="reference external" href="http://infernal.janelia.org/">Infernal</a> to scan
<a class="reference external" href="http://rfam.xfam.org/">Rfam</a> to find the homologs of known riboswitchs.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://infernal.janelia.org/">Infernal</a>.</p>
<p><a class="reference external" href="http://rfam.xfam.org/">Rfam</a>.</p>
<p>Gff files of genome annotation.</p>
<p>Fasta files of genome sequence.</p>
<p>File of <code class="docutils literal"><span class="pre">riboswitch_ID</span></code>. The file should contain Accession of Rfams, IDs and Descriptions of riboswitchs.
The format is <code class="docutils literal"><span class="pre">$ACCESSION{tab}$ID{tab}$DESCRIPTION</span></code>. You can download the file from our
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic">Github</a> (Rfam_riboswitch_ID.csv). You also can create your own one.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic riboswitch [-h] [--infernal_path INFERNAL_PATH]
                            [--riboswitch_ID RIBOSWITCH_ID]
                            [--gff_path GFF_PATH] [--tss_path TSS_PATH]
                            [--UTR_length UTR_LENGTH]
                            [--transcript_path TRANSCRIPT_PATH]
                            [--fasta_path FASTA_PATH] [--Rfam RFAM]
                            [--e_value E_VALUE] [--output_all] [--fuzzy FUZZY]
                            [--fuzzy_rbs FUZZY_RBS]
                            [--start_codon START_CODON]
                            [--max_dist_rbs MAX_DIST_RBS]
                            [--min_dist_rbs MIN_DIST_RBS]
                            [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --infernal_path INFERNAL_PATH, -if INFERNAL_PATH
                        Please assign the folder of Infernal(where is cmscan
                        and cmsearch located).
  --riboswitch_ID RIBOSWITCH_ID, -i RIBOSWITCH_ID
                        The path of the riboswitch ID of Rfam. The file is
                        include the Accession, ID and Description of
                        riboswitch in Rfam.
  --gff_path GFF_PATH, -g GFF_PATH
                        The path of annotation gff folder.
  --tss_path TSS_PATH, -t TSS_PATH
                        The path of TSS gff folder.
  --UTR_length UTR_LENGTH, -u UTR_LENGTH
                        The length of UTR. Default is 300.
  --transcript_path TRANSCRIPT_PATH, -a TRANSCRIPT_PATH
                        The path of transcript gff folder.
  --fasta_path FASTA_PATH, -f FASTA_PATH
                        The path of genome fasta folder.
  --Rfam RFAM, -R RFAM  The path of Rfam CM database.
  --e_value E_VALUE, -e E_VALUE
                        The cutoff of e value. Default is 0.001.
  --output_all, -o      One sequence may fit multiple riboswitches. If you
                        want to output all of them, please turn it on. Or it
                        will only print the best one. Default is False.
  --fuzzy FUZZY, -z FUZZY
                        It will extend some nucleotides of 3&#39; or 5&#39; end.
                        Default is 10.
  --fuzzy_rbs FUZZY_RBS, -zr FUZZY_RBS
                        How many nucleotides of ribosomal binding site allow
                        to be different with AGGAGG? Default is 2.
  --start_codon START_CODON, -ac START_CODON
                        What kinds of start coden you want to use. If you want
                        to import more than one type of start codon, please
                        use comma to separate them. Default is ATG.
  --max_dist_rbs MAX_DIST_RBS, -Mr MAX_DIST_RBS
                        The maximum distance between ribosomal binding site
                        and start codon. Default is 14.
  --min_dist_rbs MIN_DIST_RBS, -mr MIN_DIST_RBS
                        The minmum distance between ribosomal binding site and
                        start codon. Default is 5.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>All output files will be stored in <code class="docutils literal"><span class="pre">$ANNOgesic/output/riboswitch</span></code>.</p>
<p><code class="docutils literal"><span class="pre">gffs</span></code>: The gff files of riboswitchs.</p>
<p><code class="docutils literal"><span class="pre">tables</span></code>: The tables of riboswichs with more details.</p>
<p><code class="docutils literal"><span class="pre">scan_Rfam</span></code>: The results of <code class="docutils literal"><span class="pre">cmscan</span></code> of <a class="reference external" href="http://infernal.janelia.org/">Infernal</a>.
User can check the results of blast.</p>
<p><code class="docutils literal"><span class="pre">statistics</span></code>: The statistics files and figures.</p>
</div>
<div class="section" id="optimize-tsspredator">
<h2>optimize_tsspredator<a class="headerlink" href="#optimize-tsspredator" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">optimize_tsspredator</span></code> can adapt the parameter set of <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.
User just need to manual check around 200kb as a reference (we suggest the manual detected TSSs should more than 50).
Then <code class="docutils literal"><span class="pre">optimize_tsspredator</span></code> can scan whole genome based on the principle of manual detection to get the best results.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.</p>
<p>The libraries and wiggle files of TEX +/-.
Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<p>Fasta file of genome sequence.</p>
<p>Gff file of genome annotation.</p>
<p>Gff file of manual detection.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic optimize_tsspredator [-h]
                                      [--TSSpredator_path TSSPREDATOR_PATH]
                                      [--fasta_file FASTA_FILE]
                                      [--annotation_file ANNOTATION_FILE]
                                      [--wig_folder WIG_FOLDER]
                                      [--manual MANUAL]
                                      [--strain_name STRAIN_NAME]
                                      [--max_height MAX_HEIGHT]
                                      [--max_height_reduction MAX_HEIGHT_REDUCTION]
                                      [--max_factor MAX_FACTOR]
                                      [--max_factor_reduction MAX_FACTOR_REDUCTION]
                                      [--max_base_height MAX_BASE_HEIGHT]
                                      [--max_enrichment_factor MAX_ENRICHMENT_FACTOR]
                                      [--max_processing_factor MAX_PROCESSING_FACTOR]
                                      [--utr_length UTR_LENGTH] [--lib LIB]
                                      [--output_prefix OUTPUT_PREFIX]
                                      [--cluster CLUSTER] [--length LENGTH]
                                      [--core CORE] [--program PROGRAM]
                                      [--replicate_match REPLICATE_MATCH]
                                      [--steps STEPS]
                                      [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --TSSpredator_path TSSPREDATOR_PATH
                        If you want to assign the path of TSSpredator, please
                        assign here. Default is /usr/local/bin/TSSpredator.jar
  --fasta_file FASTA_FILE, -fs FASTA_FILE
                        Path of one target genome fasta file which you want to
                        opimize it.
  --annotation_file ANNOTATION_FILE, -g ANNOTATION_FILE
                        Path of one target genome annotation gff file which
                        you want to opimize it.
  --wig_folder WIG_FOLDER, -w WIG_FOLDER
                        The folder of the TEX+/- wig folder.
  --manual MANUAL, -m MANUAL
                        The file of manual checked gff file.
  --strain_name STRAIN_NAME, -n STRAIN_NAME
                        The name of the strain you want to optimize.
  --max_height MAX_HEIGHT, -he MAX_HEIGHT
                        This value relates to the minimum number of read
                        starts at a certain genomic position to be considered
                        as a TSS candidate. During optimization will be never
                        larger than this value. Default is 2.5.
  --max_height_reduction MAX_HEIGHT_REDUCTION, -rh MAX_HEIGHT_REDUCTION
                        When comparing different strains/conditions and the
                        step height threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step height threshold.
                        During optimization will be never larger than this
                        value. Default is 2.4.
  --max_factor MAX_FACTOR, -fa MAX_FACTOR
                        This is the minimum factor by which the TSS height has
                        to exceed the local expression background. During
                        optimization will be never larger than this value.
                        Default is 10.
  --max_factor_reduction MAX_FACTOR_REDUCTION, -rf MAX_FACTOR_REDUCTION
                        When comparing different strains/conditions and the
                        step factor threshold is reached in at least one
                        strain/condition, the threshold is reduced for the
                        other strains/conditions by the value set here. This
                        value must be smaller than the step factor threshold.
                        During optimization will be never larger than this
                        value. Default is 9.9.
  --max_base_height MAX_BASE_HEIGHT, -bh MAX_BASE_HEIGHT
                        This is the minimum number of reads should be mapped
                        on TSS. During optimization will be never larger than
                        this value. Default is 0.06.
  --max_enrichment_factor MAX_ENRICHMENT_FACTOR, -ef MAX_ENRICHMENT_FACTOR
                        This is the minimum enrichment factor. During
                        optimization will be never larger than this value.
                        Default is 6.0.
  --max_processing_factor MAX_PROCESSING_FACTOR, -pf MAX_PROCESSING_FACTOR
                        This is the minimum processing factor. If untreated
                        library is higher than the treated library and above
                        which the TSS candidate is considered as a processing
                        site and not annotated as detected. During
                        optimization will be never larger than this value.
                        Default is 6.0
  --utr_length UTR_LENGTH, -u UTR_LENGTH
                        The length of UTR. It is for Primary and Secondary
                        TSSs. Default is 300.
  --lib LIB, -l LIB     The libraries of wig files for TSSpredator. The format
                        is: wig_file_name:tex_treat_or_not(tex or notex):condi
                        tion_id(integer):replicate_id(alphabet):strand(+ or
                        -). If you have multiple wig files, please use comma
                        to separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --output_prefix OUTPUT_PREFIX, -p OUTPUT_PREFIX
                        The output prefix of all conditions. If you have
                        multiple conditions, please use comma to separate
                        them. For example,
                        prefix_condition1,prefix_condition2.
  --cluster CLUSTER, -cu CLUSTER
                        If the position between manual one and predicted one
                        is smaller or equal than this value, it will only
                        print one of them. Default is 2.
  --length LENGTH, -le LENGTH
                        The length of genome for running optimization. If you
                        don&#39;t want to run for whole genome, you can assign the
                        length of partial genome.Default is compare whole
                        genome.
  --core CORE, -c CORE  How many paralle runs do you want to use. Default is
                        4.
  --program PROGRAM, -t PROGRAM
                        You want to run optimize TSS or processing site.
                        Please assign &quot;TSS&quot; or &quot;Processint_site&quot;. Default is
                        TSS.
  --replicate_match REPLICATE_MATCH, -rm REPLICATE_MATCH
                        The TSS candidates should match to how many number of
                        the replicates. Default is 1.
  --steps STEPS, -s STEPS
                        How many steps do you want to run. Default is 4000
                        runs.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>Based on the programs (TSS/processing site), the output files will be stored in
<code class="docutils literal"><span class="pre">$ANNOgesic/output/TSS/optimized_TSSpredator</span></code> or <code class="docutils literal"><span class="pre">$ANNOgesic/output/processing_site/optimized_TSSpredator</span></code>.</p>
<p><code class="docutils literal"><span class="pre">stat.csv</span></code> stores the information of every step. The first column is the number of run.
The second column is the parameter set. <code class="docutils literal"><span class="pre">he</span></code> represents height; <code class="docutils literal"><span class="pre">rh</span></code> represents
height reduction; <code class="docutils literal"><span class="pre">fa</span></code> means factor; <code class="docutils literal"><span class="pre">rf</span></code> means factor reduction; <code class="docutils literal"><span class="pre">bh</span></code> is
base height; <code class="docutils literal"><span class="pre">ef</span></code> is enrichment factor; <code class="docutils literal"><span class="pre">pf</span></code> is processing factor. About the details
of parameters, please refer to <a class="reference external" href="http://it.inf.uni-tuebingen.de/?page_id=190">TSSpredator</a>.
For example, <code class="docutils literal"><span class="pre">he_2.0_rh_1.8_fa_4.4_rf_2.8_bh_0.08_ef_3.0_pf_2.6</span></code> means height is 2.0,
height reduction is 1.8, factor is 4.4, factor reduction is 2.8, base height is 0.08,
enrichment factor is 3.0 and processing factor is 2.6. The third and fourth columns are
the number of true positive. The fifth and sixth columns are true positive rate. The seventh
and eighth columns are the number of false positive. The ninth and tenth false positive rate.
The eleventh and twelfth columns are the number of false negative. The thirteenth and fourteenth
columns are missing rate.</p>
<p><code class="docutils literal"><span class="pre">best.csv</span></code> stores the best parameter set. The meanings of all columns are the same as <code class="docutils literal"><span class="pre">stat.csv</span></code>.</p>
</div>
<div class="section" id="screenshot">
<h2>screenshot<a class="headerlink" href="#screenshot" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">screenshot</span></code> will generate batch files for producing screenshot of <a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>.
When the batch files produced, user just needs to open <a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>, then presses <code class="docutils literal"><span class="pre">tools</span></code>
on the top tags and choose <code class="docutils literal"><span class="pre">run</span> <span class="pre">batch</span> <span class="pre">script</span></code>. The program will automatically produce screenshots.
Then user can refer to them easily.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="https://www.broadinstitute.org/igv">IGV</a>.</p>
<p>Gff files that user want to produce screenshots. All screenshots will be produced based on the positions of <code class="docutils literal"><span class="pre">main_gff</span></code>.
<code class="docutils literal"><span class="pre">side_gffs</span></code> are the gff files that user want to compare with <code class="docutils literal"><span class="pre">main_gff</span></code>. They will also appear in the screenshots.</p>
<p>Fasta files of genome.</p>
<p>The libraries and wiggle files. Please refer to the <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic screenshot [-h] [--main_gff MAIN_GFF] [--side_gffs SIDE_GFFS]
                            [--fasta FASTA]
                            [--frag_wig_folder FRAG_WIG_FOLDER]
                            [--tex_wig_folder TEX_WIG_FOLDER]
                            [--height HEIGHT] [--tex_libs TEX_LIBS]
                            [--frag_libs FRAG_LIBS] [--present PRESENT]
                            [--output_folder OUTPUT_FOLDER]
                            [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --main_gff MAIN_GFF, -mg MAIN_GFF
                        Screenshot will based on the position of main_gff file
                        to generate screenshot.
  --side_gffs SIDE_GFFS, -sg SIDE_GFFS
                        If you want to import some gff files to compare with
                        --main_gff, please assign here. You can input more
                        than one gff files, just use comma to separate them.
  --fasta FASTA, -f FASTA
                        The path of genome fasta folder.
  --frag_wig_folder FRAG_WIG_FOLDER, -fw FRAG_WIG_FOLDER
                        If you want to include the information of fragmented
                        wig file, please assign the folder.
  --tex_wig_folder TEX_WIG_FOLDER, -tw TEX_WIG_FOLDER
                        If you want to include the information of tex+/- wig
                        file, please assign the folder.
  --height HEIGHT, -he HEIGHT
                        You can assign the height of screenshot. Default is
                        1500.
  --tex_libs TEX_LIBS, -tl TEX_LIBS
                        If you want to include the tex+/- wig file, please
                        also assign proper format here. The format is:
                        wig_file_name:tex_treat_or_not(tex or notex):condition
                        _id(integer):replicate_id(alphabet):strand(+ or -). If
                        you have multiple wig files, please use comma to
                        separate the wig files. For example,
                        wig1:tex:1:a:+,wig2:tex:1:a:-.
  --frag_libs FRAG_LIBS, -fl FRAG_LIBS
                        If you want to include the fragmented wig file, please
                        also assign proper format here. The format is: wig_fil
                        e_name:fragmented(frag):condition_id(integer):replicat
                        e_id(alphabet):strand(+ or -). If you have multiple
                        wig files, please use comma to separate the wig files.
                        For example, wig1:frag:1:a:+,wig2:frag:1:a:-.
  --present PRESENT, -p PRESENT
                        Which type that you want to present in the screenshot.
                        expand/collapse/squish. Default is expand.
  --output_folder OUTPUT_FOLDER, -o OUTPUT_FOLDER
                        Please assign the output folder. It will create a
                        sub-folder &quot;screenshots&quot; in the output folder to store
                        the results.
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>Based on the paths of <code class="docutils literal"><span class="pre">main_gff</span></code>, <code class="docutils literal"><span class="pre">screenshot</span></code> will generate a folder - <code class="docutils literal"><span class="pre">screenshots</span></code> under the
folder of <code class="docutils literal"><span class="pre">main_gff</span></code>. All output files will be stored in this folder.</p>
<p><code class="docutils literal"><span class="pre">forward.txt</span></code> is the batch file of forward strand for running on IGV.</p>
<p><code class="docutils literal"><span class="pre">reverse.txt</span></code> is the batch file of reverse strand for running on IGV.</p>
<p><code class="docutils literal"><span class="pre">forward</span></code> is the folder for storing screenshots of forward strand.</p>
<p><code class="docutils literal"><span class="pre">reverse</span></code> is the folder for storing screenshots of reverse strand.</p>
<p>When user runs batch files on IGV. The screenshots will automatically store in <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>.
The format of filenames will be <code class="docutils literal"><span class="pre">$STRAIN:$START-$END.png</span></code>. For example, <code class="docutils literal"><span class="pre">NC_007795:1051529-1051696.png</span></code>
means the strain is NC_007795. The feature&#8217;s start point is 1051529 and the end point is
1051696.</p>
</div>
<div class="section" id="color-png">
<h2>color_png<a class="headerlink" href="#color-png" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">color_png</span></code> is a following procedure of <code class="docutils literal"><span class="pre">screenshot</span></code>. If there are many wiggle files, it will be difficult
to distinguish the tracks. Especially, when user want to compare TEX +/- libraries to check TSSs or processing sites,
it is not convenient. <code class="docutils literal"><span class="pre">color_png</span></code> can color the tracks based on TEX +/- libraries. Therefore, the user can
refer to the screenshots much easier.</p>
<ul class="simple">
<li>Pre-required tools and files</li>
</ul>
<p><a class="reference external" href="http://www.imagemagick.org/script/index.php">ImageMagick</a>.</p>
<p>The screenshots which generated from <code class="docutils literal"><span class="pre">screenshot</span></code>. Please make sure the folders of <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>
exist in the folder of <code class="docutils literal"><span class="pre">screenshots</span></code>.</p>
<ul class="simple">
<li>Arguments</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>usage: annogesic color_png [-h] [--screenshot_folder SCREENSHOT_FOLDER]
                           [--track_number TRACK_NUMBER]
                           [--ImageMagick_covert_path IMAGEMAGICK_COVERT_PATH]
                           [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given, the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --screenshot_folder SCREENSHOT_FOLDER, -f SCREENSHOT_FOLDER
                        The folder which stores the folder of &quot;screenshots&quot;
                        which generated by subcommand &quot;screenshot&quot;.
  --track_number TRACK_NUMBER, -t TRACK_NUMBER
                        How many number of tracks do you have?
  --ImageMagick_covert_path IMAGEMAGICK_COVERT_PATH, -m IMAGEMAGICK_COVERT_PATH
                        Please assign the path of &quot;convert&quot; in ImageMagick
                        package
</pre></div>
</div>
<ul class="simple">
<li>Output files</li>
</ul>
<p>The new screenshots will replace the previous ones automatically.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="test_case.html" title="Performing a test cast"
             >next</a></li>
        <li class="right" >
          <a href="docker.html" title="Dockfiles"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.1.3 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Sung-Huan Yu.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>