<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Performing an test cast &mdash; ANNOgesic 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ANNOgesic 1.0.0 documentation" href="index.html" />
    <link rel="next" title="Troubleshooting" href="troubleshooting.html" />
    <link rel="prev" title="ANNOgesic’s subcommands" href="subcommands.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="troubleshooting.html" title="Troubleshooting"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="subcommands.html" title="ANNOgesic’s subcommands"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="performing-an-test-cast">
<h1>Performing an test cast<a class="headerlink" href="#performing-an-test-cast" title="Permalink to this headline">¶</a></h1>
<p>Here we will guide you through a small test case using ANNOgesic.
You will see most subcommands of ANNOgesic. The test case is a publicly
available RNA-Seq from NCBI GEO that was part of a work by
<cite>Bischler et al. &lt;http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE67564&gt;_</cite>.
The differential RNA-seq data is of Helicobacter pylori 26695.
There will be several output files which are generated in different formats.
The CSV (tabular separated plain text files) files (opened by LibreOffice or Excel).
For GFF3 file you can use a genome browser for example <cite>IGB &lt;http://bioviz.org/igb/index.html&gt;_</cite>,
<cite>IGV &lt;https://www.broadinstitute.org/igv/&gt;_</cite> or <cite>Jbrowse &lt;http://jbrowse.org/&gt;_</cite>.
There are also several PNG and TXT files.</p>
<p>Before we start, please refer to the section of <code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">filename</span></code> and
<code class="docutils literal"><span class="pre">The</span> <span class="pre">format</span> <span class="pre">of</span> <span class="pre">libraries</span> <span class="pre">for</span> <span class="pre">import</span> <span class="pre">to</span> <span class="pre">ANNOgesic</span></code> in <code class="docutils literal"><span class="pre">subcommand</span></code>. All the details
are also in <code class="docutils literal"><span class="pre">subcommand</span></code>.</p>
<div class="section" id="generating-a-project">
<h2>Generating a project<a class="headerlink" href="#generating-a-project" title="Permalink to this headline">¶</a></h2>
<p>First of all, we need to create a working folder, we use <code class="docutils literal"><span class="pre">create</span></code> subcommand.</p>
<div class="highlight-python"><div class="highlight"><pre>annogesic create ANNOgesic
</pre></div>
</div>
<p>Then you will see</p>
<div class="highlight-python"><div class="highlight"><pre>Created folder &quot;ANNOgesic&quot; and required subfolders.
$ ls
ANNOgesic
</pre></div>
</div>
</div>
<div class="section" id="retrieving-the-input-data">
<h2>Retrieving the input data<a class="headerlink" href="#retrieving-the-input-data" title="Permalink to this headline">¶</a></h2>
<p>For our test case, we can download from <cite>NCBI &lt;ftp://ftp.ncbi.nih.gov/genomes/Bacteria/Helicobacter_pylori_26695_uid57787/&gt;_</cite>.
We can set the <code class="docutils literal"><span class="pre">$FTP_SOURCE</span></code> first</p>
<dl class="docutils">
<dt>::</dt>
<dd>FTP_SOURCE=ftp://ftp.ncbi.nih.gov/genomes/Bacteria/Helicobacter_pylori_26695_uid57787</dd>
</dl>
<p>Then download fasta files(<code class="docutils literal"><span class="pre">-f</span></code>), gff files(<code class="docutils literal"><span class="pre">-g</span></code>), gbk files(<code class="docutils literal"><span class="pre">-k</span></code>), ptt files(<code class="docutils literal"><span class="pre">-p</span></code>),
rnt files(<code class="docutils literal"><span class="pre">-r</span></code>), and transfer to embl(<code class="docutils literal"><span class="pre">-e</span></code>).</p>
<div class="highlight-python"><div class="highlight"><pre>$ annogesic get_input_files -F $FTP_SOURCE -g -f -e -k -p -r ANNOgesic
</pre></div>
</div>
<p>you can also write</p>
<div class="highlight-python"><div class="highlight"><pre>$ annogesic get_input_files -F $FTP_SOURCE -gfekpr ANNOgesic
</pre></div>
</div>
<p>Then you will get the following results</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ ls ANNOgesic/input/reference/fasta/
NC_000915.fa
$ ls ANNOgesic/input/reference/annotation/
NC_000915.embl  NC_000915.gbk  NC_000915.gff  NC_000915.ptt  NC_000915.rnt</dd>
</dl>
</div>
<div class="section" id="putting-wig-bam-files-and-reads-to-proper-location">
<h2>Putting wig, bam files and reads to proper location<a class="headerlink" href="#putting-wig-bam-files-and-reads-to-proper-location" title="Permalink to this headline">¶</a></h2>
<p>For the test case, you can download files from
<cite>here &lt;http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE67564&gt;_</cite>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP%2FSRP056%2FSRP056856/SRR1951997/SRR1951997.sra
$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP%2FSRP056%2FSRP056856/SRR1951998/SRR1951998.sra
</pre></div>
</div>
<p>Then you need to convert the SRA files to Fasta or Fastq format for mapping. You can
use <cite>SRA toolkit &lt;http://www.ncbi.nlm.nih.gov/books/NBK158900/&gt;_</cite> for that.</p>
<div class="highlight-python"><div class="highlight"><pre>$ wget http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.5.2/sratoolkit.2.5.2-ubuntu64.tar.gz
$ tar -zxvf sratoolkit.2.5.2-ubuntu64.tar.gz
$ rm sratoolkit.2.5.2-ubuntu64.tar.gz
$ ./sratoolkit.2.5.2-ubuntu64/bin/fastq-dump.2.5.2 --fasta SRR1951997.sra
$ ./sratoolkit.2.5.2-ubuntu64/bin/fastq-dump.2.5.2 --fasta SRR1951998.sra
$ rm SRR1951997.sra
$ rm SRR1951998.sra
</pre></div>
</div>
<p>Now you get the reads. You can select a part of them to reduce the running time.</p>
<p>Then we have to download the wiggle files.</p>
<dl class="docutils">
<dt>::</dt>
<dd><p class="first">wget -cP ANNOgesic/input/wigs/tex_notex <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649587/suppl/GSM1649587%5FHp26695%5FML%5FB1%5FHS1%5F%2DTEX%5Fforward%2Ewig%2Egz">ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649587/suppl/GSM1649587%5FHp26695%5FML%5FB1%5FHS1%5F%2DTEX%5Fforward%2Ewig%2Egz</a>
wget -cP ANNOgesic/input/wigs/tex_notex <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649587/suppl/GSM1649587%5FHp26695%5FML%5FB1%5FHS1%5F%2DTEX%5Freverse%2Ewig%2Egz">ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649587/suppl/GSM1649587%5FHp26695%5FML%5FB1%5FHS1%5F%2DTEX%5Freverse%2Ewig%2Egz</a>
wget -cP ANNOgesic/input/wigs/tex_notex <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649588/suppl/GSM1649588%5FHp26695%5FML%5FB1%5FHS1%5F%2BTEX%5Fforward%2Ewig%2Egz">ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649588/suppl/GSM1649588%5FHp26695%5FML%5FB1%5FHS1%5F%2BTEX%5Fforward%2Ewig%2Egz</a>
wget -cP ANNOgesic/input/wigs/tex_notex <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649588/suppl/GSM1649588%5FHp26695%5FML%5FB1%5FHS1%5F%2BTEX%5Freverse%2Ewig%2Egz">ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1649nnn/GSM1649588/suppl/GSM1649588%5FHp26695%5FML%5FB1%5FHS1%5F%2BTEX%5Freverse%2Ewig%2Egz</a>
cd ANNOgesic/input/wigs/tex_notex
gunzip <a href="#id7"><span class="problematic" id="id8">GSM1649587_Hp26695_ML_B1_HS1_</span></a>-TEX_forward.wig.gz </p>
<blockquote>
<div><a href="#id9"><span class="problematic" id="id10">GSM1649587_Hp26695_ML_B1_HS1_</span></a>-TEX_reverse.wig.gz GSM1649588_Hp26695_ML_B1_HS1_+TEX_forward.wig.gz GSM1649588_Hp26695_ML_B1_HS1_+TEX_reverse.wig.gz</div></blockquote>
<p class="last">cd ../../../../</p>
</dd>
</dl>
</div>
<div class="section" id="improving-the-reference-genome">
<h2>Improving the reference genome<a class="headerlink" href="#improving-the-reference-genome" title="Permalink to this headline">¶</a></h2>
<p>If the input data which we retrieved is exactly what you want, you can skip this step.
You just need to copy the data which you retrieved to <code class="docutils literal"><span class="pre">ANNOgesic/output/target/fasta</span></code>.</p>
<p>If the retrieved data is only the closed strain of your target strain, you may need to run
subcommand <code class="docutils literal"><span class="pre">get_target_fasta</span></code>. However, this command need the mutation table. please refer
to the user manual. Once you have the mutation table, you can improve the fasta files.</p>
<p>We use a simple example to modify our test case. The mutation table is</p>
<div class="highlight-python"><div class="highlight"><pre>#target_id  reference_id    reference_nt    position        target_nt       impact of correction    locus tag       gene    Description
NC_test.1   NC_000915.1     a       3       c               SAOUHSC_00002   dnaA    XXXXXX
NC_test.1   NC_000915.1     t       6       -       deletion                        YYYYYY
test_case2  NC_000915.1     -       6       g       insertion       SAOUHSC_00132
</pre></div>
</div>
<p>You can see the new strain will be NC_test.1 and test_case2. Although there are four fasta files
in <code class="docutils literal"><span class="pre">ANNOgesic/input/reference/fasta</span></code>, we just modify two of them. Therefore, there will be
only two fasta files in <code class="docutils literal"><span class="pre">ANNOgesic/output/target/fasta</span></code>, too.</p>
<p>Now, let&#8217;s try it</p>
<div class="highlight-python"><div class="highlight"><pre>$ annogesic get_target_fasta \
   -r ANNOgesic/input/reference/fasta \
   -o test_case1:NC_test.1 test_case2:test_case2 \
   -m ANNOgesic/input/mutation_table/mutation.csv \
   ANNOgesic
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">-r</span></code> is the folder of original fasta files. In <code class="docutils literal"><span class="pre">-o</span></code> you can assign the filename of output fasta and
which strains you want to put in it. In our case, we call the first fasta file test_case1 and the
second one test_case2. In test_case1 stores the fasta of NC_test.1 and test_case2 stores test_case2.
Now we can check the retuls.</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ head ANNOgesic/input/reference/fasta/NC_016810.fa
&gt;NC_000915.1
TGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAG
TGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAG
TGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTATAGCATCATTTTTTAAATTTAGGTATAAA
ACACCCTCAATTCAAGGGTTTTTGAGTGAGCTTTTTGCTCAAAGAATCCAAGATAGCGTTTAAAAATTTA
GGGGTGTTAGGCTCAGCGTAGAGTTTGCCAAGCTCTATGCATTCATTGATGATGATAGGGTTTTGCGTGG
GCGTGAAGCCAATTTCATACGCTCCTAAGCGTAAAATCGCCTTTTCCATGCTCCCTAATCGCTTGAAATC
CCAGTCTTTTAAATGCGGCTCGATGAGGGCGTCAATTTCATTGATTTTTTCTAACACGCCATTAAAAAGG
CTTAAAGCGAAAGCGAGTTGGTTGTTTTTAATCTTTTTTTCTTCTAACATGCTAGAAGCGATTTTTTTAA
TTTCTTCATTACCGCTCTCAAACGCATACAACAATTCAACCACAGCCCCCCTGGCTTGAGTTCGTGTCGC
$ head ANNOgesic/output/target/fasta/test_case1.fa
&gt;NC_test.1
TGcTTGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATT
AGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTG
ATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTAGTGATTA
TAGCATCATTTTTTAAATTTAGGTATAAAACACCCTCAATTCAAGGGTTTTTGAGTGAGC
TTTTTGCTCAAAGAATCCAAGATAGCGTTTAAAAATTTAGGGGTGTTAGGCTCAGCGTAG
AGTTTGCCAAGCTCTATGCATTCATTGATGATGATAGGGTTTTGCGTGGGCGTGAAGCCA
ATTTCATACGCTCCTAAGCGTAAAATCGCCTTTTCCATGCTCCCTAATCGCTTGAAATCC
CAGTCTTTTAAATGCGGCTCGATGAGGGCGTCAATTTCATTGATTTTTTCTAACACGCCA
TTAAAAAGGCTTAAAGCGAAAGCGAGTTGGTTGTTTTTAATCTTTTTTTCTTCTAACATG</dd>
</dl>
<p>In new fasta file, the third nucleotide replace from A to c. Moreover, The sixth nucleotide is deleted.
In test_case2, it also modify the fasta file based on the mutation table.</p>
<p>when you have the correct fasta files, you can map your reads to the correct fasta file.</p>
<p>If you have no mutation table, you can also use the subcommand <code class="docutils literal"><span class="pre">snp</span></code> to detect the mutation automatically.
For this subcommand, we will go through it later.</p>
</div>
<div class="section" id="generating-annotation-files">
<h2>Generating annotation files<a class="headerlink" href="#generating-annotation-files" title="Permalink to this headline">¶</a></h2>
<p>We have the fasta files now. We can use it to generate our annotation files. If the annotaion files which
you retrieved by <code class="docutils literal"><span class="pre">get_input_files</span></code> is exactly what you want, you can skip this step. You just need to
copy the annotation files to <code class="docutils literal"><span class="pre">ANNOgesic/output/target/annotation</span></code>.</p>
<p>Before you running it, you have to notice path. If you are using docker container, the path is alread setup.
You don&#8217;t need to worry about it. However, if you are build ANNOgesic by yourself, remember to assign the
path by running</p>
<dl class="docutils">
<dt>::</dt>
<dd>. $PAGIT_HOME/sourceme.pagit</dd>
</dl>
<p><code class="docutils literal"><span class="pre">$PAGIT_HOME</span></code> is the directory of PAGIT. The better way is change the environment. Or you have to run
this command everytime. For changing the environment, you just need to copy all the information of
<code class="docutils literal"><span class="pre">$PAGIT_HOME/sourceme.pagit</span></code> to <code class="docutils literal"><span class="pre">.bashrc</span></code>.</p>
<p>Now, we can try it.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic annotation_transfer \
    -re ANNOgesic/input/reference/annotation \
    -rf ANNOgesic/input/reference/fasta \
    -tf ANNOgesic/output/target/fasta \
    -e chromosome \
    -t Strain \
    -p NC_000915:NC_test.1 NC_000915:test_case2 \
    -g \
    ANNOgesic
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">-e</span></code> is the prefix of output embl files. <code class="docutils literal"><span class="pre">-t</span></code> is a program of <cite>RATT &lt;http://ratt.sourceforge.net/&gt;_</cite>.
We use <code class="docutils literal"><span class="pre">Strain</span></code> because the similarity is higher than 90%. We assign the pairs of transfer at <code class="docutils literal"><span class="pre">-p</span></code>.
The names are strains&#8217; name not filenames of fasta files. <code class="docutils literal"><span class="pre">-g</span></code> means we want to transfer the embl files
to GFF3 files and store in <code class="docutils literal"><span class="pre">ANNOgesic/output/target/annotation</span></code>.</p>
<p>When the computation is done, you can see</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/target/annotation/
test_case1.gff  test_case1.ptt  test_case1.rnt  test_case2.gff  test_case2.ptt  test_case2.rnt
$ ls ANNOgesic/output/annotation_transfer/
chromosome.NC_test.1.final.embl  chromosome.test_case2.final.embl  NC_test.1.gff  ratt_log.txt  test_case2.gff
</pre></div>
</div>
<p>In <code class="docutils literal"><span class="pre">ANNOgesic/output/target/annotation</span></code>, you can see ptt, rnt and gff files. In <code class="docutils literal"><span class="pre">ANNOgesic/output/annotation_transfer</span></code>,
you can see the results of <cite>RATT &lt;http://ratt.sourceforge.net/&gt;_</cite>. <code class="docutils literal"><span class="pre">chromosome.NC_test.1.final.embl</span></code> and
<code class="docutils literal"><span class="pre">chromosome.test_case2.final.embl</span></code> are generated by <cite>RATT &lt;http://ratt.sourceforge.net/&gt;_</cite>. Gff files are
transferred from these embl files.</p>
</div>
<div class="section" id="gene-expression-analysis">
<h2>Gene expression analysis<a class="headerlink" href="#gene-expression-analysis" title="Permalink to this headline">¶</a></h2>
<p>Now we already saw how to generate the fasta and annotation files. In order to running
the following subcommand, we have to replace the files of <code class="docutils literal"><span class="pre">ANNOgesic/output/target/annotation</span></code>
and <code class="docutils literal"><span class="pre">ANNOgesic/output/target/fasta</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>rm -rf ANNOgesic/output/target/annotation
cp -rf ANNOgesic/input/reference/annotation ANNOgesic/output/target
rm -rf ANNOgesic/output/target/fasta
cp -rf ANNOgesic/input/reference/fasta ANNOgesic/output/target
</pre></div>
</div>
<p>We do this because the wiggle file is generated by NC_000915. If you running by your data, you don&#8217;t need to do this.
Now, our fasta and annotation files are fit with wiggle files. We can other subcommands now.</p>
<p>For analyzing gene expression, we can run <code class="docutils literal"><span class="pre">expression</span> <span class="pre">analysis</span></code>. Based on this subcommand, you can
know which CDS expresses in which library or discover housekeeping gene.</p>
<p>Before running it, you can set the library first.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">tex_notex_libs</span><span class="o">=</span><span class="s">&quot;GSM1649587_Hp26695_ML_B1_HS1_-TEX_forward.wig:notex:1:a:+ </span><span class="se">\</span>
<span class="s">                GSM1649587_Hp26695_ML_B1_HS1_-TEX_reverse.wig:notex:1:a:- </span><span class="se">\</span>
<span class="s">                GSM1649588_Hp26695_ML_B1_HS1_+TEX_forward.wig:tex:1:a:+ </span><span class="se">\</span>
<span class="s">                GSM1649588_Hp26695_ML_B1_HS1_+TEX_reverse.wig:tex:1:a:-&quot;</span>
</pre></div>
</div>
<p>The command would be like</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic expression_analysis \
    -g $ANNOGESIC_FOLDER/output/target/annotation \
    -tl $tex_notex_libs \
    -tw $ANNOGESIC_FOLDER/input/wigs/tex_notex \
    -f CDS \
    -rt 1 \
    $ANNOGESIC_FOLDER
</pre></div>
</div>
<p>It will generate several gff files and statistics files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/target/annotation/for_libs/gffs
NC_000915_CDS_1_texnotex.gff  NC_000915_CDS_all_libs.gff  NC_000915_CDS_at_least_one_lib.gff  NC_000915_CDS_no_express.gff
$ ls ANNOgesic/output/target/annotation/for_libs/statistics
NC_000915_CDS.csv
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">NC_000915_CDS_1_texnotex.gff</span></code> stores the CDS which express in the library of 1_texnotex. In our definition
of library, it is <code class="docutils literal"><span class="pre">GSM1649587_Hp26695_ML_B1_HS1</span></code>. <code class="docutils literal"><span class="pre">NC_000915_CDS_all_lib.gff</span></code> is the CDS which express in all libraries.
<code class="docutils literal"><span class="pre">NC_000915_CDS_at_least_one_lib.gff</span></code> stores the CDS which express at least one library. <code class="docutils literal"><span class="pre">NC_000915_CDS_no_express.gff</span></code>
is the CDS which don&#8217;t express in any libraries.</p>
</div>
<div class="section" id="tss-and-processing-site-prediction-and-optimization">
<h2>TSS and processing site prediction and optimization<a class="headerlink" href="#tss-and-processing-site-prediction-and-optimization" title="Permalink to this headline">¶</a></h2>
<p>Before running <code class="docutils literal"><span class="pre">tsspredator</span></code>, if you want to use the optimized parameters,
you need to run <code class="docutils literal"><span class="pre">optimize_TSSpredator</span></code> first.</p>
<p>Then need to manual check some TSS. In our experience, we recommand you detect more than 50 TSS and longer than 100kb.
For test case, we can just copy the default parameter as a manual detected ones. This is only for test. If you are
running your own data and you have manual detected TSS, please use it.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic tsspredator \
    -w ANNOgesic/input/wigs/tex_notex \
    -f ANNOgesic/output/target/fasta \
    -g ANNOgesic/output/target/annotation \
    -l $tex_notex_libs \
    -p test \
    ANNOgesic
mv ANNOgesic/output/TSS/gffs/NC_000915_TSS.gff ANNOgesic/input/manual_TSS
</pre></div>
</div>
<p>Now, we have a fake manual detected TSS file. we can try optimization of TSS right now.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic optimize_tsspredator \
    -w ANNOgesic/input/wigs/tex_notex \
    -fs ANNOgesic/output/target/fasta \
    -g ANNOgesic/output/target/annotation \
    -n NC_000915.1 \
    -l $tex_notex_libs \
    -p TSS -s 25 \
    -m ANNOgesic/input/manual_TSS/NC_000915_TSS.gff  \
    ANNOgesic
...
Current Parameter:step=16_height=0.5_reduction_height=0.1_factor=8.3_reduction_factor=2.4_base_height=0.068_enrichment_factor=3.0_processing_factor=2.4
Current:TP=665  TP_rate=0.38394919168591224     FP=23   FP_rate=1.3804403604749916e-05  FN=1067 Missing_ratio=0.6160508083140878
Best Parameter:height=0.5       reduction_height=0.1    factor=8.3      reduction_factor=2.4    base_height=0.068       enrichment_factor=3.0   processing_factor=2.4
Best:TP=665     TP_rate=0.38394919168591224     FP=23   FP_rate=1.3804403604749916e-05  FN=1067 Missing_ratio=0.6160508083140878
Current Parameter:step=17_height=0.5_reduction_height=0.1_factor=9.6_reduction_factor=6.4_base_height=0.1_enrichment_factor=3.0_processing_factor=1.0
Current:TP=539  TP_rate=0.3112009237875289      FP=13   FP_rate=7.802488993989082e-06   FN=1193 Missing_ratio=0.6887990762124712
Best Parameter:height=0.5       reduction_height=0.1    factor=8.3      reduction_factor=2.4    base_height=0.068       enrichment_factor=3.0   processing_factor=2.4
Best:TP=665     TP_rate=0.38394919168591224     FP=23   FP_rate=1.3804403604749916e-05  FN=1067 Missing_ratio=0.6160508083140878
Current Parameter:step=18_height=1.4_reduction_height=0.1_factor=8.3_reduction_factor=2.8_base_height=0.1_enrichment_factor=3.0_processing_factor=1.0
Current:TP=433  TP_rate=0.25    FP=11   FP_rate=6.602106071836916e-06   FN=1299 Missing_ratio=0.75
Best Parameter:height=0.5       reduction_height=0.1    factor=8.3      reduction_factor=2.4    base_height=0.068       enrichment_factor=3.0   processing_factor=2.4
Best:TP=665     TP_rate=0.38394919168591224     FP=23   FP_rate=1.3804403604749916e-05  FN=1067 Missing_ratio=0.6160508083140878
Current Parameter:step=19_height=0.5_reduction_height=0.1_factor=8.3_reduction_factor=4.9_base_height=0.1_enrichment_factor=3.0_processing_factor=5.6
Current:TP=582  TP_rate=0.33602771362586603     FP=16   FP_rate=9.603063377217333e-06   FN=1150 Missing_ratio=0.663972286374134
Best Parameter:height=0.5       reduction_height=0.1    factor=8.3      reduction_factor=2.4    base_height=0.068       enrichment_factor=3.0   processing_factor=2.4
Best:TP=665     TP_rate=0.38394919168591224     FP=23   FP_rate=1.3804403604749916e-05  FN=1067 Missing_ratio=0.6160508083140878
...
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">optimize_TSSpredator</span></code> will compare gff files of manual checked TSS and predicted TSS to find the best parameters.
You can check the results and parameters of each step in screen. Because we just test it, we set the steps only 20.
When the program finished, you can find several files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/TSS/optimized_TSSpredator/
best.csv  log.txt  stat.csv
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">best.csv</span></code> is for the best parameters; <code class="docutils literal"><span class="pre">stat.csv</span></code> is for the parameters of each step.</p>
<p>Let&#8217;s assume the best parameters are height is 0.4, height_reduction is 0.2, factor is 2.0, factor_reduction is 0.5,
base_height is 0.0, enrichment_factor is 2.0, processing_factor is 1.5. Now we can set the parameter set for running
<code class="docutils literal"><span class="pre">tss</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic tsspredator \
    -w ANNOgesic/input/wigs/tex_notex \
    -f ANNOgesic/output/target/fasta \
    -g ANNOgesic/output/target/annotation \
    -l $tex_notex_libs \
    -p test \
    -he 0.4 \
    -rh 0.2 \
    -fa 2.0 \
    -rf 0.5 \
    -bh 0.0 \
    -ef 2.0 \
    -pf 1.5 \
    -s \
    -v \
    -m ANNOgesic/input/manual_TSS/NC_000915_TSS.gff \
    ANNOgesic
</pre></div>
</div>
<p>If you set the manual checked TSS <code class="docutils literal"><span class="pre">-m</span></code>, the subcommand will merge the manual checked TSS with predicted TSS.
If you didn&#8217;t set it, the subcommand will only produce predicted TSS. You will get gff file, MasterTable and statistic file.</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ ls ANNOgesic/output/TSS/gffs/
NC_000915_TSS.gff
$ ls ANNOgesic/output/TSS/MasterTables/MasterTable_NC_000915.1/
AlignmentStatistics.tsv  err.txt  log.txt  MasterTable.tsv  superConsensus.fa  superTSS.gff  superTSStracks.gff  test_super.fa  test_super.gff  test_TSS.gff  TSSstatistics.tsv
$ ls ANNOgesic/output/TSS/statistics/NC_000915/
stat_compare_TSSpredator_manual_NC_000915.csv  stat_gene_vali_NC_000915.csv  stat_TSS_class_NC_000915.csv  stat_TSS_libs_NC_000915.csv  TSS_class_NC_000915.1.png  TSS_venn_NC_000915.1.png</dd>
</dl>
<p>If you want to predict processing site, the procedures are the same. You just need to change the program from TSS to
processing_site (<code class="docutils literal"><span class="pre">-t</span></code>).</p>
</div>
<div class="section" id="performing-transcript-assembly">
<h2>Performing transcript assembly<a class="headerlink" href="#performing-transcript-assembly" title="Permalink to this headline">¶</a></h2>
<p>For detecting transcript boundary, transcript assembly is the basic feature.
we can use the subcommand <code class="docutils literal"><span class="pre">transcript_assembly</span></code> to do it. Normally, we strongly
recommand you to get fragmented libraries for <code class="docutils literal"><span class="pre">transcript_assembly</span></code>. However, there is no
fragmented libraries available in the database. Therefore, we only use TEX +/- to do it.</p>
<p>The command would be like the following.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic transcript_assembly \
    -g ANNOgesic/output/target/annotation \
    -nw ANNOgesic/input/wigs/tex_notex \
    -tl $tex_notex_libs \
    -rt 1 \
    -te 2 \
    -ct ANNOgesic/output/TSS/gffs \
    -cg ANNOgesic/output/target/annotation \
    ANNOgesic
</pre></div>
</div>
<p>It will generate gff files. Because we also compared with TSS and annotation files, it will generate statistics files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/transcriptome_assembly/gffs
NC_000915_transcript.gff
$ ls ANNOgesic/output/transcriptome_assembly/statistics
stat_compare_Transcriptome_assembly_CDS_NC_000915.csv  stat_compare_Transcriptome_assembly_TSS_NC_000915.csv
</pre></div>
</div>
</div>
<div class="section" id="prediction-of-terminator">
<h2>Prediction of terminator<a class="headerlink" href="#prediction-of-terminator" title="Permalink to this headline">¶</a></h2>
<p>For predicting terminator, we can use subcommand <code class="docutils literal"><span class="pre">terminator</span></code>. The command is like following.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic terminator \
    -f $ANNOGESIC_FOLDER/output/target/fasta \
    -g $ANNOGESIC_FOLDER/output/target/annotation \
    -s \
    -tw $ANNOGESIC_FOLDER/input/wigs/tex_notex \
    -a $ANNOGESIC_FOLDER/output/transcriptome_assembly/gffs \
    -tl $tex_notex_libs \
    -te 2 -rt 1 -tb \
    ANNOgesic
</pre></div>
</div>
<p>It will generate three different kinds of gff files and tables.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/terminator/gffs/
all_candidates  detect  express
$ ls ANNOgesic/output/terminator/tables
all_candidates  detect  express
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">all_candidates</span></code> is for all candidates; <code class="docutils literal"><span class="pre">express</span></code> is for the candidates which have expression;
<code class="docutils literal"><span class="pre">detect</span></code> is for the candidates which have dramatic decreasing coverage. There is a gff file or table for
each folder.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/terminator/gffs/detect
NC_000915_term.gff
$ ls ANNOgesic/output/terminator/gffs/express
NC_000915_term.gff
$ ls ANNOgesic/output/terminator/gffs/all_candidates
NC_000915_term.gff
$ ls ANNOgesic/output/terminator/tables/detect
NC_000915_term.csv
$ ls ANNOgesic/output/terminator/tables/express
NC_000915_term.csv
$ ls ANNOgesic/output/terminator/tables/all_candidates
NC_000915_term.csv
</pre></div>
</div>
<p>In transtermhp folder, there are the output files from <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/terminator/transtermhp/NC_000915.1
NC_000915.1_best_terminator_after_gene.bag  NC_000915.1_terminators.txt  NC_000915.1_terminators_within_robust_tail-to-tail_regions.t2t
</pre></div>
</div>
<p>Moreover, the statistics files are stored in <code class="docutils literal"><span class="pre">statistics</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/terminator/statistics/
stat_NC_000915.csv
</pre></div>
</div>
</div>
<div class="section" id="generating-utr">
<h2>Generating UTR<a class="headerlink" href="#generating-utr" title="Permalink to this headline">¶</a></h2>
<p>Now, we have the information of TSS, transcripts and terminators. We can detect the 5&#8217;UTR and 3&#8217;UTR easily by using
the subcommand <code class="docutils literal"><span class="pre">utr</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic utr \
    -g ANNOgesic/output/target/annotation \
    -t ANNOgesic/output/TSS/gffs \
    -a ANNOgesic/output/transcriptome_assembly/gffs \
    -e ANNOgesic/output/terminator/gffs/detect \
    ANNOgesic
</pre></div>
</div>
<p>If your TSS is not generated by ANNOgesic, please assign <code class="docutils literal"><span class="pre">-s</span></code>, it will classify the TSS to generate UTR.
The gff files and statistics files will be stored in <code class="docutils literal"><span class="pre">5UTR</span></code> and <code class="docutils literal"><span class="pre">3UTR</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/UTR/3UTR
gffs/       statistics/
$ ls ANNOgesic/output/UTR/5UTR
gffs/       statistics/
$ ls ANNOgesic/output/UTR/3UTR/gffs
NC_000915_3UTR.gff
$ ls ANNOgesic/output/UTR/5UTR/gffs
NC_000915_5UTR.gff
$ ls ANNOgesic/output/UTR/5UTR/statistics
NC_000915_all_5utr_length.png
$ ls ANNOgesic/output/UTR/3UTR/statistics
NC_000915_all_3utr_length.png
</pre></div>
</div>
<p>Until now, you have all information for defining the transcript boundary.</p>
</div>
<div class="section" id="integrating-to-operon-and-suboperon">
<h2>Integrating to operon and suboperon<a class="headerlink" href="#integrating-to-operon-and-suboperon" title="Permalink to this headline">¶</a></h2>
<p>Now, we already had TSS, transcript, terminator, CDS, UTR. We can integrate all these information to
detect operons and suboperons. You can use the subcommand <code class="docutils literal"><span class="pre">operon</span></code> to get it.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic operon \
    -g ANNOgesic/output/target/annotation \
    -t ANNOgesic/output/TSS/gffs \
    -a ANNOgesic/output/transcriptome_assembly/gffs \
    -u5 ANNOgesic/output/UTR/5UTR/gffs \
    -u3 ANNOgesic/output/UTR/3UTR/gffs \
    -e ANNOgesic/output/terminator/gffs/detect \
    -s -c \
    ANNOgesic
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">operon</span></code> will generate three folders to store gff files, tables and statistics files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/operons/
gffs  statistics  tables
$ ls ANNOgesic/output/operons/gffs/
NC_000915_all_features.gff
$ ls ANNOgesic/output/operons/tables/
operon_NC_000915.csv
$ ls ANNOgesic/output/operons/statistics/
stat_operon_NC_000915.csv
</pre></div>
</div>
</div>
<div class="section" id="prediction-of-srna-and-sorf">
<h2>Prediction of sRNA and sORF<a class="headerlink" href="#prediction-of-srna-and-sorf" title="Permalink to this headline">¶</a></h2>
<p>Based on comparison of trascripts and CDS information, we can detect intergenic sRNA. Moreover, we
have the information of TSS and processing sites. We can detect UTR derived sRNA, too. You can
get sRNA by running subcommand <code class="docutils literal"><span class="pre">srna</span></code>. Normally, we would recommand you use fragmented libraries to
compute <code class="docutils literal"><span class="pre">srna</span></code>. However, we can&#8217;t get it right now. We use TEX +/- for this test case.</p>
<p>Before ŕunning <code class="docutils literal"><span class="pre">srna</span></code>, we have to get <a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a> and
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/">nr database</a>. Please download it and store in <code class="docutils literal"><span class="pre">ANNOgesic/input/database</span></code>.
If you want to store to other folders, please change <code class="docutils literal"><span class="pre">-sd</span></code> and <code class="docutils literal"><span class="pre">-od</span></code>. Then running the following
command.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic srna \
    -d 1 2 3 4 \
    -g ANNOgesic/output/target/annotation \
    -t ANNOgesic/output/TSS/gffs \
    -p ANNOgesic/output/processing_site/gffs \
    -a ANNOgesic/output/transcriptome_assembly/gffs \
    -tw ANNOgesic/input/wigs/tex_notex \
    -f ANNOgesic/output/target/fasta \
    -m -u -fd \
    -sd ANNOgesic/input/database/sRNA_database \
    -nd ANNOgesic/input/database/nr \
    -tl $tex_notex_libs \
    -te 2 \
    -rt 1 \
    -ba -sb \
    ANNOgesic
</pre></div>
</div>
<p>Because we have no information of sORF right now, we can&#8217;t compare sRNA and sORF. If you already have
the information of sORF, you can assign <code class="docutils literal"><span class="pre">-d</span> <span class="pre">1</span> <span class="pre">2</span> <span class="pre">3</span> <span class="pre">4</span> <span class="pre">5</span></code> and <code class="docutils literal"><span class="pre">-O</span></code> for the path of sORF. It will compare
sORF ans sRNA.</p>
<p>The output of <code class="docutils literal"><span class="pre">srna</span></code> will be the following.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/sRNA/
blast_result_and_misc  gffs  log.txt  mountain_plot  sec_structure  sRNA_2d_NC_000915  sRNA_seq_NC_000915  statistics  tables
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">blast_result_and_misc</span></code> will store the results of blast; <code class="docutils literal"><span class="pre">mountain_plos</span></code> will store the mountain plots;
<code class="docutils literal"><span class="pre">sec_structure</span></code> will store the plots of secondary structure of sRNA; <code class="docutils literal"><span class="pre">statistics</span></code> will store statistics files.</p>
<p><code class="docutils literal"><span class="pre">sRNA_2d_NC_000915</span></code> and <code class="docutils literal"><span class="pre">sRNA_seq_NC_000915</span></code> are text file of sequence of sRNA and secondary structure of sRNA.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/sRNA/blast_result_and_misc/
nr_blast_NC_000915.txt  sRNA_blast_NC_000915.txt
$ ls ANNOgesic/output/sRNA/mountain_plot/NC_000915/
srna0_NC_000915.1_16651_16765_-_mountain.pdf        srna138_NC_000915.1_1496938_1497216_-_mountain.pdf  srna25_NC_000915.1_444979_445114_+_mountain.pdf  srna63_NC_000915.1_761094_761174_+_mountain.pdf
srna100_NC_000915.1_1164252_1164295_-_mountain.pdf  srna139_NC_000915.1_1502542_1502637_+_mountain.pdf  srna26_NC_000915.1_445012_445139_-_mountain.pdf  srna64_NC_000915.1_773130_773564_+_mountain.pdf
...
ls ANNOgesic/output/sRNA/sec_structure/dot_plot/NC_000915/
srna0_NC_000915.1_16651_16765_-_dp.pdf        srna138_NC_000915.1_1496938_1497216_-_dp.pdf  srna25_NC_000915.1_444979_445114_+_dp.pdf  srna63_NC_000915.1_761094_761174_+_dp.pdf
srna100_NC_000915.1_1164252_1164295_-_dp.pdf  srna139_NC_000915.1_1502542_1502637_+_dp.pdf  srna26_NC_000915.1_445012_445139_-_dp.pdf  srna64_NC_000915.1_773130_773564_+_dp.pdf
...
$ ls ANNOgesic/output/sRNA/sec_structure/sec_plot/NC_000915/
srna0_NC_000915.1_16651_16765_-_rss.pdf        srna138_NC_000915.1_1496938_1497216_-_rss.pdf  srna25_NC_000915.1_444979_445114_+_rss.pdf  srna63_NC_000915.1_761094_761174_+_rss.pdf
srna100_NC_000915.1_1164252_1164295_-_rss.pdf  srna139_NC_000915.1_1502542_1502637_+_rss.pdf  srna26_NC_000915.1_445012_445139_-_rss.pdf  srna64_NC_000915.1_773130_773564_+_rss.pdf
...
$ ls ANNOgesic/output/sRNA/statistics/
stat_sRNA_blast_class_NC_000915.csv  stat_sRNA_class_NC_000915.csv
</pre></div>
</div>
<p>For <code class="docutils literal"><span class="pre">gffs</span></code> and <code class="docutils literal"><span class="pre">tables</span></code>, they are divided by three kinds of results. <code class="docutils literal"><span class="pre">all_candidates</span></code> is for all candidates
without filtering; <code class="docutils literal"><span class="pre">best</span></code> is for the best candidates of sRNA; <code class="docutils literal"><span class="pre">for_class</span></code> is for each class of sRNA which you
imported before. For our test case, we import TSS(class 1), folding energy(class 2), blast to nr(class 3),
blast to sRNA database(class 4 for without homologs, class 5 for with homologs).</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/sRNA/gffs/
all_candidates  best  for_class
$ ls ANNOgesic/output/sRNA/tables/
all_candidates  best  for_class
$ ls ANNOgesic/output/sRNA/gffs/all_candidates/
NC_000915_sRNA.gff
$ ls ANNOgesic/output/sRNA/tables/all_candidates/
NC_000915_sRNA.csv
$ ls ANNOgesic/output/sRNA/gffs/best/
NC_000915_sRNA.gff
$ ls ANNOgesic/output/sRNA/tables/best/
NC_000915_sRNA.csv
$ ls ANNOgesic/output/sRNA/gffs/for_class/NC_000915/
class_1_all.gff                          class_1_class_2_class_3_class_5_all.gff  class_1_class_3_class_4_all.gff  class_2_all.gff                  class_2_class_4_all.gff  class_3_class_5_all.gff
class_1_class_2_all.gff                  class_1_class_2_class_4_all.gff          class_1_class_3_class_5_all.gff  class_2_class_3_all.gff          class_2_class_5_all.gff  class_4_all.gff
class_1_class_2_class_3_all.gff          class_1_class_2_class_5_all.gff          class_1_class_4_all.gff          class_2_class_3_class_4_all.gff  class_3_all.gff          class_5_all.gff
class_1_class_2_class_3_class_4_all.gff  class_1_class_3_all.gff                  class_1_class_5_all.gff          class_2_class_3_class_5_all.gff  class_3_class_4_all.gff
$ ls ANNOgesic/output/sRNA/tables/for_class/NC_000915/
class_1_all.csv                          class_1_class_2_class_3_class_5_all.csv  class_1_class_3_class_4_all.csv  class_2_all.csv                  class_2_class_4_all.csv  class_3_class_5_all.csv
class_1_class_2_all.csv                  class_1_class_2_class_4_all.csv          class_1_class_3_class_5_all.csv  class_2_class_3_all.csv          class_2_class_5_all.csv  class_4_all.csv
class_1_class_2_class_3_all.csv          class_1_class_2_class_5_all.csv          class_1_class_4_all.csv          class_2_class_3_class_4_all.csv  class_3_all.csv          class_5_all.csv
class_1_class_2_class_3_class_4_all.csv  class_1_class_3_all.csv                  class_1_class_5_all.csv          class_2_class_3_class_5_all.csv  class_3_class_4_all.csv
</pre></div>
</div>
<p>As we know, the sORF also has no annotation information and has coverage. Therefore, the potential sRNA
may be sORF not sRNA. Therefore, it is a good idea to have the information of sORF. You can use subcommand
<code class="docutils literal"><span class="pre">sorf</span></code> to get it.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic sorf \
    -g ANNOgesic/output/target/annotation \
    -t ANNOgesic/output/TSS/gffs \
    -a ANNOgesic/output/transcriptome_assembly/gffs \
    -tw ANNOgesic/input/wigs/tex_notex \
    -f ANNOgesic/output/target/fasta \
    -s ANNOgesic/output/sRNA/gffs/best \
    -tl $tex_notex_libs \
    -te 2 -rt 1 -u \
    ANNOgesic
</pre></div>
</div>
<p>Then you can get the gff files, statistics files and tables. Based on the information of TSS and
sRNA, the gff files and tables will be divided by <code class="docutils literal"><span class="pre">all_candidates</span></code> and <code class="docutils literal"><span class="pre">best</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/sORF/gffs/all_candidates/
NC_000915_sORF.gff
$ ls ANNOgesic/output/sORF/gffs/best/
NC_000915_sORF.gff
$ ls ANNOgesic/output/sORF/tables/all_candidates/
NC_000915_sORF.csv
$ ls ANNOgesic/output/sORF/tables/best/
NC_000915_sORF.csv
$ ls ANNOgesic/output/sORF/statistics/
stat_NC_000915_sORF.csv
</pre></div>
</div>
</div>
<div class="section" id="performing-srna-target-prediction">
<h2>Performing sRNA target prediction<a class="headerlink" href="#performing-srna-target-prediction" title="Permalink to this headline">¶</a></h2>
<p>Now we have the sRNA candidates. If you want to know the targets of these sRNA, you can use <code class="docutils literal"><span class="pre">srna_target</span></code>
to get these information.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic srna_target \
    -g ANNOgesic/output/target/annotation \
    -f ANNOgesic/output/target/fasta \
    -r ANNOgesic/output/sRNA/gffs/best \
    -q NC_000915.1:+:13666:13701 \
    -p both \
    ANNOgesic
</pre></div>
</div>
<p>For testing, we just did the prediction to one sRNA. If you want to compute all sRNA, you can assign <code class="docutils literal"><span class="pre">all</span></code>
to <code class="docutils literal"><span class="pre">-q</span></code>. However, it may take several days.</p>
<p><code class="docutils literal"><span class="pre">srna_target</span></code> will generate several folders.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/sRNA_targets/
merge  RNAplex  RNAup  sRNA_seqs  target_seqs
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">sRNA_seqs</span></code> and <code class="docutils literal"><span class="pre">target_seqssRNA_seqs</span></code> are for the sequences of sRNA and potential targets.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/sRNA_targets/sRNA_seqs
NC_000915.1_sRNA.fa
$ ls ANNOgesic/output/sRNA_targets/target_seqs
NC_000915.1_target.fa
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">RNAplex</span></code> and <code class="docutils literal"><span class="pre">RNAup</span></code> are for the output of <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">RNAplex and RNAup</a>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/sRNA_targets/RNAplex/NC_000915.1/
NC_000915.1_RNAplex_rank.csv  NC_000915.1_RNAplex.txt
$ ls ANNOgesic/output/sRNA_targets/RNAup/NC_000915.1/
NC_000915.1_RNAup.log  NC_000915.1_RNAup_rank.csv  NC_000915.1_RNAup.txt
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">merge</span></code> is for the merged results of <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">RNAplex and RNAup</a>.</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ ls ANNOgesic/output/sRNA_targets/merge/NC_000915.1/
NC_000915.1_merge.csv  NC_000915.1_overlap.csv</dd>
</dl>
</div>
<div class="section" id="promoter-motif-detection">
<h2>Promoter motif detection<a class="headerlink" href="#promoter-motif-detection" title="Permalink to this headline">¶</a></h2>
<p>As long as you have TSS, you can use the subcommand <code class="docutils literal"><span class="pre">promoter</span></code> to get promoter. It will based on
the class of TSS to generate the promoters. Therefore, if your TSS data is not computed by <code class="docutils literal"><span class="pre">ANNOgesic</span></code>,
you need to assign <code class="docutils literal"><span class="pre">-s</span></code> and <code class="docutils literal"><span class="pre">-g</span></code> (for annotation files). Then <code class="docutils literal"><span class="pre">promoter</span></code> will help you
to classify your TSS data.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic promoter \
    -t ANNOgesic/output/TSS/gffs \
    -f ANNOgesic/output/target/fasta \
    -w 50 2-10 \
    -p 10 \
    ANNOgesic
</pre></div>
</div>
<p>You can define the length of motif. In our test case, we use <code class="docutils literal"><span class="pre">50</span></code> and <code class="docutils literal"><span class="pre">2-10</span></code>. <code class="docutils literal"><span class="pre">2-10</span></code> means the
width is from 2 to 10.</p>
<p>Based on the class of TSS, it will generate different output files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/promoter_analysis/NC_000915
promoter_motifs_NC_000915_allstrain_all_types_2-10_nt  promoter_motifs_NC_000915_allstrain_internal_50_nt   promoter_motifs_NC_000915_allstrain_secondary_2-10_nt
promoter_motifs_NC_000915_allstrain_all_types_50_nt    promoter_motifs_NC_000915_allstrain_orphan_2-10_nt   promoter_motifs_NC_000915_allstrain_secondary_50_nt
promoter_motifs_NC_000915_allstrain_antisense_2-10_nt  promoter_motifs_NC_000915_allstrain_orphan_50_nt     promoter_motifs_NC_000915_allstrain_without_orphan_2-10_nt
promoter_motifs_NC_000915_allstrain_antisense_50_nt    promoter_motifs_NC_000915_allstrain_primary_2-10_nt  promoter_motifs_NC_000915_allstrain_without_orphan_50_nt
promoter_motifs_NC_000915_allstrain_internal_2-10_nt   promoter_motifs_NC_000915_allstrain_primary_50_nt
</pre></div>
</div>
</div>
<div class="section" id="mapping-and-detecting-of-circular-rna">
<h2>Mapping and detecting of circular RNA<a class="headerlink" href="#mapping-and-detecting-of-circular-rna" title="Permalink to this headline">¶</a></h2>
<p>You may also be interested in circular RNA. The subcommand <code class="docutils literal"><span class="pre">circrna</span></code> can help you to get the information.
it apply <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> to detect circular RNA. Because
we didn&#8217;t map the reads of test case before, we can also do it by running <code class="docutils literal"><span class="pre">circrna</span></code>. However, if
you already mapped the reads by other tools. Or you mapped read by
<a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> without <code class="docutils literal"><span class="pre">-S</span></code>. You may need re-mapping again.
If your mapping is generated by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> with <code class="docutils literal"><span class="pre">-S</span></code>,
then you can skip <code class="docutils literal"><span class="pre">-a</span></code> and assign the path of bam files to <code class="docutils literal"><span class="pre">-nb</span></code> or <code class="docutils literal"><span class="pre">-fb</span></code>. It can reduce the
running time.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic circrna \
   -f ANNOgesic/output/target/fasta \
   -p 10 \
   -g ANNOgesic/output/target/annotation \
   -cg -a \
   ANNOgesic
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">circrna</span></code> will generate several folders.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/circRNA/
circRNA_tables  gffs  segemehl_align  segemehl_splice  statistics
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">segemehl_align</span></code> and <code class="docutils literal"><span class="pre">segemehl_splice</span></code> are for the results of
<a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>. <code class="docutils literal"><span class="pre">segemehl_align</span></code> stores the bam files of
alignment and <code class="docutils literal"><span class="pre">segemehl_splice</span></code> stores the results of splice detection.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/circRNA/segemehl_align/NC_000915.1/
SRR1951997_NC_000915.1.bam  SRR1951998_NC_000915.1.bam
$ ls ANNOgesic/output/circRNA/segemehl_splice/NC_000915/
splicesites_all.bed  transrealigned_all.bed
</pre></div>
</div>
<p>The gff files, tables and statistics files are stored in the other folders.</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ ls ANNOgesic/output/circRNA/gffs/NC_000915/
NC_000915_circRNA_all.gff  NC_000915_circRNA_best.gff
$ ls ANNOgesic/output/circRNA/circRNA_tables/NC_000915/
circRNA_NC_000915.txt
$ ls ANNOgesic/output/circRNA/statistics/
stat_circRNA_NC_000915.csv</dd>
</dl>
<p><code class="docutils literal"><span class="pre">NC_000915_circRNA_all.gff</span></code> is for all circular RNA candidates and <code class="docutils literal"><span class="pre">NC_000915_circRNA_best.gff</span></code> is the
circular RNA after filering by mapping ratio and comparison of CDS.</p>
</div>
<div class="section" id="snp-calling">
<h2>SNP calling<a class="headerlink" href="#snp-calling" title="Permalink to this headline">¶</a></h2>
<p>If you want to know the SNP or mutation of your sRNA-seq data, you can use <code class="docutils literal"><span class="pre">snp</span></code> to get it.
<code class="docutils literal"><span class="pre">snp</span></code> is divided by two part. One part compares with the &#8220;reference strain&#8221; which is the
closed strain of our strain(&#8220;target strain&#8221;). You can refer to the section of <code class="docutils literal"><span class="pre">Retrieving</span> <span class="pre">the</span> <span class="pre">input</span> <span class="pre">data</span></code>.
Because you may not have time to check the mutation between &#8220;reference strain&#8221; and &#8220;target strain&#8221;,
it is a good way to detect the mutations automatically. You just need to put your alignment files of
mapping with &#8220;reference strain&#8221; into correct path. It will generate the potential sequence.
The other part is for detecting the mutations of the reads and &#8220;target strain&#8221;. In this part, you
can know the real mutations in &#8220;target strain&#8221;. Therefore, you need to put the alignment files with
&#8220;target strain&#8221; to the correct folder.</p>
<p>Before running the subcommand, we must have the bam files. Because we already generated them through
running <code class="docutils literal"><span class="pre">circrna</span></code>, we can just use them. However, please remember that the mapping function of
<code class="docutils literal"><span class="pre">circrna</span></code> is basic one. If you have other request, please do mapping by yourself.</p>
<p>For testing, we only discover the mutation for &#8220;target strain&#8221; because our mapping is based on the
fasta of &#8220;target strain&#8221; - NC_000915. Therefore, we copy the bam files to <code class="docutils literal"><span class="pre">BAMs_map_target</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>cp ANNOgesic/output/circRNA/segemehl_align/NC_000915.1/SRR195199* ANNOgesic/input/BAMs/BAMs_map_target/tex_notex
</pre></div>
</div>
<p>Then we can run the subcommand.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic snp \
    -t target \
    -p 1 2 3 \
    -tw ANNOgesic/input/BAMs/BAMs_map_target/tex_notex \
    -f ANNOgesic/output/target/fasta \
    ANNOgesic
</pre></div>
</div>
<p>If you want to compute for comparison of &#8220;reference strain&#8221; and &#8220;target strain&#8221;, you just need to change
<code class="docutils literal"><span class="pre">-t</span></code> to reference and assign the correct bam files.</p>
<p><code class="docutils literal"><span class="pre">snp</span></code> will generate two folders, <code class="docutils literal"><span class="pre">compare_reference</span></code> is for comparison of &#8220;reference strain&#8221; and
&#8220;target strain&#8221;. <code class="docutils literal"><span class="pre">validate_target</span></code> is for real mutations of &#8220;target strain&#8221;</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/SNP_calling/
compare_reference  validate_target
</pre></div>
</div>
<p>Becaues we run <code class="docutils literal"><span class="pre">validate_target</span></code>, you can see there are several folders under <code class="docutils literal"><span class="pre">validate_target</span></code>.</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ ls ANNOgesic/output/SNP_calling/validate_target/
seqs/            SNP_raw_outputs/ SNP_table/       statistics/</dd>
</dl>
<p>All folders are divided by three parts - <code class="docutils literal"><span class="pre">extend_BAQ</span></code>, <code class="docutils literal"><span class="pre">extend_BAQ</span></code> and <code class="docutils literal"><span class="pre">extend_BAQ</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/SNP_calling/validate_target/seqs/
extend_BAQ/  with_BAQ/    without_BAQ/
</pre></div>
</div>
<p>In <code class="docutils literal"><span class="pre">seqs</span></code>, there are the potential sequences.</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ ls ANNOgesic/output/SNP_calling/validate_target/seqs/with_BAQ/NC_000915.1/
NC_000915.1_NC_000915.1_1_100.fa  NC_000915.1_NC_000915.1_1_13.fa   NC_000915.1_NC_000915.1_1_179.fa  NC_000915.1_NC_000915.1_1_217.fa  NC_000915.1_NC_000915.1_1_256.fa  NC_000915.1_NC_000915.1_1_63.fa
NC_000915.1_NC_000915.1_1_101.fa  NC_000915.1_NC_000915.1_1_140.fa  NC_000915.1_NC_000915.1_1_17.fa   NC_000915.1_NC_000915.1_1_218.fa  NC_000915.1_NC_000915.1_1_25.fa   NC_000915.1_NC_000915.1_1_64.fa
....</dd>
</dl>
<p><code class="docutils literal"><span class="pre">SNP_raw_outputs</span></code> stores the output of <a href="#id11"><span class="problematic" id="id12">`Samtools and Bcftools&lt;https://github.com/samtools&gt;`_</span></a>.
<code class="docutils literal"><span class="pre">SNP_table</span></code> stores the results after filtering and the index of potential sequence.
<code class="docutils literal"><span class="pre">statistics</span></code> stores the statistics and visualization files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/SNP_calling/validate_target/SNP_raw_outputs/NC_000915.1/
NC_000915.1_extend_BAQ.vcf  NC_000915.1_with_BAQ.vcf  NC_000915.1_without_BAQ.vcf
$ ls ANNOgesic/output/SNP_calling/validate_target/SNP_table/NC_000915.1/
NC_000915.1_extend_BAQ_depth_only.vcf     NC_000915.1_with_BAQ_depth_only.vcf     NC_000915.1_without_BAQ_depth_only.vcf
NC_000915.1_extend_BAQ_depth_quality.vcf  NC_000915.1_with_BAQ_depth_quality.vcf  NC_000915.1_without_BAQ_depth_quality.vcf
NC_000915.1_extend_BAQ_seq_reference.csv  NC_000915.1_with_BAQ_seq_reference.csv  NC_000915.1_without_BAQ_seq_reference.csv
$ ls ANNOgesic/output/SNP_calling/validate_target/statistics/
NC_000915.1_extend_BAQ_NC_000915.1_SNP_QUAL.png  NC_000915.1_without_BAQ_NC_000915.1_SNP_QUAL.png  stat_NC_000915.1_with_BAQ_SNP.csv
NC_000915.1_with_BAQ_NC_000915.1_SNP_QUAL.png    stat_NC_000915.1_extend_BAQ_SNP.csv               stat_NC_000915.1_without_BAQ_SNP.csv
</pre></div>
</div>
</div>
<div class="section" id="mapping-gene-ontology">
<h2>Mapping Gene ontology<a class="headerlink" href="#mapping-gene-ontology" title="Permalink to this headline">¶</a></h2>
<p>Gene ontology is useful for understanding the function of gene product. <code class="docutils literal"><span class="pre">go_term</span></code> is the
subcommand for mapping your annotation to gene ontology. Before running <code class="docutils literal"><span class="pre">go_term</span></code>, we
need to prepare some database. First, please download
<a class="reference external" href="http://geneontology.org/page/go-slim-and-subset-guide">goslim.obo</a> and
<a class="reference external" href="http://geneontology.org/page/download-ontology">go.obo</a> and
<a class="reference external" href="http://www.uniprot.org/downloads">idmapping_selected.tab</a>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ wget -cP ANNOgesic/input/database http://www.geneontology.org/ontology/subsets/goslim_generic.obo
$ wget -cP ANNOgesic/input/database http://geneontology.org/ontology/go.obo
$ wget -cP ANNOgesic/input/database ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping_selected.tab.gz
$ gunzip ANNOgesic/input/database/idmapping_selected.tab.gz
</pre></div>
</div>
<p>Now, we have all required database. Let&#8217;s try it.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic go_term \
    -g ANNOgesic/output/target/annotation\
    ANNOgesic
</pre></div>
</div>
<p>The output of <code class="docutils literal"><span class="pre">go_term</span></code> will store in <code class="docutils literal"><span class="pre">Go_term_results</span></code>. The statistics files and
visualization files will store in <code class="docutils literal"><span class="pre">statistics</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/Go_term/Go_term_results/NC_000915/
all_strains_uniprot.csv
$ ls ANNOgesic/output/Go_term/statistics/NC_000915/
figs  stat_NC_000915.csv
$ ls ANNOgesic/output/Go_term/statistics/NC_000915/figs/
NC_000915.1_biological_process.png  NC_000915.1_cellular_component.png  NC_000915.1_molecular_function.png  NC_000915.1_three_roots.png
</pre></div>
</div>
</div>
<div class="section" id="prediction-of-subcellular-localization">
<h2>Prediction of Subcellular localization<a class="headerlink" href="#prediction-of-subcellular-localization" title="Permalink to this headline">¶</a></h2>
<p>Subcellular localization may be a useful information for analysis of protein function. For
generating the information of subcellular localization, you can use the subcommand
<code class="docutils literal"><span class="pre">subcellular_localization</span></code> to get it.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic subcellular_localization \
    -g ANNOgesic/output/target/annotation \
    -f ANNOgesic/output/target/fasta \
    -m -b positive \
    ANNOgesic
</pre></div>
</div>
<p>The output of <code class="docutils literal"><span class="pre">subcellular_localization</span></code> will generate two folder. <code class="docutils literal"><span class="pre">psortb_results</span></code> will
store the output of <a class="reference external" href="http://www.psort.org/psortb/">Psortb</a>. <code class="docutils literal"><span class="pre">statistics</span></code> will store
the statistics files and visualization files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/subcellular_localization/
psortb_results  statistics
$ ls ANNOgesic/output/subcellular_localization/psortb_results/NC_000915/
NC_000915.1_raw.txt  NC_000915_table.csv
$ ls ANNOgesic/output/subcellular_localization/statistics/NC_000915/
NC_000915_NC_000915.1.png  stat_NC_000915_sublocal.csv
</pre></div>
</div>
</div>
<div class="section" id="generating-protein-protein-interaction-network">
<h2>Generating protein-protein interaction network<a class="headerlink" href="#generating-protein-protein-interaction-network" title="Permalink to this headline">¶</a></h2>
<p>Protein-protein interaction network is an important feature for analysis of regulation.
The subcommand <code class="docutils literal"><span class="pre">ppi_network</span></code> combines STRING(database of protein-protein interaction)
and PIE(text-mining for protein-protein interaction). It can generate the protein-protein
interaction network with supported literatures. You can refer to relevance of literatures
and network to find your interesting candidates.</p>
<p>Before running the subcommand, you need to download the
<a class="reference external" href="http://string-db.org/newstring_cgi/show_download_page.pl?UserId=ReWbu8uLrfAN&amp;sessionId=_FAQBbatf7RX">species.vXXXX.txt from STRING</a></p>
<div class="highlight-python"><div class="highlight"><pre>wget -cP ANNOgesic/input/database http://string-db.org/newstring_download/species.v10.txt
</pre></div>
</div>
<p>Now, we can try the subcommand.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic ppi_network \
    -s NC_000915.ptt:&#39;Helicobacter pylori 26695 chromosome&#39;:&#39;Helicobacter pylori 26695&#39;:&#39;Helicobacter pylori&#39; \
    -p $ANNOGESIC_FOLDER/output/target/annotation \
    -d $ANNOGESIC_FOLDER/input/database/species.v10.txt \
    -q &#39;Helicobacter pylori 26695 chromosome&#39;:217:633:- &#39;Helicobacter pylori 26695 chromosome&#39;:2719:3402:+ \
    -n \
    ANNOgesic
</pre></div>
</div>
<p>If you want to get all proteins in ptt files, you just need to assign <code class="docutils literal"><span class="pre">all</span></code> in <code class="docutils literal"><span class="pre">-q</span></code>.</p>
<p><code class="docutils literal"><span class="pre">ppi_network</span></code> will generate three folders.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/PPI/
all_results/  best_results/ figures/
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">all_results</span></code> is for all interactions without filtering. <code class="docutils literal"><span class="pre">best_results</span></code> is for the interactions with
filtering of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a> score. <code class="docutils literal"><span class="pre">figures</span></code> is for
the figures of PPI network. There are two subfolders - <code class="docutils literal"><span class="pre">with_strain</span></code> and <code class="docutils literal"><span class="pre">without_strain</span></code> in these folders.
These two folders stores all information of interactions and literature scores. <code class="docutils literal"><span class="pre">with_strain</span></code> is for
the information with giving specific strain name for searching literatures. <code class="docutils literal"><span class="pre">without_strain</span></code> is for the
information without giving specific strain name for searching literatures.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/PPI/all_results/NC_000915/
NC_000915_without_strain.csv  NC_000915_with_strain.csv     without_strain/               with_strain/
$ ls ANNOgesic/output/PPI/best_results/NC_000915/
NC_000915_without_strain.csv  NC_000915_with_strain.csv  without_strain  with_strain
$ ls ANNOgesic/output/PPI/figures/NC_000915/
without_strain  with_strain
$ ls ANNOgesic/output/PPI/all_results/NC_000915/with_strain/Helicobacter\ pylori\ 26695\ chromosome/
carA_C694_01345.csv  carA_guaB.csv  carB_C694_01345.csv  guaB_C694_01345.csv  pyrB_carB.csv  pyrD_C694_01345.csv  ribD_ribH.csv
carA_carB.csv        carA_pyrB.csv  carB_guaB.csv        nusG_rpoB.csv        pyrB_guaB.csv  pyrE_pyrF.csv        rpsJ_nusB.csv
$ ls ANNOgesic/output/PPI/all_results/NC_000915/without_strain/Helicobacter\ pylori\ 26695\ chromosome/
carA_C694_01345.csv  carA_guaB.csv  carB_C694_01345.csv  guaB_C694_01345.csv  pyrB_carB.csv  pyrD_C694_01345.csv  ribD_ribH.csv
carA_carB.csv        carA_pyrB.csv  carB_guaB.csv        nusG_rpoB.csv        pyrB_guaB.csv  pyrE_pyrF.csv        rpsJ_nusB.csv
$ ls ANNOgesic/output/PPI/best_results/NC_000915/without_strain/Helicobacter\ pylori\ 26695\ chromosome/
carA_C694_01345.csv  carA_carB.csv  carA_pyrB.csv  carB_C694_01345.csv  guaB_C694_01345.csv  pyrD_C694_01345.csv
$ ls ANNOgesic/output/PPI/best_results/NC_000915/with_strain/Helicobacter\ pylori\ 26695\ chromosome/
  (It can&#39;t find any interactions)
$ ls ANNOgesic/output/PPI/figures/NC_000915/with_strain/
Helicobacter pylori 26695 chromosome
$ ls ANNOgesic/output/PPI/figures/NC_000915/with_strain/Helicobacter\ pylori\ 26695\ chromosome/
  (It can&#39;t find any interactions. Therefore, it has no figures)
$ ls ANNOgesic/output/PPI/figures/NC_000915/without_strain/Helicobacter\ pylori\ 26695\ chromosome/
HP0001_nusB.png  HP0005_pyrF.png
</pre></div>
</div>
</div>
<div class="section" id="generating-riboswitch">
<h2>Generating riboswitch<a class="headerlink" href="#generating-riboswitch" title="Permalink to this headline">¶</a></h2>
<p>If you want to know the candidates of riboswitch, you can use the subcommand <code class="docutils literal"><span class="pre">riboswitch</span></code>.
Before running <code class="docutils literal"><span class="pre">riboswitch</span></code>, you need to get the information of all riboswitch in Rfam.
You can use ours or create your own one. For testing, you can just go
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic">here</a> and downloan or copy <code class="docutils literal"><span class="pre">Rfam_riboswitch_ID.csv</span></code>.
Then you can put <code class="docutils literal"><span class="pre">Rfam_riboswitch_ID.csv</span></code> in <code class="docutils literal"><span class="pre">ANNOgesic/input/riboswitch_ID</span></code></p>
<div class="highlight-python"><div class="highlight"><pre>$ cp /path/of/Rfam_riboswitch_ID.csv ANNOgesic/input/riboswitch_ID
</pre></div>
</div>
<p>You also need to download Rfam.</p>
<dl class="docutils">
<dt>::</dt>
<dd>$ wget ANNOgesic/input/database <a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/Rfam/12.0/Rfam.tar.gz">ftp://ftp.ebi.ac.uk/pub/databases/Rfam/12.0/Rfam.tar.gz</a>
$ cd ANNOgesic/input/database
$ tar -zxvf Rfam.tar.gz
$ rm Rfam.tar.gz
$ cd ../../../</dd>
</dl>
<p>Now we can try the subcommand.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic riboswitch \
    -g ANNOgesic/output/target/annotation \
    -f ANNOgesic/output/target/fasta \
    -r \
    -i ANNOgesic/input/riboswitch_ID/Rfam_riboswitch_ID.csv \
    -R ANNOgesic/input/database/CMs/Rfam.cm \
    ANNOgesic
</pre></div>
</div>
<p>The output is the following. <code class="docutils literal"><span class="pre">gffs</span></code> is for gff files; <code class="docutils literal"><span class="pre">tables</span></code> is for tables of riboswitch;
<code class="docutils literal"><span class="pre">scan_Rfam</span></code> is for the output of scanning Rfam; <code class="docutils literal"><span class="pre">statistics</span></code> is for the statistics files.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/riboswitch/
gffs  scan_Rfam  statistics  tables
$ ls ANNOgesic/output/riboswitch/gffs/
NC_000915_RBS.gff
$ ls ANNOgesic/output/riboswitch/scan_Rfam/NC_000915/
NC_000915.1_RBS_rescan.txt  NC_000915.1_RBS.txt
$ ls ANNOgesic/output/riboswitch/tables/
NC_000915_RBS.csv
$ ls ANNOgesic/output/riboswitch/statistics/
stat_NC_000915_RBS.txt
</pre></div>
</div>
</div>
<div class="section" id="producing-the-screenshots">
<h2>Producing the screenshots<a class="headerlink" href="#producing-the-screenshots" title="Permalink to this headline">¶</a></h2>
<p>It is a good idea if we can get the screenshots of our interesting features. Then we can
check them very quickly. Therefore, ANNOgesic provide a subcommand <code class="docutils literal"><span class="pre">screenshot</span></code> for
generating screenshots.</p>
<p>Before you running it, you have to install <a class="reference external" href="https://www.broadinstitute.org/software/igv/home">IGV</a>
for generating screenshot.</p>
<p>For testing, we use TSS as main feature, sRNA and CDS information as side features.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic screenshot \
    -mg ANNOgesic/output/TSS/gffs/NC_000915_TSS.gff \
    -sg ANNOgesic/output/target/annotation/NC_000915.gff \
        ANNOgesic/output/sRNA/gffs/best/NC_000915_sRNA.gff \
    -f ANNOgesic/output/target/fasta/NC_000915.fa \
    -o ANNOgesic/output/TSS/screenshots \
    -tl $tex_notex_libs \
    -tw ANNOgesic/input/wigs/tex_notex \
ANNOgesic
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">screenshot</span></code> will generate two txt files and two folders.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/TSS/screenshots/NC_000915/
forward/     forward.txt  reverse/     reverse.txt
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">forward.txt</span></code> and <code class="docutils literal"><span class="pre">reverse.txt</span></code> are batch files for <a class="reference external" href="https://www.broadinstitute.org/software/igv/home">IGV</a>.
<code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code> are the folders for storing screenshots.</p>
<p>Now, please open <a class="reference external" href="https://www.broadinstitute.org/software/igv/home">IGV</a>. Please follow the command: Tools -&gt;
Run Batch Script -&gt; choose <code class="docutils literal"><span class="pre">forward.txt</span></code>. When it has done, please do it again for reverse strand: Tools -&gt;
Run Batch Script -&gt; choose <code class="docutils literal"><span class="pre">reverse.txt</span></code>. If you just want to test it and don&#8217;t want to wait a long time for
generating screenshots, you can delete some lines of gff files of TSS.</p>
<p>After that, you can see that there are several screenshots in <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/TSS/screenshots/NC_000915/forward
NC_000915.1:1002230-1002230.png  NC_000915.1:1245705-1245705.png  NC_000915.1:1516949-1516949.png  NC_000915.1:246369-246369.png  NC_000915.1:463673-463673.png  NC_000915.1:741002-741002.png
NC_000915.1:1002524-1002524.png  NC_000915.1:124623-124623.png    NC_000915.1:151822-151822.png    NC_000915.1:251753-251753.png  NC_000915.1:463731-463731.png  NC_000915.1:744418-744418.png
NC_000915.1:1002728-1002728.png  NC_000915.1:1249488-1249488.png  NC_000915.1:1520156-1520156.png  NC_000915.1:255496-255496.png  NC_000915.1:464179-464179.png  NC_000915.1:744551-744551.png
...

$ ls ANNOgesic/output/TSS/screenshots/NC_000915/reverse
NC_000915.1:1002215-1002215.png  NC_000915.1:1235881-1235881.png  NC_000915.1:1481470-1481470.png  NC_000915.1:179609-179609.png  NC_000915.1:467716-467716.png  NC_000915.1:767765-767765.png
NC_000915.1:1002707-1002707.png  NC_000915.1:1238472-1238472.png  NC_000915.1:1482537-1482537.png  NC_000915.1:181416-181416.png  NC_000915.1:46780-46780.png    NC_000915.1:769891-769891.png
NC_000915.1:100498-100498.png    NC_000915.1:1240517-1240517.png  NC_000915.1:1482926-1482926.png  NC_000915.1:181781-181781.png  NC_000915.1:468289-468289.png  NC_000915.1:770316-770316.png
...
</pre></div>
</div>
</div>
<div class="section" id="coloring-the-screenshots">
<h2>Coloring the screenshots<a class="headerlink" href="#coloring-the-screenshots" title="Permalink to this headline">¶</a></h2>
<p>If your tracks are many and you want to check TSS, it will be painful for distinguish the
tracks of TEX+ and TEX-. Therefore, we provide a subcommand <code class="docutils literal"><span class="pre">color_png</span></code> for coloring
your screenshots.</p>
<div class="highlight-python"><div class="highlight"><pre>python3 annogesic color_png \
    -t 2 \
    -f ANNOgesic/output/TSS \
    ANNOgesic
</pre></div>
</div>
<p>You will see the png files are not different as before. However, when you open them, the tracks are colored.</p>
<div class="highlight-python"><div class="highlight"><pre>$ ls ANNOgesic/output/TSS/screenshots/NC_000915/forward
NC_000915.1:1002230-1002230.png  NC_000915.1:1245705-1245705.png  NC_000915.1:1516949-1516949.png  NC_000915.1:246369-246369.png  NC_000915.1:463673-463673.png  NC_000915.1:741002-741002.png
NC_000915.1:1002524-1002524.png  NC_000915.1:124623-124623.png    NC_000915.1:151822-151822.png    NC_000915.1:251753-251753.png  NC_000915.1:463731-463731.png  NC_000915.1:744418-744418.png
NC_000915.1:1002728-1002728.png  NC_000915.1:1249488-1249488.png  NC_000915.1:1520156-1520156.png  NC_000915.1:255496-255496.png  NC_000915.1:464179-464179.png  NC_000915.1:744551-744551.png
...

$ ls ANNOgesic/output/TSS/screenshots/NC_000915/reverse
NC_000915.1:1002215-1002215.png  NC_000915.1:1235881-1235881.png  NC_000915.1:1481470-1481470.png  NC_000915.1:179609-179609.png  NC_000915.1:467716-467716.png  NC_000915.1:767765-767765.png
NC_000915.1:1002707-1002707.png  NC_000915.1:1238472-1238472.png  NC_000915.1:1482537-1482537.png  NC_000915.1:181416-181416.png  NC_000915.1:46780-46780.png    NC_000915.1:769891-769891.png
NC_000915.1:100498-100498.png    NC_000915.1:1240517-1240517.png  NC_000915.1:1482926-1482926.png  NC_000915.1:181781-181781.png  NC_000915.1:468289-468289.png  NC_000915.1:770316-770316.png
...
</pre></div>
</div>
<p>Now you already finished your first wonderful trip of annotation. Hopefully, you enjoy it!!</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Performing an test cast</a><ul>
<li><a class="reference internal" href="#generating-a-project">Generating a project</a></li>
<li><a class="reference internal" href="#retrieving-the-input-data">Retrieving the input data</a></li>
<li><a class="reference internal" href="#putting-wig-bam-files-and-reads-to-proper-location">Putting wig, bam files and reads to proper location</a></li>
<li><a class="reference internal" href="#improving-the-reference-genome">Improving the reference genome</a></li>
<li><a class="reference internal" href="#generating-annotation-files">Generating annotation files</a></li>
<li><a class="reference internal" href="#gene-expression-analysis">Gene expression analysis</a></li>
<li><a class="reference internal" href="#tss-and-processing-site-prediction-and-optimization">TSS and processing site prediction and optimization</a></li>
<li><a class="reference internal" href="#performing-transcript-assembly">Performing transcript assembly</a></li>
<li><a class="reference internal" href="#prediction-of-terminator">Prediction of terminator</a></li>
<li><a class="reference internal" href="#generating-utr">Generating UTR</a></li>
<li><a class="reference internal" href="#integrating-to-operon-and-suboperon">Integrating to operon and suboperon</a></li>
<li><a class="reference internal" href="#prediction-of-srna-and-sorf">Prediction of sRNA and sORF</a></li>
<li><a class="reference internal" href="#performing-srna-target-prediction">Performing sRNA target prediction</a></li>
<li><a class="reference internal" href="#promoter-motif-detection">Promoter motif detection</a></li>
<li><a class="reference internal" href="#mapping-and-detecting-of-circular-rna">Mapping and detecting of circular RNA</a></li>
<li><a class="reference internal" href="#snp-calling">SNP calling</a></li>
<li><a class="reference internal" href="#mapping-gene-ontology">Mapping Gene ontology</a></li>
<li><a class="reference internal" href="#prediction-of-subcellular-localization">Prediction of Subcellular localization</a></li>
<li><a class="reference internal" href="#generating-protein-protein-interaction-network">Generating protein-protein interaction network</a></li>
<li><a class="reference internal" href="#generating-riboswitch">Generating riboswitch</a></li>
<li><a class="reference internal" href="#producing-the-screenshots">Producing the screenshots</a></li>
<li><a class="reference internal" href="#coloring-the-screenshots">Coloring the screenshots</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="subcommands.html"
                        title="previous chapter">ANNOgesic&#8217;s subcommands</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="troubleshooting.html"
                        title="next chapter">Troubleshooting</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/test_case.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, Sung-Huan Yu.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.4</a>
      
      |
      <a href="_sources/test_case.txt"
          rel="nofollow">Page source</a></li>
    </div>

    

    
  </body>
</html>