<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tutorial of ANNOgesic &mdash; ANNOgesic 0.5.6 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.5.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ANNOgesic 0.5.6 documentation" href="index.html" />
    <link rel="next" title="Troubleshooting" href="troubleshooting.html" />
    <link rel="prev" title="ANNOgesic’s subcommands" href="subcommands.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="troubleshooting.html" title="Troubleshooting"
             accesskey="N">next</a></li>
        <li class="right" >
          <a href="subcommands.html" title="ANNOgesic’s subcommands"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.5.6 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo_annogesic.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tutorial of ANNOgesic</a><ul>
<li><a class="reference internal" href="#generating-a-project">Generating a project</a></li>
<li><a class="reference internal" href="#retrieving-the-input-data">Retrieving the input data</a></li>
<li><a class="reference internal" href="#putting-wig-bam-files-and-reads-to-proper-location">Putting wig, bam files and reads to proper location</a></li>
<li><a class="reference internal" href="#improving-the-reference-genome">Improving the reference genome</a></li>
<li><a class="reference internal" href="#generating-annotation-files">Generating annotation files</a></li>
<li><a class="reference internal" href="#tss-and-processing-site-prediction-and-optimization">TSS and processing site prediction and optimization</a></li>
<li><a class="reference internal" href="#performing-transcript-detection">Performing transcript detection</a></li>
<li><a class="reference internal" href="#prediction-of-terminator">Prediction of terminator</a></li>
<li><a class="reference internal" href="#generating-utr">Generating UTR</a></li>
<li><a class="reference internal" href="#detecting-operon-and-suboperon">Detecting operon and suboperon</a></li>
<li><a class="reference internal" href="#promoter-motif-detection">Promoter motif detection</a></li>
<li><a class="reference internal" href="#prediction-of-srna-and-sorf">Prediction of sRNA and sORF</a></li>
<li><a class="reference internal" href="#performing-srna-target-prediction">Performing sRNA target prediction</a></li>
<li><a class="reference internal" href="#mapping-and-detecting-of-circular-rna">Mapping and detecting of circular RNA</a></li>
<li><a class="reference internal" href="#snp-calling">SNP calling</a></li>
<li><a class="reference internal" href="#mapping-gene-ontology">Mapping Gene ontology</a></li>
<li><a class="reference internal" href="#prediction-of-subcellular-localization">Prediction of Subcellular localization</a></li>
<li><a class="reference internal" href="#generating-protein-protein-interaction-network">Generating protein-protein interaction network</a></li>
<li><a class="reference internal" href="#generating-riboswitch-and-rna-thermometer">Generating riboswitch and RNA thermometer</a></li>
<li><a class="reference internal" href="#detection-of-crispr">Detection of CRISPR</a></li>
<li><a class="reference internal" href="#merge-all-features-to-be-one-gff-file">Merge all features to be one gff file</a></li>
<li><a class="reference internal" href="#producing-the-screenshots">Producing the screenshots</a></li>
<li><a class="reference internal" href="#coloring-the-screenshots">Coloring the screenshots</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="subcommands.html"
                        title="previous chapter">ANNOgesic&#8217;s subcommands</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="troubleshooting.html"
                        title="next chapter">Troubleshooting</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tutorial-of-annogesic">
<h1>Tutorial of ANNOgesic<a class="headerlink" href="#tutorial-of-annogesic" title="Permalink to this headline">¶</a></h1>
<p>Here we will guide you how to use ANNOgesic through a small test case.
You will see how to run each subcommand of ANNOgesic. The test case is a public
RNA-Seq data from NCBI GEO that was part of a work by
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE38883">Dugar et al.</a>.
The differential RNA-seq data is of <strong>Campylobacter jejuni</strong> subsp. jejuni 81116.
There will be several output files which are generated in different formats -
The CSV (tabular separated plain text files) files (opened by LibreOffice or Excel), GFF3 files, TXT files and figures.
For viewing GFF3 file, you can use a genome browser, for example <a class="reference external" href="http://bioviz.org/igb/index.html">IGB</a>,
<a class="reference external" href="https://www.broadinstitute.org/igv/">IGV</a> or <a class="reference external" href="http://jbrowse.org/">Jbrowse</a>.</p>
<p>Before we start, please check <a class="reference internal" href="subcommands.html#the-format-of-filename"><span class="std std-ref">The format of filename</span></a> and
<a class="reference internal" href="subcommands.html#the-input-format-of-libraries-for-running-annogesic"><span class="std std-ref">The input format of libraries for running ANNOgesic</span></a> in
the section <a class="reference internal" href="subcommands.html#annogesic-s-subcommands"><span class="std std-ref">ANNOgesic&#8217;s subcommands</span></a>. All the details are also in <a class="reference internal" href="subcommands.html#annogesic-s-subcommands"><span class="std std-ref">ANNOgesic&#8217;s subcommands</span></a>.
Moreover, all the requirements are listed in the section <a class="reference internal" href="prerequired.html#required-tools-or-databases"><span class="std std-ref">Required tools or databases</span></a>.
For command lines which we will present later, please check the
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/tutorial_data">run.sh</a> in our Git repository.</p>
<p>If the subcommand integrates third-party software, ex: TSSpredator,
please check path of the execute file. If necessary, please assign it properly. Moreover,
if your execute path is not in your <code class="docutils literal"><span class="pre">$PATH</span></code>, please specify your execute path of ANNOgesic as well.</p>
<div class="section" id="generating-a-project">
<h2>Generating a project<a class="headerlink" href="#generating-a-project" title="Permalink to this headline">¶</a></h2>
<p>First of all, we need to create a working folder by running <code class="docutils literal"><span class="pre">create</span></code>. <code class="docutils literal"><span class="pre">-pj</span></code> represents project path.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic create -pj ANNOgesic
</pre></div>
</div>
<p>Then we will see</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>Created folder &quot;ANNOgesic&quot; and required subfolders.
$ ls
ANNOgesic
</pre></div>
</div>
</div>
<div class="section" id="retrieving-the-input-data">
<h2>Retrieving the input data<a class="headerlink" href="#retrieving-the-input-data" title="Permalink to this headline">¶</a></h2>
<p>For our test case, the input data can be downloaded from
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Campylobacter_jejuni/latest_assembly_versions/GCF_000017905.1_ASM1790v1/">NCBI</a>.
Let&#8217;s setting the <code class="docutils literal"><span class="pre">$FTP_SOURCE</span></code> first</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">FTP_SOURCE</span><span class="o">=</span><span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span><span class="n">refseq</span><span class="o">/</span><span class="n">bacteria</span><span class="o">/</span><span class="n">Campylobacter_jejuni</span><span class="o">/</span><span class="n">latest_assembly_versions</span><span class="o">/</span><span class="n">GCF_000017905</span><span class="o">.</span><span class="mi">1</span><span class="n">_ASM1790v1</span><span class="o">/</span>
</pre></div>
</div>
<p>Then, we can download fasta files(<code class="docutils literal"><span class="pre">-f</span></code>), gff files(<code class="docutils literal"><span class="pre">-g</span></code>), gbk files(<code class="docutils literal"><span class="pre">-k</span></code>), ptt files(<code class="docutils literal"><span class="pre">-p</span></code>),
rnt files(<code class="docutils literal"><span class="pre">-r</span></code>), and converts to embl(<code class="docutils literal"><span class="pre">-e</span></code>).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic get_input_files \
    -F $FTP_SOURCE \
    -g -f -e -k -p -r \
    -pj ANNOgesic
</pre></div>
</div>
<p>Then we will get following results</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/input/reference/fasta/
NC_009839.1.fa
$ ls ANNOgesic/input/reference/annotation/
NC_009839.1.embl  NC_009839.1.gbk  NC_009839.1.gff
</pre></div>
</div>
<p>If the fasta files and annotation files from NCBI is exactly what you want,
you can add <code class="docutils literal"><span class="pre">-t</span></code> for putting the files to <code class="docutils literal"><span class="pre">ANNOgesic/output/target</span></code>. Then you can skip <code class="docutils literal"><span class="pre">get_target_fasta</span></code>
and <code class="docutils literal"><span class="pre">annotation_transfer</span></code>.</p>
<p>In fact, these fasta and gff files are exactly what we want to use for the test case.
But, in order to testing <code class="docutils literal"><span class="pre">get_target_fasta</span></code> and <code class="docutils literal"><span class="pre">annotation_transfer</span></code>, we used them as references first.
After testing these subcommands, we will reorganize the data again.</p>
</div>
<div class="section" id="putting-wig-bam-files-and-reads-to-proper-location">
<h2>Putting wig, bam files and reads to proper location<a class="headerlink" href="#putting-wig-bam-files-and-reads-to-proper-location" title="Permalink to this headline">¶</a></h2>
<p>For the test case, we can download reads from
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE38883">here</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP013/SRP013869/SRR515254/SRR515254.sra
$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP013/SRP013869/SRR515255/SRR515255.sra
$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP013/SRP013869/SRR515256/SRR515256.sra
$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP013/SRP013869/SRR515257/SRR515257.sra
</pre></div>
</div>
<p>Then we need to convert SRA files to Fasta or Fastq format for mapping. We can
use <a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK158900/">SRA toolkit</a> for that.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.5.2/sratoolkit.2.5.2-ubuntu64.tar.gz
$ tar -zxvf sratoolkit.2.5.2-ubuntu64.tar.gz
$ rm sratoolkit.2.5.2-ubuntu64.tar.gz
$ ./sratoolkit.2.5.2-ubuntu64/bin/fastq-dump.2.5.2 --fasta SRR515254.sra
$ ./sratoolkit.2.5.2-ubuntu64/bin/fastq-dump.2.5.2 --fasta SRR515255.sra
$ ./sratoolkit.2.5.2-ubuntu64/bin/fastq-dump.2.5.2 --fasta SRR515256.sra
$ ./sratoolkit.2.5.2-ubuntu64/bin/fastq-dump.2.5.2 --fasta SRR515257.sra
$ mv *.fasta ANNOgesic/input/reads
$ rm SRR515254.sra SRR515255.sra SRR515256.sra SRR515257.sra
</pre></div>
</div>
<p>Now we get the reads. Then we have to download the wiggle files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/wigs/tex_notex ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM951nnn/GSM951380/suppl/GSM951380%5FLog%5F81116%5FR1%5Fminus%5FTEX%5Fin%5FNC%5F009839%5Fminus.wig.gz
$ wget -cP ANNOgesic/input/wigs/tex_notex ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM951nnn/GSM951380/suppl/GSM951380%5FLog%5F81116%5FR1%5Fminus%5FTEX%5Fin%5FNC%5F009839%5Fplus.wig.gz
$ wget -cP ANNOgesic/input/wigs/tex_notex ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM951nnn/GSM951381/suppl/GSM951381%5FLog%5F81116%5FR1%5Fplus%5FTEX%5Fin%5FNC%5F009839%5Fminus.wig.gz
$ wget -cP ANNOgesic/input/wigs/tex_notex ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM951nnn/GSM951381/suppl/GSM951381%5FLog%5F81116%5FR1%5Fplus%5FTEX%5Fin%5FNC%5F009839%5Fplus.wig.gz
$ cd ANNOgesic/input/wigs/tex_notex
$ gunzip GSM951380_Log_81116_R1_minus_TEX_in_NC_009839_minus.wig.gz \
         GSM951380_Log_81116_R1_minus_TEX_in_NC_009839_plus.wig.gz \
         GSM951381_Log_81116_R1_plus_TEX_in_NC_009839_minus.wig.gz \
         GSM951381_Log_81116_R1_plus_TEX_in_NC_009839_plus.wig.gz
$ cd ../../../../
</pre></div>
</div>
<p>If we check the wiggle files, we will find that the strain name (presented by &#8220;chrom&#8221;) is not the same as fasta or annotation gff file.
Thus, we need to change it. We can use <a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/tutorial_data">replace_seq_id.py</a> from our
Git repository. to replace the strain name properly. If the strain names in your fasta, annotation,
wiggle files are the same, you can skip this step.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/tutorial_data/replace_seq_id.py
$ python3 replace_seq_id.py -i ANNOgesic/input/wigs/tex_notex -n NC_009839.1
$ rm replace_seq_id.py
</pre></div>
</div>
<p>We only download one replicate to reduce the running time.</p>
</div>
<div class="section" id="improving-the-reference-genome">
<h2>Improving the reference genome<a class="headerlink" href="#improving-the-reference-genome" title="Permalink to this headline">¶</a></h2>
<p>Again, if the data retrieved from NCBI is exactly what you want, you can skip this step and <code class="docutils literal"><span class="pre">annotation_transfer</span></code>.</p>
<p>Although the data that we downloaded before is our real data (<code class="docutils literal"><span class="pre">ANNOgesic/input/reference</span></code>),
we will generate some new fake files via this step and <code class="docutils literal"><span class="pre">annotation_transfer</span></code> in order to
show you the function of these subcommands. After testing these two subcommands,
we can re-organize the files again.</p>
<p>Now, we assume that we need to generate fasta file of our target strain.
First of all, we need to find a close strain (fasta file and gff file can be found) of our target strain.
Then, we need to generate a mutation table between these two strains. When these files are produced,
we can run subcommand <code class="docutils literal"><span class="pre">get_target_fasta</span></code> for getting fasta file of the target strain.
For mutation table format, please check the section <a class="reference internal" href="subcommands.html#annogesic-s-subcommands"><span class="std std-ref">ANNOgesic&#8217;s subcommands</span></a>.</p>
<p>We use a simple example to modify our test case, please check
<a class="reference external" href="https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/tutorial_data/mutation.csv">mutation table</a>.
Every column of the table is separated by tab. The new strain will be NC_test.1 and test_case2. Therefore, two fasta files
will be generated in <code class="docutils literal"><span class="pre">ANNOgesic/output/target/fasta</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/mutation_table https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/tutorial_data/mutation.csv
</pre></div>
</div>
<p>Now, let&#8217;s try it</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic get_target_fasta \
   -r ANNOgesic/input/reference/fasta/NC_009839.1.fa \
   -o ANNOgesic/output/target/fasta/test_case1.fa:NC_test.1 \
      ANNOgesic/output/target/fasta/test_case2.fa:test_case2 \
   -m ANNOgesic/input/mutation_table/mutation.csv \
   -pj ANNOgesic
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">-r</span></code> is folder of the close strain fasta file. In <code class="docutils literal"><span class="pre">-o</span></code>, we assign a pairs of output filenames and
the strains that we want to put into the file. In our case, &#8220;test_case1&#8221; is the first output fasta file, and &#8220;test_case2&#8221;
is the second output fasta file. &#8220;test_case1&#8221; stores the sequence of the new strain &#8220;NC_test.1&#8221;,
and &#8220;test_case2&#8221; stores the sequence of the other new strain - &#8220;test_case2&#8221;.</p>
<p>When the running process is done, the following information will appear.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ Transfering to target fasta
  Please use the new fasta file to remapping again.
</pre></div>
</div>
<p>Since the data (<code class="docutils literal"><span class="pre">ANNOgesic/output/target/fasta</span></code>) that we generated is not real,
we can ignore the information now. However, if the new fasta file is real target one,
you have to remap again in order to get the correct alignment and coverage files.</p>
<p>Now we can check the results.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ head ANNOgesic/input/reference/fasta/NC_009839.1.fa
&gt;NC_009839.1
ATGAATCCAAATCAAATACTTGAAAATTTAAAAAAAGAATTAAGTGAAAACGAATACGAAAATTATATCGCTATCTTAAA
ATTTAACGAAAAACAAAGCAAAGCAGATTTTCTAGTCTTTAACGCTCCTAATGAGCTTTTAGCCAAATTCATACAAACAA
AATACGGTAAAAAAATTTCACATTTTTATGAAGTACAAAGCGGAAATAAAGCGAGCGTTTTGATACAAGCACAAAGTGCT
AAACAAAGTAGCAAAAGCACTAAAATCGATATCGCTCATATCAAGGCGCAAAGTACGATTTTAAATCCTTCTTTTACTTT
TGAAAGCTTTGTAGTGGGGGATTCTAACAAATACGCTTATGGAGCTTGTAAAGCTATCTCACAAAAAGACAAACTGGGAA
AACTTTATAATCCTATCTTTATCTATGGGCCTACAGGGCTTGGAAAAACGCACTTGCTTCAAGCTGTGGGAAATGCAAGT
TTGGAAATGGGAAAAAAAGTGATTTATGCTACGAGTGAAAATTTTATCAATGATTTTACTTCAAATTTAAAAAATGGCTC
TTTAGATAAATTTCACGAAAAATATAGAAATTGTGATGTTTTACTCATAGATGATGTGCAGTTTTTAGGAAAAACCGATA
AAATTCAAGAAGAATTTTTCTTTATATTTAATGAAATCAAAAATAACGATGGACAAATCATCATGACTTCAGACAATCCA
$ head ANNOgesic/output/target/fasta/test_case1.fa
&gt;NC_test.1
ATcAACCAAATCAAATACTTGAAAATTTAAAAAAAGAATTAAGTGAAAACGAATACGAAA
ATTATATCGCTATCTTAAAATTTAACGAAAAACAAAGCAAAGCAGATTTTCTAGTCTTTA
ACGCTCCTAATGAGCTTTTAGCCAAATTCATACAAACAAAATACGGTAAAAAAATTTCAC
ATTTTTATGAAGTACAAAGCGGAAATAAAGCGAGCGTTTTGATACAAGCACAAAGTGCTA
AACAAAGTAGCAAAAGCACTAAAATCGATATCGCTCATATCAAGGCGCAAAGTACGATTT
TAAATCCTTCTTTTACTTTTGAAAGCTTTGTAGTGGGGGATTCTAACAAATACGCTTATG
GAGCTTGTAAAGCTATCTCACAAAAAGACAAACTGGGAAAACTTTATAATCCTATCTTTA
TCTATGGGCCTACAGGGCTTGGAAAAACGCACTTGCTTCAAGCTGTGGGAAATGCAAGTT
TGGAAATGGGAAAAAAAGTGATTTATGCTACGAGTGAAAATTTTATCAATGATTTTACTT
</pre></div>
</div>
<p>We can see that the sequence of &#8220;NC_tetst.1&#8221; is stored in <code class="docutils literal"><span class="pre">test_case1.fa</span></code>.
Moreover, the third nucleotide is replaced from G to c. Moreover, The sixth nucleotide T is deleted.
If we check <code class="docutils literal"><span class="pre">test_case2.fa</span></code>, the modification is also according to the mutation table and our setting.</p>
<p>If we have no mutation table, we can also use subcommand <code class="docutils literal"><span class="pre">snp</span></code> to detect mutations and generate
fasta files automatically. For <code class="docutils literal"><span class="pre">snp</span></code>, we will go through it later.</p>
</div>
<div class="section" id="generating-annotation-files">
<h2>Generating annotation files<a class="headerlink" href="#generating-annotation-files" title="Permalink to this headline">¶</a></h2>
<p>We have fasta files of our target strain now. We can use them to generate our annotation files. If the annotaion files
retrieved from NCBI is exactly what you want, you can skip this step.</p>
<p>Before we running this subcommand, we have to modify environment paths of <a class="reference external" href="http://ratt.sourceforge.net/">RATT</a>.
If you execute ANNOgesic by using docker container, the path is alread setted.
If you setup ANNOgesic by yourself, please check
<a class="reference external" href="http://ratt.sourceforge.net/">RATT</a> to set your environment paths properly.</p>
<p>After setting the environment, we can try it.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">anngesic</span> <span class="n">annotation_transfer</span> \
    <span class="o">-</span><span class="n">re</span> <span class="n">ANNOgesic</span><span class="o">/</span><span class="nb">input</span><span class="o">/</span><span class="n">reference</span><span class="o">/</span><span class="n">annotation</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mf">1.</span><span class="n">embl</span> \
    <span class="o">-</span><span class="n">rf</span> <span class="n">ANNOgesic</span><span class="o">/</span><span class="nb">input</span><span class="o">/</span><span class="n">reference</span><span class="o">/</span><span class="n">fasta</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mf">1.</span><span class="n">fa</span> \
    <span class="o">-</span><span class="n">tf</span> <span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">target</span><span class="o">/</span><span class="n">fasta</span><span class="o">/</span><span class="n">test_case1</span><span class="o">.</span><span class="n">fa</span> \
        <span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">target</span><span class="o">/</span><span class="n">fasta</span><span class="o">/</span><span class="n">test_case2</span><span class="o">.</span><span class="n">fa</span> \
    <span class="o">-</span><span class="n">e</span> <span class="n">chromosome</span> \
    <span class="o">-</span><span class="n">t</span> <span class="n">Strain</span> \
    <span class="o">-</span><span class="n">p</span> <span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="p">:</span><span class="n">NC_test</span><span class="o">.</span><span class="mi">1</span> <span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="p">:</span><span class="n">test_case2</span> \
    <span class="o">-</span><span class="n">g</span> \
    <span class="o">-</span><span class="n">pj</span> <span class="n">ANNOgesic</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">-e</span></code> is prefix name of the output embl files. <code class="docutils literal"><span class="pre">-t</span></code> is a program of <a class="reference external" href="http://ratt.sourceforge.net/">RATT</a>.
We use <code class="docutils literal"><span class="pre">Strain</span></code> because the similarity is higher than 90%. For other situations, please check
<a class="reference external" href="http://ratt.sourceforge.net/">RATT</a>. In <code class="docutils literal"><span class="pre">-p</span></code>, we assign pairs of the target strains (NC_test.1 and test_case2)
and their close strains (NC_000915.1). Please be careful, the information that we assign to <code class="docutils literal"><span class="pre">-p</span></code>
is strain names not fasta filenames. <code class="docutils literal"><span class="pre">-g</span></code> means that we want to transfer the
output embl files to GFF3 files and store in <code class="docutils literal"><span class="pre">ANNOgesic/output/target/annotation</span></code>.</p>
<p>Once the transfer is done, we can see</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/target/annotation/
test_case1.gff  test_case1.ptt  test_case1.rnt  test_case2.gff  test_case2.ptt  test_case2.rnt
$ ls ANNOgesic/output/annotation_transfer/
chromosome.NC_test.1.final.embl  chromosome.test_case2.final.embl  NC_test.1.gff  ratt_log.txt  test_case2.gff
</pre></div>
</div>
<p>In <code class="docutils literal"><span class="pre">ANNOgesic/output/target/annotation</span></code>, we can find ptt, rnt and gff files. In <code class="docutils literal"><span class="pre">ANNOgesic/output/annotation_transfer</span></code>,
we can find the output of <a class="reference external" href="http://ratt.sourceforge.net/">RATT</a>.</p>
<p>We already saw how to update genome fasta and annotation files.
Now, we can re-organize our data in order to run next subcommand.
(Normally, you don&#8217;t need to do it. We re-organize the data becuase the data is only for showing the how to
run <code class="docutils literal"><span class="pre">get_target_fasta</span></code> and <code class="docutils literal"><span class="pre">annotation_transfer</span></code>. All the data is useless now. The real test case is
already downloaded via <code class="docutils literal"><span class="pre">get_input_files</span></code>. Therefore, we need to re-organize the data.)</p>
<p>Since the data in <code class="docutils literal"><span class="pre">ANNOgesic/output/target</span></code> is fake data, we can remove it.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ rm ANNOgesic/output/target/annotation/*
$ rm ANNOgesic/output/target/fasta/*
</pre></div>
</div>
<p>The real data of our query strain is stored in <code class="docutils literal"><span class="pre">ANNOgesic/input/reference</span></code> before. Thus,
we need to move the data to <code class="docutils literal"><span class="pre">ANNOgesic/output/target</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ mv ANNOgesic/input/reference/annotation/* ANNOgesic/output/target/annotation/
$ mv ANNOgesic/input/reference/fasta/* ANNOgesic/output/target/fasta/
</pre></div>
</div>
<p>Now files are re-organized, we can run following subcommands.</p>
</div>
<div class="section" id="tss-and-processing-site-prediction-and-optimization">
<h2>TSS and processing site prediction and optimization<a class="headerlink" href="#tss-and-processing-site-prediction-and-optimization" title="Permalink to this headline">¶</a></h2>
<p>Before running following subcommands, we need to setup our libraries as a correct format.
First, we set the path of wig file folder.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">WIG_FOLDER</span><span class="o">=</span><span class="s2">&quot;ANNOgesic/input/wigs/tex_notex&quot;</span>
</pre></div>
</div>
<p>Then, we can setup our libraries.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">TEX_LIBS</span><span class="o">=</span><span class="s2">&quot;$WIG_FOLDER/GSM951380_Log_81116_R1_minus_TEX_in_NC_009839_minus.wig:notex:1:a:- </span><span class="se">\</span>
<span class="s2">          $WIG_FOLDER/GSM951381_Log_81116_R1_plus_TEX_in_NC_009839_minus.wig:tex:1:a:- </span><span class="se">\</span>
<span class="s2">          $WIG_FOLDER/GSM951380_Log_81116_R1_minus_TEX_in_NC_009839_plus.wig:notex:1:a:+ </span><span class="se">\</span>
<span class="s2">          $WIG_FOLDER/GSM951381_Log_81116_R1_plus_TEX_in_NC_009839_plus.wig:tex:1:a:+&quot;</span>
</pre></div>
</div>
<p>Now, we can start to test other subcommands.
Before running <code class="docutils literal"><span class="pre">tss_processing</span></code>, if we want to use the optimized parameters,
we need to run <code class="docutils literal"><span class="pre">optimize_tss_processing</span></code> first. The optimization requires a gff file of the manual-detected TSSs.
In our experience, we recommand you to detect at least 50 TSSs and check more than 200kb of genome.</p>
<p>For the test case, you can download the <a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/tutorial_data">manual TSS file</a>
from our git repository.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/manual_TSS/ https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/tutorial_data/NC_009839_manual_TSS.gff
</pre></div>
</div>
<p>Now, we have a manual TSS gff file which is stored in <code class="docutils literal"><span class="pre">ANNOgesic/input/manual_TSS</span></code>.
we can try <code class="docutils literal"><span class="pre">optimize_tss_processing</span></code> right now (since we only check first 200kb, we set <code class="docutils literal"><span class="pre">-le</span></code> as &#8220;NC_009839.1:200000&#8221; which
means only first 200kb of NC_009839.1 is valid.).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic optimize_tss_processing \
     -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
     -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
     -tl $TEX_LIBS \
     -p TSS -s 25 \
     -m ANNOgesic/input/manual_TSS/NC_009839_manual_TSS.gff \
     -le NC_009839.1:200000 \
     -rt all_1 \
     -pj ANNOgesic
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">optimize_tss_processing</span></code> will compare manual checked TSSs with predicted TSSs to search the best parameters.
Results of the different parameters will be printed in the screen. we only set 25 runs for testing.
Once the optimization is done, you can find several files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/TSS/optimized_TSSpredator/
best_NC_009839.1.csv  log.txt  stat_NC_009839.1.csv
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">best_NC_009839.1.csv</span></code> is for the best parameters; <code class="docutils literal"><span class="pre">stat_NC_009839.1.csv</span></code> is for parameters of each step.</p>
<p>Now, we assume the best parameters are following: height is 0.4, height_reduction is 0.1, factor is 1.7, factor_reduction is 0.2,
base_height is 0.039, enrichment_factor is 1.1, processing_factor is 4.5. We can set these parameters for running
<code class="docutils literal"><span class="pre">tss</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic tss_processing \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -tl $TEX_LIBS \
    -p test \
    -he 0.4 \
    -rh 0.1 \
    -fa 1.7 \
    -rf 0.2 \
    -bh 0.039 \
    -ef 1.1 \
    -pf 4.5 \
    -v \
    -rt all_1 \
    -le NC_009839.1:200000 \
    -m ANNOgesic/input/manual_TSS/NC_009839_manual_TSS.gff \
    -pj ANNOgesic
</pre></div>
</div>
<p>We assign the manual-checked TSS gff file to <code class="docutils literal"><span class="pre">-m</span></code>. Therefore, the output gff file contains the manual-detected TSSs and predicted TSSs.
If we didn&#8217;t assign it, Only the predicted TSSs will be included in output gff file.
The output files are gff file, MasterTable and statistic files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/TSS/
MasterTables  configs  gffs  optimized_TSSpredator  statistics
$ ls ANNOgesic/output/TSS/configs/
config_NC_009839.1.ini
$ ls ANNOgesic/output/TSS/gffs/
NC_009839.1_TSS.gff
$ ls ANNOgesic/output/TSS/MasterTables/MasterTable_NC_009839.1/
AlignmentStatistics.tsv  err.txt  log.txt  MasterTable.tsv  superConsensus.fa  superTSS.gff  superTSStracks.gff  test_super.fa  test_super.gff  test_TSS.gff  TSSstatistics.tsv
$ ls ANNOgesic/output/TSS/statistics/NC_009839.1/
TSS_class_NC_009839.1.png  TSSstatistics.tsv               stat_TSS_libs_NC_009839.1.csv                    stat_gene_vali_NC_009839.1.csv
TSS_venn_NC_009839.1.png   stat_TSS_class_NC_009839.1.csv  stat_compare_TSSpredator_manual_NC_009839.1.csv
</pre></div>
</div>
<p>If we want to predict processing sites, the procedures are the same. we just need to change the program from TSS to
processing_site (<code class="docutils literal"><span class="pre">-t</span></code>) and assign the proper parameter sets. We assume the best parameter sets are following:
height is 0.2, height_reduction is 0.1, factor is 2.0, factor_reduction is 0.5,
base_height is 0.009, enrichment_factor is 1.2, processing_factor is 1.5.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic tss_processing \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -tl $TEX_LIBS \
    -p test \
    -he 0.2 \
    -rh 0.1 \
    -fa 2.0 \
    -rf 0.5 \
    -bh 0.009 \
    -ef 1.2 \
    -pf 1.5 \
    -rt all_1 \
    -t processing_site \
    -pj ANNOgesic
</pre></div>
</div>
<p>The output files are following:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/processing_site/
configs  gffs  MasterTables  statistics
$ ls ANNOgesic/output/processing_site/configs/
config_NC_009839.1.ini
$ ls ANNOgesic/output/processing_site/gffs/
NC_009839.1_processing.gff
$ ls ANNOgesic/output/processing_site/MasterTables/MasterTable_NC_009839.1/
AlignmentStatistics.tsv  err.txt  log.txt  MasterTable.tsv  superConsensus.fa  superTSS.gff  superTSStracks.gff  test_super.fa  test_super.gff  test_TSS.gff  TSSstatistics.tsv
$ ls ANNOgesic/output/processing_site/statistics/NC_009839.1/
TSSstatistics.tsv  processing_class_NC_009839.1.png  processing_venn_NC_009839.1.png  stat_processing_class_NC_009839.1.csv  stat_processing_libs_NC_009839.1.csv
</pre></div>
</div>
<p>Since we use TSSpredator to detect processing site, the files in
<code class="docutils literal"><span class="pre">ANNOgesic/output/processing_site/MasterTables/MasterTable_NC_009839.1/</span></code> are for processing site not for TSS.</p>
</div>
<div class="section" id="performing-transcript-detection">
<h2>Performing transcript detection<a class="headerlink" href="#performing-transcript-detection" title="Permalink to this headline">¶</a></h2>
<p>transcript detection is a basic precedure for detecting transcript boundary.
we can use subcommand <code class="docutils literal"><span class="pre">transcript</span></code> to do it. Normally, we strongly
recommand that the user should provide fragmented libraries. Because dRNA-Seq usually loses some information
of 3&#8217;end. However, we only use TEX +/- for testing.</p>
<p>The command is like following:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic transcript \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -tl $TEX_LIBS \
    -rt all_1 \
    -cf gene CDS \
    -ct ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -pj ANNOgesic
</pre></div>
</div>
<p>The output files are gff files, tables and statistic files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/transcript/gffs
NC_009839.1_transcript.gff
$ ls ANNOgesic/output/transcript/tables
NC_009839.1_transcript.csv
$ ls ANNOgesic/output/transcript/statistics
NC_009839.1_length_all.png  NC_009839.1_length_less_2000.png  stat_compare_transcript_TSS_NC_009839.1.csv  stat_compare_transcript_genome_NC_009839.1.csv
</pre></div>
</div>
</div>
<div class="section" id="prediction-of-terminator">
<h2>Prediction of terminator<a class="headerlink" href="#prediction-of-terminator" title="Permalink to this headline">¶</a></h2>
<p>We can use subcommand <code class="docutils literal"><span class="pre">terminator</span></code> to detect terminators. <code class="docutils literal"><span class="pre">terminator</span></code> integrates <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/RNAfold.1.html">RNAfold</a>
for computing secondary structure of potential terminators. Therefore, this process will take a while. The command is like following:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic terminator \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -tl $TEX_LIBS \
    -rt all_1 -tb \
    -pj ANNOgesic
</pre></div>
</div>
<p>Four different kinds of gff files and tables will be generated.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/terminator/gffs/
all_candidates  best  express non_express
$ ls ANNOgesic/output/terminator/tables
all_candidates  best  express non_express
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">all_candidates</span></code> is for all candidates; <code class="docutils literal"><span class="pre">express</span></code> is for the candidates which reveal gene expression;
<code class="docutils literal"><span class="pre">best</span></code> is for the candidates which reveal gene expression and their coverage shows significant decreasing;
<code class="docutils literal"><span class="pre">non_express</span></code> is for the candidates which have no expression.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/terminator/gffs/best
NC_009839.1_term.gff
$ ls ANNOgesic/output/terminator/gffs/express
NC_009839.1_term.gff
$ ls ANNOgesic/output/terminator/gffs/all_candidates
NC_009839.1_term.gff
$ ls ANNOgesic/output/terminator/gffs/non_express
NC_009839.1_term.gff
$ ls ANNOgesic/output/terminator/tables/best
NC_009839.1_term.csv
$ ls ANNOgesic/output/terminator/tables/express
NC_009839.1_term.csv
$ ls ANNOgesic/output/terminator/tables/all_candidates
NC_009839.1_term.csv
$ ls ANNOgesic/output/terminator/tables/non_express
NC_009839.1_term.csv
</pre></div>
</div>
<p>In transtermhp folder, output files of <a class="reference external" href="http://transterm.cbcb.umd.edu/">TranstermHP</a> can be found.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/terminator/transtermhp/NC_009839.1
NC_009839.1_best_terminator_after_gene.bag  NC_009839.1_terminators.txt  NC_009839.1_terminators_within_robust_tail-to-tail_regions.t2t
</pre></div>
</div>
<p>Moreover, statistic files are stored in <code class="docutils literal"><span class="pre">statistics</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/terminator/statistics/
stat_NC_009839.1.csv                                               stat_compare_terminator_transcript_NC_009839.1_best.csv
stat_compare_terminator_transcript_NC_009839.1_all_candidates.csv  stat_compare_terminator_transcript_NC_009839.1_express.csv
</pre></div>
</div>
</div>
<div class="section" id="generating-utr">
<h2>Generating UTR<a class="headerlink" href="#generating-utr" title="Permalink to this headline">¶</a></h2>
<p>Now, we have the information of TSSs, transcripts and terminators. We can detect the 5&#8217;UTRs and 3&#8217;UTRs easily by using
subcommand <code class="docutils literal"><span class="pre">utr</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic utr \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -t ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -e ANNOgesic/output/terminator/gffs/best/NC_009839.1_term.gff \
    -pj ANNOgesic
</pre></div>
</div>
<p>If the TSS gff file is not generated by ANNOgesic, please assign <code class="docutils literal"><span class="pre">-s</span></code>,  the TSSs can be classified for generating UTRs.
Output gff files and statistic files will be stored in <code class="docutils literal"><span class="pre">5UTR</span></code> and <code class="docutils literal"><span class="pre">3UTR</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/UTR/3UTR
gffs/       statistics/
$ ls ANNOgesic/output/UTR/5UTR
gffs/       statistics/
$ ls ANNOgesic/output/UTR/3UTR/gffs
NC_009839.1_3UTR.gff
$ ls ANNOgesic/output/UTR/5UTR/gffs
NC_009839.1_5UTR.gff
$ ls ANNOgesic/output/UTR/5UTR/statistics
NC_009839.1_all_5utr_length.png
$ ls ANNOgesic/output/UTR/3UTR/statistics
NC_009839.1_all_3utr_length.png
</pre></div>
</div>
<p>Now, we have all information for defining the transcript boundary.</p>
</div>
<div class="section" id="detecting-operon-and-suboperon">
<h2>Detecting operon and suboperon<a class="headerlink" href="#detecting-operon-and-suboperon" title="Permalink to this headline">¶</a></h2>
<p>We already had TSSs, transcripts, terminators, CDSs, UTRs. We can integrate all these feature to
detect operons and suboperons by executing subcommand <code class="docutils literal"><span class="pre">operon</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic operon \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -t ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -u5 ANNOgesic/output/UTR/5UTR/gffs/NC_009839.1_5UTR.gff \
    -u3 ANNOgesic/output/UTR/3UTR/gffs/NC_009839.1_3UTR.gff \
    -e ANNOgesic/output/terminator/gffs/best/NC_009839.1_term.gff \
    -pj ANNOgesic
</pre></div>
</div>
<p>Three folders will be generated to store gff files, tables and statistics files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/operons/
gffs  statistics  tables
$ ls ANNOgesic/output/operons/gffs/
NC_009839.1_operon.gff
$ ls ANNOgesic/output/operons/tables/
NC_009839.1_operon.csv
$ ls ANNOgesic/output/operons/statistics/
stat_NC_009839.1_operon.csv
</pre></div>
</div>
</div>
<div class="section" id="promoter-motif-detection">
<h2>Promoter motif detection<a class="headerlink" href="#promoter-motif-detection" title="Permalink to this headline">¶</a></h2>
<p>As long as we have TSSs, we can use subcommand <code class="docutils literal"><span class="pre">promoter</span></code> to get promoters. The promoters can be detected
by different types of the TSSs. Therefore, if the TSSs gff files are not generated by <code class="docutils literal"><span class="pre">ANNOgesic</span></code>,
you need to add <code class="docutils literal"><span class="pre">-s</span></code> and assign corresponding genome annotation file to <code class="docutils literal"><span class="pre">-g</span></code>.
Now, let try <code class="docutils literal"><span class="pre">promoter</span></code> by running MEME and GLAM2 (<code class="docutils literal"><span class="pre">-p</span></code> is assigned by &#8220;both&#8221; in default. If you want to only run
MEME or GLAM2, please assign &#8220;meme&#8221; or &#8220;glam2&#8221; to <code class="docutils literal"><span class="pre">-p</span></code>), the process may take a while.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic promoter \
    -t ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -w 45 2-10 \
    -pj ANNOgesic
</pre></div>
</div>
<p>We define the length of the motifs as <code class="docutils literal"><span class="pre">50</span></code> and <code class="docutils literal"><span class="pre">2-10</span></code>. <code class="docutils literal"><span class="pre">2-10</span></code> means the width can be from 2 to 10.</p>
<p>Based on different types of the TSSs and the length of the motif, numerous output files will be generated.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/promoter_analysis/
NC_009839.1/ fasta_class
$ ls ANNOgesic/output/promoter_analysis/fasta_class/NC_009839.1
NC_009839.1_allstrain_all_types.fa  NC_009839.1_allstrain_internal.fa  NC_009839.1_allstrain_primary.fa    NC_009839.1_allstrain_without_orphan.fa
NC_009839.1_allstrain_antisense.fa  NC_009839.1_allstrain_orphan.fa    NC_009839.1_allstrain_secondary.fa
$ ls ANNOgesic/output/promoter_analysis/NC_009839.1
MEME GLAM2
$ ls ANNOgesic/output/promoter_analysis/NC_009839.1/MEME
promoter_motifs_NC_009839.1_allstrain_all_types_2-10_nt  promoter_motifs_NC_009839.1_allstrain_internal_45_nt   promoter_motifs_NC_009839.1_allstrain_secondary_2-10_nt
promoter_motifs_NC_009839.1_allstrain_all_types_45_nt    promoter_motifs_NC_009839.1_allstrain_orphan_2-10_nt   promoter_motifs_NC_009839.1_allstrain_secondary_45_nt
promoter_motifs_NC_009839.1_allstrain_antisense_2-10_nt  promoter_motifs_NC_009839.1_allstrain_orphan_45_nt     promoter_motifs_NC_009839.1_allstrain_without_orphan_2-10_nt
promoter_motifs_NC_009839.1_allstrain_antisense_45_nt    promoter_motifs_NC_009839.1_allstrain_primary_2-10_nt  promoter_motifs_NC_009839.1_allstrain_without_orphan_45_nt
promoter_motifs_NC_009839.1_allstrain_internal_2-10_nt   promoter_motifs_NC_009839.1_allstrain_primary_45_nt
$ ls ANNOgesic/output/promoter_analysis/NC_009839.1/GLAM2
promoter_motifs_NC_009839.1_allstrain_all_types_2-10_nt  promoter_motifs_NC_009839.1_allstrain_internal_45_nt   promoter_motifs_NC_009839.1_allstrain_secondary_2-10_nt
promoter_motifs_NC_009839.1_allstrain_all_types_45_nt    promoter_motifs_NC_009839.1_allstrain_orphan_2-10_nt   promoter_motifs_NC_009839.1_allstrain_secondary_45_nt
promoter_motifs_NC_009839.1_allstrain_antisense_2-10_nt  promoter_motifs_NC_009839.1_allstrain_orphan_45_nt     promoter_motifs_NC_009839.1_allstrain_without_orphan_2-10_nt
promoter_motifs_NC_009839.1_allstrain_antisense_45_nt    promoter_motifs_NC_009839.1_allstrain_primary_2-10_nt  promoter_motifs_NC_009839.1_allstrain_without_orphan_45_nt
promoter_motifs_NC_009839.1_allstrain_internal_2-10_nt   promoter_motifs_NC_009839.1_allstrain_primary_45_nt
$ ls ANNOgesic/output/promoter_analysis/NC_009839.1/MEME/promoter_motifs_NC_009839.1_allstrain_all_types_45_nt/
logo1.eps  logo2.eps  logo3.eps  logo4.eps  logo5.eps  logo_rc1.eps  logo_rc2.eps  logo_rc3.eps  logo_rc4.eps  logo_rc5.eps  meme.csv   meme.txt
logo1.png  logo2.png  logo3.png  logo4.png  logo5.png  logo_rc1.png  logo_rc2.png  logo_rc3.png  logo_rc4.png  logo_rc5.png  meme.html  meme.xml
$ ls ANNOgesic/output/promoter_analysis/NC_009839.1/GLAM2/promoter_motifs_NC_009839.1_allstrain_all_types_45_nt/
glam2.csv   glam2.txt   logo1.eps  logo2.png  logo4.eps  logo5.png  logo7.eps  logo8.png  logo_ssc10.eps  logo_ssc1.png  logo_ssc3.eps  logo_ssc4.png  logo_ssc6.eps  logo_ssc7.png  logo_ssc9.eps
glam2.html  logo10.eps  logo1.png  logo3.eps  logo4.png  logo6.eps  logo7.png  logo9.eps  logo_ssc10.png  logo_ssc2.eps  logo_ssc3.png  logo_ssc5.eps  logo_ssc6.png  logo_ssc8.eps  logo_ssc9.png
glam2.meme  logo10.png  logo2.eps  logo3.png  logo5.eps  logo6.png  logo8.eps  logo9.png  logo_ssc1.eps   logo_ssc2.png  logo_ssc4.eps  logo_ssc5.png  logo_ssc7.eps  logo_ssc8.png
</pre></div>
</div>
</div>
<div class="section" id="prediction-of-srna-and-sorf">
<h2>Prediction of sRNA and sORF<a class="headerlink" href="#prediction-of-srna-and-sorf" title="Permalink to this headline">¶</a></h2>
<p>Based on trascripts, genome annotation and coverage information, sRNAs can be detected. Moreover, we
have TSSs and processing sites which can be used for detecting UTR-derived sRNAs as well. Now, we can
get sRNAs by running subcommand <code class="docutils literal"><span class="pre">srna</span></code>. Normally, we recommand that the user inputs fragmented libraries as well.
Here, we only use TEX +/- for testing.</p>
<p>For running <code class="docutils literal"><span class="pre">srna</span></code>, we can apply several filters to improve the detection. These filters are <code class="docutils literal"><span class="pre">tss</span></code>, <code class="docutils literal"><span class="pre">sec_str</span></code>,
<code class="docutils literal"><span class="pre">blast_nr</span></code>, <code class="docutils literal"><span class="pre">blast_srna</span></code>, <code class="docutils literal"><span class="pre">promoter</span></code>, <code class="docutils literal"><span class="pre">term</span></code>, <code class="docutils literal"><span class="pre">sorf</span></code>. Normally, <code class="docutils literal"><span class="pre">tss</span></code>, <code class="docutils literal"><span class="pre">sec_str</span></code>,
<code class="docutils literal"><span class="pre">blast_nr</span></code>, <code class="docutils literal"><span class="pre">blast_srna</span></code> are recommaned to used.</p>
<p>Before running <code class="docutils literal"><span class="pre">srna</span></code>, we have to get sRNA database (we can use <a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a>) and
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/">nr database</a> (if you have not downloaded before).
We can download fasta file of <a class="reference external" href="http://www.bac-srna.org/BSRD/index.jsp">BSRD</a> from our
<a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/database">Git repository</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/database/ https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/database/sRNA_database_BSRD.fa
</pre></div>
</div>
<p>If you already had sRNA database in other folders, please assign your path of databases to <code class="docutils literal"><span class="pre">-sd</span></code>.
If your databse is formated before, you can remove <code class="docutils literal"><span class="pre">-sf</span></code>.
In order to use the recommaned filters, we have to download
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/">nr database</a> (takes a while). If you already had it,
you can skip this step.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/database/ ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz
$ gunzip ANNOgesic/input/database/nr.gz
$ mv ANNOgesic/input/database/nr ANNOgesic/input/database/nr.fa
</pre></div>
</div>
<p>If your nr database is in other folders, please assign your path <code class="docutils literal"><span class="pre">-nd</span></code>.
You can also remove <code class="docutils literal"><span class="pre">-nf</span></code> if your database is already formated.
Now, we can use the recommanded filters to run <code class="docutils literal"><span class="pre">srna</span></code>, but it may takes several hours.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic srna \
    -d tss blast_srna blast_nr sec_str \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -t ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -p ANNOgesic/output/processing_site/gffs/NC_009839.1_processing.gff \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -tf ANNOgesic/output/terminator/gffs/best/NC_009839.1_term.gff \
    -pt ANNOgesic/output/promoter_analysis/NC_009839.1/promoter_motifs_NC_009839.1_allstrain_all_types_45_nt/meme.csv \
    -pn MOTIF_1 \
    -m \
    -u \
    -sf \
    -nf \
    -sd ANNOgesic/input/database/sRNA_database_BSRD \
    -nd ANNOgesic/input/database/nr \
    -tl $TEX_LIBS \
    -rt all_1 \
    -pj ANNOgesic
</pre></div>
</div>
<p>If you have sORF information, you can also assign path of the sORF gff folder to <code class="docutils literal"><span class="pre">-O</span></code>.
Then, comparison of sRNAs and sORFs can be done.</p>
<p>Output files are following.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sRNA/
blast_result_and_misc  gffs  log.txt  mountain_plot  sRNA_2d_NC_009839.1  sRNA_seq_NC_009839.1  sec_structure  statistics  tables
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">blast_result_and_misc</span></code> stores the results of blast; <code class="docutils literal"><span class="pre">mountain_plot</span></code> stores mountain plots of sRNAs;
<code class="docutils literal"><span class="pre">sec_structure</span></code> stores the plots of the sRNA secondary structures; <code class="docutils literal"><span class="pre">statistics</span></code> stores statistic files.</p>
<p><code class="docutils literal"><span class="pre">sRNA_2d_NC_009839.1</span></code> and <code class="docutils literal"><span class="pre">sRNA_seq_NC_009839.1</span></code> are text files of sRNA sequences and secondary structures.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sRNA/blast_result_and_misc/
nr_blast_NC_009839.1.txt  sRNA_blast_NC_009839.1.txt
$ ls ANNOgesic/output/sRNA/mountain_plot/NC_009839.1/
srna0_NC_009839.1_36954_37044_-_mountain.pdf     srna25_NC_009839.1_854600_854673_-_mountain.pdf    srna40_NC_009839.1_1091155_1091251_-_mountain.pdf  srna56_NC_009839.1_1440826_1441414_+_mountain.pdf
srna10_NC_009839.1_248098_248257_-_mountain.pdf  srna26_NC_009839.1_879881_880088_-_mountain.pdf    srna41_NC_009839.1_1097654_1097750_-_mountain.pdf  srna57_NC_009839.1_1448211_1448306_+_mountain.pdf
...

$ ls ANNOgesic/output/sRNA/sec_structure/dot_plot/NC_009839.1/
srna0_NC_009839.1_36954_37044_-_dp.pdf     srna25_NC_009839.1_854600_854673_-_dp.pdf    srna40_NC_009839.1_1091155_1091251_-_dp.pdf  srna56_NC_009839.1_1440826_1441414_+_dp.pdf
srna10_NC_009839.1_248098_248257_-_dp.pdf  srna26_NC_009839.1_879881_880088_-_dp.pdf    srna41_NC_009839.1_1097654_1097750_-_dp.pdf  srna57_NC_009839.1_1448211_1448306_+_dp.pdf
...

$ ls ANNOgesic/output/sRNA/sec_structure/sec_plot/NC_009839.1/
rna0_NC_009839.1_36954_37044_-_rss.pdf     srna25_NC_009839.1_854600_854673_-_rss.pdf    srna40_NC_009839.1_1091155_1091251_-_rss.pdf  srna56_NC_009839.1_1440826_1441414_+_rss.pdf
srna10_NC_009839.1_248098_248257_-_rss.pdf  srna26_NC_009839.1_879881_880088_-_rss.pdf    srna41_NC_009839.1_1097654_1097750_-_rss.pdf  srna57_NC_009839.1_1448211_1448306_+_rss.pdf
...

$ ls ANNOgesic/output/sRNA/statistics/
stat_NC_009839.1_sRNA_blast.csv  stat_sRNA_class_NC_009839.1.csv
</pre></div>
</div>
<p>In <code class="docutils literal"><span class="pre">gffs</span></code> and <code class="docutils literal"><span class="pre">tables</span></code>, three different folders are generated. <code class="docutils literal"><span class="pre">all_candidates</span></code> is for all candidates
without filtering; <code class="docutils literal"><span class="pre">best</span></code> is for the candidates after filtering;
<code class="docutils literal"><span class="pre">for_class</span></code> is for different sRNA types based on <code class="docutils literal"><span class="pre">stat_sRNA_class_NC_009839.1.csv</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sRNA/gffs/
all_candidates  best  for_class
$ ls ANNOgesic/output/sRNA/tables/
all_candidates  best  for_class
$ ls ANNOgesic/output/sRNA/gffs/all_candidates/
NC_009839.1_sRNA.gff
$ ls ANNOgesic/output/sRNA/tables/all_candidates/
NC_009839.1_sRNA.csv
$ ls ANNOgesic/output/sRNA/gffs/best/
NC_009839.1_sRNA.gff
$ ls ANNOgesic/output/sRNA/tables/best/
NC_009839.1_sRNA.csv
$ ls ANNOgesic/output/sRNA/gffs/for_class/NC_009839.1/
class_1_all.gff                                          class_1_class_2_class_7_all.gff                  class_2_all.gff                                  class_3_all.gff
class_1_class_2_all.gff                                  class_1_class_3_all.gff                          class_2_class_3_all.gff                          class_3_class_4_all.gff
...

$ ls ANNOgesic/output/sRNA/tables/for_class/NC_009839.1/
class_1_all.csv                                          class_1_class_2_class_7_all.csv                  class_2_all.csv                                  class_3_all.csv
class_1_class_2_all.csv                                  class_1_class_3_all.csv                          class_2_class_3_all.csv                          class_3_class_4_all.csv
...
</pre></div>
</div>
<p>As we know, expressed region without annotation may be sORF as well.
In order to get information of sORFs, we can use subcommand <code class="docutils literal"><span class="pre">sorf</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic sorf \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -t ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -s ANNOgesic/output/sRNA/gffs/best/NC_009839.1_sRNA.gff \
    -tl $TEX_LIBS \
    -rt all_1 -u \
    -pj ANNOgesic
</pre></div>
</div>
<p>For generating best candidates, some filters can be assigned
(ex: with ribosome binding site, with TSS, without overlap with sRNA, etc.).
After running <code class="docutils literal"><span class="pre">sorf</span></code>, gff files, statistic files and tables of the sORF will be generated. <code class="docutils literal"><span class="pre">all_candidates</span></code>
stores the gff files and tables without filtering; <code class="docutils literal"><span class="pre">best</span></code> stores the gff_files and tables with filtering.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sORF/gffs/all_candidates/
NC_009839.1_sORF.gff
$ ls ANNOgesic/output/sORF/gffs/best/
NC_009839.1_sORF.gff
$ ls ANNOgesic/output/sORF/tables/all_candidates/
NC_009839.1_sORF.csv
$ ls ANNOgesic/output/sORF/tables/best/
NC_009839.1_sORF.csv
$ ls ANNOgesic/output/sORF/statistics/
stat_NC_009839.1_sORF.csv
</pre></div>
</div>
</div>
<div class="section" id="performing-srna-target-prediction">
<h2>Performing sRNA target prediction<a class="headerlink" href="#performing-srna-target-prediction" title="Permalink to this headline">¶</a></h2>
<p>Now we have sRNA candidates. If we want to know targets of these sRNAs, we can use <code class="docutils literal"><span class="pre">srna_target</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic srna_target \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -r ANNOgesic/output/sRNA/gffs/best/NC_009839.1_sRNA.gff \
    -q NC_009839.1:36954:37044:- \
    -p both \
    -pj ANNOgesic
</pre></div>
</div>
<p>For testing, we only assign one sRNA to do the prediction. You can also assign several of sRNAs like
<code class="docutils literal"><span class="pre">NC_009839.1:36954:37044:-</span> <span class="pre">NC_009839.1:75845:75990:+</span></code>. If you want to compute all sRNAs, you
can assign <code class="docutils literal"><span class="pre">all</span></code> to <code class="docutils literal"><span class="pre">-q</span></code> (may take several days).</p>
<p>Several output folders will be generated.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sRNA_targets/
merge  RNAplex  RNAup  sRNA_seqs  target_seqs
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">sRNA_seqs</span></code> and <code class="docutils literal"><span class="pre">target_seqs</span></code> are for sequences of the sRNAs and the potential targets.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sRNA_targets/sRNA_seqs
NC_009839.1_sRNA.fa
$ ls ANNOgesic/output/sRNA_targets/target_seqs
NC_009839.1_target.fa
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">RNAplex</span></code> and <code class="docutils literal"><span class="pre">RNAup</span></code> are for output of <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/">RNAplex and RNAup</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sRNA_targets/RNAplex/NC_009839.1/
NC_009839.1_RNAplex_rank.csv  NC_009839.1_RNAplex.txt
$ ls ANNOgesic/output/sRNA_targets/RNAup/NC_009839.1/
NC_009839.1_RNAup.log  NC_009839.1_RNAup_rank.csv  NC_009839.1_RNAup.txt
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">merge</span></code> is for the merged results of <a class="reference external" href="http://www.tbi.univie.ac.at/RNA/RNAplex.1.html">RNAplex</a> and
<a class="reference external" href="http://www.tbi.univie.ac.at/RNA/RNAup.1.html">RNAup</a>. <code class="docutils literal"><span class="pre">NC_009839.1_merge.csv</span></code>  contains all results of the
both methods. <code class="docutils literal"><span class="pre">NC_009839.1_overlap.csv</span></code> only stores candidates which are top 20 (default) in the both methods.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/sRNA_targets/merge/NC_009839.1/
NC_009839.1_merge.csv  NC_009839.1_overlap.csv
</pre></div>
</div>
</div>
<div class="section" id="mapping-and-detecting-of-circular-rna">
<h2>Mapping and detecting of circular RNA<a class="headerlink" href="#mapping-and-detecting-of-circular-rna" title="Permalink to this headline">¶</a></h2>
<p>You may also be interested in circular RNAs. The subcommand <code class="docutils literal"><span class="pre">circrna</span></code> can help us to get circular RNAs by
using <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>. Since
we didn&#8217;t map reads of the test case before, we can also do mapping by running <code class="docutils literal"><span class="pre">circrna</span></code>. If you already mapped
the reads by <a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> with <code class="docutils literal"><span class="pre">-S</span></code>, then you can
remove <code class="docutils literal"><span class="pre">-a</span></code> and add path of the bam files to <code class="docutils literal"><span class="pre">-nb</span></code> or <code class="docutils literal"><span class="pre">-fb</span></code>. However,
if you mapped the reads by other tools or you mapped the reads by
<a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a> without <code class="docutils literal"><span class="pre">-S</span></code>, Unfortunately,
you have to re-map the reads again. You can assign parallel (<code class="docutils literal"><span class="pre">-p</span></code>) for mapping.</p>
<p>In normal situation, the reads should be directly given to <code class="docutils literal"><span class="pre">circrna</span></code>. However, we just want to test the
subcommand. Thus, we can reduce the running time by selecting the subset of reads (first 50000) for only testing.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ head -n 50000 ANNOgesic/input/reads/SRR515254.fasta &gt; ANNOgesic/input/reads/SRR515254_50000.fasta
$ head -n 50000 ANNOgesic/input/reads/SRR515255.fasta &gt; ANNOgesic/input/reads/SRR515255_50000.fasta
$ head -n 50000 ANNOgesic/input/reads/SRR515256.fasta &gt; ANNOgesic/input/reads/SRR515256_50000.fasta
$ head -n 50000 ANNOgesic/input/reads/SRR515257.fasta &gt; ANNOgesic/input/reads/SRR515257_50000.fasta
$ rm ANNOgesic/input/reads/SRR515254.fasta
$ rm ANNOgesic/input/reads/SRR515255.fasta
$ rm ANNOgesic/input/reads/SRR515256.fasta
$ rm ANNOgesic/input/reads/SRR515257.fasta
</pre></div>
</div>
<p>Now, we can try <code class="docutils literal"><span class="pre">circrna</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic circrna \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -p 10 \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -a \
    -rp ANNOgesic/input/reads/SRR515254_50000.fasta \
        ANNOgesic/input/reads/SRR515255_50000.fasta \
        ANNOgesic/input/reads/SRR515256_50000.fasta \
        ANNOgesic/input/reads/SRR515257_50000.fasta \
    -pj ANNOgesic
</pre></div>
</div>
<p>If you can&#8217;t find testrealign.x, please refer to <a class="reference internal" href="prerequired.html#required-tools-or-databases"><span class="std std-ref">Required tools or databases</span></a>.
Several output folders will be generated.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/circRNA/
circRNA_tables  gffs  segemehl_align  segemehl_splice statistics
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">segemehl_align</span></code> and <code class="docutils literal"><span class="pre">segemehl_splice</span></code> are for results of
<a class="reference external" href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">Segemehl</a>. <code class="docutils literal"><span class="pre">segemehl_align</span></code> stores Bam files of
the alignment and <code class="docutils literal"><span class="pre">segemehl_splice</span></code> stores results of the splice detection.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/circRNA/segemehl_align/NC_009839.1/
SRR515254_50000_NC_009839.1.bam  SRR515256_50000_NC_009839.1.bam
SRR515255_50000_NC_009839.1.bam  SRR515257_50000_NC_009839.1.bam
$ ls ANNOgesic/output/circRNA/segemehl_splice/NC_009839.1/
splicesites_all.bed  transrealigned_all.bed
</pre></div>
</div>
<p>Gff files, tables and statistic files are stored in <code class="docutils literal"><span class="pre">gffs</span></code>, <code class="docutils literal"><span class="pre">circRNA_tables</span></code> and <code class="docutils literal"><span class="pre">statistics</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/circRNA/gffs/NC_009839.1/
NC_009839.1_circRNA_all.gff  NC_009839.1_circRNA_best.gff
$ ls ANNOgesic/output/circRNA/circRNA_tables/NC_009839.1/
NC_009839.1_circRNA_all.csv  NC_009839.1_circRNA_best.csv
$ ls ANNOgesic/output/circRNA/statistics/
stat_circRNA_NC_009839.1.csv
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">NC_009839.1_circRNA_all.gff</span></code> and <code class="docutils literal"><span class="pre">NC_009839.1_circRNA_all.csv</span></code> store all circular RNAs without filtering.
<code class="docutils literal"><span class="pre">NC_009839.1_circRNA_best.gff</span></code> and <code class="docutils literal"><span class="pre">NC_009839.1_circRNA_best.csv</span></code> store
the circular RNAs after filering. In our case, there are some circular RNAs can be detected, but no circular RNAs
can exist after filtering.</p>
</div>
<div class="section" id="snp-calling">
<h2>SNP calling<a class="headerlink" href="#snp-calling" title="Permalink to this headline">¶</a></h2>
<p>If we want to know SNPs or mutations of our RNA-seq data, we can use <code class="docutils literal"><span class="pre">snp</span></code> to achieve this purpose.
<code class="docutils literal"><span class="pre">snp</span></code> is compose of two parts. One part is for obtaining the differences between our query strain (&#8220;target strain&#8221;)
and the close strain of our query strain (&#8220;reference strain&#8221;). If we have no fasta file of our &#8220;target strain&#8221;,
this part will be very useful. We just need to map reads of the &#8220;target strain&#8221; on fasta file of the &#8220;reference strain&#8221;. Then
using <code class="docutils literal"><span class="pre">snp</span></code> can automatically detect differences between the &#8220;target strain&#8221; and the &#8220;reference strain&#8221;.
Furthermore, potential fasta files of the &#8220;target strain&#8221; can be generated automatically as well.
The other part is for detecting SNPs or mutations of the &#8220;target strain&#8221;. In this part,
you can know real mutations of the &#8220;target strain&#8221;.</p>
<p>Before running the subcommand, bam files are required. Since we already generated them through
running <code class="docutils literal"><span class="pre">circrna</span></code>, we can just need to put them to right place. Please remember that the mapping function of
<code class="docutils literal"><span class="pre">circrna</span></code> is very basic.</p>
<p>Now, we can try to detect mutations of the &#8220;target strain&#8221; because our mapping is based on
fasta file of the &#8220;target strain&#8221; - NC_000915. Therefore, we copy the bam files to <code class="docutils literal"><span class="pre">BAMs_map_target</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cp ANNOgesic/output/circRNA/segemehl_align/NC_009839.1/SRR51525* ANNOgesic/input/BAMs/BAMs_map_target/tex_notex
</pre></div>
</div>
<p>Then we can run the subcommand with three programs &#8211; <code class="docutils literal"><span class="pre">extend_BAQ</span></code>, <code class="docutils literal"><span class="pre">with_BAQ</span></code> and <code class="docutils literal"><span class="pre">without_BAQ</span></code>, and sample number
for this test case is 1 (<code class="docutils literal"><span class="pre">-ms</span></code>).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic snp \
    -t target \
    -p with_BAQ without_BAQ extend_BAQ \
    -ms 1 \
    -b ANNOgesic/input/BAMs/BAMs_map_target/tex_notex/SRR515254_50000_NC_009839.1.bam \
       ANNOgesic/input/BAMs/BAMs_map_target/tex_notex/SRR515255_50000_NC_009839.1.bam \
       ANNOgesic/input/BAMs/BAMs_map_target/tex_notex/SRR515256_50000_NC_009839.1.bam \
       ANNOgesic/input/BAMs/BAMs_map_target/tex_notex/SRR515257_50000_NC_009839.1.bam \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -pj ANNOgesic
</pre></div>
</div>
<p>If you want to compare between the &#8220;reference strain&#8221; and the &#8220;target strain&#8221;, you just need to change
<code class="docutils literal"><span class="pre">-t</span></code> to <code class="docutils literal"><span class="pre">reference</span></code> and assign correct bam files.</p>
<p>Two output folders will be generated, <code class="docutils literal"><span class="pre">compare_reference</span></code> is for results of the comparison between the &#8220;reference strain&#8221;
and the &#8220;target strain&#8221;, <code class="docutils literal"><span class="pre">validate_target</span></code> is for results of detecting mutations of the &#8220;target strain&#8221;.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/SNP_calling/
compare_reference  validate_target
</pre></div>
</div>
<p>Since we run <code class="docutils literal"><span class="pre">validate_target</span></code>,  the output folders are produced under <code class="docutils literal"><span class="pre">validate_target</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/SNP_calling/validate_target/
SNP_raw_outputs  SNP_table  seqs  statistics
</pre></div>
</div>
<p>The output folders are compose of three parts - <code class="docutils literal"><span class="pre">extend_BAQ</span></code>, <code class="docutils literal"><span class="pre">with_BAQ</span></code> and <code class="docutils literal"><span class="pre">without_BAQ</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/SNP_calling/validate_target/seqs/
extend_BAQ/  with_BAQ/    without_BAQ/
</pre></div>
</div>
<p>In <code class="docutils literal"><span class="pre">seqs</span></code>, the potential sequences can be found.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/SNP_calling/validate_target/seqs/with_BAQ/NC_009839.1/
NC_009839.1_NC_009839.1_1_1.fa
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">SNP_raw_outputs</span></code> stores output of <a class="reference external" href="https://github.com/samtools">Samtools and Bcftools</a>.
<code class="docutils literal"><span class="pre">SNP_table</span></code> stores results after filtering and the indices of potential sequence
(potential sequences are stored in <code class="docutils literal"><span class="pre">seqs</span></code>).
<code class="docutils literal"><span class="pre">statistics</span></code> stores the statistic files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/SNP_calling/validate_target/SNP_raw_outputs/NC_009839.1/
NC_009839.1_extend_BAQ.vcf  NC_009839.1_with_BAQ.vcf  NC_009839.1_without_BAQ.vcf
$ ls ANNOgesic/output/SNP_calling/validate_target/SNP_table/NC_009839.1/
NC_009839.1_extend_BAQ_best.vcf     NC_009839.1_with_BAQ_best.vcf     NC_009839.1_without_BAQ_best.vcf
NC_009839.1_extend_BAQ_seq_reference.csv  NC_009839.1_with_BAQ_seq_reference.csv  NC_009839.1_without_BAQ_seq_reference.csv
$ ls ANNOgesic/output/SNP_calling/validate_target/statistics/
NC_009839.1_extend_BAQ_NC_009839.1_SNP_QUAL_best.png  NC_009839.1_with_BAQ_NC_009839.1_SNP_QUAL_raw.png      stat_NC_009839.1_extend_BAQ_SNP_best.csv  stat_NC_009839.1_with_BAQ_SNP_raw.csv
NC_009839.1_extend_BAQ_NC_009839.1_SNP_QUAL_raw.png   NC_009839.1_without_BAQ_NC_009839.1_SNP_QUAL_best.png  stat_NC_009839.1_extend_BAQ_SNP_raw.csv   stat_NC_009839.1_without_BAQ_SNP_best.csv
NC_009839.1_with_BAQ_NC_009839.1_SNP_QUAL_best.png    NC_009839.1_without_BAQ_NC_009839.1_SNP_QUAL_raw.png   stat_NC_009839.1_with_BAQ_SNP_best.csv    stat_NC_009839.1_without_BAQ_SNP_raw.csv
</pre></div>
</div>
</div>
<div class="section" id="mapping-gene-ontology">
<h2>Mapping Gene ontology<a class="headerlink" href="#mapping-gene-ontology" title="Permalink to this headline">¶</a></h2>
<p>Gene ontology is useful for understanding function of gene products.
Implementing <code class="docutils literal"><span class="pre">go_term</span></code> can map our annotations to gene ontology. Before running <code class="docutils literal"><span class="pre">go_term</span></code>, we
need to prepare some databases. First, please download
<a class="reference external" href="http://geneontology.org/page/go-slim-and-subset-guide">goslim.obo</a> and
<a class="reference external" href="http://geneontology.org/page/download-ontology">go.obo</a> and
<a class="reference external" href="http://www.uniprot.org/downloads">idmapping_selected.tab</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/database http://www.geneontology.org/ontology/subsets/goslim_generic.obo
$ wget -cP ANNOgesic/input/database http://geneontology.org/ontology/go.obo
$ wget -cP ANNOgesic/input/database ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping_selected.tab.gz
$ gunzip ANNOgesic/input/database/idmapping_selected.tab.gz
</pre></div>
</div>
<p>Now, we have all required databases. We can also import information of the transcripts to
generate results which only included the expressed CDS.</p>
<p>Let&#8217;s try it.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic go_term \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -pj ANNOgesic
</pre></div>
</div>
<p>Output of <code class="docutils literal"><span class="pre">go_term</span></code> are stored in <code class="docutils literal"><span class="pre">GO_term_results</span></code>. The statistic files and
figures are stored in <code class="docutils literal"><span class="pre">statistics</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/GO_term/
all_CDS  expressed_CDS
$ ls ANNOgesic/output/GO_term/all_CDS/
Go_term_results  statistics
$ ls ANNOgesic/output/GO_term/all_CDS/GO_term_results/NC_009839.1/
all_strains_uniprot.csv
$ ls ANNOgesic/output/GO_term/all_CDS/statistics/NC_009839.1/
figs  stat_NC_009839.1.csv
$ ls ANNOgesic/output/GO_term/all_CDS/statistics/NC_009839.1/figs/
NC_009839.1_biological_process.png  NC_009839.1_cellular_component.png  NC_009839.1_molecular_function.png  NC_009839.1_three_roots.png
</pre></div>
</div>
</div>
<div class="section" id="prediction-of-subcellular-localization">
<h2>Prediction of Subcellular localization<a class="headerlink" href="#prediction-of-subcellular-localization" title="Permalink to this headline">¶</a></h2>
<p>Subcellular localization is also a useful information for analysis of protein function. For
detecting subcellular localization, we can use the subcommand
<code class="docutils literal"><span class="pre">subcellular_localization</span></code>. Like <code class="docutils literal"><span class="pre">go_term</span></code>, we can also import
information of the transcript to generate results which only included the expressed CDS.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic subcellular_localization \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -m -b negative \
    -pj ANNOgesic
</pre></div>
</div>
<p>Two output folders will be generated. <code class="docutils literal"><span class="pre">psortb_results</span></code> stores output
of <a class="reference external" href="http://www.psort.org/psortb/">Psortb</a>. <code class="docutils literal"><span class="pre">statistics</span></code> stores
statistic files and figures.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/subcellular_localization/
all_CDS  expressed_CDS
$ ls ANNOgesic/output/subcellular_localization/all_CDS/
psortb_results  statistics
$ ls ANNOgesic/output/subcellular_localization/all_CDS/psortb_results/NC_009839.1/
NC_009839.1_raw.txt  NC_009839.1_table.csv
$ ls ANNOgesic/output/subcellular_localization/all_CDS/statistics/NC_009839.1/
NC_009839.1_NC_009839.1_sublocal.png  stat_NC_009839.1_sublocal.csv
</pre></div>
</div>
</div>
<div class="section" id="generating-protein-protein-interaction-network">
<h2>Generating protein-protein interaction network<a class="headerlink" href="#generating-protein-protein-interaction-network" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">ppi_network</span></code> can detect protein-protein interaction from <a class="reference external" href="http://string-db.org/">STRING</a>
(database of protein-protein interaction) and searching the literatures by implementing
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a>
(text-mining for protein-protein interaction). Therefore, <code class="docutils literal"><span class="pre">ppi_network</span></code> can generate protein-protein
interaction networks with supported literatures.</p>
<p>Before running the subcommand, you need to download
<a class="reference external" href="http://string-db.org/cgi/download.pl">species.v{$VERSIO}.txt from STRING</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/database http://string-db.org/newstring_download/species.v10.txt
</pre></div>
</div>
<p>Now, we can try the subcommand.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic ppi_network \
    -s NC_009839.1.gff:NC_009839.1:&#39;Campylobacter jejuni 81176&#39;:&#39;Campylobacter jejuni&#39; \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -d ANNOgesic/input/database/species.v10.txt \
    -q NC_009839.1:70579:71463:+ NC_009839.1:102567:103973:+ \
    -n \
    -pj ANNOgesic
</pre></div>
</div>
<p>We only detected for two proteins. If you want to detectc for all proteins in ptt files,
you can easily assign <code class="docutils literal"><span class="pre">all</span></code> in <code class="docutils literal"><span class="pre">-q</span></code>.</p>
<p>Three output folders will be generated.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/PPI/
all_results/  best_results/ figures/
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">all_results</span></code> is for all interactions without filtering. <code class="docutils literal"><span class="pre">best_results</span></code> is for the interactions with
the high <a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Wilbur/IRET/PIE/">PIE</a> score. <code class="docutils literal"><span class="pre">figures</span></code> is for
figures of the protein-protein interaction networks. There are two subfolders - <code class="docutils literal"><span class="pre">with_strain</span></code> and <code class="docutils literal"><span class="pre">without_strain</span></code> in
<code class="docutils literal"><span class="pre">figures</span></code>. These two folders store all information of the interactions and literature scores.
<code class="docutils literal"><span class="pre">with_strain</span></code> is for results with assignning specific strain name for searching literatures.
<code class="docutils literal"><span class="pre">without_strain</span></code> is for results without giving specific strain name for searching literatures.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/PPI/all_results/PPI_NC_009839.1/
NC_009839.1_without_strain.csv  NC_009839.1_with_strain.csv     without_strain/               with_strain/
$ ls ANNOgesic/output/PPI/best_results/PPI_NC_009839.1/
NC_009839.1_without_strain.csv  NC_009839.1_with_strain.csv  without_strain  with_strain
$ ls ANNOgesic/output/PPI/figures/PPI_NC_009839.1/
without_strain  with_strain
$ ls ANNOgesic/output/PPI/all_results/PPI_NC_009839.1/with_strain/NC_009839.1/
flgB_flgD.csv    flgE_flgD.csv  flgF_fliG.csv  flgG_fliG.csv  fliG_fliF.csv
flgE-1_flgD.csv  flgF_flgC.csv  flgG_flgC.csv  flgI_flgH.csv  pyrB_ansA.csv
$ ls ANNOgesic/output/PPI/all_results/PPI_NC_009839.1/without_strain/NC_009839.1/
flgB_flgD.csv    flgE_flgD.csv  flgF_fliG.csv  flgG_fliG.csv  fliG_fliF.csv
flgE-1_flgD.csv  flgF_flgC.csv  flgG_flgC.csv  flgI_flgH.csv  pyrB_ansA.csv
$ ls ANNOgesic/output/PPI/best_results/PPI_NC_009839.1/without_strain/NC_009839.1/
flgB_flgD.csv    flgE_flgD.csv  flgG_flgC.csv  fliG_fliF.csv
flgE-1_flgD.csv  flgF_flgC.csv  flgI_flgH.csv
$ ls ANNOgesic/output/PPI/best_results/PPI_NC_009839.1/with_strain/NC_009839.1/
fliG_fliF.csv
$ ls ANNOgesic/output/PPI/figures/PPI_NC_009839.1/with_strain/NC_009839.1/
C8J_RS00250_flgD.png
$ ls ANNOgesic/output/PPI/figures/PPI_NC_009839.1/without_strain/NC_009839.1/
C8J_RS00250_flgD.png
</pre></div>
</div>
</div>
<div class="section" id="generating-riboswitch-and-rna-thermometer">
<h2>Generating riboswitch and RNA thermometer<a class="headerlink" href="#generating-riboswitch-and-rna-thermometer" title="Permalink to this headline">¶</a></h2>
<p>If we want to detect riboswitches and RNA thermometer, we can use subcommand <code class="docutils literal"><span class="pre">riboswitch_thermometer</span></code>.
Before running it, we need to get information of the known riboswitches and RNA thermometer in Rfam.
The <a class="reference external" href="https://github.com/Sung-Huan/ANNOgesic/tree/master/database">riboswitches and RNA thermometer files</a>
can be downloaded them from our Git repository.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/riboswitch_ID/ https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/database/Rfam_riboswitch_ID.csv
$ wget -cP ANNOgesic/input/RNA_thermometer_ID/ https://raw.githubusercontent.com/Sung-Huan/ANNOgesic/master/database/Rfam_RNA_thermometer_ID.csv
</pre></div>
</div>
<p>We also need to download <a class="reference external" href="http://rfam.xfam.org/">Rfam</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ wget -cP ANNOgesic/input/database ftp://ftp.ebi.ac.uk/pub/databases/Rfam/12.0/Rfam.tar.gz
$ cd ANNOgesic/input/database
$ tar -zxvf Rfam.tar.gz
$ rm Rfam.tar.gz
$ cd ../../../
</pre></div>
</div>
<p>Now we can try the subcommand.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic riboswitch_thermometer \
    -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -ri ANNOgesic/input/riboswitch_ID/Rfam_riboswitch_ID.csv \
    -ti ANNOgesic/input/RNA_thermometer_ID/Rfam_RNA_thermometer_ID.csv \
    -R ANNOgesic/input/database/Rfam/CMs/Rfam.cm \
    -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
    -t ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -pj ANNOgesic
</pre></div>
</div>
<p>Output files are following, <code class="docutils literal"><span class="pre">gffs</span></code> stores gff files of the riboswitchs / RNA_thermometer;
<code class="docutils literal"><span class="pre">tables</span></code> stores tables of the riboswitchs / RNA_thermometer;
<code class="docutils literal"><span class="pre">scan_Rfam</span></code> stores output files of scanning Rfam; <code class="docutils literal"><span class="pre">statistics</span></code> is for statistic files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/riboswitch/
gffs  scan_Rfam  statistics  tables
$ ls ANNOgesic/output/riboswitch/gffs/
NC_009839.1_riboswitch.gff
$ ls ANNOgesic/output/riboswitch/scan_Rfam/NC_009839.1/
NC_009839.1_riboswitch_prescan.txt  NC_009839.1_riboswitch_scan.txt
$ ls ANNOgesic/output/riboswitch/tables/
NC_009839.1_riboswitch.csv
$ ls ANNOgesic/output/riboswitch/statistics/
stat_NC_009839.1_riboswitch.txt
$ ls ANNOgesic/output/RNA_thermometer/
gffs  scan_Rfam  statistics  tables
$ ls ANNOgesic/output/RNA_thermometer/gffs/
NC_009839.1_RNA_thermometer.gff
$ ls ANNOgesic/output/RNA_thermometer/scan_Rfam/NC_009839.1/
NC_009839.1_RNA_thermometer_prescan.txt  NC_009839.1_RNA_thermometer_scan.txt
$ ls ANNOgesic/output/RNA_thermometer/tables/
NC_009839.1_RNA_thermometer.csv
$ ls ANNOgesic/output/RNA_thermometer/statistics/
stat_NC_009839.1_RNA_thermometer.txt
</pre></div>
</div>
</div>
<div class="section" id="detection-of-crispr">
<h2>Detection of CRISPR<a class="headerlink" href="#detection-of-crispr" title="Permalink to this headline">¶</a></h2>
<p>CRISPR is an unique features for research of immunology. <code class="docutils literal"><span class="pre">crispr</span></code> is a useful subcommand for CRISPR detectiion.
<code class="docutils literal"><span class="pre">crispr</span></code> integrates <a class="reference external" href="http://www.room220.com/crt/">CRT</a> and compare genome
annotation to remove false positive. Let&#8217;s try it.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic crispr \
   -g ANNOgesic/output/target/annotation/NC_009839.1.gff \
   -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
   -pj ANNOgesic
</pre></div>
</div>
<p>Output are as following, <code class="docutils literal"><span class="pre">CRT_output</span></code> stores output of <a class="reference external" href="http://www.room220.com/crt/">CRT</a>;
<code class="docutils literal"><span class="pre">gffs</span></code> stores gff files of the CRISPRs; <code class="docutils literal"><span class="pre">statistics</span></code> is for statistic files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/crispr/
CRT_output  gffs  statistics
$ ls ANNOgesic/output/crispr/CRT_output
NC_009839.1.txt
$ ls ANNOgesic/output/crispr/gffs
all_candidates  best
$ ls ANNOgesic/output/crispr/gffs/all_candidates
NC_009839.1_CRISPR.gff
$ ls ANNOgesic/output/crispr/gffs/best
NC_009839.1_CRISPR.gff
$ ls ANNOgesic/output/crispr/statistics
NC_009839.1.csv
</pre></div>
</div>
</div>
<div class="section" id="merge-all-features-to-be-one-gff-file">
<h2>Merge all features to be one gff file<a class="headerlink" href="#merge-all-features-to-be-one-gff-file" title="Permalink to this headline">¶</a></h2>
<p>Now, we generated all features that ANNOgesic can provide. Sometimes, merging all features into
one gff file is useful. <code class="docutils literal"><span class="pre">merge_features</span></code> is the subcommand to achieve this purpose.
Moreover, <code class="docutils literal"><span class="pre">merge_features</span></code> can search parent transcript to each feature that
you assigned. the relationship between all features can be revealed.</p>
<p>Now let&#8217;s do it. We merge all features that we have.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ALL_FEATURES</span><span class="o">=</span><span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">TSS</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_TSS</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">target</span><span class="o">/</span><span class="n">annotation</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mf">1.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">UTR</span><span class="o">/</span><span class="mi">5</span><span class="n">UTR</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_5UTR</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">UTR</span><span class="o">/</span><span class="mi">3</span><span class="n">UTR</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_3UTR</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">terminator</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">best</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_term</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">processing_site</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_processing</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">sRNA</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">best</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_sRNA</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">sORF</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">best</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_sORF</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">riboswitch</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_riboswitch</span><span class="o">.</span><span class="n">gff</span> \
<span class="n">ANNOgesic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">crispr</span><span class="o">/</span><span class="n">gffs</span><span class="o">/</span><span class="n">best</span><span class="o">/</span><span class="n">NC_009839</span><span class="o">.</span><span class="mi">1</span><span class="n">_CRISPR</span><span class="o">.</span><span class="n">gff</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic merge_features \
   -a ANNOgesic/output/transcript/gffs/NC_009839.1_transcript.gff \
   -of $ALL_FEATURES\
   -s NC_009839.1 \
   -pj ANNOgesic
</pre></div>
</div>
<p>Output gff file is stored in <code class="docutils literal"><span class="pre">merge_all_features</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/merge_all_features/
NC_009839.1_merge_features.gff
</pre></div>
</div>
</div>
<div class="section" id="producing-the-screenshots">
<h2>Producing the screenshots<a class="headerlink" href="#producing-the-screenshots" title="Permalink to this headline">¶</a></h2>
<p>It is a good idea if we can get screenshots of our interesting features. Then we can
check them very quickly. Therefore, ANNOgesic provides a subcommand <code class="docutils literal"><span class="pre">screenshot</span></code> for
generating screenshots.</p>
<p>Before we running it, we have to install <a class="reference external" href="https://www.broadinstitute.org/software/igv/home">IGV</a>.</p>
<p>For testing, we use TSSs as main feature, sRNAs and CDSs as side features.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic screenshot \
    -mg ANNOgesic/output/TSS/gffs/NC_009839.1_TSS.gff \
    -sg ANNOgesic/output/target/annotation/NC_009839.1.gff \
        ANNOgesic/output/sRNA/gffs/best/NC_009839.1_sRNA.gff \
    -f ANNOgesic/output/target/fasta/NC_009839.1.fa \
    -o ANNOgesic/output/TSS \
    -tl $TEX_LIBS \
    -pj ANNOgesic
</pre></div>
</div>
<p>Two txt files and two folders will be generated.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/TSS/screenshots/NC_009839.1/
forward/     forward.txt  reverse/     reverse.txt
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">forward.txt</span></code> and <code class="docutils literal"><span class="pre">reverse.txt</span></code> are batch files for running in <a class="reference external" href="https://www.broadinstitute.org/software/igv/home">IGV</a>.
<code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code> are the folders for storing screenshots.</p>
<p>Since there are numerous candidates, we can only generate several ones in order to reduce the running time for testing.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ head -n 30 ANNOgesic/output/TSS/screenshots/NC_009839.1/forward.txt &gt; ANNOgesic/output/TSS/screenshots/NC_009839.1/forward_6_cases.txt
$ head -n 30 ANNOgesic/output/TSS/screenshots/NC_009839.1/reverse.txt &gt; ANNOgesic/output/TSS/screenshots/NC_009839.1/reverse_6_cases.txt
</pre></div>
</div>
<p>Now, please open <a class="reference external" href="https://www.broadinstitute.org/software/igv/home">IGV</a> and follow the procedures: Tools -&gt;
Run Batch Script -&gt; choose <code class="docutils literal"><span class="pre">forward_6_cases.txt</span></code>. Once it is done, please do it again for reverse strand: Tools -&gt;
Run Batch Script -&gt; choose <code class="docutils literal"><span class="pre">reverse_6_cases.txt</span></code>. If you just want to generate the screenshots for all candidates,
you can run <code class="docutils literal"><span class="pre">forward.txt</span></code> and <code class="docutils literal"><span class="pre">reverse.txt</span></code>. Please be careful, if you use Docker container, the path may be not correct.</p>
<p>As soon as the gneration of the screenshots is done,
we can see that there are several screenshots in <code class="docutils literal"><span class="pre">forward</span></code> and <code class="docutils literal"><span class="pre">reverse</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/TSS/screenshots/NC_009839.1/forward
NC_009839.1:1396-1396.png  NC_009839.1:14812-14812.png  NC_009839.1:6676-6676.png  NC_009839.1:6680-6680.png  NC_009839.1:8098-8098.png  NC_009839.1:9295-9295.png
$ ls ANNOgesic/output/TSS/screenshots/NC_009839.1/reverse
NC_009839.1:15670-15670.png  NC_009839.1:18053-18053.png  NC_009839.1:18360-18360.png  NC_009839.1:2199-2199.png  NC_009839.1:4463-4463.png  NC_009839.1:856-856.png
</pre></div>
</div>
</div>
<div class="section" id="coloring-the-screenshots">
<h2>Coloring the screenshots<a class="headerlink" href="#coloring-the-screenshots" title="Permalink to this headline">¶</a></h2>
<p>If we have numerous samples and we want to check TSSs, Distinguishing the
tracks of TEX+ and TEX- will be painful. Therefore, we provide a subcommand <code class="docutils literal"><span class="pre">color_png</span></code> to color
our screenshots.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ annogesic color_png \
    -t 2 \
    -f ANNOgesic/output/TSS \
    -pj ANNOgesic
</pre></div>
</div>
<p>We will see output filenames are the same as before. However, when we open the figures, the tracks are colored.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ls ANNOgesic/output/TSS/screenshots/NC_009839.1/forward
NC_009839.1:1396-1396.png  NC_009839.1:14812-14812.png  NC_009839.1:6676-6676.png  NC_009839.1:6680-6680.png  NC_009839.1:8098-8098.png  NC_009839.1:9295-9295.png
$ ls ANNOgesic/output/TSS/screenshots/NC_009839.1/reverse
NC_009839.1:15670-15670.png  NC_009839.1:18053-18053.png  NC_009839.1:18360-18360.png  NC_009839.1:2199-2199.png  NC_009839.1:4463-4463.png  NC_009839.1:856-856.png
</pre></div>
</div>
<p>Now we already finished the first wonderful trip of ANNOgesic. Hopefully, you enjoy it!</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="troubleshooting.html" title="Troubleshooting"
             >next</a></li>
        <li class="right" >
          <a href="subcommands.html" title="ANNOgesic’s subcommands"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ANNOgesic 0.5.6 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Sung-Huan Yu.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.3.
    </div>
  </body>
</html>